---
title: "Causal models of diversity"
author: "Gustavo V. Barroso"
date: "`r Sys.Date()`"
output:
  pdf_document:
  toc: true
number_sections: true
toc_depth: 1
---

```{r setup, include=F, message=F, warning=F}

knitr::opts_chunk$set(echo=TRUE)

library(data.table)
library(tidyverse)
library(cowplot)
library(scales)
library(GenomicRanges)

library(MASS)
library(lmtest)
library(nlme)
library(car)
library(interactions)
library(ppcor)
library(Hmisc)
library(MVN)
library(heplots)
library(parallel) # mcapply

library(survcomp) # for fisher's C test, combine.test()
library(piecewiseSEM)
library(semTools)
library(graph)
library(ggm)
library(lavaan)
library(dagitty)
library(ggdag)
library(magrittr)
library(ggpubr)
library(GGally)

library(reticulate)
use_python("/usr/bin/python3")

scale.4d <- function(x) sprintf("%.4f", x) # for adjusting precision in plots

knitr::opts_knit$set(root.dir="~/Desktop/sim_data")
#knitr::opts_knit$set(root.dir="~/Data/bgs_lmr/sim_data")
```

# Building chromosome models

We write a table with parameters shaping genomic landscapes
(mutation, recombination, functional elements)

Later, we will assume a single DFE for all elements.
Thus |s| can be seen as "constant" (see DAGs below).

```{r, build-chr-models, include=F, message=F, warning=F}

# Gamma dist params
shapes_rate_u <- c(5, 10, 30, 100)
shapes_rate_r <- c(1, 3, 10)
mean_r <- 1e-8
mean_u <- 1.25e-8

# Geom dist params
avg_rec_spans <- c(1e+3, 1e+4, 1e+5)
avg_mut_spans <- c(1e+3, 1e+4, 1e+5)
exon_lengths <- 1000
num_exons <- c(5000, 10000, 25000)
L <- 1e+8
mult_r_u <- c(0, 0.05, 0.1, 0.25) # non-mutagenic recomb. is a special case

p_stay <- 0.25 # related to auto-correlation in the distribution of elements
# note since the mean length of intergenic > exon_lengths, p_stay > 0 => AR > 0

models <- setDT(crossing(L, p_stay, num_exons, exon_lengths, mult_r_u,
                         mean_r, mean_u, shapes_rate_u, shapes_rate_r,
                         avg_rec_spans, avg_mut_spans))
models$model <- 1:nrow(models)

Sys.setenv(num_models=nrow(models))
Sys.setenv(num_reps=10)

fwrite(models, "models.csv")
models <- fread("models.csv")
```

```{bash, mk-dirs, engine.opts='-i', include=T, eval=T, results=F, warning=F}

source ~/.bashrc

for ((m=1; m <= $num_models; m++))
do
  mkdir model_$m
  cd model_$m
  for ((r=1; r <= $num_reps; r++))
  do
    mkdir rep_$r
  done
  cd ..
done

mkdir Coeff_lm
mkdir Correlations
mkdir DAGs
mkdir Manhattan
mkdir R2
mkdir SEM
```

We use the models to assemble the genomic maps

```{r, build-chr-maps, include=F, message=F, warning=F}

for(m in 1:num_models) {
  print(paste(m, Sys.Date(), Sys.time()), sep="\t")
  bin_size <- 1e+3 # "basal" scale, larger windows are built after
  L <- models$L[m]
  V <- L - models$num_exons[m] * models$exon_lengths[m]
  mult <- models$mult_r_u[m]
    
  for(rep in 1:num_reps) {
    # constrained elements, call them genes or exons 
    intergenic_lengths <- rgeom(n=models$num_exons[m], 
                                prob=models$num_exons[m]/V) 
    
    while(sum(intergenic_lengths) < V) {
      intergenic_lengths <- c(intergenic_lengths, 
                              rgeom(n=1, prob=models$num_exons[m]/V))
    }
    if(sum(intergenic_lengths) > V) {
      intergenic_lengths <- intergenic_lengths[cumsum(intergenic_lengths) < V]
      intergenic_lengths <- c(intergenic_lengths, V - sum(intergenic_lengths))
    }
    intergenic_lengths <- intergenic_lengths[intergenic_lengths>0]
    
    chr_elements <- c(rbind(intergenic_lengths, 
                        rep(models$exon_lengths[m],
                            models$num_exons[m])))

    # Introducing clusters of exonic and integenic sites (for "realism" ++)
    # this function avoids problems with sample from vector with length == 1
    resample <- function(x) x[sample.int(length(x))][1] 
    p_stay <- models$p_stay[m] # symmetric auto-correlation
    is_exonic <- chr_elements[1] == models$exon_lengths[m]
    
    new_elements <- numeric(length=length(chr_elements)) # init
    new_elements[1] <- chr_elements[1]
    chr_elements <- chr_elements[-1]
    
    for(i in 2:length(new_elements)) {
      if(!is_exonic) {
        if(any(chr_elements != models$exon_lengths[m])) {
          val <- resample(chr_elements[chr_elements!=models$exon_lengths[m]])
          idx <- which(chr_elements==val)[1] # first occurrence of val in elems
          
          new_elements[i] <- val
          chr_elements <- chr_elements[-idx] # removes sampled item
        } else { # we ran out of intergenic regions
          is_exonic <- T
          p_stay <- 1
          
          val <- resample(chr_elements[chr_elements==models$exon_lengths[m]]) 
          idx <- which(chr_elements==val)[1] # first occurrence of val in elems
          new_elements[i] <- val
          chr_elements <- chr_elements[-idx] # removes sampled item
        }
      }
      else {
        if(any(chr_elements == models$exon_lengths[m])) {
          val <- resample(chr_elements[chr_elements==models$exon_lengths[m]])
          idx <- which(chr_elements==val)[1] # first occurrence of val in elems
          
          new_elements[i] <- val
          chr_elements <- chr_elements[-idx] # removes sampled item
        } else { # we ran out of exons
          is_exonic <- F
          p_stay <- 1
          
          val <- resample(chr_elements[chr_elements!=models$exon_lengths[m]])
          idx <- which(chr_elements==val)[1] # first occurrence of val in elems
          
          new_elements[i] <- val
          chr_elements <- chr_elements[-idx] # removes sampled item
        }
      }
    
      if(runif(1, 0, 1) > p_stay) { # switch from/to exonic/intergenic
        is_exonic <- !is_exonic
      }
    }
    
    elements <- setDT(bind_cols(chr="chr1",
    		              dplyr::lag(cumsum(new_elements), n=1, default=0),
    		              dplyr::lag(cumsum(new_elements), n=0, default=0)))
    names(elements) <- c("#chrom", "chromStart", "chromEnd")
    elements <- filter(elements, start < L)
    elements$end[nrow(elements)] <- L
    elements$selected <- as.integer(elements$end - 
elements$start == models$exon_lengths[m])
    
    # recombination map
    rec_spans <- rgeom(n=ceiling(L/models$avg_rec_spans[m]), 
                       prob=1/models$avg_rec_spans[m])
    while(sum(rec_spans) < L) {
      rec_spans <- c(rec_spans, rgeom(n=1, prob=1/models$avg_rec_spans[m]))
    }
    if(sum(rec_spans) > L) {
      rec_spans <- rec_spans[cumsum(rec_spans) < L]
      rec_spans <- c(rec_spans, L - sum(rec_spans))
    }
    rec_spans <- rec_spans[rec_spans>0]
    
    rmap <- suppressMessages(setDT(bind_cols(chr="chr1",
            		  dplyr::lag(cumsum(rec_spans), n=1, default=0),
    		          dplyr::lag(cumsum(rec_spans), n=0, default=0))))  
    names(rmap) <- c("#chrom", "chromStart", "chromEnd")
    rs <- rgamma(n=nrow(rmap),
                 shape=models$shapes_rate_r[m],
                 rate=models$shapes_rate_r[m])
    rmap$r <- rs * models$mean_r[m]
    fwrite(rmap, paste("model_", m, "/rep_", rep, "/rec_map.csv.gz", sep=""))

    # mutation map
    mut_spans <- rgeom(n=ceiling(L/models$avg_mut_spans[m]), 
                       prob=1/models$avg_mut_spans[m])
    while(sum(mut_spans) < L) {
      mut_spans <- c(mut_spans, rgeom(n=1, prob=1/models$avg_mut_spans[m]))
    }
    if(sum(mut_spans) > L) {
      mut_spans <- mut_spans[cumsum(mut_spans) < L]
      mut_spans <- c(mut_spans, L - sum(mut_spans))
    }
    mut_spans <- mut_spans[mut_spans>0]
    
    mmap <- suppressMessages(setDT(bind_cols(chr="chr1",
             		  dplyr::lag(cumsum(mut_spans), n=1, default=0),
    		          dplyr::lag(cumsum(mut_spans), n=0, default=0))))
    names(mmap) <- c("#chrom", "chromStart", "chromEnd")
    mus <- rgamma(n=nrow(mmap),
                	shape=models$shapes_rate_u[m],
                	rate=models$shapes_rate_u[m])
    mmap$u <- mus * models$mean_u[m]
    
    # binning
    bins <- rep(bin_size, floor(L / bin_size))
    bins <- suppressMessages(bind_cols(chr="chr1",
                      dplyr::lag(cumsum(bins), n=1, default=0),
                      dplyr::lag(cumsum(bins), n=0, default=0)))
    names(bins) <- c("#chrom", "chromStart", "chromEnd")
    bins$start <- bins$start + 1 # so that GRanges understands the intervals
    gr_bins <- makeGRangesFromDataFrame(bins)
    
    rmap2 <- rmap
    rmap2$start <- rmap2$start + 1 # so that GRanges understands the intervals
    rec_gr <- makeGRangesFromDataFrame(rmap2, keep.extra.columns=T)
    hits <- findOverlaps(query=gr_bins, subject=rec_gr) 
    hit_list <- as.data.frame(hits)
    names(hit_list) <- c("bins", "rec_bins")
    rmap2$rec_bins <- 1:nrow(rmap2)
    rec_bins <- merge(rmap2, hit_list) %>%
                group_by(bins) %>%
                mutate(avg_rec=weighted.mean(r, end-start))
    rec_bins <- rec_bins %>% distinct(bins, .keep_all=T) 
    
    mmap2 <- mmap 
    mmap2$start <- mmap2$start + 1 # so that GRanges understands the intervals
    mut_gr <- makeGRangesFromDataFrame(mmap2, keep.extra.columns=T)
    
    # transform mut map with rec map
    mmap2$mut_bins <- 1:nrow(mmap2)
    hits <- findOverlaps(query=rec_gr, subject=mut_gr) 
    hit_list <- as.data.frame(hits)
    names(hit_list) <- c("rec_bins", "mut_bins")
    mmap3 <- merge(mmap2, hit_list) 
    mmap3 <- dplyr::select(rmap2, -c("chr", "start", "end")) %>% 
             merge(., mmap3, by="rec_bins")
    mmap3$u <- mmap3$u + mmap3$r * mult
    mmap3 <- mmap3 %>% 
      distinct(mut_bins, .keep_all=T) %>% 
      dplyr::select(names(mmap))
    mmap3$u <- unique(models$mean_u) * (mmap3$u / mean(mmap3$u)) # re-center
    mmap2 <- mmap3
    mmap2$start <- mmap2$start - 1
    fwrite(mmap2, paste("model_", m, "/rep_", rep, "/mut_map.csv.gz", sep=""))
    
    # binning mut
    mut_gr <- makeGRangesFromDataFrame(mmap3, keep.extra.columns=T) # over-write
    hits <- findOverlaps(query=gr_bins, subject=mut_gr) 
    hit_list <- as.data.frame(hits)
    names(hit_list) <- c("bins", "mut_bins")
    mmap3$mut_bins <- 1:nrow(mmap3)
    mut_bins <- merge(mmap3, hit_list, by="mut_bins") %>%
                group_by(bins) %>%
                mutate(avg_mut=weighted.mean(u, end-start))
    mut_bins <- mut_bins %>% distinct(bins, .keep_all=T) 
    
    # avg mut rate within each element
    elements2 <- elements 
    elements2$start <- elements2$start + 1 # so GRanges understands intervals
    elem_gr <- makeGRangesFromDataFrame(elements2, keep.extra.columns=T)
    hits <- findOverlaps(query=elem_gr, subject=mut_gr) 
    hit_list <- as.data.frame(hits)
    names(hit_list) <- c("elems", "mut_bins")
    x <- merge(hit_list, mmap3) 
    y <- x %>% group_by(elems) %>% mutate(avg_mut=weighted.mean(u, end-start)) 
    x <- distinct(y, elems, .keep_all=T)
    elements$u <- x$avg_mut
    
    fwrite(elements, paste("model_", m,
                           "/rep_", rep, 
                           "/elements.csv.gz", sep=""))
    
    # binning exons (number of exonic sites per 1kb window)
    bins$bins_1kb <- 1:nrow(bins)
    elem_gr <- makeGRangesFromDataFrame(dplyr::select(elements, -u), 
                                        keep.extra.columns=T) 
    hits <- findOverlaps(query=elem_gr, subject=gr_bins) 
    hit_list <- as.data.frame(hits)
    names(hit_list) <- c("elem_bins", "bins_1kb")
    elements2$elem_bins <- 1:nrow(elements2)
    elem_bins <- merge(elements2, hit_list, by="elem_bins")
    elem_bins <- merge(dplyr::select(elem_bins, -chr), bins, by="bins_1kb") 
    names(elem_bins) <- c("bins_1kb", "elem_bins", "start_elem", "end_elem",
                          "selected", "chr", "start_1kb", "end_1kb")
    
    # find 1st, last, and middle occurrences of selected elems within 1 kb bins
    elem_bins <- elem_bins %>% mutate(isFirst=as.integer(selected==1 &
(is.na(dplyr::lag(selected)) | dplyr::lag(selected)!=1)))
    elem_bins <- elem_bins %>% mutate(isLast=as.integer(selected==1 &
(is.na(dplyr::lag(selected)) | dplyr::lead(selected)!=1)))
    elem_bins <- elem_bins %>% mutate(isMiddle=as.integer(selected==1 &
(dplyr::lag(selected)==1) & dplyr::lead(selected)==1))
    
    # cleaning corners
    elem_bins$isFirst[is.na(elem_bins$isFirst)] <- 0
    elem_bins$isMiddle[is.na(elem_bins$isMiddle)] <- 0
    elem_bins$isLast[is.na(elem_bins$isLast)] <- 0
    
    # assigns exonic counts to each 1kb bin
    elem_bins$leftCount <- elem_bins$isFirst * 
(elem_bins$end_1kb - elem_bins$start_elem + 1)
    elem_bins$rightCount <- elem_bins$isLast * 
(elem_bins$end_elem - elem_bins$start_1kb + 1)
    elem_bins$midCount <- elem_bins$isMiddle * models$exon_lengths[m] 
    elem_bins$elemCount <- elem_bins$leftCount + 
                           elem_bins$rightCount +
                           elem_bins$midCount
    
    elem_bins_2 <- elem_bins %>%
                   group_by(bins_1kb) %>% 
                   mutate(exonic_sites=sum(elemCount)) %>% 
                   distinct(bins_1kb, .keep_all=T) %>%
                   dplyr::select(chr, start_1kb, end_1kb, exonic_sites)
    
    # corrects the double counting of middle elements assigned to 2 1kb bins
    elem_bins_2[elem_bins_2$exonic_sites==2*models$exon_lengths[m],]$exonic_sites <-
      models$exon_lengths[m]

    maps_1kb <- bind_cols(bins,
                          rec_bins["avg_rec"],
                          mut_bins["avg_mut"],
                          elem_bins_2["exonic_sites"])
    
    maps_1kb$start <- maps_1kb$start - 1 # back
    maps_1kb$bin_10kb <- (maps_1kb$bins_1kb - 1) %/% 10
    maps_1kb$bin_100kb <- (maps_1kb$bins_1kb - 1) %/% 100
    maps_1kb$bin_1Mb <- (maps_1kb$bins_1kb - 1) %/% 1000
    maps_10kb <- maps_1kb %>% group_by(bin_10kb) %>% 
                 summarise_at(c("avg_mut", "avg_rec", "exonic_sites"), mean)
    maps_100kb <- maps_1kb %>% group_by(bin_100kb) %>% 
                  summarise_at(c("avg_mut", "avg_rec", "exonic_sites"), mean)
    maps_1Mb <- maps_1kb %>% group_by(bin_1Mb) %>% 
                summarise_at(c("avg_mut", "avg_rec", "exonic_sites"), mean)
    
    # for number of exonic sites we want their sum in each bin
    maps_10kb$exonic_sites <- maps_10kb$exonic_sites * 10
    maps_100kb$exonic_sites <- maps_100kb$exonic_sites * 100
    maps_1Mb$exonic_sites <- maps_1Mb$exonic_sites * 1000
    
    bins_1kb <- rep(1e+3, floor(L / 1e+3))
    bins_1kb <- suppressMessages(bind_cols(chr="chr1",
      dplyr::lag(cumsum(bins_1kb), n=1, default=0),
      dplyr::lag(cumsum(bins_1kb), n=0, default=0)))
    names(bins_1kb) <- c("chr", "start", "end")
    
    bins_10kb <- rep(1e+4, floor(L / 1e+4))
    bins_10kb <- suppressMessages(bind_cols(chr="chr1",
      dplyr::lag(cumsum(bins_10kb), n=1, default=0),
      dplyr::lag(cumsum(bins_10kb), n=0, default=0)))
    names(bins_10kb) <- c("chr", "start", "end")
    
    bins_100kb <- rep(1e+5, floor(L / 1e+5))
    bins_100kb <- suppressMessages(bind_cols(chr="chr1",
      dplyr::lag(cumsum(bins_100kb), n=1, default=0),
      dplyr::lag(cumsum(bins_100kb), n=0, default=0)))
    names(bins_100kb) <- c("chr", "start", "end")
    
    bins_1Mb <- rep(1e+6, floor(L / 1e+6))
    bins_1Mb <- suppressMessages(bind_cols(chr="chr1",
      dplyr::lag(cumsum(bins_1Mb), n=1, default=0),
      dplyr::lag(cumsum(bins_1Mb), n=0, default=0)))
    names(bins_1Mb) <- c("chr", "start", "end")
    
    maps_1kb <- cbind.data.frame(bins_1kb, dplyr::select(maps_1kb,
                                                         avg_rec, 
                                                         avg_mut,
                                                         exonic_sites))
    maps_10kb <- cbind.data.frame(bins_10kb, dplyr::select(maps_10kb,
                                                           avg_rec, 
                                                           avg_mut,
                                                           exonic_sites))
    maps_100kb <- cbind.data.frame(bins_100kb, dplyr::select(maps_100kb,
                                                             avg_rec, 
                                                             avg_mut,
                                                             exonic_sites))
    maps_1Mb <- cbind.data.frame(bins_1Mb, dplyr::select(maps_1Mb,
                                                         avg_rec,
                                                         avg_mut,
                                                         exonic_sites))
    
    fwrite(maps_1kb, paste("model_", m,
                           "/rep_", rep,
                           "/maps_1kb.csv.gz", sep=""))
    fwrite(maps_10kb, paste("model_", m,
                            "/rep_", rep,
                            "/maps_10kb.csv.gz", sep=""))
    fwrite(maps_100kb, paste("model_", m,
                             "/rep_", rep,
                             "/maps_100kb.csv.gz", sep=""))
    fwrite(maps_1Mb, paste("model_", m,
                           "/rep_", rep,
                           "/maps_1Mb.csv.gz", sep=""))
  }
}
```

NOTE: we used predict_B.py, a script that emplys bgshr functionality to output
a table of predictions in 1 kb windows.

For reference, the HTC condor submission script used (submitted on May 19 2025):
```{bash, chtc-sub, engine.opts='-i', include=F, eval=F}

log = reports/job_$(Cluster)_$(Process).log
error = reports/job_$(Cluster)_$(Process).err
output = reports/job_$(Cluster)_$(Process).out


requirements = (Target.HasCHTCStaging == true)
container_image = file:///staging/valadaresbar/bgshr.sif


executable = ../../predict_B.sh


arguments = \ 
    --cores 32 \
    --block_size 5000 \ 
    --save_corrections \
    -n 6 \
    --shapes \
        0.215 \
    --scales \
        0.023964 \
    --p_neus \
        0.302 \
    -a elements.csv.gz \
    -t lookup_tbl_equilibrium.csv.gz \
    -r rec_map.csv.gz \
    -u mut_map.csv.gz \
    -o tbl_1kb.csv.gz \
    -Ne 10000
    

transfer_input_files = \
    ../../predict_B.py, \
    ../../lookup_tbl_equilibrium.csv.gz, \
    elements.csv.gz, \
    rec_map.csv.gz, \
    mut_map.csv.gz
    

should_transfer_files = YES
when_to_transfer_output = ON_EXIT


request_cpus = 32
request_memory = 32GB
request_disk = 2GB
```

# Specifying causal models

```{r dags-viz, include=F, message=F, warning=F}

# Barroso & Dutheil 2023
DAG0 <- dagify(
  pi ~ t,
  pi ~ u,
  t ~ S,
  t ~ D,
  t ~ r
)

DD0p <- DAG0 %>% 
  tidy_dagitty(layout="time_ordered") %>%
  arrange(name) %>% 
  ggplot(aes(x=x, y=y, xend=xend, yend=yend)) +
  geom_dag_node() +
  geom_dag_edges() +
  theme_dag() + 
  geom_dag_text(parse=TRUE, 
                label = c("D", "r", expression(pi),
                          "S", expression(tau), expression(mu)))
  
save_plot("DAGs/DAG_0.svg", DD0p)
save_plot("DAGs/DAG_0.png", DD0p)

DAG1 <- dagify(
  pi ~ D,
  pi ~ B,
  pi ~ u,
  B ~ S,
  B ~ u,
  B ~ r,
  u ~ r
)

DD1p <- DAG1 %>% 
  tidy_dagitty(layout="auto") %>%
  arrange(name) %>% 
  ggplot(aes(x=x, y=y, xend=xend, yend=yend)) +
  geom_dag_node() +
  geom_dag_edges() +
  theme_dag() + 
  geom_dag_text(parse=TRUE, 
                label = c("B", "D", "S", expression(pi), "r", expression(mu)))
  
save_plot("DAGs/DAG_1.svg", DD1p)
save_plot("DAGs/DAG_1.png", DD1p)

DAG2 <- dagify(
  pi_s ~ B,
  pi_s ~ u,
  B ~ pi_n,
  B ~ r,
  B ~ S,
  u ~ r,
  pi_n ~ u,
  pi_n ~ S
)

DD2p <- DAG2 %>% 
  tidy_dagitty(layout="fr") %>%
  arrange(name) %>% 
  ggplot(aes(x=x, y=y, xend=xend, yend=yend)) +
  geom_dag_node() +
  geom_dag_edges() +
  theme_dag() + 
  geom_dag_text(parse=TRUE, 
                label = c("B", "S", 
                          expression(pi[N]), expression(pi[S]), 
                          "r", expression(mu)))
  
save_plot("DAGs/DAG_2.svg", DD2p)
save_plot("DAGs/DAG_2.png", DD2p)

DAG3 <- dagify(
  pi_s ~ B,
  pi_s ~ u,
  B ~ pi_n,
  B ~ LD,
  LD ~ S,
  LD ~ r,
  pi_n ~ S,
  u ~ r,
  pi_n ~ u
)

DD3p <- DAG3 %>% 
  tidy_dagitty(layout="auto") %>%
  arrange(name) %>% 
  ggplot(aes(x=x, y=y, xend=xend, yend=yend)) +
  geom_dag_node() +
  geom_dag_edges() +
  theme_dag() + 
  geom_dag_text(parse=TRUE, 
                label = c("B", "LD", "S", 
                          expression(pi[N]), 
                          expression(pi[S]), "r", expression(mu)))

save_plot("DAGs/DAG_3.svg", DD3p)
save_plot("DAGs/DAG_3.png", DD3p)

# with interference
DAG4 <- dagify(
  pi_s ~ B,
  pi_s ~ u,
  B ~ pi_n,
  B ~ LD,
  Ne ~ B,
  LD ~ Ne,
  pi_n ~ Ne,
  S ~ Ne,
  pi_s ~ Ne,
  LD ~ S,
  LD ~ r,
  pi_n ~ S,
  u ~ r,
  pi_n ~ u
)

DD4p <- DAG4 %>% 
  tidy_dagitty(layout="auto") %>%
  arrange(name) %>% 
  ggplot(aes(x=x, y=y, xend=xend, yend=yend)) +
  geom_dag_node() +
  geom_dag_edges() +
  theme_dag() + 
  geom_dag_text(parse=TRUE, 
                label = c("B", "LD", "Ne", "S",
                          expression(pi[N]), 
                          expression(pi[S]), "r", expression(mu)))

save_plot("DAGs/DAG_4.svg", DD4p)
save_plot("DAGs/DAG_4.png", DD4p)

DAG5 <- dagify(
  pi_s ~ B,
  pi_s ~ u,
  B ~ pi_n,
  B ~ r,
  pi_n ~ E,
  u ~ r,
  pi_n ~ u,
  pi ~ pi_s,
  pi ~ pi_n
)

# assuming the strength of selection is constant across functional elements
# the problem is that B and piN cannot affect each other directly
DD5p <- DAG5 %>% 
  tidy_dagitty(layout="auto") %>%
  arrange(name) %>% 
  ggplot(aes(x=x, y=y, xend=xend, yend=yend)) +
  geom_dag_node() +
  geom_dag_edges() +
  theme_dag() + 
  geom_dag_text(parse=TRUE, 
                label = c("B", "E", expression(pi),
                          expression(pi[N]), 
                          expression(pi[S]),
                          "r", expression(mu)))

save_plot("DAGs/DAG_5.svg", DD5p)
save_plot("DAGs/DAG_5.png", DD5p)

# the DAGs that we will test with d-sep
DAG6 <- dagify(
  pi_s ~ B,
  pi_s ~ u,
  B ~ E,
  B ~ r,
  B ~ u,
  u ~ r
)

DD6p <- DAG6 %>% 
  tidy_dagitty(layout="auto") %>%
  arrange(name) %>% 
  ggplot(aes(x=x, y=y, xend=xend, yend=yend)) +
  geom_dag_node() +
  geom_dag_edges() +
  theme_dag() + 
  geom_dag_text(parse=TRUE, 
                label = c("B", "E", expression(pi[S]),
                          "r", expression(mu)))

save_plot("DAGs/DAG_6.svg", DD6p)
save_plot("DAGs/DAG_6.png", DD6p)

# non-mutagenic rec
DAG7 <- dagify(
  pi_s ~ B,
  pi_s ~ u,
  B ~ E,
  B ~ r,
  B ~ u
)

DD7p <- DAG7 %>% 
  tidy_dagitty(layout="auto") %>%
  arrange(name) %>% 
  ggplot(aes(x=x, y=y, xend=xend, yend=yend)) +
  geom_dag_node() +
  geom_dag_edges() +
  theme_dag() + 
  geom_dag_text(parse=TRUE, 
                label = c("B", "E", expression(pi[S]),
                          "r", expression(mu)))

save_plot("DAGs/DAG_7.svg", DD7p)
save_plot("DAGs/DAG_7.png", DD7p)
```

The DAGs that we will test with the data via d-sep (based on DD6p and DD7p).
NOTE: '~' operator means 'is a affected by'
Selection is constant across exons.
We set variable names to match column names in the output of the python chunk.

```{r dags-models, include=F, message=F, warning=F}
# neutral pi
## non-mutagenic rec
D1g <- DAG(
  B~avg_rec,
  B~exonic_sites,
  B~avg_mut,
  pi_neu_sim~B,
  pi_neu_sim~avg_mut
)

## mutagenic rec
D2g <- DAG(
  avg_mut~avg_rec,
  B~avg_rec,
  B~exonic_sites,
  B~avg_mut,
  pi_neu_sim~B,
  pi_neu_sim~avg_mut
)

# sim_pi == total pi (neutral + simulated)
## non-mutagenic rec
D3g <- DAG(
  B~avg_rec,
  B~exonic_sites,
  B~avg_mut,
  sim_pi~B,
  sim_pi~avg_mut
)

## mutagenic rec
D4g <- DAG(
  avg_mut~avg_rec,
  B~avg_rec,
  B~exonic_sites,
  B~avg_mut,
  sim_pi~B,
  sim_pi~avg_mut
)

bs1 <- basiSet(D1g)
bs2 <- basiSet(D2g)
bs3 <- basiSet(D3g)
bs4 <- basiSet(D4g)
```

# Simulated data

These simulations are based purely on (naive) moments++ predictions using 
artificially constructed maps.

Therefore, there is no interference in these data.

TODO: adapt script to interference iterations and to introduce mut noise in expectations of pi total, neu and del
eg
S = 100 # diploid sample size
pi_sim <- numeric(length=nrow(human_filtered_10kb))
for(i in 1:S) { 
    pi_sim <- pi_sim + rbinom(size=1e+4, 
                              n=nrow(human_filtered_10kb),
                              prob=human_filtered_10kb$exp_pi) / 1e+4
}

## Setting up

Computing CVs and plotting landscapes

```{r plots-lands, include=F, message=F, warning=F}

cv_u_cv_B <- as.data.frame(matrix(ncol=3, nrow=num_models))
names(cv_u_cv_B) <- c(1e+4, 1e+5, 1e+6)

for(m in 1:num_models) {
  
  print(Sys.time())
  cat(paste("Visualizing data from model", m, "...\n"))
  
  cv_u_cv_B_10kb <- 0
  cv_u_cv_B_100kb <- 0
  cv_u_cv_B_1Mb <- 0
  
  for(rep in 1:num_reps) {
    
    cat(paste("Rep. ", rep, "\n"))
    
    # (unlike in the SEM chunk,) here we keep all cols in the maps
    # because it helps with plotting
    m1kb <- fread(paste("model_", m, "/rep_", rep, "/tbl_1kb.csv.gz", sep=""))
    data_cols <- names(m1kb)[4:ncol(m1kb)]
    
    m1kb$bin_10kb <- (1:nrow(m1kb) - 1) %/% 10
    m1kb$bin_100kb <- (1:nrow(m1kb) - 1) %/% 100
    m1kb$bin_1Mb <- (1:nrow(m1kb) - 1) %/% 1000
    
    m10kb <- m1kb %>%
             group_by(bin_10kb) %>% 
             summarise_at(data_cols, mean) %>%
             dplyr::select(., -bin_10kb)
    m10kb$start <- seq(1, by=1e+4, length.out=nrow(m10kb)) 
    m10kb$end <- seq(1e+4, by=1e+4, length.out=nrow(m10kb)) 
      
    m100kb <- m1kb %>%
              group_by(bin_100kb) %>% 
              summarise_at(data_cols, mean) %>%
              dplyr::select(., -bin_100kb)
    m100kb$start <- seq(1, by=1e+5, length.out=nrow(m100kb)) 
    m100kb$end <- seq(1e+5, by=1e+5, length.out=nrow(m100kb)) 
    
    m1Mb <- m1kb %>%
            group_by(bin_1Mb) %>% 
            summarise_at(data_cols, mean) %>%
            dplyr::select(., -bin_1Mb)
    m1Mb$start <- seq(1, by=1e+6, length.out=nrow(m1Mb)) 
    m1Mb$end <- seq(1e+5, by=1e+6, length.out=nrow(m1Mb)) 
    
    cv_u_cv_B_10kb <- cv_u_cv_B_10kb + sd(m10kb$avg_mut) /
      mean(m10kb$avg_mut) / sd(m10kb$B) / mean(m10kb$B)
    cv_u_cv_B_100kb <- cv_u_cv_B_100kb + sd(m100kb$avg_mut) / 
      mean(m100kb$avg_mut) / sd(m100kb$B) / mean(m100kb$B)
    cv_u_cv_B_1Mb <- cv_u_cv_B_1Mb + sd(m1Mb$avg_mut) /
      mean(m1Mb$avg_mut) / sd(m1Mb$B) / mean(m1Mb$B)

    # plots binned maps 10kb - 1Mb 
    m10kb_m <- pivot_longer(m10kb, 
                            cols=c("sim_pi", "pi_neu_sim"),
                            names_to="diversity")
    pi_10kb <- ggplot(data=m10kb_m, 
                      aes(x=start/1e+6, 
                          y=value, 
                          color=diversity)) +
      geom_point(data=m10kb_m, shape=1) + geom_point() + theme_bw() +
      scale_y_continuous(breaks=pretty_breaks(), 
                         labels=scale.4d) +
      scale_color_discrete(name=NULL,
                           labels=c("pi tot., pi neu."),
                           type=c("purple3", "green4")) +
      labs(title="10 kb", x=NULL, y=expression(pi)) + 
      theme(axis.title.x=element_text(size=16),
            axis.text.x=element_blank(),
            axis.text.y=element_text(size=16),
            plot.title=element_text(size=20),
            legend.position="inside", 
            legend.position.inside=c(0.925, 0.925)) 
    
    u_10kb <- ggplot(data=m10kb, 
                     aes(x=start / 1e+6, 
                         y=avg_mut)) + 
      geom_line(data=m10kb) + geom_point() + theme_bw() +
      scale_y_continuous(breaks=pretty_breaks(), 
                         labels=scientific) +
      labs(title=NULL, x=NULL, y=expression(mu)) +
      theme(axis.title.x=element_text(size=16),
            axis.text.x=element_blank(),
            axis.text.y=element_text(size=16),
            plot.title=element_text(size=20),
            legend.position="none")
    
    B_10kb <- ggplot(data=m10kb, 
                     aes(x=start / 1e+6,
                         y=B)) + 
      geom_line(data=m10kb) + geom_point() + theme_bw() +
      scale_y_continuous(breaks=pretty_breaks(), 
                         labels=scientific) +
      labs(title=NULL, x=NULL, y="B") +
      theme(axis.title.x=element_text(size=16),
            axis.text.x=element_blank(),
            axis.text.y=element_text(size=16),
            plot.title=element_text(size=20),
            legend.position="none")
    
    r_10kb <- ggplot(data=m10kb, 
                     aes(x=start / 1e+6,
                         y=avg_rec)) +
      geom_line(data=m10kb) + geom_point() + theme_bw() +
      scale_y_continuous(breaks=pretty_breaks(),
                         labels=scientific) +
      labs(title=NULL, x="Position (Mb)", y="r") +
      theme(axis.title.x=element_text(size=16),
            axis.text.x=element_text(size=12),
            axis.text.y=element_text(size=16),
            plot.title=element_text(size=20), 
            legend.position="none")
    
    lands_10kb <- plot_grid(pi_10kb, 
                            u_10kb, 
                            B_10kb, 
                            r_10kb,
                            align='v', ncol=1, rel_heights=c(1, 1, 1, 1.25))

    m100kb_m <- pivot_longer(m100kb, 
                             cols=c("sim_pi", "pi_neu_sim"),
                             names_to="diversity")
    pi_100kb <- ggplot(data=m100kb_m, 
                       aes(x=start / 1e+6, 
                           y=value,
                           color=diversity)) +
      geom_line(data=m100kb_m) + geom_point() + theme_bw() +
      scale_y_continuous(breaks=pretty_breaks(), 
                         labels=scale.4d) +
      scale_color_discrete(name=NULL, 
                           labels=c("pi tot., pi neu."),
                           type=c("purple3", "green4")) +
      labs(title="100 kb", x=NULL, y=NULL) +
      theme(axis.title.x=element_text(size=16),
            axis.text.x=element_blank(),
            axis.text.y=element_blank(),
            plot.title=element_text(size=20),
            legend.position="inside", 
            legend.position.inside=c(0.925, 0.925)) 
    
    u_100kb <- ggplot(data=m100kb,
                      aes(x=start / 1e+6,
                          y=avg_mut))+
      geom_line(data=m100kb) + geom_point() + theme_bw() +
      scale_y_continuous(breaks=pretty_breaks(), 
                         labels=scientific) +
      labs(title=NULL, x=NULL, y=NULL) +
      theme(axis.title.x=element_text(size=16),
            axis.text.x=element_blank(),
            axis.text.y=element_blank(),
            plot.title=element_text(size=20),
            legend.position="none")
    
    B_100kb <- ggplot(data=m100kb, 
                      aes(x=start / 1e+6,
                          y=B))+
      geom_line(data=m100kb) + geom_point() + theme_bw() +
      scale_y_continuous(breaks=pretty_breaks(),
                         labels=scientific) +
      labs(title=NULL, x=NULL, y=NULL) +
      theme(axis.title.x=element_text(size=16),
            axis.text.x=element_blank(),
            axis.text.y=element_blank(),
            plot.title=element_text(size=20),
            legend.position="none")
    
    r_100kb <- ggplot(data=m100kb, 
                      aes(x=start / 1e+6,
                          y=avg_rec)) +
      geom_line(data=m100kb) + geom_point() + theme_bw() +
      scale_y_continuous(breaks=pretty_breaks(), 
                         labels=scientific) +
      labs(title=NULL, x="Position (Mb)", y=NULL) +
      theme(axis.title.x=element_text(size=16),
            axis.text.x=element_text(size=12),
            axis.text.y=element_blank(),
            plot.title=element_text(size=20),
            legend.position="none")
    
    lands_100kb <- plot_grid(pi_100kb,
                             u_100kb,
                             B_100kb, 
                             r_100kb, 
                             align='v', ncol=1, rel_heights=c(1, 1, 1, 1.25))
    
    m1Mb_m <- pivot_longer(m1Mb, 
                           cols=c("sim_pi", "pi_neu_sim"),
                           names_to="diversity")
    pi_1Mb <- ggplot(data=m1Mb_m, 
                     aes(x=start/1e+6,
                        y=value, 
                        color=diversity)) +
      geom_line(data=m1Mb_m) + geom_point() + theme_bw() +
      scale_y_continuous(breaks=pretty_breaks(), 
                         labels=scale.4d) +
      scale_color_discrete(name=NULL, 
                           labels=c("pi tot., pi neu."),
                           type=c("purple3", "green4")) +
      labs(title="1 Mb", x=NULL, y=NULL) +
      theme(axis.title.x=element_text(size=16),
            axis.text.x=element_blank(),
            axis.text.y=element_blank(),
            plot.title=element_text(size=20),
            legend.position="inside", 
            legend.position.inside=c(0.925, 0.925)) 
    
    u_1Mb <- ggplot(data=m1Mb, 
                    aes(x=start / 1e+6,
                        y=avg_mut))+
      geom_line(data=m1Mb) + geom_point() + theme_bw() +
      scale_y_continuous(breaks=pretty_breaks(), 
                         labels=scientific) +
      labs(title=NULL, x=NULL, y=NULL) +
      theme(axis.title.x=element_text(size=16),
            axis.text.x=element_blank(),
            axis.text.y=element_blank(),
            plot.title=element_text(size=20),
            legend.position="none")
    
    B_1Mb <- ggplot(data=m1Mb, 
                    aes(x=start / 1e+6, 
                        y=B))+
      geom_line(data=m1Mb) + geom_point() + theme_bw() +
      scale_y_continuous(breaks=pretty_breaks(),
                         labels=scientific) +
      labs(title=NULL, x=NULL, y=NULL) +
      theme(axis.title.x=element_text(size=16),
            axis.text.x=element_blank(),
            axis.text.y=element_blank(),
            plot.title=element_text(size=20),
            legend.position="none")
    
    r_1Mb <- ggplot(data=m1Mb, 
                    aes(x=start / 1e+6,
                        y=avg_rec)) +
      geom_line(data=m1Mb) + geom_point() + theme_bw() +
      scale_y_continuous(breaks=pretty_breaks(), 
                         labels=scientific) +
      labs(title=NULL, x="Position (Mb)", y=NULL) +
      theme(axis.title.x=element_text(size=16),
            axis.text.x=element_text(size=12),
            axis.text.y=element_blank(),
            plot.title=element_text(size=20),
            legend.position="none")
    
    lands_1Mb <- plot_grid(pi_1Mb,
                           u_1Mb,
                           B_1Mb,
                           r_1Mb, 
                           align='v', ncol=1, rel_heights=c(1, 1, 1, 1.25))

    lands_scales <- plot_grid(lands_10kb, lands_100kb, lands_1Mb, nrow=1)
    #save_plot(paste("model_", m, "/rep_", rep, "/maps.png", sep=""),
    #          lands_scales, base_height=12, base_width=24)
  }
  
  cv_u_cv_B[m, 2] <- cv_u_cv_B_10kb / num_reps
  cv_u_cv_B[m, 3] <- cv_u_cv_B_100kb / num_reps
  cv_u_cv_B[m, 4] <- cv_u_cv_B_1Mb / num_reps
}

cv_u_cv_B$model <- 1:nrow(cv_u_cv_B)
cv_models <- merge(cv_u_cv_B, models, by="model")
cv_models <- pivot_longer(cv_models, cols=as.character(c(1e+4, 
                                                         1e+5, 
                                                         1e+6)),
                          names_to="scale", values_to="cv_u_cv_B")
cv_models$scale <- as.numeric(cv_models$scale)

fwrite(cv_models, "cv_models.csv")
```

## D-separation 

First we evaluate the goodness of fit of DAGs 1-4 via d-separation.
Unlike the SEM approach from later chunks, these tests include the number of
exonic sites in the DAG, because they are more flexible (Spearman pcors).

DAG 1: non-mutagenic recombination, pi neutral
```{r d-sep-DAG1, include=F, message=F, warning=F}

basis_set <- bs1
weights <- rep(1, length=length(basis_set))
    
dsep_c_10kb <- as.data.frame(matrix(nrow=num_models, ncol=num_reps))
dsep_c_100kb <- as.data.frame(matrix(nrow=num_models, ncol=num_reps))
dsep_c_1Mb <- as.data.frame(matrix(nrow=num_models, ncol=num_reps))

# NOTE: interpreting Fisher's C test is tricky because of auto-correlation
# within each of the genomic maps. Looking at larger scales is more informative
names(dsep_c_10kb) <- paste("rep_", 1:num_reps, sep="")
names(dsep_c_100kb) <- paste("rep_", 1:num_reps, sep="")
names(dsep_c_1Mb) <- paste("rep_", 1:num_reps, sep="")

pb <- txtProgressBar(min=0, max=num_models, style=3)  
for(m in 1:num_models) {
  setTxtProgressBar(pb, m)
  for(rep in 1:num_reps) {
    
    maps_1kb <- fread(paste("model_", m, "/rep_", rep, 
                            "/tbl_1kb.csv.gz", sep=""))
    data_cols <- names(maps_1kb)[4:ncol(maps_1kb)]
    
    maps_1kb$bin_10kb <- (1:nrow(maps_1kb) - 1) %/% 10
    maps_1kb$bin_100kb <- (1:nrow(maps_1kb) - 1) %/% 100
    maps_1kb$bin_1Mb <- (1:nrow(maps_1kb) - 1) %/% 1000
    
    maps_10kb <- maps_1kb %>%
                 group_by(bin_10kb) %>% 
                 summarise_at(data_cols, mean) %>%
                 dplyr::select(., -bin_10kb) %>% as.data.table()
    
    maps_100kb <- maps_1kb %>%
                  group_by(bin_100kb) %>% 
                  summarise_at(data_cols, mean) %>%
                  dplyr::select(., -bin_100kb) %>% as.data.table()
    
    maps_1Mb <- maps_1kb %>%
                group_by(bin_1Mb) %>% 
                summarise_at(data_cols, mean) %>%
                dplyr::select(., -bin_1Mb) %>% as.data.table()
    
    # this filtering step is to remove auto-correlation at smaller scales
    maps_10kb <- dplyr::filter(maps_10kb, row_number() %% 100 == 0) 
    maps_100kb <- dplyr::filter(maps_10kb, row_number() %% 10 == 0) 
    
    # init as NAs for easy handling of matrix singularity in ppcor() (see below)
    pvals_10kb <- rep(NA, length(basis_set))
    pvals_100kb <- rep(NA, length(basis_set))
    pvals_1Mb <- rep(NA, length(basis_set))
    
    for(i in 1:length(basis_set)) {
      
      var1 <- basis_set[[i]][1]
      var2 <- basis_set[[i]][2]
      
      cond <- NULL
      if(length(basis_set[[i]]) > 2) {
        cond <- basis_set[[i]][3:length(basis_set[[i]])]
      }
      
      stopifnot(dSep(D1g, var1, var2, cond))
      
      # pcor.test requires matrix inversions which can potentially cause errors
      # we let these cases remain NA in the pval vectors
      if(length(basis_set[[i]]) > 2) {
        tryCatch({
          pvals_10kb[i] <- ppcor::pcor.test(maps_10kb[, ..var1], 
                                            maps_10kb[, ..var2], 
                                            maps_10kb[, ..cond],
                                            method="spearman")$p.value
        }, error=function(e){cat("WARNING :",conditionMessage(e), "\n")})
        tryCatch({
          pvals_100kb[i] <- ppcor::pcor.test(maps_100kb[, ..var1], 
                                             maps_100kb[, ..var2], 
                                             maps_100kb[, ..cond],
                                             method="spearman")$p.value
        }, error=function(e){cat("WARNING :",conditionMessage(e), "\n")})
        tryCatch({
          pvals_1Mb[i] <- ppcor::pcor.test(maps_1Mb[, ..var1], 
                                           maps_1Mb[, ..var2], 
                                           maps_1Mb[, ..cond],
                                           method="spearman")$p.value
        }, error=function(e){cat("WARNING :",conditionMessage(e), "\n")})
      } else {
        pvals_10kb[i] <- cor.test(maps_10kb[, ..var1][[1]],
                                  maps_10kb[, ..var2][[1]],
                                  method="spearman")$p.value
        
        pvals_100kb[i] <- cor.test(maps_100kb[, ..var1][[1]],
                                   maps_100kb[, ..var2][[1]],
                                   method="spearman")$p.value
        
        pvals_1Mb[i] <- cor.test(maps_1Mb[, ..var1][[1]],
                                 maps_1Mb[, ..var2][[1]],
                                 method="spearman")$p.value
      }
    }

    dsep_c_10kb[m, rep] <- combine.test(pvals_10kb, weights, "fisher", na.rm=T)
    dsep_c_100kb[m, rep] <- combine.test(pvals_100kb, weights, "fisher", na.rm=T)
    dsep_c_1Mb[m, rep] <- combine.test(pvals_1Mb, weights, "fisher", na.rm=T)
  }
}
close(pb)

pval_dsep <- rbind.data.frame(dsep_c_10kb,
                              dsep_c_100kb,
                              dsep_c_1Mb)
pval_dsep$model <- rep(1:num_models, 3)
pval_dsep$DAG <- 1
pval_dsep$scale <- c(rep(1e+4, num_models),
                     rep(1e+5, num_models),
                     rep(1e+6, num_models))

fwrite(pval_dsep, "pval_dsep1.csv")
```

DAG 2: mutagenic recombination, pi neutral
```{r d-sep-DAG2, include=F, message=F, warning=F}

basis_set <- bs2
weights <- rep(1, length=length(basis_set))
    
dsep_c_10kb <- as.data.frame(matrix(nrow=num_models, ncol=num_reps))
dsep_c_100kb <- as.data.frame(matrix(nrow=num_models, ncol=num_reps))
dsep_c_1Mb <- as.data.frame(matrix(nrow=num_models, ncol=num_reps))

# NOTE: interpreting Fisher's C test is tricky because of auto-correlation
# within each of the genomic maps. Looking at larger scales is more informative
names(dsep_c_10kb) <- paste("rep_", 1:num_reps, sep="")
names(dsep_c_100kb) <- paste("rep_", 1:num_reps, sep="")
names(dsep_c_1Mb) <- paste("rep_", 1:num_reps, sep="")

pb <- txtProgressBar(min=0, max=num_models, style=3)  
for(m in 1:num_models) {
  setTxtProgressBar(pb, m)
  for(rep in 1:num_reps) {
    
    maps_1kb <- fread(paste("model_", m, "/rep_", rep, 
                            "/tbl_1kb.csv.gz", sep=""))
    data_cols <- names(maps_1kb)[4:ncol(maps_1kb)]
    
    maps_1kb$bin_10kb <- (1:nrow(maps_1kb) - 1) %/% 10
    maps_1kb$bin_100kb <- (1:nrow(maps_1kb) - 1) %/% 100
    maps_1kb$bin_1Mb <- (1:nrow(maps_1kb) - 1) %/% 1000
    
    maps_10kb <- maps_1kb %>%
                 group_by(bin_10kb) %>% 
                 summarise_at(data_cols, mean) %>%
                 dplyr::select(., -bin_10kb) %>% as.data.table()
    
    maps_100kb <- maps_1kb %>%
                  group_by(bin_100kb) %>% 
                  summarise_at(data_cols, mean) %>%
                  dplyr::select(., -bin_100kb) %>% as.data.table()
    
    maps_1Mb <- maps_1kb %>%
                group_by(bin_1Mb) %>% 
                summarise_at(data_cols, mean) %>%
                dplyr::select(., -bin_1Mb) %>% as.data.table()
    
    # this filtering step is to remove auto-correlation at smaller scales
    maps_10kb <- dplyr::filter(maps_10kb, row_number() %% 100 == 0) 
    maps_100kb <- dplyr::filter(maps_100kb, row_number() %% 10 == 0) 
    
    pvals_10kb <- numeric(length=length(basis_set))
    pvals_100kb <- numeric(length=length(basis_set))
    pvals_1Mb <- numeric(length=length(basis_set))
    
    for(i in 1:length(basis_set)) {
      
      var1 <- basis_set[[i]][1]
      var2 <- basis_set[[i]][2]
      
      cond <- NULL
      if(length(basis_set[[i]]) > 2) {
        cond <- basis_set[[i]][3:length(basis_set[[i]])]
      }
      
      dSep(D1g, var1, var2, cond)
      
      if(length(basis_set[[i]]) > 2) {
         pvals_10kb[i] <- ppcor::pcor.test(maps_10kb[, ..var1], 
                                           maps_10kb[, ..var2], 
                                           maps_10kb[, ..cond],
                                           method="spearman")$p.value
         
         pvals_100kb[i] <- ppcor::pcor.test(maps_100kb[, ..var1], 
                                            maps_100kb[, ..var2], 
                                            maps_100kb[, ..cond],
                                            method="spearman")$p.value
         
         pvals_1Mb[i] <- ppcor::pcor.test(maps_1Mb[, ..var1], 
                                          maps_1Mb[, ..var2], 
                                          maps_1Mb[, ..cond],
                                          method="spearman")$p.value
         
      } else {
        pvals_10kb[i] <- cor.test(maps_10kb[, ..var1][[1]],
                                  maps_10kb[, ..var2][[1]],
                                  method="spearman")$p.value
        
        pvals_100kb[i] <- cor.test(maps_100kb[, ..var1][[1]],
                                   maps_100kb[, ..var2][[1]],
                                   method="spearman")$p.value
        
        pvals_1Mb[i] <- cor.test(maps_1Mb[, ..var1][[1]],
                                 maps_1Mb[, ..var2][[1]],
                                 method="spearman")$p.value
      }
    }

    dsep_c_10kb[m, rep] <- combine.test(pvals_10kb, weights, "fisher")
    dsep_c_100kb[m, rep] <- combine.test(pvals_100kb, weights, "fisher")
    dsep_c_1Mb[m, rep] <- combine.test(pvals_1Mb, weights, "fisher")
  }
}
close(pb)

pval_dsep <- rbind.data.frame(dsep_c_10kb,
                              dsep_c_100kb,
                              dsep_c_1Mb)
pval_dsep$model <- rep(1:num_models, 3)
pval_dsep$DAG <- 2
pval_dsep$scale <- c(rep(1e+4, num_models),
                     rep(1e+5, num_models),
                     rep(1e+6, num_models))

fwrite(pval_dsep, "pval_dsep2.csv")
```

DAG 3: non-mutagenic recombination, simulated pi (mutational variance)
```{r d-sep-DAG3, include=F, message=F, warning=F}

basis_set <- bs3
weights <- rep(1, length=length(basis_set))
    
dsep_c_10kb <- as.data.frame(matrix(nrow=num_models, ncol=num_reps))
dsep_c_100kb <- as.data.frame(matrix(nrow=num_models, ncol=num_reps))
dsep_c_1Mb <- as.data.frame(matrix(nrow=num_models, ncol=num_reps))

# NOTE: interpreting Fisher's C test is tricky because of auto-correlation
# within each of the genomic maps. Looking at larger scales is more informative
names(dsep_c_10kb) <- paste("rep_", 1:num_reps, sep="")
names(dsep_c_100kb) <- paste("rep_", 1:num_reps, sep="")
names(dsep_c_1Mb) <- paste("rep_", 1:num_reps, sep="")

pb <- txtProgressBar(min=0, max=num_models, style=3)  
for(m in 1:num_models) {
  setTxtProgressBar(pb, m)
  for(rep in 1:num_reps) {
    
    maps_1kb <- fread(paste("model_", m, "/rep_", rep, 
                            "/tbl_1kb.csv.gz", sep=""))
    data_cols <- names(maps_1kb)[4:ncol(maps_1kb)]
    
    maps_1kb$bin_10kb <- (1:nrow(maps_1kb) - 1) %/% 10
    maps_1kb$bin_100kb <- (1:nrow(maps_1kb) - 1) %/% 100
    maps_1kb$bin_1Mb <- (1:nrow(maps_1kb) - 1) %/% 1000
    
    maps_10kb <- maps_1kb %>%
                 group_by(bin_10kb) %>% 
                 summarise_at(data_cols, mean) %>%
                 dplyr::select(., -bin_10kb) %>% as.data.table()
    
    maps_100kb <- maps_1kb %>%
                  group_by(bin_100kb) %>% 
                  summarise_at(data_cols, mean) %>%
                  dplyr::select(., -bin_100kb) %>% as.data.table()
    
    maps_1Mb <- maps_1kb %>%
                group_by(bin_1Mb) %>% 
                summarise_at(data_cols, mean) %>%
                dplyr::select(., -bin_1Mb) %>% as.data.table()
    
    # this filtering step is to remove auto-correlation at smaller scales
    maps_10kb <- dplyr::filter(maps_10kb, row_number() %% 100 == 0) 
    maps_100kb <- dplyr::filter(maps_100kb, row_number() %% 10 == 0) 
    
    pvals_10kb <- numeric(length=length(basis_set))
    pvals_100kb <- numeric(length=length(basis_set))
    pvals_1Mb <- numeric(length=length(basis_set))
    
    for(i in 1:length(basis_set)) {
      
      var1 <- basis_set[[i]][1]
      var2 <- basis_set[[i]][2]
      
      cond <- NULL
      if(length(basis_set[[i]]) > 2) {
        cond <- basis_set[[i]][3:length(basis_set[[i]])]
      }
      
      dSep(D3g, var1, var2, cond)
      
      if(length(basis_set[[i]]) > 2) {
         pvals_10kb[i] <- ppcor::pcor.test(maps_10kb[, ..var1], 
                                           maps_10kb[, ..var2], 
                                           maps_10kb[, ..cond],
                                           method="spearman")$p.value
         
         pvals_100kb[i] <- ppcor::pcor.test(maps_100kb[, ..var1], 
                                            maps_100kb[, ..var2], 
                                            maps_100kb[, ..cond],
                                            method="spearman")$p.value
         
         pvals_1Mb[i] <- ppcor::pcor.test(maps_1Mb[, ..var1], 
                                          maps_1Mb[, ..var2], 
                                          maps_1Mb[, ..cond],
                                          method="spearman")$p.value
         
      } else {
        pvals_10kb[i] <- cor.test(maps_10kb[, ..var1][[1]],
                                  maps_10kb[, ..var2][[1]],
                                  method="spearman")$p.value
        
        pvals_100kb[i] <- cor.test(maps_100kb[, ..var1][[1]],
                                   maps_100kb[, ..var2][[1]],
                                   method="spearman")$p.value
        
        pvals_1Mb[i] <- cor.test(maps_1Mb[, ..var1][[1]],
                                 maps_1Mb[, ..var2][[1]],
                                 method="spearman")$p.value
      }
    }

    dsep_c_10kb[m, rep] <- combine.test(pvals_10kb, weights, "fisher")
    dsep_c_100kb[m, rep] <- combine.test(pvals_100kb, weights, "fisher")
    dsep_c_1Mb[m, rep] <- combine.test(pvals_1Mb, weights, "fisher")
  }
}
close(pb)

pval_dsep <- rbind.data.frame(dsep_c_10kb,
                              dsep_c_100kb,
                              dsep_c_1Mb)
pval_dsep$model <- rep(1:num_models, 3)
pval_dsep$DAG <- 3
pval_dsep$scale <- c(rep(1e+4, num_models),
                     rep(1e+5, num_models),
                     rep(1e+6, num_models))

fwrite(pval_dsep, "pval_dsep3.csv")
```

DAG 4: mutagenic recombination, simulated pi (mutational variance)
```{r d-sep-DAG4, include=F, message=F, warning=F}

basis_set <- bs4
weights <- rep(1, length=length(basis_set))
    
dsep_c_10kb <- as.data.frame(matrix(nrow=num_models, ncol=num_reps))
dsep_c_100kb <- as.data.frame(matrix(nrow=num_models, ncol=num_reps))
dsep_c_1Mb <- as.data.frame(matrix(nrow=num_models, ncol=num_reps))

# NOTE: interpreting Fisher's C test is tricky because of auto-correlation
# within each of the genomic maps. Looking at larger scales is more informative
names(dsep_c_10kb) <- paste("rep_", 1:num_reps, sep="")
names(dsep_c_100kb) <- paste("rep_", 1:num_reps, sep="")
names(dsep_c_1Mb) <- paste("rep_", 1:num_reps, sep="")

pb <- txtProgressBar(min=0, max=num_models, style=3)  
for(m in 1:num_models) {
  setTxtProgressBar(pb, m)
  for(rep in 1:num_reps) {
    
    maps_1kb <- fread(paste("model_", m, "/rep_", rep, 
                            "/tbl_1kb.csv.gz", sep=""))
    data_cols <- names(maps_1kb)[4:ncol(maps_1kb)]
    
    maps_1kb$bin_10kb <- (1:nrow(maps_1kb) - 1) %/% 10
    maps_1kb$bin_100kb <- (1:nrow(maps_1kb) - 1) %/% 100
    maps_1kb$bin_1Mb <- (1:nrow(maps_1kb) - 1) %/% 1000
    
    maps_10kb <- maps_1kb %>%
                 group_by(bin_10kb) %>% 
                 summarise_at(data_cols, mean) %>%
                 dplyr::select(., -bin_10kb) %>% as.data.table()
    
    maps_100kb <- maps_1kb %>%
                  group_by(bin_100kb) %>% 
                  summarise_at(data_cols, mean) %>%
                  dplyr::select(., -bin_100kb) %>% as.data.table()
    
    maps_1Mb <- maps_1kb %>%
                group_by(bin_1Mb) %>% 
                summarise_at(data_cols, mean) %>%
                dplyr::select(., -bin_1Mb) %>% as.data.table()
    
    # this filtering step is to remove auto-correlation at smaller scales
    maps_10kb <- dplyr::filter(maps_10kb, row_number() %% 100 == 0) 
    maps_100kb <- dplyr::filter(maps_100kb, row_number() %% 10 == 0) 
    
    pvals_10kb <- numeric(length=length(basis_set))
    pvals_100kb <- numeric(length=length(basis_set))
    pvals_1Mb <- numeric(length=length(basis_set))
    
    for(i in 1:length(basis_set)) {
      
      var1 <- basis_set[[i]][1]
      var2 <- basis_set[[i]][2]
      
      cond <- NULL
      if(length(basis_set[[i]]) > 2) {
        cond <- basis_set[[i]][3:length(basis_set[[i]])]
      }
      
      dSep(D4g, var1, var2, cond)
      
      if(length(basis_set[[i]]) > 2) {
         pvals_10kb[i] <- ppcor::pcor.test(maps_10kb[, ..var1], 
                                           maps_10kb[, ..var2], 
                                           maps_10kb[, ..cond],
                                           method="spearman")$p.value
         
         pvals_100kb[i] <- ppcor::pcor.test(maps_100kb[, ..var1], 
                                            maps_100kb[, ..var2], 
                                            maps_100kb[, ..cond],
                                            method="spearman")$p.value
         
         pvals_1Mb[i] <- ppcor::pcor.test(maps_1Mb[, ..var1], 
                                          maps_1Mb[, ..var2], 
                                          maps_1Mb[, ..cond],
                                          method="spearman")$p.value
         
      } else {
        pvals_10kb[i] <- cor.test(maps_10kb[, ..var1][[1]],
                                  maps_10kb[, ..var2][[1]],
                                  method="spearman")$p.value
        
        pvals_100kb[i] <- cor.test(maps_100kb[, ..var1][[1]],
                                   maps_100kb[, ..var2][[1]],
                                   method="spearman")$p.value
        
        pvals_1Mb[i] <- cor.test(maps_1Mb[, ..var1][[1]],
                                 maps_1Mb[, ..var2][[1]],
                                 method="spearman")$p.value
      }
    }

    dsep_c_10kb[m, rep] <- combine.test(pvals_10kb, weights, "fisher")
    dsep_c_100kb[m, rep] <- combine.test(pvals_100kb, weights, "fisher")
    dsep_c_1Mb[m, rep] <- combine.test(pvals_1Mb, weights, "fisher")
  }
}
close(pb)

pval_dsep <- rbind.data.frame(dsep_c_10kb,
                              dsep_c_100kb,
                              dsep_c_1Mb)
pval_dsep$model <- rep(1:num_models, 3)
pval_dsep$DAG <- 4
pval_dsep$scale <- c(rep(1e+4, num_models),
                     rep(1e+5, num_models),
                     rep(1e+6, num_models))

fwrite(pval_dsep, "pval_dsep4.csv")
```

Manhattan plot with d-separation tests.

```{r d-sep-plot, include=F, message=F, warning=F}

# combines tables from previous chunks.
pval_dsep1 <- fread("pval_dsep1.csv") # non-mutagenic recomb., neutral pi
pval_dsep2 <- fread("pval_dsep2.csv") # mutagenic recomb., neutral pi
pval_dsep3 <- fread("pval_dsep3.csv") # non-mutagenic recomb., total pi
pval_dsep4 <- fread("pval_dsep4.csv") # mutagenic recomb., total pi

m_dsep1 <- pivot_longer(merge(pval_dsep1, models, by="model"),
                        cols=paste("rep", 1:num_reps, sep="_"),
                        names_to="replicate", values_to="pval")

p1 <- ggplot(data=m_dsep1, 
             aes(x=model, 
                 y=-log(pval), 
                 shape=as.factor(shapes_rate_u), 
                 color=as.factor(mult_r_u))) +
      geom_hline(aes(yintercept=-log10(0.05/num_reps)), 
                 linetype="dashed", color="grey1") +
      geom_point(size=1.5) + theme_bw() + 
      facet_wrap(~scale, ncol=1, labeller=label_both) +
      scale_shape_manual(values=c(0, 1, 2, 3)) +
      scale_color_manual(values=c("brown1", "purple3", "green4", "cyan3")) +
      scale_y_continuous(breaks=pretty_breaks()) +
      scale_x_continuous(breaks=c(0, 432, 864, 1296)) + 
      labs(title="H0: data fits DAG 1 (non-mutagenic rec., Pi neu.)",
           x="Model ID", y="-log10(p-value)",
           shape="Shape=Rate Mut. Gamma", color="Mutagenic Rec.") +
      theme(axis.title=element_text(size=16), 
            axis.text=element_text(size=12), 
            axis.text.x=element_text(size=12),
            legend.text=element_text(size=16),
            legend.title=element_text(size=16),
            legend.position="bottom")
save_plot("Manhattan/manhattan_d-sep_dag1.png", p1, 
          base_height=10, base_width=15)

m_dsep2 <- pivot_longer(merge(pval_dsep2, models, by="model"),
                        cols=paste("rep", 1:num_reps, sep="_"),
                        names_to="replicate", values_to="pval")

p2 <- ggplot(data=m_dsep2, 
             aes(x=model, 
                 y=-log(pval), 
                 shape=as.factor(shapes_rate_u), 
                 color=as.factor(mult_r_u))) +
      geom_hline(aes(yintercept=-log10(0.05/num_reps)), 
                 linetype="dashed", color="grey1") +
      geom_point(size=1.5) + theme_bw() + 
      facet_wrap(~scale, ncol=1, labeller=label_both) +
      scale_shape_manual(values=c(0, 1, 2, 3)) +
      scale_color_manual(values=c("brown1", "purple3", "green4", "cyan3")) +
      scale_y_continuous(breaks=pretty_breaks()) +
      scale_x_continuous(breaks=c(0, 432, 864, 1296)) + 
      labs(title="H0: data fits DAG 2 (mutagenic rec., Pi neu.)", 
           x="Model ID", y="-log10(p-value)",
           shape="Shape=Rate Mut. Gamma", color="Mutagenic Rec.") +
      theme(axis.title=element_text(size=16), 
            axis.text=element_text(size=12), 
            axis.text.x=element_text(size=12),
            legend.text=element_text(size=16),
            legend.title=element_text(size=16),
            legend.position="bottom")
save_plot("Manhattan/manhattan_d-sep_dag2.png", p2, 
          base_height=10, base_width=15)

m_dsep3 <- pivot_longer(merge(pval_dsep3, models, by="model"),
                        cols=paste("rep", 1:num_reps, sep="_"),
                        names_to="replicate", values_to="pval")

p3 <- ggplot(data=m_dsep3, 
             aes(x=model, 
                 y=-log(pval), 
                 shape=as.factor(shapes_rate_u), 
                 color=as.factor(mult_r_u))) +
      geom_hline(aes(yintercept=-log10(0.05/num_reps)), 
                 linetype="dashed", color="grey1") +
      geom_point(size=1.5) + theme_bw() + 
      facet_wrap(~scale, ncol=1, labeller=label_both) +
      scale_shape_manual(values=c(0, 1, 2, 3)) +
      scale_color_manual(values=c("brown1", "purple3", "green4", "cyan3")) +
      scale_y_continuous(breaks=pretty_breaks()) +
      scale_x_continuous(breaks=c(0, 432, 864, 1296)) + 
      labs(title="H0: data fits DAG 3 (non-mutagenic rec., Pi total (neu+del))",
           x="Model ID", y="-log10(p-value)",
           shape="Shape=Rate Mut. Gamma", color="Mutagenic Rec.") +
      theme(axis.title=element_text(size=16), 
            axis.text=element_text(size=12), 
            axis.text.x=element_text(size=12),
            legend.text=element_text(size=16),
            legend.title=element_text(size=16),
            legend.position="bottom")
save_plot("Manhattan/manhattan_d-sep_dag3.png", p3, 
          base_height=10, base_width=15)

m_dsep4 <- pivot_longer(merge(pval_dsep4, models, by="model"),
                        cols=paste("rep", 1:num_reps, sep="_"),
                        names_to="replicate", values_to="pval")

p4 <- ggplot(data=m_dsep4, 
             aes(x=model, 
                 y=-log(pval), 
                 shape=as.factor(shapes_rate_u), 
                 color=as.factor(mult_r_u))) +
      geom_hline(aes(yintercept=-log10(0.05/num_reps)), 
                 linetype="dashed", color="grey1") +
      geom_point(size=1.5) + theme_bw() + 
      facet_wrap(~scale, ncol=1, labeller=label_both) +
      scale_shape_manual(values=c(0, 1, 2, 3)) +
      scale_color_manual(values=c("brown1", "purple3", "green4", "cyan3")) +
      scale_y_continuous(breaks=pretty_breaks()) +
      scale_x_continuous(breaks=c(0, 432, 864, 1296)) + 
      labs(title="H0: data fits DAG 4 (mutagenic rec., Pi total (neu+del))",
           x="Model ID", y="-log10(p-value)",
           shape="Shape=Rate Mut. Gamma", color="Mutagenic Rec.") +
      theme(axis.title=element_text(size=16), 
            axis.text=element_text(size=12), 
            axis.text.x=element_text(size=12),
            legend.text=element_text(size=16),
            legend.title=element_text(size=16),
            legend.position="bottom")
save_plot("Manhattan/manhattan_d-sep_dag4.png", p4,
          base_height=10, base_width=15)
```

## Classic statistical models

Computing simple correlations and fitting linear regression models

```{r lms, include=F, message=F, warning=F}

for(m in 1:num_models) {
  print(paste(m, Sys.Date(), Sys.time()), sep="\t")
  for(rep in 1:num_reps) {
    
    maps_1kb <- fread(paste("model_", m, "/rep_", rep, 
                            "/tbl_1kb.csv.gz", sep=""))
    data_cols <- names(maps_1kb)[4:ncol(maps_1kb)]
    
    maps_1kb$bin_10kb <- (1:nrow(maps_1kb) - 1) %/% 10
    maps_1kb$bin_100kb <- (1:nrow(maps_1kb) - 1) %/% 100
    maps_1kb$bin_1Mb <- (1:nrow(maps_1kb) - 1) %/% 1000
    
    maps_10kb <- maps_1kb %>%
                 group_by(bin_10kb) %>% 
                 summarise_at(data_cols, mean) %>%
                 dplyr::select(., -bin_10kb)
    
    maps_100kb <- maps_1kb %>%
                  group_by(bin_100kb) %>% 
                  summarise_at(data_cols, mean) %>%
                  dplyr::select(., -bin_100kb)
    
    maps_1Mb <- maps_1kb %>%
                group_by(bin_1Mb) %>% 
                summarise_at(data_cols, mean) %>%
                dplyr::select(., -bin_1Mb)
    
    # correlation matrices, filtering to reduce auto-correlation
    mat=Hmisc::rcorr(as.matrix(filter(maps_10kb, row_number() %% 100 == 0)),
                     type="spearman") 
    tbl_10kb <- mat$r
    tbl_10kb[upper.tri(tbl_10kb)] <- mat$P[upper.tri(mat$P)]
    suppressMessages(fwrite(tbl_10kb, 
                            paste("model_", m, "/rep_", rep, 
                                  "/pval_upper_spearman_lower_10kb.csv", 
                                  sep="")))

    mat=Hmisc::rcorr(as.matrix(maps_100kb), type="spearman")
    tbl_100kb <- mat$r
    tbl_100kb[upper.tri(tbl_100kb)] <- mat$P[upper.tri(mat$P)]
    suppressMessages(fwrite(tbl_100kb, 
                            paste("model_", m, "/rep_", rep, 
                                  "/pval_upper_spearman_lower_100kb.csv", 
                                  sep="")))

    mat=Hmisc::rcorr(as.matrix(maps_1Mb), type="spearman")
    tbl_1Mb <- mat$r
    tbl_1Mb[upper.tri(tbl_1Mb)] <- mat$P[upper.tri(mat$P)]
    suppressMessages(fwrite(tbl_1Mb,
                            paste("model_", m, "/rep_", rep, 
                                   "/pval_upper_spearman_lower_1Mb.csv", 
                                   sep="")))
    
    # table for storing total R^2 values straight from the lm function ()
    r2_raw <- as.data.frame(matrix(ncol=1, nrow=3))
    names(r2_raw) <- "r2_typeI"
    r2_raw$scale <- c(1e+4, 1e+5, 1e+6) # rows are bin sizes
    
    # tables for storing correlations 
    corr_tbl <- as.data.frame(matrix(ncol=2, nrow=3))
    names(corr_tbl) <- c("u", "B")
    corr_tbl$scale <- c(1e+4, 1e+5, 1e+6) # rows are bin sizes
    
    # table for storing p-values of tests related to lm assumptions
    tests_tbl <- as.data.frame(matrix(ncol=3, nrow=3))
    names(tests_tbl) <- c("DW_test_p", "HMC_test_p", "Shapiro_test_p")
    tests_tbl$scale <- c(1e+4, 1e+5, 1e+6) # rows are bin sizes
    
    corr_tbl$B[1] <- cor.test(maps_10kb$B, 
                              maps_10kb$sim_pi, 
                              method="pearson")$estimate 
    corr_tbl$B[2] <- cor.test(maps_100kb$B,
                              maps_100kb$sim_pi,
                              method="pearson")$estimate  
    corr_tbl$B[3] <- cor.test(maps_1Mb$B,
                              maps_1Mb$sim_pi, 
                              method="pearson")$estimate 
    
    corr_tbl$u[1] <- cor.test(maps_10kb$avg_mut, 
                              maps_10kb$sim_pi,
                              method="pearson")$estimate 
    corr_tbl$u[2] <- cor.test(maps_100kb$avg_mut,
                              maps_100kb$sim_pi, 
                              method="pearson")$estimate  
    corr_tbl$u[3] <- cor.test(maps_1Mb$avg_mut,
                              maps_1Mb$sim_pi,
                              method="pearson")$estimate 
    
    fwrite(corr_tbl, paste("model_", m, "/rep_", rep, 
                           "/pearson_pi_tbl.csv", sep=""))
    
    # tables for storing estimates from linear models (rows are bin sizes)
    coeff_tbl <- as.data.frame(matrix(ncol=3, nrow=3))
    pval_tbl <- as.data.frame(matrix(ncol=3, nrow=3))
    names(coeff_tbl) <- c("u", "B", "B:u")
    names(pval_tbl) <- c("u", "B", "B:u")
    coeff_tbl$scale <- c(1e+4, 1e+5, 1e+6)
    pval_tbl$scale <- c(1e+4, 1e+5, 1e+6)
    
    # VIFs
    vif_tbl <- as.data.frame(matrix(ncol=2, nrow=3))
    names(vif_tbl) <- c("u", "B")
    vif_tbl$scale <- c(1e+4, 1e+5, 1e+6)
    
    # standardizing variables to help interpretation of linear coefficients
    std_10kb <- as.data.frame(apply(maps_10kb, 2,
                                    function(x) (x-mean(x)) / sd(x)))
    std_10kb$bin <- 1:nrow(std_10kb)
    
    std_100kb <- as.data.frame(apply(maps_100kb, 2,
                                     function(x) (x-mean(x)) / sd(x)))
    std_100kb$bin <- 1:nrow(std_100kb)
    
    std_1Mb <- as.data.frame(apply(maps_1Mb, 2,
                                   function(x) (x-mean(x)) / sd(x)))
    std_1Mb$bin <- 1:nrow(std_1Mb)
    
    # NOTE: GLS to model autocorrelation at lower scales gets too expensive
    
    #################################
    # sim_pi => total pi
    # 10 kb, filtering to remove auto-correlation
    m_10kb <- lm(sim_pi ~ (B + avg_mut) ^ 2,
                 data=filter(std_10kb, row_number() %% 100 == 0))
    summary(m_10kb)
    r2_raw$r2_typeI[1] <- summary(m_10kb)$r.squared * 100
    
    v10kb <- suppressMessages(vif(m_10kb, type='predictor'))
    vif_tbl$B[1] <- v10kb$GVIF[1]
    vif_tbl$u[1] <- v10kb$GVIF[2]
    
    dw_10kb <- dwtest(m_10kb)
    hmc_10kb <- hmctest(m_10kb)
    shapiro_10kb <- shapiro.test(resid(m_10kb)) 
    tests_tbl$DW_test_p[1] <- dw_10kb$p.value
    tests_tbl$HMC_test_p[1] <- hmc_10kb$p.value
    tests_tbl$Shapiro_test_p[1] <- shapiro_10kb$p.value
    
    coeff_tbl$u[1] <- m_10kb$coefficients[3]
    coeff_tbl$B[1] <- m_10kb$coefficients[2]
    coeff_tbl$`B:u`[1] <- m_10kb$coefficients[4]
    
    pval_tbl$u[1] <- summary(m_10kb)$coefficients[3,4]
    pval_tbl$B[1] <- summary(m_10kb)$coefficients[2,4]
    pval_tbl$`B:u`[1] <- summary(m_10kb)$coefficients[4,4]
    
    r2_tbl <- as.data.frame(matrix(ncol=3, nrow=3))
    names(r2_tbl) <- c("u", "B", "B:u")
    r2_tbl$scale <- c(1e+4, 1e+5, 1e+6) 
    
    anova.pi <- Anova(m_10kb)
    apiss <- anova.pi$"Sum Sq"
    anova.pi$VarExp <- apiss / sum(apiss)
    
    r2_tbl$`B:u`[1] <- anova.pi$VarExp[3] * 100
    r2_tbl$u[1] <- anova.pi$VarExp[2] * 100
    r2_tbl$B[1] <- anova.pi$VarExp[1] * 100
    
    # 100 kb
    m_100kb <- lm(sim_pi ~ (B + avg_mut) ^ 2, 
                  data=filter(std_100kb, row_number() %% 10 == 0))
    summary(m_100kb)
    r2_raw$r2_typeI[2] <- summary(m_100kb)$r.squared * 100
    
    v100kb <- suppressMessages(vif(m_100kb, type = 'predictor'))
    vif_tbl$B[2] <- v100kb$GVIF[1]
    vif_tbl$u[2] <- v100kb$GVIF[2]
    
    dw_100kb <- dwtest(m_100kb)
    hmc_100kb <- hmctest(m_100kb)
    shapiro_100kb <- shapiro.test(resid(m_100kb))
    tests_tbl$DW_test_p[2] <- dw_100kb$p.value
    tests_tbl$HMC_test_p[2] <- hmc_100kb$p.value
    tests_tbl$Shapiro_test_p[2] <- shapiro_100kb$p.value
    
    coeff_tbl$u[2] <- m_100kb$coefficients[3]
    coeff_tbl$B[2] <- m_100kb$coefficients[2]
    coeff_tbl$`B:u`[2] <- m_100kb$coefficients[4]
    
    pval_tbl$u[2] <- summary(m_100kb)$coefficients[3,4]
    pval_tbl$B[2] <- summary(m_100kb)$coefficients[2,4]
    pval_tbl$`B:u`[2] <- summary(m_100kb)$coefficients[4,4]
    
    anova.pi <- Anova(m_100kb)
    apiss <- anova.pi$"Sum Sq"
    anova.pi$VarExp <- apiss / sum(apiss)
    
    r2_tbl$`B:u`[2] <- anova.pi$VarExp[3] * 100
    r2_tbl$u[2] <- anova.pi$VarExp[2] * 100
    r2_tbl$B[2] <- anova.pi$VarExp[1] * 100
    
    # 1 Mb
    m_1Mb <- lm(sim_pi ~ (B + avg_mut) ^ 2, data=std_1Mb)
    summary(m_1Mb)
    r2_raw$r2_typeI[3] <- summary(m_1Mb)$r.squared * 100
    
    v1Mb <- suppressMessages(vif(m_1Mb, type = 'predictor'))
    vif_tbl$B[3] <- v1Mb$GVIF[1]
    vif_tbl$u[3] <- v1Mb$GVIF[2]
    
    dw_1Mb <- dwtest(m_1Mb)
    hmc_1Mb <- hmctest(m_1Mb)
    shapiro_1Mb <- shapiro.test(resid(m_1Mb))
    tests_tbl$DW_test_p[3] <- dw_1Mb$p.value
    tests_tbl$HMC_test_p[3] <- hmc_1Mb$p.value
    tests_tbl$Shapiro_test_p[3] <- shapiro_1Mb$p.value
    
    coeff_tbl$u[3] <- m_1Mb$coefficients[3]
    coeff_tbl$B[3] <- m_1Mb$coefficients[2]
    coeff_tbl$`B:u`[3] <- m_1Mb$coefficients[4]
    
    pval_tbl$u[3] <- summary(m_1Mb)$coefficients[3,4]
    pval_tbl$B[3] <- summary(m_1Mb)$coefficients[2,4]
    pval_tbl$`B:u`[3] <- summary(m_1Mb)$coefficients[4,4]
    
    anova.pi <- Anova(m_1Mb)
    apiss <- anova.pi$"Sum Sq"
    anova.pi$VarExp <- apiss / sum(apiss)
    
    r2_tbl$`B:u`[3] <- anova.pi$VarExp[3] * 100
    r2_tbl$u[3] <- anova.pi$VarExp[2] * 100
    r2_tbl$B[3] <- anova.pi$VarExp[1] * 100
    
    fwrite(r2_raw, paste("model_", m, "/rep_", rep, 
                         "/r2_raw_tbl_pi_total.csv", sep=""))
    fwrite(coeff_tbl, paste("model_", m, "/rep_", rep, 
                         "/coeff_tbl_pi_total.csv", sep=""))
    fwrite(pval_tbl, paste("model_", m, "/rep_", rep, 
                         "/pval_tbl_pi_total.csv", sep=""))
    fwrite(r2_tbl, paste("model_", m, "/rep_", rep, 
                         "/r2_typeII_tbl_pi_total.csv", sep=""))
    fwrite(vif_tbl, paste("model_", m, "/rep_", rep, 
                         "/vif_tbl.csv", sep=""))
    fwrite(tests_tbl, paste("model_", m, "/rep_", rep, 
                         "/tests_tbl_pi_total.csv", sep=""))

    # 10 kb
    df_10kb <- cbind.data.frame(resid(m_10kb), fitted(m_10kb))
    names(df_10kb) <- c("residuals", "fitted")
    p_10kb <- ggplot(data=df_10kb, aes(x=residuals, y=fitted)) + 
             geom_point(size=2, shape=1) + 
             theme_bw() + labs(title="10 kb", x=NULL, y=NULL) +
             theme(axis.title.x=element_text(size=16),
                   axis.text.x=element_text(size=12),
                   axis.text.y=element_text(size=12),
                   axis.title.y=element_text(size=16),
                   plot.title=element_text(size=20)) 
    q_10kb <- ggplot(data=df_10kb, aes(x=residuals)) + 
             geom_histogram(aes(y=..density..), bins=30) + 
             stat_function(fun=dnorm, color="red",
                           args=list(mean=mean(df_10kb$residuals), 
                                    sd=sd(df_10kb$residuals))) +
             theme_bw() + labs(title=NULL, x="Residuals", y=NULL) +
             theme(axis.title.x=element_text(size=16),
                   axis.text.x=element_text(size=12),
                   axis.text.y=element_text(size=12),
                   axis.title.y=element_text(size=16),
                   plot.title=element_text(size=20)) 
    
    lm_10kb <- plot_grid(p_10kb, q_10kb, ncol=1)
    
    # 100 kb
    df_100kb <- cbind.data.frame(resid(m_100kb), fitted(m_100kb))
    names(df_100kb) <- c("residuals", "fitted")
    p_100kb <- ggplot(data=df_100kb, aes(x=residuals, y=fitted)) + 
             geom_point(size=2, shape=1) + 
             theme_bw() + labs(title="100 kb", x=NULL, y=NULL) +
             theme(axis.title.x=element_text(size=16),
                   axis.text.x=element_text(size=12),
                   axis.text.y=element_text(size=12),
                   axis.title.y=element_text(size=16),
                   plot.title=element_text(size=20)) 
    q_100kb <- ggplot(data=df_100kb, aes(x=residuals)) + 
             geom_histogram(aes(y=..density..), bins=15) + 
             stat_function(fun=dnorm, color="red",
                           args=list(mean=mean(df_100kb$residuals), 
                                    sd=sd(df_100kb$residuals))) +
             theme_bw() + labs(title=NULL, x="Residuals", y=NULL) +
             theme(axis.title.x=element_text(size=16),
                   axis.text.x=element_text(size=12),
                   axis.text.y=element_text(size=12),
                   axis.title.y=element_text(size=16),
                   plot.title=element_text(size=20)) 
    
    lm_100kb <- plot_grid(p_100kb, q_100kb, ncol=1)
    
    # 1 Mb
    df_1Mb <- cbind.data.frame(resid(m_1Mb), fitted(m_1Mb))
    names(df_1Mb) <- c("residuals", "fitted")
    p_1Mb <- ggplot(data=df_1Mb, aes(x=residuals, y=fitted)) + 
             geom_point(size=2, shape=1) + 
             theme_bw() + labs(title="1 Mb", x=NULL, y=NULL) +
             theme(axis.title.x=element_text(size=16),
                   axis.text.x=element_text(size=12),
                   axis.text.y=element_text(size=12),
                   axis.title.y=element_text(size=16),
                   plot.title=element_text(size=20)) 
    q_1Mb <- ggplot(data=df_1Mb, aes(x=residuals)) + 
             geom_histogram(aes(y=..density..), bins=7) + 
             stat_function(fun=dnorm, color="red",
                           args=list(mean=mean(df_1Mb$residuals), 
                                    sd=sd(df_1Mb$residuals))) +
             theme_bw() + labs(title=NULL, x="Residuals", y=NULL) +
             theme(axis.title.x=element_text(size=16),
                   axis.text.x=element_text(size=12),
                   axis.text.y=element_text(size=12),
                   axis.title.y=element_text(size=16),
                   plot.title=element_text(size=20)) 
    
    lm_1Mb <- plot_grid(p_1Mb, q_1Mb, ncol=1)
    
    lm_plots <- plot_grid(lm_10kb, lm_100kb, lm_1Mb, nrow=1)
    save_plot(paste("model_", m, 
                    "/rep_", rep,
                    "/lm_plots_pi_total.svg", sep=""),
              lm_plots, base_height=10, base_width=20)
    
    ################################# 
    # neutral pi
    # 10 kb, filtering to remove auto-correlation
    m_10kb <- lm(pi_neu_sim ~ (B + avg_mut) ^ 2,
                 data=filter(std_10kb, row_number() %% 100 == 0))
    summary(m_10kb)
    r2_raw$r2_typeI[1] <- summary(m_10kb)$r.squared * 100
    
    v10kb <- suppressMessages(vif(m_10kb, type='predictor'))
    vif_tbl$B[1] <- v10kb$GVIF[1]
    vif_tbl$u[1] <- v10kb$GVIF[2]
    
    dw_10kb <- dwtest(m_10kb)
    hmc_10kb <- hmctest(m_10kb)
    shapiro_10kb <- shapiro.test(resid(m_10kb)) 
    tests_tbl$DW_test_p[1] <- dw_10kb$p.value
    tests_tbl$HMC_test_p[1] <- hmc_10kb$p.value
    tests_tbl$Shapiro_test_p[1] <- shapiro_10kb$p.value
    
    coeff_tbl$u[1] <- m_10kb$coefficients[3]
    coeff_tbl$B[1] <- m_10kb$coefficients[2]
    coeff_tbl$`B:u`[1] <- m_10kb$coefficients[4]
    
    pval_tbl$u[1] <- summary(m_10kb)$coefficients[3, 4]
    pval_tbl$B[1] <- summary(m_10kb)$coefficients[2, 4]
    pval_tbl$`B:u`[1] <- summary(m_10kb)$coefficients[4, 4]
    
    r2_tbl <- as.data.frame(matrix(ncol=3, nrow=3))
    names(r2_tbl) <- c("u", "B", "B:u")
    r2_tbl$scale <- c(1e+4, 1e+5, 1e+6) 
    
    tryCatch({ # to handle residual sum of squares == 0 
      anova.pi <- Anova(m_10kb)
      apiss <- anova.pi$"Sum Sq"
      anova.pi$VarExp <- apiss / sum(apiss)
    
      r2_tbl$`B:u`[1] <- anova.pi$VarExp[3] * 100
      r2_tbl$u[1] <- anova.pi$VarExp[2] * 100
      r2_tbl$B[1] <- anova.pi$VarExp[1] * 100
    }, error=function(e){cat("WARNING :", conditionMessage(e), "\n")})
    
    # 100 kb
    m_100kb <- lm(pi_neu_sim ~ (B + avg_mut) ^ 2, std_100kb)
    summary(m_100kb)
    r2_raw$r2_typeI[2] <- summary(m_100kb)$r.squared * 100
    
    v100kb <- suppressMessages(vif(m_100kb, type = 'predictor'))
    vif_tbl$B[2] <- v100kb$GVIF[1]
    vif_tbl$u[2] <- v100kb$GVIF[2]
    
    dw_100kb <- dwtest(m_100kb)
    hmc_100kb <- hmctest(m_100kb)
    shapiro_100kb <- shapiro.test(resid(m_100kb))
    tests_tbl$DW_test_p[2] <- dw_100kb$p.value
    tests_tbl$HMC_test_p[2] <- hmc_100kb$p.value
    tests_tbl$Shapiro_test_p[2] <- shapiro_100kb$p.value
    
    coeff_tbl$u[2] <- m_100kb$coefficients[3]
    coeff_tbl$B[2] <- m_100kb$coefficients[2]
    coeff_tbl$`B:u`[2] <- m_100kb$coefficients[4]
    
    pval_tbl$u[2] <- summary(m_100kb)$coefficients[3,4]
    pval_tbl$B[2] <- summary(m_100kb)$coefficients[2,4]
    pval_tbl$`B:u`[2] <- summary(m_100kb)$coefficients[4,4]
    
    tryCatch({ # to handle residual sum of squares == 0 
      anova.pi <- Anova(m_100kb)
      apiss <- anova.pi$"Sum Sq"
      anova.pi$VarExp <- apiss / sum(apiss)
    
      r2_tbl$`B:u`[2] <- anova.pi$VarExp[3] * 100
      r2_tbl$u[2] <- anova.pi$VarExp[2] * 100
      r2_tbl$B[2] <- anova.pi$VarExp[1] * 100
    }, error=function(e){cat("WARNING :", conditionMessage(e), "\n")})
    
    # 1 Mb
    m_1Mb <- lm(pi_neu_sim ~ (B + avg_mut) ^ 2, data=std_1Mb)
    summary(m_1Mb)
    r2_raw$r2_typeI[3] <- summary(m_1Mb)$r.squared * 100
    
    v1Mb <- suppressMessages(vif(m_1Mb, type = 'predictor'))
    vif_tbl$B[3] <- v1Mb$GVIF[1]
    vif_tbl$u[3] <- v1Mb$GVIF[2]
    
    dw_1Mb <- dwtest(m_1Mb)
    hmc_1Mb <- hmctest(m_1Mb)
    shapiro_1Mb <- shapiro.test(resid(m_1Mb))
    tests_tbl$DW_test_p[3] <- dw_1Mb$p.value
    tests_tbl$HMC_test_p[3] <- hmc_1Mb$p.value
    tests_tbl$Shapiro_test_p[3] <- shapiro_1Mb$p.value
    
    coeff_tbl$u[3] <- m_1Mb$coefficients[3]
    coeff_tbl$B[3] <- m_1Mb$coefficients[2]
    coeff_tbl$`B:u`[3] <- m_1Mb$coefficients[4]
    
    pval_tbl$u[3] <- summary(m_1Mb)$coefficients[3,4]
    pval_tbl$B[3] <- summary(m_1Mb)$coefficients[2,4]
    pval_tbl$`B:u`[3] <- summary(m_1Mb)$coefficients[4,4]
    
    tryCatch({ # to handle residual sum of squares == 0 
      anova.pi <- Anova(m_1Mb)
      apiss <- anova.pi$"Sum Sq"
      anova.pi$VarExp <- apiss / sum(apiss)
    
      r2_tbl$`B:u`[3] <- anova.pi$VarExp[3] * 100
      r2_tbl$u[3] <- anova.pi$VarExp[2] * 100
      r2_tbl$B[3] <- anova.pi$VarExp[1] * 100
    }, error=function(e){cat("WARNING :", conditionMessage(e), "\n")})
    
    fwrite(r2_raw, paste("model_", m, "/rep_", rep, 
                         "/r2_raw_tbl_pi_neu.csv", sep=""))
    fwrite(coeff_tbl, paste("model_", m, "/rep_", rep, 
                         "/coeff_tbl_pi_neu.csv", sep=""))
    fwrite(pval_tbl, paste("model_", m, "/rep_", rep, 
                         "/pval_tbl_pi_neu.csv", sep=""))
    fwrite(r2_tbl, paste("model_", m, "/rep_", rep, 
                         "/r2_typeII_tbl_pi_neu.csv", sep=""))
    fwrite(vif_tbl, paste("model_", m, "/rep_", rep, 
                         "/vif_tbl.csv", sep=""))
    fwrite(tests_tbl, paste("model_", m, "/rep_", rep, 
                         "/tests_tbl_pi_neu.csv", sep=""))

    # 10 kb
    df_10kb <- cbind.data.frame(resid(m_10kb), fitted(m_10kb))
    names(df_10kb) <- c("residuals", "fitted")
    p_10kb <- ggplot(data=df_10kb, aes(x=residuals, y=fitted)) + 
             geom_point(size=2, shape=1) + 
             theme_bw() + labs(title="10 kb", x=NULL, y=NULL) +
             theme(axis.title.x=element_text(size=16),
                   axis.text.x=element_text(size=12),
                   axis.text.y=element_text(size=12),
                   axis.title.y=element_text(size=16),
                   plot.title=element_text(size=20)) 
    q_10kb <- ggplot(data=df_10kb, aes(x=residuals)) + 
             geom_histogram(aes(y=..density..), bins=30) + 
             stat_function(fun=dnorm, color="red",
                           args=list(mean=mean(df_10kb$residuals), 
                                    sd=sd(df_10kb$residuals))) +
             theme_bw() + labs(title=NULL, x="Residuals", y=NULL) +
             theme(axis.title.x=element_text(size=16),
                   axis.text.x=element_text(size=12),
                   axis.text.y=element_text(size=12),
                   axis.title.y=element_text(size=16),
                   plot.title=element_text(size=20)) 
    
    lm_10kb <- plot_grid(p_10kb, q_10kb, ncol=1)
    
    # 100 kb
    df_100kb <- cbind.data.frame(resid(m_100kb), fitted(m_100kb))
    names(df_100kb) <- c("residuals", "fitted")
    p_100kb <- ggplot(data=df_100kb, aes(x=residuals, y=fitted)) + 
             geom_point(size=2, shape=1) + 
             theme_bw() + labs(title="100 kb", x=NULL, y=NULL) +
             theme(axis.title.x=element_text(size=16),
                   axis.text.x=element_text(size=12),
                   axis.text.y=element_text(size=12),
                   axis.title.y=element_text(size=16),
                   plot.title=element_text(size=20)) 
    q_100kb <- ggplot(data=df_100kb, aes(x=residuals)) + 
             geom_histogram(aes(y=..density..), bins=15) + 
             stat_function(fun=dnorm, color="red",
                           args=list(mean=mean(df_100kb$residuals), 
                                    sd=sd(df_100kb$residuals))) +
             theme_bw() + labs(title=NULL, x="Residuals", y=NULL) +
             theme(axis.title.x=element_text(size=16),
                   axis.text.x=element_text(size=12),
                   axis.text.y=element_text(size=12),
                   axis.title.y=element_text(size=16),
                   plot.title=element_text(size=20)) 
    
    lm_100kb <- plot_grid(p_100kb, q_100kb, ncol=1)
    
    # 1 Mb
    df_1Mb <- cbind.data.frame(resid(m_1Mb), fitted(m_1Mb))
    names(df_1Mb) <- c("residuals", "fitted")
    p_1Mb <- ggplot(data=df_1Mb, aes(x=residuals, y=fitted)) + 
             geom_point(size=2, shape=1) + 
             theme_bw() + labs(title="1 Mb", x=NULL, y=NULL) +
             theme(axis.title.x=element_text(size=16),
                   axis.text.x=element_text(size=12),
                   axis.text.y=element_text(size=12),
                   axis.title.y=element_text(size=16),
                   plot.title=element_text(size=20)) 
    q_1Mb <- ggplot(data=df_1Mb, aes(x=residuals)) + 
             geom_histogram(aes(y=..density..), bins=7) + 
             stat_function(fun=dnorm, color="red",
                           args=list(mean=mean(df_1Mb$residuals), 
                                    sd=sd(df_1Mb$residuals))) +
             theme_bw() + labs(title=NULL, x="Residuals", y=NULL) +
             theme(axis.title.x=element_text(size=16),
                   axis.text.x=element_text(size=12),
                   axis.text.y=element_text(size=12),
                   axis.title.y=element_text(size=16),
                   plot.title=element_text(size=20)) 
    
    lm_1Mb <- plot_grid(p_1Mb, q_1Mb, ncol=1)
    
    lm_plots <- plot_grid(lm_10kb, lm_100kb, lm_1Mb, nrow=1)
    save_plot(paste("model_", m, 
                    "/rep_", rep, 
                    "/lm_plots_pi_neu.svg", sep=""),
              lm_plots, base_height=10, base_width=20)
  }
}
```

Preparing tables with simple correlations

```{r prep-corrs, include=F, message=F, warning=F}

r_vs_u_10kb <- as.data.frame(matrix(nrow=num_models, ncol=num_reps))
r_vs_u_100kb <- as.data.frame(matrix(nrow=num_models, ncol=num_reps))
r_vs_u_1Mb <- as.data.frame(matrix(nrow=num_models, ncol=num_reps))

r_vs_B_10kb <- as.data.frame(matrix(nrow=num_models, ncol=num_reps))
r_vs_B_100kb <- as.data.frame(matrix(nrow=num_models, ncol=num_reps))
r_vs_B_1Mb <- as.data.frame(matrix(nrow=num_models, ncol=num_reps))

r_vs_sim_pi_10kb <- as.data.frame(matrix(nrow=num_models, ncol=num_reps))
r_vs_sim_pi_100kb <- as.data.frame(matrix(nrow=num_models, ncol=num_reps))
r_vs_sim_pi_1Mb <- as.data.frame(matrix(nrow=num_models, ncol=num_reps))

u_vs_B_10kb <- as.data.frame(matrix(nrow=num_models, ncol=num_reps))
u_vs_B_100kb <- as.data.frame(matrix(nrow=num_models, ncol=num_reps))
u_vs_B_1Mb <- as.data.frame(matrix(nrow=num_models, ncol=num_reps))

u_vs_sim_pi_10kb <- as.data.frame(matrix(nrow=num_models, ncol=num_reps))
u_vs_sim_pi_100kb <- as.data.frame(matrix(nrow=num_models, ncol=num_reps))
u_vs_sim_pi_1Mb <- as.data.frame(matrix(nrow=num_models, ncol=num_reps))

B_vs_exonic_sites_10kb <- as.data.frame(matrix(nrow=num_models, ncol=num_reps))
B_vs_exonic_sites_100kb <- as.data.frame(matrix(nrow=num_models, ncol=num_reps))
B_vs_exonic_sites_1Mb <- as.data.frame(matrix(nrow=num_models, ncol=num_reps))

B_vs_sim_pi_10kb <- as.data.frame(matrix(nrow=num_models, ncol=num_reps))
B_vs_sim_pi_100kb <- as.data.frame(matrix(nrow=num_models, ncol=num_reps))
B_vs_sim_pi_1Mb <- as.data.frame(matrix(nrow=num_models, ncol=num_reps))

pb <- txtProgressBar(min=0, max=num_models, style=3)
for(m in 1:num_models) {
  setTxtProgressBar(pb, m)
  for(rep in 1:num_reps) {
    # correlations in lower part of triangular matrices, pvalues in upper
    cors_10kb <- fread(paste("model_", m, "/rep_", rep, 
                            "/pval_upper_spearman_lower_10kb.csv", sep=""))
    
    cors_100kb <- fread(paste("model_", m, "/rep_", rep, 
                            "/pval_upper_spearman_lower_100kb.csv", sep=""))
    
    cors_1Mb <- fread(paste("model_", m, "/rep_", rep, 
                            "/pval_upper_spearman_lower_1Mb.csv", sep=""))
    
    row <- which(names(cors_10kb) == "avg_mut")
    r_vs_u_10kb[m, rep] <- cors_10kb[row, "avg_rec"]
    r_vs_u_100kb[m, rep] <- cors_100kb[row, "avg_rec"]
    r_vs_u_1Mb[m, rep] <- cors_1Mb[row, "avg_rec"]
    
    row <- which(names(cors_10kb) == "B")
    r_vs_B_10kb[m, rep] <- cors_10kb[row, "avg_rec"]
    r_vs_B_100kb[m, rep] <- cors_100kb[row, "avg_rec"]
    r_vs_B_1Mb[m, rep] <- cors_1Mb[row, "avg_rec"]
    
    row <- which(names(cors_10kb) == "sim_pi")
    r_vs_sim_pi_10kb[m, rep] <- cors_10kb[row, "avg_rec"]
    r_vs_sim_pi_100kb[m, rep] <- cors_100kb[row, "avg_rec"]
    r_vs_sim_pi_1Mb[m, rep] <- cors_1Mb[row, "avg_rec"]
    
    row <- which(names(cors_10kb) == "B")
    u_vs_B_10kb[m, rep] <- cors_10kb[row, "avg_mut"]
    u_vs_B_100kb[m, rep] <- cors_100kb[row, "avg_mut"]
    u_vs_B_1Mb[m, rep] <- cors_1Mb[row, "avg_mut"]
    
    row <- which(names(cors_10kb) == "sim_pi")
    u_vs_sim_pi_10kb[m, rep] <- cors_10kb[row, "avg_mut"]
    u_vs_sim_pi_100kb[m, rep] <- cors_100kb[row, "avg_mut"]
    u_vs_sim_pi_1Mb[m, rep] <- cors_1Mb[row, "avg_mut"]
    
    row <- which(names(cors_10kb) == "B")
    B_vs_exonic_sites_10kb[m, rep] <- cors_10kb[row, "exonic_sites"]
    B_vs_exonic_sites_100kb[m, rep] <- cors_100kb[row, "exonic_sites"]
    B_vs_exonic_sites_1Mb[m, rep] <- cors_1Mb[row, "exonic_sites"]
    
    row <- which(names(cors_10kb) == "sim_pi")
    B_vs_sim_pi_10kb[m, rep] <- cors_10kb[row, "B"]
    B_vs_sim_pi_100kb[m, rep] <- cors_100kb[row, "B"]
    B_vs_sim_pi_1Mb[m, rep] <- cors_1Mb[row, "B"]
  }
}
close(pb)

# r vs mu
names(r_vs_u_10kb) <- paste("rep_", 1:num_reps, sep="")
r_vs_u_10kb$model <- 1:num_models
r_vs_u_10kb$scale <- 1e+4

names(r_vs_u_100kb) <- paste("rep_", 1:num_reps, sep="")
r_vs_u_100kb$model <- 1:num_models
r_vs_u_100kb$scale <- 1e+5

names(r_vs_u_1Mb) <- paste("rep_", 1:num_reps, sep="")
r_vs_u_1Mb$model <- 1:num_models
r_vs_u_1Mb$scale <- 1e+6

r_vs_u <- rbind.data.frame(r_vs_u_10kb,
                           r_vs_u_100kb,
                           r_vs_u_1Mb)
fwrite(r_vs_u, "corr_r_u.csv")

# r vs B
names(r_vs_B_10kb) <- paste("rep_", 1:num_reps, sep="")
r_vs_B_10kb$model <- 1:num_models
r_vs_B_10kb$scale <- 1e+4

names(r_vs_B_100kb) <- paste("rep_", 1:num_reps, sep="")
r_vs_B_100kb$model <- 1:num_models
r_vs_B_100kb$scale <- 1e+5

names(r_vs_B_1Mb) <- paste("rep_", 1:num_reps, sep="")
r_vs_B_1Mb$model <- 1:num_models
r_vs_B_1Mb$scale <- 1e+6

r_vs_B <- rbind.data.frame(r_vs_B_10kb,
                           r_vs_B_100kb,
                           r_vs_B_1Mb)
fwrite(r_vs_B, "corr_r_B.csv")

# r vs pi
names(r_vs_sim_pi_10kb) <- paste("rep_", 1:num_reps, sep="")
r_vs_sim_pi_10kb$model <- 1:num_models
r_vs_sim_pi_10kb$scale <- 1e+4

names(r_vs_sim_pi_100kb) <- paste("rep_", 1:num_reps, sep="")
r_vs_sim_pi_100kb$model <- 1:num_models
r_vs_sim_pi_100kb$scale <- 1e+5

names(r_vs_sim_pi_1Mb) <- paste("rep_", 1:num_reps, sep="")
r_vs_sim_pi_1Mb$model <- 1:num_models
r_vs_sim_pi_1Mb$scale <- 1e+6

r_vs_sim_pi <- rbind.data.frame(r_vs_sim_pi_10kb,
                               r_vs_sim_pi_100kb,
                               r_vs_sim_pi_1Mb)
fwrite(r_vs_sim_pi, "corr_r_sim_pi.csv")

# mu x B
names(u_vs_B_10kb) <- paste("rep_", 1:num_reps, sep="")
u_vs_B_10kb$model <- 1:num_models
u_vs_B_10kb$scale <- 1e+4

names(u_vs_B_100kb) <- paste("rep_", 1:num_reps, sep="")
u_vs_B_100kb$model <- 1:num_models
u_vs_B_100kb$scale <- 1e+5

names(u_vs_B_1Mb) <- paste("rep_", 1:num_reps, sep="")
u_vs_B_1Mb$model <- 1:num_models
u_vs_B_1Mb$scale <- 1e+6

u_vs_B <- rbind.data.frame(u_vs_B_10kb,
                           u_vs_B_100kb,
                           u_vs_B_1Mb)
fwrite(u_vs_B, "corr_u_B.csv")

# mu x pi
names(u_vs_sim_pi_10kb) <- paste("rep_", 1:num_reps, sep="")
u_vs_sim_pi_10kb$model <- 1:num_models
u_vs_sim_pi_10kb$scale <- 1e+4

names(u_vs_sim_pi_100kb) <- paste("rep_", 1:num_reps, sep="")
u_vs_sim_pi_100kb$model <- 1:num_models
u_vs_sim_pi_100kb$scale <- 1e+5

names(u_vs_sim_pi_1Mb) <- paste("rep_", 1:num_reps, sep="")
u_vs_sim_pi_1Mb$model <- 1:num_models
u_vs_sim_pi_1Mb$scale <- 1e+6

u_vs_sim_pi <- rbind.data.frame(u_vs_sim_pi_10kb,
                               u_vs_sim_pi_100kb,
                               u_vs_sim_pi_1Mb)
fwrite(u_vs_sim_pi, "corr_u_sim_pi.csv")

# B vs pi
names(B_vs_sim_pi_10kb) <- paste("rep_", 1:num_reps, sep="")
B_vs_sim_pi_10kb$model <- 1:num_models
B_vs_sim_pi_10kb$scale <- 1e+4

names(B_vs_sim_pi_100kb) <- paste("rep_", 1:num_reps, sep="")
B_vs_sim_pi_100kb$model <- 1:num_models
B_vs_sim_pi_100kb$scale <- 1e+5

names(B_vs_sim_pi_1Mb) <- paste("rep_", 1:num_reps, sep="")
B_vs_sim_pi_1Mb$model <- 1:num_models
B_vs_sim_pi_1Mb$scale <- 1e+6

B_vs_sim_pi <- rbind.data.frame(B_vs_sim_pi_10kb,
                               B_vs_sim_pi_100kb,
                               B_vs_sim_pi_1Mb)
fwrite(B_vs_sim_pi, "corr_B_sim_pi.csv")

# B vs E
names(B_vs_exonic_sites_10kb) <- paste("rep_", 1:num_reps, sep="")
B_vs_exonic_sites_10kb$model <- 1:num_models
B_vs_exonic_sites_10kb$scale <- 1e+4

names(B_vs_exonic_sites_100kb) <- paste("rep_", 1:num_reps, sep="")
B_vs_exonic_sites_100kb$model <- 1:num_models
B_vs_exonic_sites_100kb$scale <- 1e+5

names(B_vs_exonic_sites_1Mb) <- paste("rep_", 1:num_reps, sep="")
B_vs_exonic_sites_1Mb$model <- 1:num_models
B_vs_exonic_sites_1Mb$scale <- 1e+6

B_vs_exonic_sites <- rbind.data.frame(B_vs_exonic_sites_10kb,
                                      B_vs_exonic_sites_100kb,
                                      B_vs_exonic_sites_1Mb)
fwrite(B_vs_exonic_sites, "corr_B_E.csv")
```

Preparing tables with p-values

```{r prep-corrs, include=F, message=F, warning=F}

r_vs_u_10kb <- as.data.frame(matrix(nrow=num_models, ncol=num_reps))
r_vs_u_100kb <- as.data.frame(matrix(nrow=num_models, ncol=num_reps))
r_vs_u_1Mb <- as.data.frame(matrix(nrow=num_models, ncol=num_reps))

r_vs_B_10kb <- as.data.frame(matrix(nrow=num_models, ncol=num_reps))
r_vs_B_100kb <- as.data.frame(matrix(nrow=num_models, ncol=num_reps))
r_vs_B_1Mb <- as.data.frame(matrix(nrow=num_models, ncol=num_reps))

r_vs_sim_pi_10kb <- as.data.frame(matrix(nrow=num_models, ncol=num_reps))
r_vs_sim_pi_100kb <- as.data.frame(matrix(nrow=num_models, ncol=num_reps))
r_vs_sim_pi_1Mb <- as.data.frame(matrix(nrow=num_models, ncol=num_reps))

u_vs_B_10kb <- as.data.frame(matrix(nrow=num_models, ncol=num_reps))
u_vs_B_100kb <- as.data.frame(matrix(nrow=num_models, ncol=num_reps))
u_vs_B_1Mb <- as.data.frame(matrix(nrow=num_models, ncol=num_reps))

u_vs_sim_pi_10kb <- as.data.frame(matrix(nrow=num_models, ncol=num_reps))
u_vs_sim_pi_100kb <- as.data.frame(matrix(nrow=num_models, ncol=num_reps))
u_vs_sim_pi_1Mb <- as.data.frame(matrix(nrow=num_models, ncol=num_reps))

B_vs_exonic_sites_10kb <- as.data.frame(matrix(nrow=num_models, ncol=num_reps))
B_vs_exonic_sites_100kb <- as.data.frame(matrix(nrow=num_models, ncol=num_reps))
B_vs_exonic_sites_1Mb <- as.data.frame(matrix(nrow=num_models, ncol=num_reps))

B_vs_sim_pi_10kb <- as.data.frame(matrix(nrow=num_models, ncol=num_reps))
B_vs_sim_pi_100kb <- as.data.frame(matrix(nrow=num_models, ncol=num_reps))
B_vs_sim_pi_1Mb <- as.data.frame(matrix(nrow=num_models, ncol=num_reps))

pb <- txtProgressBar(min=0, max=num_models, style=3)
for(m in 1:num_models) {
  setTxtProgressBar(pb, m)
  for(rep in 1:num_reps) {
    # correlations in lower part of triangular matrices, pvalues in upper
    pval_10kb <- fread(paste("model_", m, "/rep_", rep, 
                            "/pval_upper_spearman_lower_10kb.csv", sep=""))
    
    pval_100kb <- fread(paste("model_", m, "/rep_", rep, 
                            "/pval_upper_spearman_lower_100kb.csv", sep=""))
    
    pval_1Mb <- fread(paste("model_", m, "/rep_", rep, 
                            "/pval_upper_spearman_lower_1Mb.csv", sep=""))
    
    row <- which(names(pval_10kb) == "avg_rec")
    r_vs_u_10kb[m, rep] <- pval_10kb[row, "avg_mut"]
    r_vs_u_100kb[m, rep] <- pval_100kb[row, "avg_mut"]
    r_vs_u_1Mb[m, rep] <- pval_1Mb[row, "avg_mut"]
    
    row <- which(names(pval_10kb) == "avg_rec")
    r_vs_B_10kb[m, rep] <- pval_10kb[row, "B"]
    r_vs_B_100kb[m, rep] <- pval_100kb[row, "B"]
    r_vs_B_1Mb[m, rep] <- pval_1Mb[row, "B"]
    
    row <- which(names(pval_10kb) == "avg_rec")
    r_vs_sim_pi_10kb[m, rep] <- pval_10kb[row, "sim_pi"]
    r_vs_sim_pi_100kb[m, rep] <- pval_100kb[row, "sim_pi"]
    r_vs_sim_pi_1Mb[m, rep] <- pval_1Mb[row, "sim_pi"]
    
    row <- which(names(pval_10kb) == "avg_mut")
    u_vs_B_10kb[m, rep] <- pval_10kb[row, "B"]
    u_vs_B_100kb[m, rep] <- pval_100kb[row, "B"]
    u_vs_B_1Mb[m, rep] <- pval_1Mb[row, "B"]
    
    row <- which(names(pval_10kb) == "avg_mut")
    u_vs_sim_pi_10kb[m, rep] <- pval_10kb[row, "sim_pi"]
    u_vs_sim_pi_100kb[m, rep] <- pval_100kb[row, "sim_pi"]
    u_vs_sim_pi_1Mb[m, rep] <- pval_1Mb[row, "sim_pi"]
    
    row <- which(names(pval_10kb) == "exonic_sites")
    B_vs_exonic_sites_10kb[m, rep] <- pval_10kb[row, "B"]
    B_vs_exonic_sites_100kb[m, rep] <- pval_100kb[row, "B"]
    B_vs_exonic_sites_1Mb[m, rep] <- pval_1Mb[row, "B"]
    
    row <- which(names(pval_10kb) == "B")
    B_vs_sim_pi_10kb[m, rep] <- pval_10kb[row, "sim_pi"]
    B_vs_sim_pi_100kb[m, rep] <- pval_100kb[row, "sim_pi"]
    B_vs_sim_pi_1Mb[m, rep] <- pval_1Mb[row, "sim_pi"]
  }
}
close(pb)

# r vs mu
names(r_vs_u_10kb) <- paste("rep_", 1:num_reps, sep="")
r_vs_u_10kb$model <- 1:num_models
r_vs_u_10kb$scale <- 1e+4

names(r_vs_u_100kb) <- paste("rep_", 1:num_reps, sep="")
r_vs_u_100kb$model <- 1:num_models
r_vs_u_100kb$scale <- 1e+5

names(r_vs_u_1Mb) <- paste("rep_", 1:num_reps, sep="")
r_vs_u_1Mb$model <- 1:num_models
r_vs_u_1Mb$scale <- 1e+6

r_vs_u <- rbind.data.frame(r_vs_u_10kb,
                           r_vs_u_100kb,
                           r_vs_u_1Mb)
fwrite(r_vs_u, "pval_r_u.csv")

# r vs B
names(r_vs_B_10kb) <- paste("rep_", 1:num_reps, sep="")
r_vs_B_10kb$model <- 1:num_models
r_vs_B_10kb$scale <- 1e+4

names(r_vs_B_100kb) <- paste("rep_", 1:num_reps, sep="")
r_vs_B_100kb$model <- 1:num_models
r_vs_B_100kb$scale <- 1e+5

names(r_vs_B_1Mb) <- paste("rep_", 1:num_reps, sep="")
r_vs_B_1Mb$model <- 1:num_models
r_vs_B_1Mb$scale <- 1e+6

r_vs_B <- rbind.data.frame(r_vs_B_10kb,
                           r_vs_B_100kb,
                           r_vs_B_1Mb)
fwrite(r_vs_B, "pval_r_B.csv")

# r vs pi
names(r_vs_sim_pi_10kb) <- paste("rep_", 1:num_reps, sep="")
r_vs_sim_pi_10kb$model <- 1:num_models
r_vs_sim_pi_10kb$scale <- 1e+4

names(r_vs_sim_pi_100kb) <- paste("rep_", 1:num_reps, sep="")
r_vs_sim_pi_100kb$model <- 1:num_models
r_vs_sim_pi_100kb$scale <- 1e+5

names(r_vs_sim_pi_1Mb) <- paste("rep_", 1:num_reps, sep="")
r_vs_sim_pi_1Mb$model <- 1:num_models
r_vs_sim_pi_1Mb$scale <- 1e+6

r_vs_sim_pi <- rbind.data.frame(r_vs_sim_pi_10kb,
                               r_vs_sim_pi_100kb,
                               r_vs_sim_pi_1Mb)
fwrite(r_vs_sim_pi, "pval_r_sim_pi.csv")

# mu x B
names(u_vs_B_10kb) <- paste("rep_", 1:num_reps, sep="")
u_vs_B_10kb$model <- 1:num_models
u_vs_B_10kb$scale <- 1e+4

names(u_vs_B_100kb) <- paste("rep_", 1:num_reps, sep="")
u_vs_B_100kb$model <- 1:num_models
u_vs_B_100kb$scale <- 1e+5

names(u_vs_B_1Mb) <- paste("rep_", 1:num_reps, sep="")
u_vs_B_1Mb$model <- 1:num_models
u_vs_B_1Mb$scale <- 1e+6

u_vs_B <- rbind.data.frame(u_vs_B_10kb,
                           u_vs_B_100kb,
                           u_vs_B_1Mb)
fwrite(u_vs_B, "pval_u_B.csv")

# mu x pi
names(u_vs_sim_pi_10kb) <- paste("rep_", 1:num_reps, sep="")
u_vs_sim_pi_10kb$model <- 1:num_models
u_vs_sim_pi_10kb$scale <- 1e+4

names(u_vs_sim_pi_100kb) <- paste("rep_", 1:num_reps, sep="")
u_vs_sim_pi_100kb$model <- 1:num_models
u_vs_sim_pi_100kb$scale <- 1e+5

names(u_vs_sim_pi_1Mb) <- paste("rep_", 1:num_reps, sep="")
u_vs_sim_pi_1Mb$model <- 1:num_models
u_vs_sim_pi_1Mb$scale <- 1e+6

u_vs_sim_pi <- rbind.data.frame(u_vs_sim_pi_10kb,
                               u_vs_sim_pi_100kb,
                               u_vs_sim_pi_1Mb)
fwrite(u_vs_sim_pi, "pval_u_sim_pi.csv")

# B vs pi
names(B_vs_sim_pi_10kb) <- paste("rep_", 1:num_reps, sep="")
B_vs_sim_pi_10kb$model <- 1:num_models
B_vs_sim_pi_10kb$scale <- 1e+4

names(B_vs_sim_pi_100kb) <- paste("rep_", 1:num_reps, sep="")
B_vs_sim_pi_100kb$model <- 1:num_models
B_vs_sim_pi_100kb$scale <- 1e+5

names(B_vs_sim_pi_1Mb) <- paste("rep_", 1:num_reps, sep="")
B_vs_sim_pi_1Mb$model <- 1:num_models
B_vs_sim_pi_1Mb$scale <- 1e+6

B_vs_sim_pi <- rbind.data.frame(B_vs_sim_pi_10kb,
                               B_vs_sim_pi_100kb,
                               B_vs_sim_pi_1Mb)
fwrite(B_vs_sim_pi, "pval_B_sim_pi.csv")

# B vs E
names(B_vs_exonic_sites_10kb) <- paste("rep_", 1:num_reps, sep="")
B_vs_exonic_sites_10kb$model <- 1:num_models
B_vs_exonic_sites_10kb$scale <- 1e+4

names(B_vs_exonic_sites_100kb) <- paste("rep_", 1:num_reps, sep="")
B_vs_exonic_sites_100kb$model <- 1:num_models
B_vs_exonic_sites_100kb$scale <- 1e+5

names(B_vs_exonic_sites_1Mb) <- paste("rep_", 1:num_reps, sep="")
B_vs_exonic_sites_1Mb$model <- 1:num_models
B_vs_exonic_sites_1Mb$scale <- 1e+6

B_vs_exonic_sites <- rbind.data.frame(B_vs_exonic_sites_10kb,
                                      B_vs_exonic_sites_100kb,
                                      B_vs_exonic_sites_1Mb)
fwrite(B_vs_exonic_sites, "pval_B_E.csv")
```

Plotting the simple correlations

```{r plot-corrs, include=F, message=F, warning=F}

models <- fread("models.csv")

# B vs E
## cors
B_vs_exonic_sites <- fread("corr_B_E.csv")
B_vs_E_c <- merge(B_vs_exonic_sites, 
                  models, 
                  by="model")
m_B_vs_E_c <- pivot_longer(B_vs_E_c, 
                           cols=starts_with("rep"),
                           names_to="replicate",
                           values_to="cor")
## pvals
pvals <- fread("pval_B_E.csv")
B_vs_E_p <- merge(pvals, 
                  models, 
                  by="model")
m_B_vs_E_p <- pivot_longer(B_vs_E_p, 
                           cols=starts_with("rep"),
                           names_to="replicate",
                           values_to="pval")

m_B_vs_E <- left_join(m_B_vs_E_c, m_B_vs_E_p)

p <- ggplot(data=filter(m_B_vs_E, pval < 0.01),
            aes(x=model, 
                y=cor, 
                color=as.factor(shapes_rate_r), 
                shape=as.factor(avg_rec_spans))) +
      geom_point(size=1.5) + theme_bw() + 
      geom_hline(aes(yintercept=0), 
                 linetype="dashed", color="grey1") +
      facet_wrap(~scale, ncol=1, labeller=label_both) +
      scale_shape_manual(values=c(0, 1, 2, 3)) +
      scale_color_manual(values=c("brown1", "purple3", "green4", "cyan3")) +
      scale_y_continuous(breaks=pretty_breaks()) +
      scale_x_continuous(breaks=c(0, 432, 864, 1296)) + 
      labs(title=paste("Simple Spearman correlations between",
                       "B and % Exonic, pval < 1%"),
           x="Model ID", y="Spearman's rho",
           color="Shape Rate Rec.", shape="Avg. Rec. Spans") +
      theme(axis.title=element_text(size=16), 
            axis.text=element_text(size=12), 
            axis.text.x=element_text(size=12),
            legend.text=element_text(size=16),
            legend.title=element_text(size=16),
            legend.position="bottom")
save_plot("Correlations/corr_B_E.png", p, base_height=10, base_width=16)

# B vs pi
## cors
B_vs_pi <- fread("corr_B_sim_pi.csv")
B_vs_pi_c <- merge(B_vs_pi, 
                   models,
                   by="model")
m_B_vs_pi_c <- pivot_longer(B_vs_pi_c, 
                          cols=starts_with("rep"),
                          names_to="replicate",
                          values_to="cor")
## pvals
pvals <- fread("pval_B_sim_pi.csv")
B_vs_pi_p <- merge(pvals, 
                   models, 
                   by="model")
m_B_vs_pi_p <- pivot_longer(B_vs_pi_p, 
                            cols=starts_with("rep"),
                            names_to="replicate",
                            values_to="pval")

m_B_vs_pi <- left_join(m_B_vs_pi_c, m_B_vs_pi_p)

p <- ggplot(data=filter(m_B_vs_pi, pval < 0.01),
            aes(x=model,
                y=cor, 
                color=as.factor(shapes_rate_u), 
                shape=as.factor(avg_rec_spans))) +
      geom_point(size=1.5) + theme_bw() + 
      geom_hline(aes(yintercept=0), 
                 linetype="dashed", color="grey1") +
      facet_wrap(~scale, ncol=1, labeller=label_both) +
      scale_shape_manual(values=c(0, 1, 2, 3)) +
      scale_color_manual(values=c("brown1", "purple3", "green4", "cyan3")) +
      scale_y_continuous(breaks=pretty_breaks()) +
      scale_x_continuous(breaks=c(0, 432, 864, 1296)) + 
      labs(title=paste("Simple Spearman correlations between",
                       "B and pi, pval < 1%"),
           x="Model ID", y="Spearman's rho",
           color="Shape Rate Mut.", shape="Avg. Rec. Span") +
      theme(axis.title=element_text(size=16), 
            axis.text=element_text(size=12), 
            axis.text.x=element_text(size=12),
            legend.text=element_text(size=16),
            legend.title=element_text(size=16),
            legend.position="bottom")
save_plot("Correlations/corr_B_pi.png", p, base_height=10, base_width=16)

# mut vs pi
## cors
u_vs_pi <- fread("corr_u_sim_pi.csv")
u_vs_pi_c <- merge(u_vs_pi,
                   models,
                   by="model")
m_u_vs_pi_c <- pivot_longer(u_vs_pi_c, 
                            cols=starts_with("rep"),
                            names_to="replicate",
                            values_to="cor")

## pvals
pvals <- fread("pval_u_sim_pi.csv")
u_vs_pi_p <- merge(pvals, 
                   models, 
                   by="model")
m_u_vs_pi_p <- pivot_longer(u_vs_pi_p, 
                            cols=starts_with("rep"),
                            names_to="replicate",
                            values_to="pval")

m_u_vs_pi <- left_join(m_u_vs_pi_c, m_u_vs_pi_p)

p <- ggplot(data=filter(m_u_vs_pi, pval < 0.01),
            aes(x=model,
                y=cor, 
                color=as.factor(shapes_rate_u), 
                shape=as.factor(avg_mut_spans))) +
      geom_point(size=1.5) + theme_bw() + 
      geom_hline(aes(yintercept=0), 
                 linetype="dashed", color="grey1") +
      facet_wrap(~scale, ncol=1, labeller=label_both) +
      scale_shape_manual(values=c(0, 1, 2, 3)) +
      scale_color_manual(values=c("brown1", "purple3", "green4", "cyan3")) +
      scale_y_continuous(breaks=pretty_breaks()) +
      scale_x_continuous(breaks=c(0, 432, 864, 1296)) + 
      labs(title=paste("Simple Spearman correlations between",
                       "mutation rate and pi, pval < 1%"),
           x="Model ID", y="Spearman's rho",
           shape="Avg. Mu Span", color="Shape Rate Mu") +
      theme(axis.title=element_text(size=16), 
            axis.text=element_text(size=12), 
            axis.text.x=element_text(size=12),
            legend.text=element_text(size=16),
            legend.title=element_text(size=16),
            legend.position="bottom")
save_plot("Correlations/corr_mu_pi.png", p, base_height=10, base_width=16)

# mu vs B
u_vs_B <- fread("corr_u_B.csv")
u_vs_B_c <- merge(u_vs_B, 
                  models, 
                  by="model")
m_u_vs_B_c <- pivot_longer(u_vs_B_c, 
                           cols=starts_with("rep"),
                           names_to="replicate",
                           values_to="cor")

## pvals
pvals <- fread("pval_u_B.csv")
u_vs_B_p <- merge(pvals, 
                  models, 
                  by="model")
m_u_vs_B_p <- pivot_longer(u_vs_B_p, 
                           cols=starts_with("rep"),
                           names_to="replicate",
                           values_to="pval")

m_u_vs_B <- left_join(m_u_vs_B_c, m_u_vs_B_p)

p <- ggplot(data=filter(m_u_vs_B, pval < 0.01),
            aes(x=model,
                y=cor, 
                color=as.factor(mult_r_u), 
                shape=as.factor(avg_mut_spans))) +
      geom_point(size=1.5) + theme_bw() + 
      geom_hline(aes(yintercept=0), 
                 linetype="dashed", color="grey1") +
      facet_wrap(~scale, ncol=1, labeller=label_both) +
      scale_shape_manual(values=c(0, 1, 2, 3)) +
      scale_color_manual(values=c("brown1", "purple3", "green4", "cyan3")) +
      scale_y_continuous(breaks=pretty_breaks()) +
      scale_x_continuous(breaks=c(0, 432, 864, 1296)) + 
      labs(title=paste("Simple Spearman correlations between",
                       "mu and B, pval < 1%"),
           x="Model ID", y="Spearman's rho",
           shape="Avg. Mu Span", color="Mult. Rec Mu") +
      theme(axis.title=element_text(size=16), 
            axis.text=element_text(size=12), 
            axis.text.x=element_text(size=12),
            legend.text=element_text(size=16),
            legend.title=element_text(size=16),
            legend.position="bottom")
save_plot("Correlations/corr_mu_B.png", p, base_height=10, base_width=16)

# r vs B
## cors
r_vs_B <- fread("corr_r_B.csv")
r_vs_B_c <- merge(r_vs_B,
                 models, 
                 by="model")

m_r_vs_B_c <- pivot_longer(r_vs_B_c, 
                           cols=starts_with("rep"),
                           names_to="replicate",
                           values_to="cor")
## pvals
pvals <- fread("pval_r_B.csv")
r_vs_B_p <- merge(pvals,
                  models, 
                  by="model")

m_r_vs_B_p <- pivot_longer(r_vs_B_p, 
                           cols=starts_with("rep"),
                           names_to="replicate",
                           values_to="pval")

m_r_vs_B <- left_join(m_r_vs_B_c, m_r_vs_B_p)

p <- ggplot(data=filter(m_r_vs_B, pval < 0.01),
            aes(x=model,
                y=cor, 
                color=as.factor(shapes_rate_r), 
                shape=as.factor(avg_rec_spans))) +
      geom_point(size=1.5) + theme_bw() + 
      geom_hline(aes(yintercept=0), 
                 linetype="dashed", color="grey1") +
      facet_wrap(~scale, ncol=1, labeller=label_both) +
      scale_shape_manual(values=c(0, 1, 2, 3)) +
      scale_color_manual(values=c("brown1", "purple3", "green4", "cyan3")) +
      scale_y_continuous(breaks=pretty_breaks()) +
      scale_x_continuous(breaks=c(0, 432, 864, 1296)) + 
      labs(title=paste("Simple Spearman correlations between",
                       "recombination and B, pval < 1%"),
           x="Model ID", y="Spearman's rho",
           shape="Avg. Rec. Span", color="Shapes Rate Rec.") +
      theme(axis.title=element_text(size=16), 
            axis.text=element_text(size=12), 
            axis.text.x=element_text(size=12),
            legend.text=element_text(size=16),
            legend.title=element_text(size=16),
            legend.position="bottom")
save_plot("Correlations/corr_r_B.png", p, base_height=10, base_width=16)

# r vs pi
## cors
r_vs_pi <- fread("corr_r_sim_pi.csv")
r_vs_pi_c <- merge(r_vs_pi, 
                   models, 
                   by="model")
m_r_vs_pi_c <- pivot_longer(r_vs_pi_c, 
                            cols=starts_with("rep"),
                            names_to="replicate",
                            values_to="cor")
## pvals
pvals <- fread("pval_r_sim_pi.csv")
r_vs_pi_p <- merge(pvals, 
                   models, 
                   by="model")
m_r_vs_pi_p <- pivot_longer(r_vs_pi_p, 
                            cols=starts_with("rep"),
                            names_to="replicate",
                            values_to="pval")

m_r_vs_pi <- left_join(m_r_vs_pi_c, m_r_vs_pi_p)

p <- ggplot(data=filter(m_r_vs_pi, pval < 0.01),
            aes(x=model,
                y=cor, 
                color=as.factor(mult_r_u), 
                shape=as.factor(avg_rec_spans))) +
      geom_point(size=1.5) + theme_bw() + 
      geom_hline(aes(yintercept=0), 
                 linetype="dashed", color="grey1") +
      facet_wrap(~scale, ncol=1, labeller=label_both) +
      scale_shape_manual(values=c(0, 1, 2, 3)) +
      scale_color_manual(values=c("brown1", "purple3", "green4", "cyan3")) +
      scale_y_continuous(breaks=pretty_breaks()) +
      scale_x_continuous(breaks=c(0, 432, 864, 1296)) + 
      labs(title=paste("Simple Spearman correlations between",
                       "recombination and pi, pval < 1%"),
           x="Model ID", y="Spearman's rho",
           shape="Avg. Rec Span", color="Mult. Rec Mu") +
      theme(axis.title=element_text(size=16), 
            axis.text=element_text(size=12), 
            axis.text.x=element_text(size=12),
            legend.text=element_text(size=16),
            legend.title=element_text(size=16),
            legend.position="bottom")
save_plot("Correlations/corr_r_pi.png", p, base_height=10, base_width=16)

# r vs mu
## cors
r_vs_u <- fread("corr_r_u.csv")
r_vs_u_c <- merge(r_vs_u, 
                  models, 
                  by="model")
m_r_vs_u_c <- pivot_longer(r_vs_u_c, 
                           cols=starts_with("rep"),
                           names_to="replicate",
                           values_to="cor")

## pvals
pvals <- fread("pval_r_u.csv")
r_vs_u_p <- merge(pvals,
                  models, 
                  by="model")
m_r_vs_u_p <- pivot_longer(r_vs_u_p, 
                           cols=starts_with("rep"),
                           names_to="replicate",
                           values_to="pval")

m_r_vs_u <- left_join(m_r_vs_u_c, m_r_vs_u_p)

p <- ggplot(data=filter(m_r_vs_u, pval < 0.01),
            aes(x=model,
                y=cor, 
                color=as.factor(mult_r_u), 
                shape=as.factor(avg_rec_spans))) +
      geom_point(size=1.5) + theme_bw() + 
      geom_hline(aes(yintercept=0), 
                 linetype="dashed", color="grey1") +
      facet_wrap(~scale, ncol=1, labeller=label_both) +
      scale_shape_manual(values=c(0, 1, 2, 3)) +
      scale_color_manual(values=c("brown1", "purple3", "green4", "cyan3")) +
      scale_y_continuous(breaks=pretty_breaks()) +
      scale_x_continuous(breaks=c(0, 432, 864, 1296)) + 
      labs(title=paste("Simple Spearman correlations between",
                       "recombination and mutation, pval < 1%"),
           x="Model ID", y="Spearman's rho",
           shape="Avg. Rec. Spans", color="Mult. Rec. Mut.") +
      theme(axis.title=element_text(size=16), 
            axis.text=element_text(size=12), 
            axis.text.x=element_text(size=12),
            legend.text=element_text(size=16),
            legend.title=element_text(size=16),
            legend.position="bottom")
save_plot("Correlations/corr_r_mu.png", p, base_height=10, base_width=16)
```

Assembling tables from linear regression (R-sqrd and linear coefficients)

```{r tbl-lm, include=F, message=F, warning=F}

# visualization of R^2 and linear coefficients across models
r2_tbl_neu <- data.table()
r2_tbl_sim <- data.table()
r2_tbl_raw <- data.table()

coeff_tbl_neu <- data.table()
coeff_tbl_sim <- data.table()

pb <- txtProgressBar(min=1, max=num_models, style=3)
for(i in 1:num_models) {
  
  setTxtProgressBar(pb, i)
  
  # converting num_rep tables (3 x 3: 10kb, 100kb, 1Mb x u, B, B:u)
  # into 3 tables of dimension num_reps x 3 (u, B, B:u)
  r2_10kb <- as.data.frame(matrix(nrow=num_reps, ncol=4))
  r2_100kb <- as.data.frame(matrix(nrow=num_reps, ncol=4))
  r2_1Mb <- as.data.frame(matrix(nrow=num_reps, ncol=4))
  
  coeff_10kb <- as.data.frame(matrix(nrow=num_reps, ncol=4))
  coeff_100kb <- as.data.frame(matrix(nrow=num_reps, ncol=4))
  coeff_1Mb <- as.data.frame(matrix(nrow=num_reps, ncol=4))
  
  r2_raw_10kb <- as.data.frame(matrix(nrow=num_reps, ncol=2))
  r2_raw_100kb <- as.data.frame(matrix(nrow=num_reps, ncol=2))
  r2_raw_1Mb <- as.data.frame(matrix(nrow=num_reps, ncol=2))
  
  names(r2_raw_10kb) <- c("pi_tot", "pi_neu")
  names(r2_raw_100kb) <- c("pi_tot", "pi_neu")
  names(r2_raw_1Mb) <- c("pi_tot", "pi_neu")

  # neutral
  for(rep in 1:num_reps) {
    tmp <- fread(paste("model_", i,
                       "/rep_", rep, 
                       "/r2_typeII_tbl_pi_neu.csv", sep=""))
    
    r2_10kb[rep,] <- tmp[1,]
    r2_100kb[rep,] <- tmp[2,]
    r2_1Mb[rep,] <- tmp[3,]
    
    tmp <- fread(paste("model_", i,
                       "/rep_", rep, 
                       "/coeff_tbl_pi_neu.csv", sep=""))
    
    coeff_10kb[rep,] <- tmp[1,]
    coeff_100kb[rep,] <- tmp[2,]
    coeff_1Mb[rep,] <- tmp[3,]
  }
  
  tbl_r2 <- rbind.data.frame(r2_10kb, r2_100kb, r2_1Mb)
  names(tbl_r2) <- c("u", "B", "B:u", "scale")
  tbl_r2$model <- i
  tbl_r2$rep <- rep(1:10, 3)
  
  tbl_coeff <- rbind.data.frame(coeff_10kb, coeff_100kb, coeff_1Mb)
  names(tbl_coeff) <- c("u", "B", "B:u", "scale")
  tbl_coeff$model <- i
  tbl_coeff$rep <- rep(1:10, 3)

  r2_tbl_neu <- rbind.data.frame(r2_tbl_neu, tbl_r2)
  coeff_tbl_neu <- rbind.data.frame(coeff_tbl_neu, tbl_coeff)
  
  # with mutational variance
  for(rep in 1:num_reps) {
    tmp <- fread(paste("model_", i, 
                       "/rep_", rep, 
                       "/r2_typeII_tbl_pi_total.csv", sep=""))
    
    r2_10kb[rep,] <- tmp[1,]
    r2_100kb[rep,] <- tmp[2,]
    r2_1Mb[rep,] <- tmp[3,]
    
    tmp <- fread(paste("model_", i, 
                       "/rep_", rep, 
                       "/coeff_tbl_pi_total.csv", sep=""))
    
    coeff_10kb[rep,] <- tmp[1,]
    coeff_100kb[rep,] <- tmp[2,]
    coeff_1Mb[rep,] <- tmp[3,]
  }
  
  tbl_r2 <- rbind.data.frame(r2_10kb, r2_100kb, r2_1Mb)
  names(tbl_r2) <- c("u", "B", "B:u", "scale")
  tbl_r2$model <- i
  tbl_r2$rep <- rep(1:10, 3)
  
  tbl_coeff <- rbind.data.frame(coeff_10kb, coeff_100kb, coeff_1Mb)
  names(tbl_coeff) <- c("u", "B", "B:u", "scale")
  tbl_coeff$model <- i
  tbl_coeff$rep <- rep(1:10, 3)
  
  coeff_tbl_sim <- rbind.data.frame(coeff_tbl_sim, tbl_coeff)
  r2_tbl_sim <- rbind.data.frame(r2_tbl_sim, tbl_r2)
  
  # raw R^2 (from lm() or type I ANOVA)
  ## sim pi
  for(rep in 1:num_reps) {
    tmp <- fread(paste("model_", i,
                       "/rep_", rep,
                       "/r2_raw_tbl_pi_total.csv", sep=""))
    
    r2_raw_10kb[rep, 1] <- tmp[1, 1]
    r2_raw_100kb[rep, 1] <- tmp[2, 1]
    r2_raw_1Mb[rep, 1] <- tmp[3, 1]
  }
  
  ## pi neutral
  for(rep in 1:num_reps) {
    tmp <- fread(paste("model_", i,
                       "/rep_", rep,
                       "/r2_raw_tbl_pi_neu.csv", sep=""))
    
    r2_raw_10kb[rep, 2] <- tmp[1, 1]
    r2_raw_100kb[rep, 2] <- tmp[2, 1]
    r2_raw_1Mb[rep, 2] <- tmp[3, 1]
  }
  
  r2_raw_10kb$scale <- 1e+4
  r2_raw_100kb$scale <- 1e+5
  r2_raw_1Mb$scale <- 1e+6
  
  tbl_r2_raw <- rbind.data.frame(r2_raw_10kb,
                                 r2_raw_100kb,
                                 r2_raw_1Mb)
  names(tbl_r2_raw) <- c("Pi_Tot", "Pi_Neu", "scale")
  tbl_r2_raw$model <- i
  tbl_r2_raw$rep <- rep(1:10, 3)
  
  r2_tbl_raw <- rbind.data.frame(r2_tbl_raw, tbl_r2_raw)
}
close(pb)

r2s_neu <- r2_tbl_neu %>% group_by(model, scale) %>% 
                          summarize_at(vars(u, B, `B:u`), 
                                       list(avg=mean, sd=sd))

r2s_sim <- r2_tbl_sim %>% group_by(model, scale) %>% 
                          summarize_at(vars(u, B, `B:u`), 
                                       list(avg=mean, sd=sd))

r2s_raw <- r2_tbl_raw %>% group_by(model, scale) %>% 
  summarize_at(vars(Pi_Neu, Pi_Tot), 
               list(avg=mean, sd=sd))

cs_neu <- coeff_tbl_neu %>% group_by(model, scale) %>% 
  summarize_at(vars(u, B, `B:u`), 
               list(avg=mean, sd=sd))

cs_sim <- coeff_tbl_sim %>% group_by(model, scale) %>% 
  summarize_at(vars(u, B, `B:u`), 
               list(avg=mean, sd=sd))

fwrite(r2s_raw, "r2_raw_tbl_models.csv")
fwrite(r2s_neu, "r2_neu_tbl_models.csv")
fwrite(r2s_sim, "r2_sim_tbl_models.csv")

fwrite(cs_neu, "coeff_neu_tbl_models.csv")
fwrite(cs_sim, "coeff_sim_tbl_models.csv")
```

Plotting coefficients and R^2 arising from linear regression
We want to compare them to those obtained with SEM

```{r plot-lm-1, include=F, message=F, warning=F}

r2s_raw <- fread("r2_raw_tbl_models.csv")
r2s_neu <- fread("r2_neu_tbl_models.csv")
r2s_sim <- fread("r2_sim_tbl_models.csv")

cs_neu <- fread("coeff_neu_tbl_models.csv")
cs_sim <- fread("coeff_sim_tbl_models.csv")

# plots raw r2
r2s_raw_f <- dplyr::select(r2s_raw, -starts_with("B:u")) %>%
  dplyr::select(., -ends_with("sd")) 

r2_raw_models <- merge(models, r2s_raw_f, by="model")
m_r2_raw_models <- pivot_longer(r2_raw_models,
                                cols=starts_with("Pi_"), 
                                names_to="var")

for(ars in unique(m_r2_raw_models$avg_rec_spans)) {
  for(srr in unique(m_r2_raw_models$shapes_rate_r)) {
    for(mru in unique(m_r2_raw_models$mult_r_u)) {

      b1 <- ggplot(data=filter(m_r2_raw_models, 
                               mult_r_u==mru,
                               shapes_rate_r==srr, 
                               avg_rec_spans==ars),
                   aes(x=scale/1e+3, 
                       y=value, 
                       shape=var,
                       color=as.factor(num_exons))) +
        geom_line() + geom_point(size=4) + theme_bw() + 
        facet_grid(avg_mut_spans~shapes_rate_u, labeller=label_both) +
        scale_shape_manual(values=c(0, 1), name=NULL, labels=c("Neu", "Tot"))+
        scale_color_discrete(name="num exons", 
                             type=c("purple3", "green4", "brown1")) +
        scale_x_continuous(breaks=unique(m_r2_raw_models$scale/1e+3), 
                           trans="log10") +
        scale_y_continuous(breaks=pretty_breaks()) +
        labs(title=paste("Total variance explained on pi"),
             x="Map Scale (kb)", y="Total Variance Explained (%)") +
        theme(axis.title=element_text(size=16), 
              axis.text=element_text(size=12), 
              axis.text.x=element_text(size=12),
              legend.text=element_text(size=16),
              legend.title=element_text(size=16),
              legend.position="bottom")
      
      fout <- paste("R2/r2_total_ars", ars, 
                    "_srr", srr, 
                    "_mru", mru,
                    ".png", sep="")
      save_plot(fout, b1, base_height=9, base_width=12)
    }
  }
}

# plots of R^2 per variable per model
pb <- txtProgressBar(min=1, max=num_models, style=3)
for(i in 1:num_models) {
  
  setTxtProgressBar(pb, i)
  
  molten_avgs <- pivot_longer(r2s_neu,
                              cols=ends_with("avg"),
                              names_to="variable")
  p <- ggplot(data=filter(molten_avgs, model==i), 
              aes(x=scale/1e+3, y=value, shape=variable)) +
    geom_line() + theme_bw() + geom_point(size=4) +
    scale_shape_manual(values=c(0, 1, 8),
                       name=NULL, labels=c("B", "B:u", "u")) +
    scale_x_continuous(breaks=unique(molten_avgs$scale/1e+3), trans="log10") +
    scale_y_continuous(breaks=pretty_breaks()) +
    labs(title=paste("Model", i, "(Mean over replicates)"),
         x="Map Scale (kb)", y="Variance Explained (%)") +
    theme(axis.title=element_text(size=16), 
          axis.text=element_text(size=12), 
          axis.text.x=element_text(size=12),
          legend.text=element_text(size=16),
          legend.position="bottom")
  
  molten_sds <- pivot_longer(r2s_neu, 
                             cols=ends_with("sd"), 
                             names_to="variable")
  q <- ggplot(data=filter(molten_sds, model==i), 
              aes(x=scale/1e+3, y=value, shape=variable)) +
    geom_line() + theme_bw() + geom_point(size=4) +
    scale_shape_manual(values=c(0, 1, 8),
                       name=NULL, labels=c("B", "B:u", "u")) +
    scale_x_continuous(breaks=unique(molten_sds$scale/1e+3), trans="log10") +
    scale_y_continuous(breaks=pretty_breaks()) +
    labs(title=paste("Model", i, "(SD over replicates)"),
         x="Map Scale (kb)", y=NULL) +
    theme(axis.title=element_text(size=16), 
          axis.text=element_text(size=12), 
          axis.text.x=element_text(size=12),
          legend.text=element_text(size=16), 
          legend.position="bottom")
  
  save_plot(paste("model_", i, "/pi_neu_sim_r2.png", sep=""), plot_grid(p, q, nrow=1), 
            base_height=10, base_width=15)
  
  molten_avgs <- pivot_longer(r2s_sim, 
                              cols=ends_with("avg"), 
                              names_to="variable")
  p <- ggplot(data=filter(molten_avgs, model==i), 
              aes(x=scale/1e+3, y=value, shape=variable)) +
    geom_line() + theme_bw() + geom_point(size=4) +
    scale_shape_manual(values=c(0, 1, 8),
                       name=NULL, labels=c("B", "B:u", "u")) +
    scale_x_continuous(breaks=unique(molten_avgs$scale/1e+3), trans="log10") +
    scale_y_continuous(breaks=pretty_breaks()) +
    labs(title=paste("Model", i, "(Mean over replicates)"),
         x="Map Scale (kb)", y="Variance explained (%)") +
    theme(axis.title=element_text(size=16), 
          axis.text=element_text(size=12), 
          axis.text.x=element_text(size=12),
          legend.text=element_text(size=16),
          legend.position="bottom")
  
  molten_sds <- pivot_longer(r2s_sim, 
                             cols=ends_with("sd"),
                             names_to="variable")
  q <- ggplot(data=filter(molten_sds, model==i), 
              aes(x=scale/1e+3, y=value, shape=variable)) +
    geom_line() + theme_bw() + geom_point(size=4) +
    scale_shape_manual(values=c(0, 1, 8),
                       name=NULL, labels=c("B", "B:u", "u")) +
    scale_x_continuous(breaks=unique(molten_sds$scale/1e+3), trans="log10") +
    scale_y_continuous(breaks=pretty_breaks()) +
    labs(title=paste("Model", i, "(SD over replicates)"),
         x="Map Scale (kb)", y=NULL) +
    theme(axis.title=element_text(size=16), 
          axis.text=element_text(size=12), 
          axis.text.x=element_text(size=12),
          legend.text=element_text(size=16), 
          legend.position="bottom")
  
  save_plot(paste("model_", i, "/pi_total_r2.png", sep=""), 
            plot_grid(p, q, nrow=1), 
            base_height=10, base_width=15)
}
close(pb)

# plots of coeffs per variable per model
pb <- txtProgressBar(min=1, max=num_models, style=3)
for(i in 1:num_models) {
  
  setTxtProgressBar(pb, i)
  
  molten_avgs <- pivot_longer(cs_neu, 
                              cols=ends_with("avg"), 
                              names_to="variable")
  p <- ggplot(data=filter(molten_avgs, model==i), 
              aes(x=scale/1e+3, y=value, shape=variable)) +
    geom_line() + theme_bw() + geom_point(size=4) +
    scale_shape_manual(values=c(0, 1, 8),
                       name=NULL, labels=c("B", "B:u", "u")) +
    scale_x_continuous(breaks=unique(molten_avgs$scale/1e+3), trans="log10") +
    scale_y_continuous(breaks=pretty_breaks()) +
    labs(title=paste("Model", i, "(Mean over replicates)"),
         x="Map Scale (kb)", y="Linear Coefficient") +
    theme(axis.title=element_text(size=16), 
          axis.text=element_text(size=12), 
          axis.text.x=element_text(size=12),
          legend.text=element_text(size=16),
          legend.position="bottom")
  
  molten_sds <- pivot_longer(cs_neu, 
                             cols=ends_with("sd"),
                             names_to="variable")
  q <- ggplot(data=filter(molten_sds, model==i), 
              aes(x=scale/1e+3, y=value, shape=variable)) +
    geom_line() + theme_bw() + geom_point(size=4) +
    scale_shape_manual(values=c(0, 1, 8),
                       name=NULL, labels=c("B", "B:u", "u")) +
    scale_x_continuous(breaks=unique(molten_sds$scale/1e+3), trans="log10") +
    scale_y_continuous(breaks=pretty_breaks()) +
    labs(title=paste("Model", i, "(SD over replicates)"),
         x="Map Scale (kb)", y=NULL) +
    theme(axis.title=element_text(size=16), 
          axis.text=element_text(size=12), 
          axis.text.x=element_text(size=12),
          legend.text=element_text(size=16), 
          legend.position="bottom")
  
  save_plot(paste("model_", i, "/pi_neu_sim_coeff.png", sep=""),
            plot_grid(p, q, nrow=1), 
            base_height=10, base_width=15)
  
  molten_avgs <- pivot_longer(cs_sim, 
                              cols=ends_with("avg"), 
                              names_to="variable")
  p <- ggplot(data=filter(molten_avgs, model==i), 
              aes(x=scale/1e+3, y=value, shape=variable)) +
    geom_line() + theme_bw() + geom_point(size=4) +
    scale_shape_manual(values=c(0, 1, 8),
                       name=NULL, labels=c("B", "B:u", "u")) +
    scale_x_continuous(breaks=unique(molten_avgs$scale/1e+3), trans="log10") +
    scale_y_continuous(breaks=pretty_breaks()) +
    labs(title=paste("Model", i, "(Mean over replicates)"),
         x="Map Scale (kb)", y="Linear Coefficient") +
    theme(axis.title=element_text(size=16), 
          axis.text=element_text(size=12), 
          axis.text.x=element_text(size=12),
          legend.text=element_text(size=16),
          legend.position="bottom")
  
  molten_sds <- pivot_longer(cs_sim, 
                             cols=ends_with("sd"),
                             names_to="variable")
  q <- ggplot(data=filter(molten_sds, model==i), 
              aes(x=scale/1e+3, y=value, shape=variable)) +
    geom_line() + theme_bw() + geom_point(size=4) +
    scale_shape_manual(values=c(0, 1, 8),
                       name=NULL, labels=c("B", "B:u", "u")) +
    scale_x_continuous(breaks=unique(molten_sds$scale/1e+3), trans="log10") +
    scale_y_continuous(breaks=pretty_breaks()) +
    labs(title=paste("Model", i, "(SD over replicates)"),
         x="Map Scale (kb)", y=NULL) +
    theme(axis.title=element_text(size=16), 
          axis.text=element_text(size=12), 
          axis.text.x=element_text(size=12),
          legend.text=element_text(size=16), 
          legend.position="bottom")
  
  save_plot(paste("model_", i, "/pi_total_coeff.png", sep=""),
            plot_grid(p, q, nrow=1), 
            base_height=10, base_width=15)
}
close(pb)
```

Combining data across models to facilitate visualization

```{r lm-tbls, include=F, message=F, warning=F}

r2s_raw <- fread("r2_raw_tbl_models.csv")
r2s_neu <- fread("r2_neu_tbl_models.csv")
r2s_sim <- fread("r2_sim_tbl_models.csv")

cs_neu <- fread("coeff_neu_tbl_models.csv")
cs_sim <- fread("coeff_sim_tbl_models.csv")

# pi neutral
## R2
r2s_neu_f <- dplyr::select(r2s_neu, -starts_with("B:u")) %>%
             dplyr::select(., -ends_with("sd"))

r2_neu_models <- merge(models, r2s_neu_f, by="model")
r2_neu_models$source_pi <- "neu"
m_r2_neu_models <- pivot_longer(r2_neu_models, 
                                cols=ends_with("avg"),
                                names_to="var")

## coeffs
coeff_neu_f <- dplyr::select(cs_neu, -starts_with("B:u")) %>%
               dplyr::select(., -ends_with("sd")) %>%
               mutate(., ratio = u_avg / B_avg)

coeff_neu_models <- merge(models, coeff_neu_f, by="model")
coeff_neu_models$source_pi <- "neu"
m_coeff_neu_models <- pivot_longer(coeff_neu_models, 
                                   cols=c("ratio", ends_with("avg")),
                                   names_to="var")

# simulated pi
## R2
r2s_sim_f <- dplyr::select(r2s_sim, -starts_with("B:u")) %>%
             dplyr::select(., -ends_with("sd")) 

r2_sim_models <- merge(models, r2s_sim_f, by="model")
r2_sim_models$source_pi <- "sim"
m_r2_sim_models <- pivot_longer(r2_sim_models,
                                cols=ends_with("avg"),
                                names_to="var")

## coeffs
coeff_sim_f <- dplyr::select(cs_sim, -starts_with("B:u")) %>%
               dplyr::select(., -ends_with("sd")) %>%
               mutate(., ratio=u_avg / B_avg)

coeff_sim_models <- merge(models, coeff_sim_f, by="model")
coeff_sim_models$source_pi <- "sim"
m_coeff_sim_models <- pivot_longer(coeff_sim_models,
                                   cols=c("ratio", ends_with("avg")),
                                   names_to="var")

m_coeff_models <- rbind.data.frame(m_coeff_neu_models, m_coeff_sim_models)
m_r2_models <- rbind.data.frame(m_r2_neu_models, m_r2_sim_models)

fwrite(m_coeff_models, "coeff_models.csv.gz")
fwrite(m_r2_models, "r2_models.csv.gz")
```

Plotting variance explained 

```{r plots-lm-r2, include=F, message=F, warning=F}

m_r2_models <- fread("r2_models.csv.gz")
cv_models <- fread("cv_models.csv")

for(e in unique(m_r2_models$num_exons)) {
  for(ars in unique(m_r2_models$avg_rec_spans)) {
    for(srr in unique(m_r2_models$shapes_rate_r)) {
      for(mru in unique(m_r2_models$mult_r_u)) {
        p <- ggplot(data=filter(m_r2_models, 
                                num_exons==e, 
                                avg_rec_spans==ars, 
                                shapes_rate_r==srr, 
                                mult_r_u==mru),
           aes(x=scale/1e+3, y=value, color=var, shape=source_pi)) +
           geom_line() + geom_point(size=4) + theme_bw() + 
           facet_grid(avg_mut_spans~shapes_rate_u, labeller=label_both) +
           scale_shape_manual(values=c(0, 1, 2), name="pi")+
           scale_color_discrete(name=NULL, labels=c("B", "u"), 
                                type=c("purple3", "green4")) +
           scale_x_continuous(breaks=unique(m_r2_models$scale/1e+3), 
                              trans="log10") +
           scale_y_continuous(breaks=pretty_breaks()) +
           labs(title="Variance explained per variable in linear regression",
                x="Map Scale (kb)", y="Variance Explained (%)") +
           theme(axis.title=element_text(size=16), 
                 axis.text=element_text(size=12), 
                 axis.text.x=element_text(size=12),
                 legend.text=element_text(size=16),
                 legend.title=element_text(size=16),
                 legend.position="bottom")
        
          fout <- paste("R2/r2_vars_lm_exons", e, 
                        "_ars", ars, 
                        "_srr", srr, 
                        "_mru", mru,
                        ".png", sep="")
          save_plot(fout, p, base_height=10, base_width=18)
      }
    }
  }
}

r2_cv_models <- left_join(m_r2_models, cv_models )

for(e in unique(m_r2_models$num_exons)) {
  for(ars in unique(m_r2_models$avg_rec_spans)) {
    for(srr in unique(m_r2_models$shapes_rate_r)) {
      for(mru in unique(m_r2_models$mult_r_u)) {
        p <- ggplot(data=filter(r2_cv_models,
                                num_exons==e, 
                                avg_rec_spans==ars, 
                                shapes_rate_r==srr, 
                                mult_r_u==mru),
                     aes(x=cv_u_cv_B, 
                         y=value, 
                         shape=var,
                         color=source_pi)) +
          geom_line() + geom_point(size=4) + theme_bw() + 
          facet_grid(avg_mut_spans~scale, labeller=label_both) +
          scale_shape_manual(values=c(0, 1, 2), name=NULL, labels=c("B", "u"))+
          scale_color_discrete(name="source pi", type=c("purple3", "green4")) +
          scale_x_continuous(breaks=c(1, 10, 100), trans="log10") +
          scale_y_continuous(breaks=pretty_breaks()) +
          labs(title="Variance explained per variable in linear regression",
               x="CV(u) / CV(B)", y="Variance Explained (%)") +
          theme(axis.title=element_text(size=16), 
                axis.text=element_text(size=12), 
                axis.text.x=element_text(size=12),
                legend.text=element_text(size=16),
                legend.title=element_text(size=16),
                legend.position="bottom")
        
        fout <- paste("R2/r2_cvs_lm_exons", e, 
                      "_ars", ars, 
                      "_srr", srr, 
                      "_mru", mru,
                      ".png", sep="")
        save_plot(fout, p, base_height=10, base_width=18)
      }
    }
  }
}
```

Plotting (standardized) linear coefficients

```{r plots-lm-coeffs, include=F, message=F, warning=F}

m_coeff_models <- fread("coeff_models.csv.gz")
cv_models <- fread("cv_models.csv")

for(e in unique(m_coeff_models$num_exons)) {
  for(ars in unique(m_coeff_models$avg_rec_spans)) {
    for(srr in unique(m_coeff_models$shapes_rate_r)) {
      for(mru in unique(m_coeff_models$mult_r_u)) {
        p <- ggplot(data=filter(m_coeff_models, 
                                var!="ratio",
                                num_exons==e, 
                                avg_rec_spans==ars, 
                                shapes_rate_r==srr, 
                                mult_r_u==mru),
           aes(x=scale/1e+3, y=value, color=var, shape=source_pi)) +
           geom_line() + geom_point(size=4) + theme_bw() + 
           facet_grid(avg_mut_spans~shapes_rate_u, labeller=label_both) +
           scale_shape_manual(values=c(0, 1, 2), name="pi")+
           scale_color_discrete(name=NULL, labels=c("B", "u"), 
                              type=c("purple3", "green4")) +
           scale_x_continuous(breaks=unique(m_r2_models$scale/1e+3), 
                              trans="log10") +
           scale_y_continuous(breaks=pretty_breaks()) +
           labs(title="Standardized linear coefficients",
                x="Map Scale (kb)", y="Variance Explained (%)") +
           theme(axis.title=element_text(size=16), 
                 axis.text=element_text(size=12), 
                 axis.text.x=element_text(size=12),
                 legend.text=element_text(size=16),
                 legend.title=element_text(size=16),
                 legend.position="bottom")
        
          fout <- paste("Coeff_lm/coeff_vars_lm_exons", e, 
                        "_ars", ars, 
                        "_srr", srr, 
                        "_mru", mru,
                        ".png", sep="")
          save_plot(fout, p, base_height=10, base_width=18)
      }
    }
  }
}

coeff_cv_models <- left_join(m_coeff_models, cv_models)

for(e in unique(coeff_cv_models$num_exons)) {
  for(ars in unique(coeff_cv_models$avg_rec_spans)) {
    for(srr in unique(coeff_cv_models$shapes_rate_r)) {
      for(mru in unique(coeff_cv_models$mult_r_u)) {
        p <- ggplot(data=filter(coeff_cv_models,
                                var!="ratio",
                                num_exons==e, 
                                avg_rec_spans==ars, 
                                shapes_rate_r==srr, 
                                mult_r_u==mru),
                     aes(x=cv_u_cv_B, 
                         y=value, 
                         color=var,
                         shape=source_pi)) +
          geom_line() + geom_point(size=4) + theme_bw() + 
          facet_grid(avg_mut_spans~scale, labeller=label_both) +
          scale_shape_manual(values=c(0, 1, 2), name="pi")+
          scale_color_discrete(name=NULL, labels=c("B", "u"), 
                               type=c("purple3", "green4")) +
          scale_x_continuous(breaks=c(1, 10, 100), trans="log10") +
          scale_y_continuous(breaks=pretty_breaks()) +
          labs(title="Standardized linear coefficients",
               x="CV(u) / CV(B)", y="Standardized linear coefficients") +
          theme(axis.title=element_text(size=16), 
                axis.text=element_text(size=12), 
                axis.text.x=element_text(size=12),
                legend.text=element_text(size=16),
                legend.title=element_text(size=16),
                legend.position="bottom")
        
        fout <- paste("Coeff_lm/coeff_cvs_lm_exons", e, 
                      "_ars", ars, 
                      "_srr", srr, 
                      "_mru", mru,
                      ".png", sep="")
        save_plot(fout, p, base_height=10, base_width=18)
      }
    }
  }
}

# plotting ratios of std. linear coefficients
p <- ggplot(data=filter(coeff_cv_models,
                        var=="ratio",
                        source_pi=="neu"),
                     aes(x=cv_u_cv_B, 
                         y=value, 
                         color=as.factor(100*num_exons*exon_lengths/L),
                         shape=as.factor(scale))) +
          geom_line() + geom_point(size=4) + theme_bw() + 
          geom_hline(yintercept=1, linetype="dashed", color="grey1") +
          geom_vline(xintercept=1, linetype="dashed", color="grey1") +
          facet_grid(shapes_rate_r~avg_rec_spans, labeller=label_both) +
          scale_shape_manual(values=c(0, 1, 2, 3), name="Bin Size")+
          scale_color_discrete(name="Exon %",
                               type=c("brown1", "purple3", "green4")) +
          scale_x_continuous(breaks=c(0.25, 1, 10, 50), trans="log10") +
          scale_y_continuous(breaks=c(0.1, 1, 10), trans="log10") +
          labs(title="Ratios of std. linear coefficients",
               x="CV(u) / CV(B)", y="Coeff(u) / Coeff(B)") +
          theme(axis.title=element_text(size=16), 
                axis.text=element_text(size=12), 
                axis.text.x=element_text(size=12),
                legend.text=element_text(size=16),
                legend.title=element_text(size=16),
                legend.position="bottom")
save_plot("Coeff_lm/ratios_lmcoeff_cv.png", p, base_height=12, base_width=18)
```

Meta linear models

The goal is to understand the effect of genomic parameters on the effect of
mu and B on pi.

```{r meta-lm, include=F, message=F, warning=F}

# meta linear models (pi neu)
## 10 kb
std_r2_10kb <- as.data.frame(r2_neu_models %>% 
                              filter(., scale==1e+4) %>% 
                              dplyr::select(c(u_avg, B_avg,
                                              num_exons,
                                              mult_r_u,
                                              shapes_rate_u,
                                              shapes_rate_r,
                                              avg_mut_spans,
                                              avg_rec_spans)) %>%
                              apply(., 2, function(x) (x-mean(x)) / sd(x)))

m_u_10kb <- lm(u_avg ~ (num_exons + shapes_rate_u + 
                        shapes_rate_r + mult_r_u +
                        avg_mut_spans + avg_rec_spans)^2,
               data=std_r2_10kb)
m_B_10kb <- lm(B_avg ~ (num_exons + shapes_rate_u +
                        shapes_rate_r + mult_r_u +
                        avg_mut_spans + avg_rec_spans)^2,
               data=std_r2_10kb)

summary(m_u_10kb)
summary(m_B_10kb)

## 100 kb
std_r2_100kb <- as.data.frame(r2_neu_models %>% 
                              filter(., scale==1e+5) %>% 
                              dplyr::select(c(u_avg, B_avg,
                                              num_exons,
                                              mult_r_u,
                                              shapes_rate_u,
                                              shapes_rate_r,
                                              avg_mut_spans,
                                              avg_rec_spans)) %>%
                              apply(., 2, function(x) (x-mean(x)) / sd(x)))

m_u_100kb <- lm(u_avg ~ (num_exons + shapes_rate_u + 
                         shapes_rate_r + mult_r_u +
                         avg_mut_spans + avg_rec_spans)^2, 
                data=std_r2_100kb)
m_B_100kb <- lm(B_avg ~ (num_exons + shapes_rate_u + 
                         shapes_rate_r + mult_r_u +
                         avg_mut_spans + avg_rec_spans)^2,
                data=std_r2_100kb)

summary(m_u_100kb)
summary(m_B_100kb)

## 1 Mb
std_r2_1Mb <- as.data.frame(r2_neu_models %>% 
                            filter(., scale==1e+6) %>% 
                            dplyr::select(c(u_avg, B_avg,
                                            num_exons,
                                            mult_r_u,
                                            shapes_rate_u,
                                            shapes_rate_r,
                                            avg_mut_spans,
                                            avg_rec_spans)) %>%
                            apply(., 2, function(x) (x-mean(x)) / sd(x)))

m_u_1Mb <- lm(u_avg ~ (num_exons + shapes_rate_u + 
                       shapes_rate_r + mult_r_u +
                       avg_mut_spans + avg_rec_spans)^2,
              data=std_r2_1Mb)
m_B_1Mb <- lm(B_avg ~ (num_exons + shapes_rate_u + 
                       shapes_rate_r + mult_r_u +
                       avg_mut_spans + avg_rec_spans)^2,
              data=std_r2_1Mb)

summary(m_u_1Mb)
summary(m_B_1Mb)

# mut
meta_lm_u <- rbind.data.frame(m_u_10kb$coefficients,
                              m_u_100kb$coefficients,
                              m_u_1Mb$coefficients)
names(meta_lm_u) <- names(m_u_10kb$coefficients)

meta_lm_u <- dplyr::select(meta_lm_u, c("shapes_rate_u",
                                        "avg_mut_spans",
                                        "num_exons",
                                        "mult_r_u",
                                        "shapes_rate_r",
                                        "avg_rec_spans"))

meta_lm_u$bin_size <- c(1e+4, 1e+5, 1e+6)
meta_lm_u$response <- "Mu"

mmlm_mu <- pivot_longer(meta_lm_u, cols=names(meta_lm_u)[1:(ncol(meta_lm_u)-2)],
                        names_to="var", values_to="coeff")

# B-val
meta_lm_B <- rbind.data.frame(m_B_10kb$coefficients,
                              m_B_100kb$coefficients,
                              m_B_1Mb$coefficients)
names(meta_lm_B) <- names(m_B_10kb$coefficients)

meta_lm_B <- dplyr::select(meta_lm_B, c("shapes_rate_u",
                                        "avg_mut_spans",
                                        "num_exons",
                                        "mult_r_u",
                                        "shapes_rate_r",
                                        "avg_rec_spans"))

meta_lm_B$bin_size <- c(1e+4, 1e+5, 1e+6)
meta_lm_B$response <- "B"

mmlm_B <- pivot_longer(meta_lm_B, cols=names(meta_lm_B)[1:(ncol(meta_lm_B)-2)],
                       names_to="var", values_to="coeff")

mmlm_tbl <- rbind.data.frame(mmlm_mu, mmlm_B)

meta_lm <- ggplot(data=mmlm_tbl, 
                  aes(x=var, 
                      y=coeff, 
                      shape=as.factor(bin_size),
                      color=response)) +
  geom_point(size=4) + theme_bw() +
  geom_hline(yintercept=0, linetype="dashed") +
  scale_y_continuous(breaks=pretty_breaks()) +
  scale_color_discrete(name=NULL, type=c("purple3", "green4")) +
  scale_shape_manual(values=c(0, 1, 2, 4), name="Bin Size") +
  labs(title="Meta Linear Model for Variance Explained by 'Response'",
       x="Variable", y="Coefficient") +
  theme(axis.title=element_text(size=16), 
        axis.text=element_text(size=12), 
        axis.text.x=element_text(angle=90, size=12, vjust=0.5, hjust=1.0),
        legend.text=element_text(size=16),
        legend.title=element_text(size=16),
        legend.position="bottom")
save_plot("meta_lm_pi_neu_sim.png", meta_lm, base_height=8, base_width=10)

# pi sim
## 10 kb
std_r2_10kb <- as.data.frame(r2_sim_models %>% 
                             filter(., scale==1e+4) %>% 
                             dplyr::select(c(u_avg, B_avg,
                                             num_exons,
                                             mult_r_u,
                                             shapes_rate_u,
                                             shapes_rate_r,
                                             avg_mut_spans,
                                             avg_rec_spans)) %>%
                             apply(., 2, function(x) (x-mean(x)) / sd(x)))

m_u_10kb <- lm(u_avg ~ (num_exons + shapes_rate_u + 
                        shapes_rate_r + mult_r_u +
                        avg_mut_spans + avg_rec_spans)^2,
               data=std_r2_10kb)
m_B_10kb <- lm(B_avg ~ (num_exons + shapes_rate_u + 
                        shapes_rate_r + mult_r_u +
                        avg_mut_spans + avg_rec_spans)^2,
               data=std_r2_10kb)

summary(m_u_10kb)
summary(m_B_10kb)

## 100 kb
std_r2_100kb <- as.data.frame(r2_sim_models %>% 
                              filter(., scale==1e+5) %>% 
                              dplyr::select(c(u_avg, B_avg,
                                              num_exons,
                                              mult_r_u,
                                              shapes_rate_u,
                                              shapes_rate_r,
                                              avg_mut_spans,
                                              avg_rec_spans)) %>%
                              apply(., 2, function(x) (x-mean(x)) / sd(x)))

m_u_100kb <- lm(u_avg ~ (num_exons + shapes_rate_u + 
                         shapes_rate_r + mult_r_u +
                         avg_mut_spans + avg_rec_spans)^2, 
                data=std_r2_100kb)
m_B_100kb <- lm(B_avg ~ (num_exons + shapes_rate_u + 
                         shapes_rate_r + mult_r_u +
                         avg_mut_spans + avg_rec_spans)^2, 
                data=std_r2_100kb)

summary(m_u_100kb)
summary(m_B_100kb)

## 1 Mb
std_r2_1Mb <- as.data.frame(r2_sim_models %>% 
                            filter(., scale==1e+6) %>% 
                            dplyr::select(c(u_avg, B_avg,
                                            num_exons,
                                            mult_r_u,
                                            shapes_rate_u,
                                            shapes_rate_r,
                                            avg_mut_spans,
                                            avg_rec_spans)) %>%
                            apply(., 2, function(x) (x-mean(x)) / sd(x)))

m_u_1Mb <- lm(u_avg ~ (num_exons + shapes_rate_u + 
                       shapes_rate_r + mult_r_u +
                       avg_mut_spans + avg_rec_spans)^2,
              data=std_r2_1Mb)
m_B_1Mb <- lm(B_avg ~ (num_exons + shapes_rate_u +
                       shapes_rate_r + mult_r_u +
                       avg_mut_spans + avg_rec_spans)^2,
              data=std_r2_1Mb)

summary(m_u_1Mb)
summary(m_B_1Mb)

# mut
meta_lm_u <- rbind.data.frame(m_u_10kb$coefficients,
                              m_u_100kb$coefficients,
                              m_u_1Mb$coefficients)
names(meta_lm_u) <- names(m_u_10kb$coefficients)

meta_lm_u <- dplyr::select(meta_lm_u, c("shapes_rate_u",
                                        "avg_mut_spans",
                                        "num_exons",
                                        "mult_r_u",
                                        "shapes_rate_r",
                                        "avg_rec_spans"))
meta_lm_u$bin_size <- c(1e+4, 1e+5, 1e+6)
meta_lm_u$response <- "Mu"

mmlm_mu <- pivot_longer(meta_lm_u, cols=names(meta_lm_u)[1:(ncol(meta_lm_u)-2)],
                        names_to="var", values_to="coeff")

# B-val
meta_lm_B <- rbind.data.frame(m_B_10kb$coefficients,
                              m_B_100kb$coefficients,
                              m_B_1Mb$coefficients)
names(meta_lm_B) <- names(m_B_10kb$coefficients)

meta_lm_B <- dplyr::select(meta_lm_B, c("shapes_rate_u",
                                        "avg_mut_spans",
                                        "num_exons",
                                        "mult_r_u",
                                        "shapes_rate_r",
                                        "avg_rec_spans"))
meta_lm_B$bin_size <- c(1e+4, 1e+5, 1e+6)
meta_lm_B$response <- "B"

mmlm_B <- pivot_longer(meta_lm_B, cols=names(meta_lm_B)[1:(ncol(meta_lm_B)-2)],
                       names_to="var", values_to="coeff")

mmlm_tbl <- rbind.data.frame(mmlm_mu, mmlm_B)

meta_lm <- ggplot(data=mmlm_tbl,
                  aes(x=var, 
                      y=coeff, 
                      shape=as.factor(bin_size), 
                      color=response)) +
  geom_point(size=4) + theme_bw() +
  geom_hline(yintercept=0, linetype="dashed") +
  scale_y_continuous(breaks=pretty_breaks()) +
  scale_color_discrete(name=NULL, type=c("purple3", "green4")) +
  scale_shape_manual(values=c(0, 1, 2, 4), name="Bin Size") +
  labs(title="Meta Linear Model for Variance Explained by 'Response'",
       x="Variable", y="Coefficient") +
  theme(axis.title=element_text(size=16), 
        axis.text=element_text(size=12), 
        axis.text.x=element_text(angle=90, size=12, vjust=0.5, hjust=1.0),
        legend.text=element_text(size=16),
        legend.title=element_text(size=16),
        legend.position="bottom")
save_plot("meta_lm_pi_total_sim.png", meta_lm, base_height=8, base_width=10)
```

## Path Models

Specifying the DAGs in lavaan format
Variable names should match those in the data frames
NOTE: 'del_sites', a transformation of 'exonic_sites,' is computed later on

TODO: test the fit of the nested mutagenic/non-mutagenic recombination models

```{r DAGs-lavaan, include=F, message=F, warning=F}

# non-mutagenic recomb., pi neutral
DAG1 <- "
pi_neu_sim ~ start(0.5)*c1*B
pi_neu_sim ~ start(0.5)*c2*avg_mut
B ~ start(-0.25)*c3*avg_mut
B ~ start(-0.5)*c6*del_sites
B ~ start(0.2)*c4*avg_rec
avg_mut ~~ 0*avg_rec
# fixing exogenous residual variances since we standardize variables?
avg_mut ~~ 1*avg_mut
avg_rec ~~ 1*avg_rec
del_sites ~~ 1*del_sites
# st
# defining indirect and total effects
rec_B_pi:=c1*c4
mu_B_pi:=c1*c3
total_mu_pi:=c2 + mu_B_pi
"

# mutagenic recomb., pi neutral
DAG2 <- "
pi_neu_sim ~ start(0.5)*c1*B
pi_neu_sim ~ start(0.5)*c2*avg_mut
B ~ start(-0.25)*c3*avg_mut
B ~ start(-0.5)*c6*del_sites
B ~ start(0.2)*c4*avg_rec
avg_mut ~ start(0.1)*c5*avg_rec
# fixing exogenous residual variance since we standardize variables?
avg_rec ~~ 1*avg_rec
del_sites ~~ 1*del_sites
# defining indirect and total effects
mu_B_pi:=c1*c3
rec_mu_B:=c3*c5
rec_B_pi:=c1*c4
rec_mu_pi:=c2*c5
rec_mu_B_pi:=mu_B_pi*c5
total_mu_pi:=c2 + mu_B_pi
total_rec_pi:=rec_B_pi + rec_mu_pi + rec_mu_B_pi
"

# non-mutagenic recomb., pi sim
DAG3 <- "
sim_pi ~ start(0.5)*c1*B
sim_pi ~ start(0.5)*c2*avg_mut
B ~ start(-0.25)*c3*avg_mut
B ~ start(-0.5)*c6*del_sites
B ~ start(0.2)*c4*avg_rec
avg_mut ~~ 0*avg_rec
# fixing exogenous residual variances since we standardize variables?
avg_mut ~~ 1*avg_mut
avg_rec ~~ 1*avg_rec
del_sites ~~ 1*del_sites
# defining indirect and total effects
rec_B_pi:=c1*c4
mu_B_pi:=c1*c3
total_mu_pi:=c2 + mu_B_pi
"

# mutagenic recomb., pi sim
DAG4 <- "
sim_pi ~ start(0.5)*c1*B
sim_pi ~ start(0.5)*c2*avg_mut
B ~ start(-0.25)*c3*avg_mut
B ~ start(-0.5)*c6*del_sites
B ~ start(0.2)*c4*avg_rec
avg_mut ~ start(0.3)*c5*avg_rec
# fixing exogenous residual variance since we standardize variables?
avg_rec ~~ 1*avg_rec
del_sites ~~ 1*del_sites
# defining indirect and total effects
mu_B_pi:=c1*c3
rec_mu_B:=c3*c5
rec_B_pi:=c1*c4
rec_mu_pi:=c2*c5
rec_mu_B_pi:=mu_B_pi*c5
total_mu_pi:=c2 + mu_B_pi
total_rec_pi:=rec_B_pi + rec_mu_pi + rec_mu_B_pi
"

# same as DAGs 3 and 4, but for real data (obs pi = avg_pi)
DAG5 <- "
avg_pi ~ start(0.5)*c1*B
avg_pi ~ start(0.5)*c2*avg_mut
B ~ start(-0.25)*c3*avg_mut
B ~ start(-0.5)*c6*del_sites
B ~ start(0.2)*c4*avg_rec
avg_mut ~~ 0*avg_rec
# fixing exogenous residual variances since we standardize variables?
avg_mut ~~ 1*avg_mut
avg_rec ~~ 1*avg_rec
del_sites ~~ 1*del_sites
# defining indirect and total effects
rec_B_pi:=c1*c4
mu_B_pi:=c1*c3
total_mu_pi:=c2 + mu_B_pi
"

# mutagenic recomb., pi sim
DAG6 <- "
avg_pi ~ start(0.5)*c1*B
avg_pi ~ start(0.5)*c2*avg_mut
B ~ start(-0.25)*c3*avg_mut
B ~ start(-0.5)*c6*del_sites
B ~ start(0.2)*c4*avg_rec
avg_mut ~ start(0.3)*c5*avg_rec
# fixing exogenous residual variance since we standardize variables?
avg_rec ~~ 1*avg_rec
del_sites ~~ 1*del_sites
# defining indirect and total effects
mu_B_pi:=c1*c3
rec_mu_B:=c3*c5
rec_B_pi:=c1*c4
rec_mu_pi:=c2*c5
rec_mu_B_pi:=mu_B_pi*c5
total_mu_pi:=c2 + mu_B_pi
total_rec_pi:=rec_B_pi + rec_mu_pi + rec_mu_B_pi
"
```

Fitting SEM models
Since data are simulated, there is no error in the maps and 
SEM reduces to path models

```{r SEM, include=F, message=F, warning=F}

# matching data frame structure with d-sep Fisher's C p-value
pval_sem1_10kb <- as.data.frame(matrix(nrow=num_models, ncol=num_reps))
pval_sem1_100kb <- as.data.frame(matrix(nrow=num_models, ncol=num_reps))
pval_sem1_1Mb <- as.data.frame(matrix(nrow=num_models, ncol=num_reps))

pval_sem2_10kb <- as.data.frame(matrix(nrow=num_models, ncol=num_reps))
pval_sem2_100kb <- as.data.frame(matrix(nrow=num_models, ncol=num_reps))
pval_sem2_1Mb <- as.data.frame(matrix(nrow=num_models, ncol=num_reps))

pval_sem3_10kb <- as.data.frame(matrix(nrow=num_models, ncol=num_reps))
pval_sem3_100kb <- as.data.frame(matrix(nrow=num_models, ncol=num_reps))
pval_sem3_1Mb <- as.data.frame(matrix(nrow=num_models, ncol=num_reps))

pval_sem4_10kb <- as.data.frame(matrix(nrow=num_models, ncol=num_reps))
pval_sem4_100kb <- as.data.frame(matrix(nrow=num_models, ncol=num_reps))
pval_sem4_1Mb <- as.data.frame(matrix(nrow=num_models, ncol=num_reps))

for(m in 1:num_models) {
  print(paste(m, Sys.Date(), Sys.time()), sep="\t")
  for(rep in 1:num_reps) {
    maps_1kb <- fread(paste("model_", m, "/rep_", rep, 
                            "/tbl_1kb.csv.gz", sep=""))
    data_cols <- names(maps_1kb)[4:ncol(maps_1kb)]
    
    maps_1kb$bin_10kb <- (1:nrow(maps_1kb) - 1) %/% 10
    maps_1kb$bin_100kb <- (1:nrow(maps_1kb) - 1) %/% 100
    maps_1kb$bin_1Mb <- (1:nrow(maps_1kb) - 1) %/% 1000
    
    maps_10kb <- maps_1kb %>%
                 group_by(bin_10kb) %>% 
                 summarise_at(data_cols, mean) %>%
                 dplyr::select(., -bin_10kb)
    
    maps_100kb <- maps_1kb %>%
                  group_by(bin_100kb) %>% 
                  summarise_at(data_cols, mean) %>%
                  dplyr::select(., -bin_100kb)
    
    maps_1Mb <- maps_1kb %>%
                group_by(bin_1Mb) %>% 
                summarise_at(data_cols, mean) %>%
                dplyr::select(., -bin_1Mb)
    
    # gets sum of relevant deleterious sites (around each genomic window)
    # we will use it as a more "meaningful" measure of local deleterious sites
    ## 10 kb
    maps_10kb$cum_rec <- cumsum(maps_10kb$avg_rec * 1e+4)
    # matrix with pairwise recombination distances
    prd <- as.data.frame(abs(outer(maps_10kb$cum_rec, 
                                   maps_10kb$cum_rec, 
                                   "-")))
    names(prd) <- 1:ncol(prd)
    for(d in 1:nrow(prd)) {
      prd[d, d] <- maps_10kb$avg_rec[d] * 1e+4 # filling in main diagonal
    }
    
    # here a threshold of 1e-2 is used otherwise memory explodes
    relevant_exons <- apply(prd, 2, function(x) x < 1e-2)
    exons_per_bin <- apply(relevant_exons, 1, function(x) which(x))
    maps_10kb$del_sites <- unlist(mclapply(mc.cores=16, exons_per_bin,
                           function(x) 
                           return(sum(maps_10kb$exonic_sites[x]))))# * 1/prd[x, x]))))
    
    ## 100 kb
    maps_100kb$cum_rec <- cumsum(maps_100kb$avg_rec * 1e+5)
    # matrix with pairwise recombination distances
    prd <- as.data.frame(abs(outer(maps_100kb$cum_rec, 
                                   maps_100kb$cum_rec, 
                                   "-")))
    names(prd) <- 1:ncol(prd)
    for(d in 1:nrow(prd)) {
      prd[d, d] <- maps_100kb$avg_rec[d] * 1e+5 # filling in diagonal
    }
    
    relevant_exons <- apply(prd, 2, function(x) x < 1e-2)
    exons_per_bin <- apply(relevant_exons, 1, function(x) which(x))
    maps_100kb$del_sites <- unlist(mclapply(mc.cores=16, exons_per_bin,
                         function(x) 
                         return(sum(maps_100kb$exonic_sites[x]))))# * 1/prd[x, x]))))
    
    ## 1 Mb
    maps_1Mb$cum_rec <- cumsum(maps_1Mb$avg_rec * 1e+6)
    # matrix with pairwise recombination distances
    prd <- as.data.frame(abs(outer(maps_1Mb$cum_rec, 
                                   maps_1Mb$cum_rec, 
                                   "-")))
    names(prd) <- 1:ncol(prd)
    for(d in 1:nrow(prd)) {
      prd[d, d] <- maps_1Mb$avg_rec[d] * 1e+6 # filling in diagonal
    }
    
    relevant_exons <- apply(prd, 2, function(x) x < 2.5e-2)
    exons_per_bin <- apply(relevant_exons, 1, function(x) which(x))
    maps_1Mb$del_sites <- unlist(lapply(exons_per_bin, function(x) 
                          return(sum(maps_1Mb$exonic_sites[x]))))# * 1/prd[x, x]))))
           
    # MVN tests -- only include endogenous variables (mu, B, pi) (Bolen, 1989)
    ## Filtering bins to avoid auto-correlation
    ## and keeping only one pi variable to avoid numerically singular matrices
    mvn_10kb <- mvn(data=dplyr::select(filter(maps_10kb, 
                                              row_number() %% 100==0,
                                              del_sites > 0),
                         c(avg_mut, B, sim_pi)),
                    transform="log",
                    univariatePlot="none",
                    multivariatePlot="none",
                    mvnTest="hz")
    
    mvn_100kb <- mvn(data=dplyr::select(filter(maps_100kb, 
                                               row_number() %% 10==0,
                                               del_sites > 0),
                          c(avg_mut, B, sim_pi)),
                     transform="log",
                     univariatePlot="none",
                     multivariatePlot="none",
                     mvnTest="hz")
    
    mvn_1Mb <- mvn(data=dplyr::select(filter(maps_1Mb, del_sites > 0),
                         c(avg_mut, B, sim_pi)),
                    transform="log",
                    univariatePlot="none",
                    multivariatePlot="none",
                    mvnTest="hz",
                    showOutliers=T)
    
    mvn_tbl <- data.frame(pval=c(mvn_10kb$multivariateNormality$`p value`,
                                 mvn_100kb$multivariateNormality$`p value`,
                                 mvn_1Mb$multivariateNormality$`p value`),
                          scale=c(1e+4, 1e+5, 1e+6))
    
    fwrite(mvn_tbl, paste("model_", m, "/rep_", rep, "/mvn_tests.csv", sep=""))
    
    # log-transforming and standardizing as well as
    # filtering to remove autocorrelation and outliers (del_sites > 0)
    filt_10kb <- filter(maps_10kb, del_sites > 0, row_number() %% 100 == 0) 
    std_10kb <- as.data.frame(apply(apply(dplyr::select(filt_10kb,
      c(B, avg_rec, del_sites, avg_mut, pi_neu_sim, sim_pi)), 2, log),
      2, function(x) (x-mean(x)) / sd(x)))
    
    filt_100kb <- filter(maps_100kb, del_sites > 0, row_number() %% 10 == 0) 
    std_100kb <- as.data.frame(apply(apply(dplyr::select(filt_100kb,
      c(B, avg_rec, del_sites, avg_mut, pi_neu_sim, sim_pi)), 2, log),
      2, function(x) (x-mean(x)) / sd(x)))
    
    filt_1Mb <- filter(maps_1Mb, del_sites > 0) 
    std_1Mb <- as.data.frame(apply(apply(dplyr::select(filt_1Mb,
      c(B, avg_rec, del_sites, avg_mut, pi_neu_sim, sim_pi)), 2, log),
      2, function(x) (x-mean(x)) / sd(x)))
    
    # Q-Q plots, also keeping only endogenous variales (Bolen, 1989 apud Shipley)
    svg(paste("model_", m, "/rep_", rep, "/qq_10kb.svg", sep=""), 
        width=9, height=7)
    heplots::cqplot(dplyr::select(std_10kb, c(avg_mut, 
                                              B, 
                                              sim_pi)))
    dev.off()
    
    svg(paste("model_", m, "/rep_", rep, "/qq_100kb.svg", sep=""),
        width=9, height=7)
    heplots::cqplot(dplyr::select(std_100kb, c(avg_mut, 
                                               B, 
                                               sim_pi)))
    dev.off()
    
    svg(paste("model_", m, "/rep_", rep, "/qq_1Mb.svg", sep=""),
        width=9, height=7)
    heplots::cqplot(dplyr::select(std_1Mb, c(avg_mut,
                                             B, 
                                             sim_pi)))
    dev.off()
    
    # pairwise scatter plots
    svg(paste("model_", m, "/rep_", rep, "/scatter_10kb.svg", sep=""), 
        width=9, height=7)
    pairs(std_10kb)
    dev.off()
    
    svg(paste("model_", m, "/rep_", rep, "/scatter_100kb.svg", sep=""),
        width=9, height=7)
    pairs(std_100kb)
    dev.off()
    
    svg(paste("model_", m, "/rep_", rep, "/scatter_1Mb.svg", sep=""),
        width=9, height=7)
    pairs(std_1Mb)
    dev.off()
    
    # using lavaan to fit path models, DAG1
    sem1_10kb <- sem(DAG1, data=std_10kb, estimator="GLS")
    sem1_100kb <- sem(DAG1, data=std_100kb, estimator="GLS")
    sem1_1Mb <- sem(DAG1, data=std_1Mb, estimator="GLS")
    
    # storing p-values
    # error handling to bypass cases where lavaan fails to converge the optim
    # such cases can be identified as NA entries in the pval table
    tryCatch({
      pval_sem1_10kb[m, rep] <- fitMeasures(sem1_10kb)[5]
    }, error=function(e){cat("WARNING :", conditionMessage(e), "\n")})
    
    tryCatch({
      pval_sem1_100kb[m, rep] <- fitMeasures(sem1_100kb)[5]
    }, error=function(e){cat("WARNING :", conditionMessage(e), "\n")})
    
    tryCatch({
      pval_sem1_1Mb[m, rep] <- fitMeasures(sem1_1Mb)[5]
    }, error=function(e){cat("WARNING :", conditionMessage(e), "\n")})
    
    # r-sqrd of endogenous variables, low for B because
    # its determinants extend beyond the focal window
    sem1_r2 <- rbind.data.frame(inspect(sem1_10kb, "rsquare"),
                                inspect(sem1_100kb, "rsquare"),
                                inspect(sem1_1Mb, "rsquare"))
    names(sem1_r2) <- c("pi_neu_sim", "B")
    fwrite(sem1_r2, paste("model_", m, 
                          "/rep_", rep, 
                          "/sem_DAG1_r2.csv", sep=""))
    
    fwrite(parameterestimates(sem1_10kb, ci=T), 
           paste("model_", m, 
                 "/rep_", rep, 
                 "/lavaan_DAG1_10kb.csv", sep=""))
    
    fwrite(parameterestimates(sem1_100kb, ci=T), 
           paste("model_", m, 
                 "/rep_", rep,
                 "/lavaan_DAG1_100kb.csv", sep=""))
    
    fwrite(parameterestimates(sem1_1Mb, ci=T), 
           paste("model_", m, 
                 "/rep_", rep, 
                 "/lavaan_DAG1_1Mb.csv", sep=""))
    
    # DAG 2
    sem2_10kb <- sem(DAG2, data=std_10kb, estimator="GLS")
    sem2_100kb <- sem(DAG2, data=std_100kb, estimator="GLS")
    sem2_1Mb <- sem(DAG2, data=std_1Mb, estimator="GLS")
    
    # error handling to bypass cases where lavaan fails to converge the optim
    # such cases can be identified as NA entries in the pval table
    tryCatch({
      pval_sem2_10kb[m, rep] <- fitMeasures(sem2_10kb)[5]
    }, error=function(e){cat("WARNING :", conditionMessage(e), "\n")})
    
    tryCatch({
      pval_sem2_100kb[m, rep] <- fitMeasures(sem2_100kb)[5]
    }, error=function(e){cat("WARNING :", conditionMessage(e), "\n")})
    
    tryCatch({
      pval_sem2_1Mb[m, rep] <- fitMeasures(sem2_1Mb)[5]
    }, error=function(e){cat("WARNING :", conditionMessage(e), "\n")})
    
    sem2_r2 <- rbind.data.frame(inspect(sem2_10kb, "rsquare"),
                                inspect(sem2_100kb, "rsquare"),
                                inspect(sem2_1Mb, "rsquare"))
    names(sem2_r2) <- c("pi_neu_sim", "B", "avg_mut")
    fwrite(sem2_r2, paste("model_", m,
                          "/rep_", rep,
                          "/sem_DAG2_r2.csv", sep=""))
    
    fwrite(parameterestimates(sem2_10kb, ci=T), 
           paste("model_", m, 
                 "/rep_", rep, 
                 "/lavaan_DAG2_10kb.csv", sep=""))
    
    fwrite(parameterestimates(sem2_100kb, ci=T), 
           paste("model_", m, 
                 "/rep_", rep, 
                 "/lavaan_DAG2_100kb.csv", sep=""))
    
    fwrite(parameterestimates(sem2_1Mb, ci=T), 
           paste("model_", m, 
                 "/rep_", rep, 
                 "/lavaan_DAG2_1Mb.csv", sep=""))
    
    # DAG 3
    sem3_10kb <- sem(DAG3, data=std_10kb, estimator="GLS")
    sem3_100kb <- sem(DAG3, data=std_100kb, estimator="GLS")
    sem3_1Mb <- sem(DAG3, data=std_1Mb, estimator="GLS")
    
    # error handling to bypass cases where lavaan fails to converge the optim
    # such cases can be identified as NA entries in the pval table
    tryCatch({
      pval_sem3_10kb[m, rep] <- fitMeasures(sem3_10kb)[5]
    }, error=function(e){cat("WARNING :", conditionMessage(e), "\n")})
    
    tryCatch({
      pval_sem3_100kb[m, rep] <- fitMeasures(sem3_100kb)[5]
    }, error=function(e){cat("WARNING :", conditionMessage(e), "\n")})
    
    tryCatch({
      pval_sem3_1Mb[m, rep] <- fitMeasures(sem3_1Mb)[5]
    }, error=function(e){cat("WARNING :", conditionMessage(e), "\n")})
    
    sem3_r2 <- rbind.data.frame(inspect(sem3_10kb, "rsquare"),
                                inspect(sem3_100kb, "rsquare"),
                                inspect(sem3_1Mb, "rsquare"))
    names(sem3_r2) <- c("pi_sim", "B")
    fwrite(sem3_r2, paste("model_", m,
                          "/rep_", rep,
                          "/sem_DAG3_r2.csv", sep=""))
    
    fwrite(parameterestimates(sem3_10kb, ci=T), 
           paste("model_", m, 
                 "/rep_", rep, 
                 "/lavaan_DAG3_10kb.csv", sep=""))
    
    fwrite(parameterestimates(sem3_100kb, ci=T), 
           paste("model_", m,
                 "/rep_", rep, 
                 "/lavaan_DAG3_100kb.csv", sep=""))
    
    fwrite(parameterestimates(sem3_1Mb, ci=T), 
           paste("model_", m, 
                 "/rep_", rep, 
                 "/lavaan_DAG3_1Mb.csv", sep=""))
    
    # DAG 4
    sem4_10kb <- sem(DAG4, data=std_10kb, estimator="GLS")
    sem4_100kb <- sem(DAG4, data=std_100kb, estimator="GLS")
    sem4_1Mb <- sem(DAG4, data=std_1Mb, estimator="GLS")
    
    # error handling to bypass cases where lavaan fails to converge the optim
    # such cases can be identified as NA entries in the pval table
    tryCatch({
      pval_sem4_10kb[m, rep] <- fitMeasures(sem4_10kb)[5]
    }, error=function(e){cat("WARNING :", conditionMessage(e), "\n")})
    
    tryCatch({
      pval_sem4_100kb[m, rep] <- fitMeasures(sem4_100kb)[5]
    }, error=function(e){cat("WARNING :", conditionMessage(e), "\n")})
    
    tryCatch({
      pval_sem4_1Mb[m, rep] <- fitMeasures(sem4_1Mb)[5]
    }, error=function(e){cat("WARNING :", conditionMessage(e), "\n")})
    
    sem4_r2 <- rbind.data.frame(inspect(sem4_10kb, "rsquare"),
                                inspect(sem4_100kb, "rsquare"),
                                inspect(sem4_1Mb, "rsquare"))
    names(sem4_r2) <- c("pi_sim", "B", "avg_mut")
    fwrite(sem4_r2, paste("model_", m,
                          "/rep_", rep,
                          "/sem_DAG4_r2.csv", sep=""))
    
    fwrite(parameterestimates(sem4_10kb, ci=T), 
           paste("model_", m, 
                 "/rep_", rep, 
                 "/lavaan_DAG4_10kb.csv", sep=""))
    
    fwrite(parameterestimates(sem4_100kb, ci=T), 
           paste("model_", m, 
                 "/rep_", rep,
                 "/lavaan_DAG4_100kb.csv", sep=""))
    
    fwrite(parameterestimates(sem4_1Mb, ci=T), 
           paste("model_", m, 
                 "/rep_", rep, 
                 "/lavaan_DAG4_1Mb.csv", sep=""))
  }
}

pval_sem1 <- rbind.data.frame(pval_sem1_10kb,
                              pval_sem1_100kb,
                              pval_sem1_1Mb)
pval_sem1$DAG <- 1
pval_sem1$model <- rep(1:num_models, 3)
pval_sem1$scale <- c(rep(1e+4, num_models),
                     rep(1e+5, num_models),
                     rep(1e+6, num_models))
  
pval_sem2 <- rbind.data.frame(pval_sem2_10kb,
                              pval_sem2_100kb,
                              pval_sem2_1Mb)
pval_sem2$DAG <- 2
pval_sem2$model <- rep(1:num_models, 3)
pval_sem2$scale <- c(rep(1e+4, num_models),
                     rep(1e+5, num_models),
                     rep(1e+6, num_models))

pval_sem3 <- rbind.data.frame(pval_sem3_10kb,
                              pval_sem3_100kb,
                              pval_sem3_1Mb)
pval_sem3$DAG <- 3
pval_sem3$model <- rep(1:num_models, 3)
pval_sem3$scale <- c(rep(1e+4, num_models),
                     rep(1e+5, num_models),
                     rep(1e+6, num_models))

pval_sem4 <- rbind.data.frame(pval_sem4_10kb,
                              pval_sem4_100kb,
                              pval_sem4_1Mb)
pval_sem4$DAG <- 4
pval_sem4$model <- rep(1:num_models, 3)
pval_sem4$scale <- c(rep(1e+4, num_models),
                     rep(1e+5, num_models),
                     rep(1e+6, num_models))

pval_sem <- rbind.data.frame(pval_sem1, pval_sem2, pval_sem3, pval_sem4)
names(pval_sem)[1:10] <- 1:num_reps

fwrite(pval_sem, "pval_sem.csv.gz")
```

Manhattan plots matching those obtained with d-separation 

```{r manhattan-plots-SEM, include=F, message=F, warning=F}

mvn_tests_10kb <- as.data.frame(matrix(nrow=num_models, ncol=num_reps))
mvn_tests_100kb <- as.data.frame(matrix(nrow=num_models, ncol=num_reps))
mvn_tests_1Mb <- as.data.frame(matrix(nrow=num_models, ncol=num_reps))

names(mvn_tests_10kb) <- 1:num_reps
names(mvn_tests_100kb) <- 1:num_reps
names(mvn_tests_1Mb) <-1:num_reps

# collecting pvalues from MVN tests
pb <- txtProgressBar(min=0, max=num_models, style=3)
for(m in 1:num_models) {
  setTxtProgressBar(pb, m)
  for(rep in 1:num_reps) {
    mvn <- fread(paste("model_", m,
                       "/rep_", rep,
                       "/mvn_tests.csv", sep=""))
    
    mvn_tests_10kb[m, rep] <- mvn$pval[1]
    mvn_tests_100kb[m, rep] <- mvn$pval[2]
    mvn_tests_1Mb[m, rep] <- mvn$pval[3]
  }
}
close(pb)

mvn_tests_10kb$scale <- 1e+4
mvn_tests_100kb$scale <- 1e+5
mvn_tests_1Mb$scale <- 1e+6

mvn_tests_10kb$model <- 1:num_models
mvn_tests_100kb$model <- 1:num_models
mvn_tests_1Mb$model <- 1:num_models

mvn_tests <- rbind.data.frame(mvn_tests_10kb, 
                              mvn_tests_100kb,
                              mvn_tests_1Mb)

m_mvn <- pivot_longer(mvn_tests, 
                      cols=as.character(1:num_reps), 
                      names_to="replicate",
                      values_to="mvn_p")

pval_sem <- fread("pval_sem.csv.gz")
m_pval_sem <- pivot_longer(merge(pval_sem, models, by="model"),
                           cols=as.character(1:num_reps),
                           names_to="replicate", 
                           values_to="pval")

fwrite(m_pval_sem, "m_pval_sem.csv.gz")
df <- merge(m_pval_sem, m_mvn, by=c("model", "replicate", "scale"))

# manhattan plots with shape of points indicating significance (NAs are dropped)
p1 <- ggplot(data=filter(df, DAG==1),
             aes(x=model,
                 y=-log(pval), 
                 shape=mvn_p > 0.01, 
                 color=as.factor(mult_r_u))) +
      geom_hline(aes(yintercept=-log10(0.05/num_reps)), 
                 linetype="dashed", color="grey1") +
      geom_point(size=1.5) + theme_bw() + 
      facet_wrap(~scale, ncol=1, labeller=label_both) +
      scale_shape_manual(values=c(1, 4)) +
      scale_color_manual(values=c("brown1", "purple3", "green4", "cyan3")) +
      scale_y_continuous(breaks=pretty_breaks()) +
      scale_x_continuous(breaks=c(0, 432, 864, 1296)) + 
      labs(title="H0: data fits DAG 1 (non-mutagenic rec., Pi Neu.)",
           x="Model ID", y="-log10(p-value)",
           shape="MVN p-val > 1%", color="Mutagenic Rec.") +
      theme(axis.title=element_text(size=16), 
            axis.text=element_text(size=12), 
            axis.text.x=element_text(size=12),
            legend.text=element_text(size=16),
            legend.title=element_text(size=16),
            legend.position="bottom")
save_plot("Manhattan/manhattan_SEM_dag1.png", p1, 
          base_height=10, base_width=15)

p2 <- ggplot(data=filter(df, DAG==2),
             aes(x=model, 
                 y=-log(pval), 
                 shape=mvn_p > 0.01, 
                 color=as.factor(mult_r_u))) +
      geom_hline(aes(yintercept=-log10(0.05/num_reps)), 
                 linetype="dashed", color="grey1") +
      geom_point(size=1.5) + theme_bw() + 
      facet_wrap(~scale, ncol=1, labeller=label_both) +
      scale_shape_manual(values=c(1, 4)) +
      scale_color_manual(values=c("brown1", "purple3", "green4", "cyan3")) +
      scale_y_continuous(breaks=pretty_breaks()) +
      scale_x_continuous(breaks=c(0, 432, 864, 1296)) + 
      labs(title="H0: data fits DAG 2 (mutagenic rec., Pi Neu.)",
           x="Model ID", y="-log10(p-value)",
           shape="MVN p-val > 1%", color="Mutagenic Rec.") +
      theme(axis.title=element_text(size=16), 
            axis.text=element_text(size=12), 
            axis.text.x=element_text(size=12),
            legend.text=element_text(size=16),
            legend.title=element_text(size=16),
            legend.position="bottom")
save_plot("Manhattan/manhattan_SEM_dag2.png", p2, 
          base_height=10, base_width=15)

p3 <- ggplot(data=filter(df, DAG==3),
             aes(x=model, 
                 y=-log(pval), 
                 shape=mvn_p > 0.01, 
                 color=as.factor(mult_r_u))) +
      geom_hline(aes(yintercept=-log10(0.05/num_reps)), 
                 linetype="dashed", color="grey1") +
      geom_point(size=1.5) + theme_bw() + 
      facet_wrap(~scale, ncol=1, labeller=label_both) +
      scale_shape_manual(values=c(1, 4)) +
      scale_color_manual(values=c("brown1", "purple3", "green4", "cyan3")) +
      scale_y_continuous(breaks=pretty_breaks()) +
      scale_x_continuous(breaks=c(0, 432, 864, 1296)) + 
      labs(title="H0: data fits DAG 3 (non-mutagenic rec., Pi Tot.)",
           x="Model ID", y="-log10(p-value)",
           shape="MVN p-val > 1%", color="Mutagenic Rec.") +
      theme(axis.title=element_text(size=16), 
            axis.text=element_text(size=12), 
            axis.text.x=element_text(size=12),
            legend.text=element_text(size=16),
            legend.title=element_text(size=16),
            legend.position="bottom")
save_plot("Manhattan/manhattan_SEM_dag3.png", p3,
          base_height=10, base_width=15)

p4 <- ggplot(data=filter(df, DAG==4),
             aes(x=model,
                 y=-log(pval), 
                 shape=mvn_p > 0.01, 
                 color=as.factor(mult_r_u))) +
      geom_hline(aes(yintercept=-log10(0.05/num_reps)), 
                 linetype="dashed", color="grey1") +
      geom_point(size=1.5) + theme_bw() + 
      facet_wrap(~scale, ncol=1, labeller=label_both) +
      scale_shape_manual(values=c(1, 4)) +
      scale_color_manual(values=c("brown1", "purple3", "green4", "cyan3")) +
      scale_y_continuous(breaks=pretty_breaks()) +
      scale_x_continuous(breaks=c(0, 432, 864, 1296)) + 
      labs(title="H0: data fits DAG 4 (mutagenic rec., Pi Tot.)",
           x="Model ID", y="-log10(p-value)",
           shape="MVN p-val > 1%", color="Mutagenic Rec.") +
      theme(axis.title=element_text(size=16), 
            axis.text=element_text(size=12), 
            axis.text.x=element_text(size=12),
            legend.text=element_text(size=16),
            legend.title=element_text(size=16),
            legend.position="bottom")
save_plot("Manhattan/manhattan_SEM_dag4.png", p4, 
          base_height=10, base_width=15)

# DAG 1, bracketing MVN p-value
p5 <- ggplot(data=filter(df, DAG==1, mvn_p < 1e-2),
             aes(x=model, 
                 y=-log(pval), 
                 alpha=mvn_p,  
                 color=as.factor(mult_r_u))) +
      geom_hline(aes(yintercept=-log10(0.05/num_reps)), 
                 linetype="dashed", color="grey1") +
      geom_point(size=1.5) + theme_bw() + 
      facet_wrap(~scale, ncol=1, labeller=label_both) +
      scale_shape_manual(values=c(1, 4)) +
      scale_color_manual(values=c("brown1", "purple3", "green4", "cyan3")) +
      scale_y_continuous(breaks=pretty_breaks()) +
      scale_x_continuous(breaks=c(0, 432, 864, 1296)) + 
      labs(title="H0: data fits DAG 1 (non-mutagenic rec., Pi Neu.) | MVN p < 1%",
           x="Model ID", y="-log10(p-value)",
           color="Mutagenic Rec.") +
      theme(axis.title=element_text(size=16), 
            axis.text=element_text(size=12), 
            axis.text.x=element_text(size=12),
            legend.text=element_text(size=16),
            legend.title=element_text(size=16),
            legend.position="bottom")
save_plot("Manhattan/manhattan_SEM_dag1_mvnpL.png", p5, 
          base_height=10, base_width=15)

p6 <- ggplot(data=filter(df, DAG==1, mvn_p > 1e-2 & mvn_p < 1e-1),
             aes(x=model,
                 y=-log(pval), 
                 alpha=mvn_p,  
                 color=as.factor(mult_r_u))) +
      geom_hline(aes(yintercept=-log10(0.05/num_reps)), 
                 linetype="dashed", color="grey1") +
      geom_point(size=1.5) + theme_bw() + 
      facet_wrap(~scale, ncol=1, labeller=label_both) +
      scale_shape_manual(values=c(1, 4)) +
      scale_color_manual(values=c("brown1", "purple3", "green4", "cyan3")) +
      scale_y_continuous(breaks=pretty_breaks()) +
      scale_x_continuous(breaks=c(0, 432, 864, 1296)) + 
      labs(title=paste("H0: data fits DAG 1 (non-mutagenic rec., Pi Neu.)",
                       "| 10% < MVN p > 1%"),
           x="Model ID", y="-log10(p-value)",
           color="Mutagenic Rec.") +
      theme(axis.title=element_text(size=16), 
            axis.text=element_text(size=12), 
            axis.text.x=element_text(size=12),
            legend.text=element_text(size=16),
            legend.title=element_text(size=16),
            legend.position="bottom")
save_plot("Manhattan/manhattan_SEM_dag1_mvnpM.png", p6,
          base_height=10, base_width=15)

p7 <- ggplot(data=filter(df, DAG==1, mvn_p > 1e-1),
             aes(x=model,
                 y=-log(pval), 
                 alpha=mvn_p,   
                 color=as.factor(mult_r_u))) +
      geom_hline(aes(yintercept=-log10(0.05/num_reps)), 
                 linetype="dashed", color="grey1") +
      geom_point(size=1.5) + theme_bw() + 
      facet_wrap(~scale, ncol=1, labeller=label_both) +
      scale_shape_manual(values=c(1, 4)) +
      scale_color_manual(values=c("brown1", "purple3", "green4", "cyan3")) +
      scale_y_continuous(breaks=pretty_breaks()) +
      scale_x_continuous(breaks=c(0, 432, 864, 1296)) + 
      labs(title="H0: data fits DAG 1 (non-mutagenic rec., Pi Neu.) | MVN p > 10%",
           x="Model ID", y="-log10(p-value)",
           color="Mutagenic Rec.") +
      theme(axis.title=element_text(size=16), 
            axis.text=element_text(size=12), 
            axis.text.x=element_text(size=12),
            legend.text=element_text(size=16),
            legend.title=element_text(size=16),
            legend.position="bottom")
save_plot("Manhattan/manhattan_SEM_dag1_mvnpH.png", p7,
          base_height=10, base_width=15)

# DAG 2, bracketing MVN p-value
p8 <- ggplot(data=filter(df, DAG==2, mvn_p < 1e-2),
             aes(x=model,
                 y=-log(pval), 
                 alpha=mvn_p,   
                 color=as.factor(mult_r_u))) +
      geom_hline(aes(yintercept=-log10(0.05/num_reps)), 
                 linetype="dashed", color="grey1") +
      geom_point(size=1.5) + theme_bw() + 
      facet_wrap(~scale, ncol=1, labeller=label_both) +
      scale_shape_manual(values=c(1, 4)) +
      scale_color_manual(values=c("brown1", "purple3", "green4", "cyan3")) +
      scale_y_continuous(breaks=pretty_breaks()) +
      scale_x_continuous(breaks=c(0, 432, 864, 1296)) + 
      labs(title="H0: data fits DAG 2 (mutagenic rec., Pi Neu.) | MVN p < 1%",
           x="Model ID", y="-log10(p-value)",
           color="Mutagenic Rec.") +
      theme(axis.title=element_text(size=16), 
            axis.text=element_text(size=12), 
            axis.text.x=element_text(size=12),
            legend.text=element_text(size=16),
            legend.title=element_text(size=16),
            legend.position="bottom")
save_plot("Manhattan/manhattan_SEM_dag2_mvnpL.png", p8,
          base_height=10, base_width=15)

p9 <- ggplot(data=filter(df, DAG==2, mvn_p > 1e-2 & mvn_p < 1e-1),
             aes(x=model,
                 y=-log(pval), 
                 alpha=mvn_p,   
                 color=as.factor(mult_r_u))) +
      geom_hline(aes(yintercept=-log10(0.05/num_reps)), 
                 linetype="dashed", color="grey1") +
      geom_point(size=1.5) + theme_bw() + 
      facet_wrap(~scale, ncol=1, labeller=label_both) +
      scale_shape_manual(values=c(1, 4)) +
      scale_color_manual(values=c("brown1", "purple3", "green4", "cyan3")) +
      scale_y_continuous(breaks=pretty_breaks()) +
      scale_x_continuous(breaks=c(0, 432, 864, 1296)) + 
      labs(title="H0: data fits DAG 2 (mutagenic rec., Pi Neu.) | 10% < MVN p > 1%",
           x="Model ID", y="-log10(p-value)",
           color="Mutagenic Rec.") +
      theme(axis.title=element_text(size=16), 
            axis.text=element_text(size=12), 
            axis.text.x=element_text(size=12),
            legend.text=element_text(size=16),
            legend.title=element_text(size=16),
            legend.position="bottom")
save_plot("Manhattan/manhattan_SEM_dag2_mvnpM.png", p9, 
          base_height=10, base_width=15)

p10 <- ggplot(data=filter(df, DAG==2, mvn_p > 1e-1),
             aes(x=model,
                 y=-log(pval), 
                 alpha=mvn_p,   
                 color=as.factor(mult_r_u))) +
      geom_hline(aes(yintercept=-log10(0.05/num_reps)), 
                 linetype="dashed", color="grey1") +
      geom_point(size=1.5) + theme_bw() + 
      facet_wrap(~scale, ncol=1, labeller=label_both) +
      scale_shape_manual(values=c(1, 4)) +
      scale_color_manual(values=c("brown1", "purple3", "green4", "cyan3")) +
      scale_y_continuous(breaks=pretty_breaks()) +
      scale_x_continuous(breaks=c(0, 432, 864, 1296)) + 
      labs(title="H0: data fits DAG 2 (mutagenic rec., Pi Neu.) | MVN p > 10%",
           x="Model ID", y="-log10(p-value)",
           color="Mutagenic Rec.") +
      theme(axis.title=element_text(size=16), 
            axis.text=element_text(size=12), 
            axis.text.x=element_text(size=12),
            legend.text=element_text(size=16),
            legend.title=element_text(size=16),
            legend.position="bottom")
save_plot("Manhattan/manhattan_SEM_dag2_mvnpH.png", p10, 
          base_height=10, base_width=15)

# DAG 3, bracketing MVN p-value
p11 <- ggplot(data=filter(df, DAG==3, mvn_p < 1e-2),
             aes(x=model, 
                 y=-log(pval), 
                 alpha=mvn_p,   
                 color=as.factor(mult_r_u))) +
      geom_hline(aes(yintercept=-log10(0.05/num_reps)), 
                 linetype="dashed", color="grey1") +
      geom_point(size=1.5) + theme_bw() + 
      facet_wrap(~scale, ncol=1, labeller=label_both) +
      scale_shape_manual(values=c(1, 4)) +
      scale_color_manual(values=c("brown1", "purple3", "green4", "cyan3")) +
      scale_y_continuous(breaks=pretty_breaks()) +
      scale_x_continuous(breaks=c(0, 432, 864, 1296)) + 
      labs(title="H0: data fits DAG 3 (non-mutagenic rec., Pi Tot.) | MVN p < 1%",
           x="Model ID", y="-log10(p-value)",
           color="Mutagenic Rec.") +
      theme(axis.title=element_text(size=16), 
            axis.text=element_text(size=12), 
            axis.text.x=element_text(size=12),
            legend.text=element_text(size=16),
            legend.title=element_text(size=16),
            legend.position="bottom")
save_plot("Manhattan/manhattan_SEM_dag3_mvnpL.png", p11, 
          base_height=10, base_width=15)

p12 <- ggplot(data=filter(df, DAG==3, mvn_p > 1e-2 & mvn_p < 1e-1),
             aes(x=model,
                 y=-log(pval), 
                 alpha=mvn_p,   
                 color=as.factor(mult_r_u))) +
      geom_hline(aes(yintercept=-log10(0.05/num_reps)), 
                 linetype="dashed", color="grey1") +
      geom_point(size=1.5) + theme_bw() + 
      facet_wrap(~scale, ncol=1, labeller=label_both) +
      scale_shape_manual(values=c(1, 4)) +
      scale_color_manual(values=c("brown1", "purple3", "green4", "cyan3")) +
      scale_y_continuous(breaks=pretty_breaks()) +
      scale_x_continuous(breaks=c(0, 432, 864, 1296)) + 
      labs(title="H0: data fits DAG 3 (non-mutagenic rec., Pi Tot.) | 10% < MVN p > 1%",
           x="Model ID", y="-log10(p-value)",
           color="Mutagenic Rec.") +
      theme(axis.title=element_text(size=16), 
            axis.text=element_text(size=12), 
            axis.text.x=element_text(size=12),
            legend.text=element_text(size=16),
            legend.title=element_text(size=16),
            legend.position="bottom")
save_plot("Manhattan/manhattan_SEM_dag3_mvnpM.png", p12,
          base_height=10, base_width=15)

p13 <- ggplot(data=filter(df, DAG==3, mvn_p > 1e-1),
             aes(x=model, 
                 y=-log(pval),  
                 alpha=mvn_p,  
                 color=as.factor(mult_r_u))) +
      geom_hline(aes(yintercept=-log10(0.05/num_reps)), 
                 linetype="dashed", color="grey1") +
      geom_point(size=1.5) + theme_bw() + 
      facet_wrap(~scale, ncol=1, labeller=label_both) +
      scale_shape_manual(values=c(1, 4)) +
      scale_color_manual(values=c("brown1", "purple3", "green4", "cyan3")) +
      scale_y_continuous(breaks=pretty_breaks()) +
      scale_x_continuous(breaks=c(0, 432, 864, 1296)) + 
      labs(title="H0: data fits DAG 3 (non-mutagenic rec., Pi Tot.) | MVN p > 10%",
           x="Model ID", y="-log10(p-value)",
           color="Mutagenic Rec.") +
      theme(axis.title=element_text(size=16), 
            axis.text=element_text(size=12), 
            axis.text.x=element_text(size=12),
            legend.text=element_text(size=16),
            legend.title=element_text(size=16),
            legend.position="bottom")
save_plot("Manhattan/manhattan_SEM_dag3_mvnpH.png", p13, 
          base_height=10, base_width=15)

# DAG 4, bracketing MVN p-value
p14 <- ggplot(data=filter(df, DAG==4, mvn_p < 1e-2),
             aes(x=model,
                 y=-log(pval), 
                 alpha=mvn_p,  
                 color=as.factor(mult_r_u))) +
      geom_hline(aes(yintercept=-log10(0.05/num_reps)), 
                 linetype="dashed", color="grey1") +
      geom_point(size=1.5) + theme_bw() + 
      facet_wrap(~scale, ncol=1, labeller=label_both) +
      scale_shape_manual(values=c(1, 4)) +
      scale_color_manual(values=c("brown1", "purple3", "green4", "cyan3")) +
      scale_y_continuous(breaks=pretty_breaks()) +
      scale_x_continuous(breaks=c(0, 432, 864, 1296)) + 
      labs(title="H0: data fits DAG 4 (mutagenic rec., Pi Tot.) | MVN p < 1%",
           x="Model ID", y="-log10(p-value)",
           color="Mutagenic Rec.") +
      theme(axis.title=element_text(size=16), 
            axis.text=element_text(size=12), 
            axis.text.x=element_text(size=12),
            legend.text=element_text(size=16),
            legend.title=element_text(size=16),
            legend.position="bottom")
save_plot("Manhattan/manhattan_SEM_dag4_mvnpL.png", p14,
          base_height=10, base_width=15)

p15 <- ggplot(data=filter(df, DAG==4, mvn_p > 1e-2 & mvn_p < 1e-1),
             aes(x=model, 
                 y=-log(pval),  
                 alpha=mvn_p,  
                 color=as.factor(mult_r_u))) +
      geom_hline(aes(yintercept=-log10(0.05/num_reps)), 
                 linetype="dashed", color="grey1") +
      geom_point(size=1.5) + theme_bw() + 
      facet_wrap(~scale, ncol=1, labeller=label_both) +
      scale_shape_manual(values=c(1, 4)) +
      scale_color_manual(values=c("brown1", "purple3", "green4", "cyan3")) +
      scale_y_continuous(breaks=pretty_breaks()) +
      scale_x_continuous(breaks=c(0, 432, 864, 1296)) + 
      labs(title="H0: data fits DAG 4 (mutagenic rec., Pi Tot.) | 10% < MVN p > 1%",
           x="Model ID", y="-log10(p-value)",
           color="Mutagenic Rec.") +
      theme(axis.title=element_text(size=16), 
            axis.text=element_text(size=12), 
            axis.text.x=element_text(size=12),
            legend.text=element_text(size=16),
            legend.title=element_text(size=16),
            legend.position="bottom")
save_plot("Manhattan/manhattan_SEM_dag4_mvnpM.png", p15, 
          base_height=10, base_width=15)

p16 <- ggplot(data=filter(df, DAG==4, mvn_p > 1e-1),
             aes(x=model, 
                 y=-log(pval),  
                 alpha=mvn_p,  
                 color=as.factor(mult_r_u))) +
      geom_hline(aes(yintercept=-log10(0.05/num_reps)), 
                 linetype="dashed", color="grey1") +
      geom_point(size=1.5) + theme_bw() + 
      facet_wrap(~scale, ncol=1, labeller=label_both) +
      scale_shape_manual(values=c(1, 4)) +
      scale_color_manual(values=c("brown1", "purple3", "green4", "cyan3")) +
      scale_y_continuous(breaks=pretty_breaks()) +
      scale_x_continuous(breaks=c(0, 432, 864, 1296)) + 
      labs(title="H0: data fits DAG 4 (mutagenic rec., Pi Tot.) | MVN p > 10%",
           x="Model ID", y="-log10(p-value)",
           color="Mutagenic Rec.") +
      theme(axis.title=element_text(size=16), 
            axis.text=element_text(size=12), 
            axis.text.x=element_text(size=12),
            legend.text=element_text(size=16),
            legend.title=element_text(size=16),
            legend.position="bottom")
save_plot("Manhattan/manhattan_SEM_dag4_mvnpH.png", p16,
          base_height=10, base_width=15)
```

R-squared SEM plots 

```{r r2-plots-SEM, include=F, message=F, warning=F}

# DAG 1
r2_dag1 <- list(length=num_models)
pb <- txtProgressBar(min=0, max=num_models, style=3)
for(m in 1:num_models) {
  setTxtProgressBar(pb, m)
  r2_reps <- list(length=num_reps)
  for(rep in 1:num_reps) {
    df <- fread(paste("model_", m, 
                      "/rep_", rep,
                      "/sem_DAG1_r2.csv", sep=""))
    df$DAG <- 1
    df$model <- m
    df$replicate <- rep
    df$scale <- c(1e+4, 1e+5, 1e+6)
    
    r2_reps[[rep]] <- df
  }
  tbl <- do.call(rbind.data.frame, r2_reps)
  r2_dag1[[m]] <- tbl
}
close(pb)

r2_sem_dag1 <- do.call(rbind.data.frame, r2_dag1)

# DAG 2
r2_dag2 <- list(length=num_models)
pb <- txtProgressBar(min=0, max=num_models, style=3)
for(m in 1:num_models) {
  setTxtProgressBar(pb, m)
  r2_reps <- list(length=num_reps)
  for(rep in 1:num_reps) {
    df <- fread(paste("model_", m, 
                      "/rep_", rep, 
                      "/sem_DAG2_r2.csv", sep=""))
    df$DAG <- 2
    df$model <- m
    df$replicate <- rep
    df$scale <- c(1e+4, 1e+5, 1e+6)
    
    r2_reps[[rep]] <- df
  }
  tbl <- do.call(rbind.data.frame, r2_reps)
  r2_dag2[[m]] <- tbl
}
close(pb)

r2_sem_dag2 <- do.call(rbind.data.frame, r2_dag2)

# DAG 3
r2_dag3 <- list(length=num_models)
pb <- txtProgressBar(min=0, max=num_models, style=3)
for(m in 1:num_models) {
  setTxtProgressBar(pb, m)
  r2_reps <- list(length=num_reps)
  for(rep in 1:num_reps) {
    df <- fread(paste("model_", m, 
                      "/rep_", rep,
                      "/sem_DAG3_r2.csv", sep=""))
    df$DAG <- 3
    df$model <- m
    df$replicate <- rep
    df$scale <- c(1e+4, 1e+5, 1e+6)
    
    r2_reps[[rep]] <- df
  }
  tbl <- do.call(rbind.data.frame, r2_reps)
  r2_dag3[[m]] <- tbl
}
close(pb)

r2_sem_dag3 <- do.call(rbind.data.frame, r2_dag3)

# DAG 4
r2_dag4 <- list(length=num_models)
pb <- txtProgressBar(min=0, max=num_models, style=3)
for(m in 1:num_models) {
  setTxtProgressBar(pb, m)
  r2_reps <- list(length=num_reps)
  for(rep in 1:num_reps) {
    df <- fread(paste("model_", m,
                      "/rep_", rep, 
                      "/sem_DAG4_r2.csv", sep=""))
    df$DAG <- 4
    df$model <- m
    df$replicate <- rep
    df$scale <- c(1e+4, 1e+5, 1e+6)
    
    r2_reps[[rep]] <- df
  }
  tbl <- do.call(rbind.data.frame, r2_reps)
  r2_dag4[[m]] <- tbl
}
close(pb)

r2_sem_dag4 <- do.call(rbind.data.frame, r2_dag4)

# putting it together
r2_nonmut_rec <- rbind.data.frame(r2_sem_dag1, r2_sem_dag3) # B, pi endogenous
r2_mut_rec <- rbind.data.frame(r2_sem_dag2, r2_sem_dag4) # u, B, pi endogenous

# true model: mutagenic rec
m_r2_sem_1 <- merge(r2_mut_rec, models, by="model") %>%
              group_by(model, scale, DAG) %>%
              mutate(mean_rsqrd_mut=mean(avg_mut),
                     sd_rsqrd_mut=sd(avg_mut),
                     mean_rsqrd_pi=mean(pi_neu_sim),
                     sd_rsqrd_pi=sd(pi_neu_sim),
                     mean_rsqrd_B=mean(B),
                     sd_rsqrd_B=sd(B)) %>% ungroup %>%
              distinct(model, scale, DAG, .keep_all=TRUE) %>%
              pivot_longer(., cols=c("mean_rsqrd_mut",
                                     "mean_rsqrd_B", 
                                     "mean_rsqrd_pi"),
                           names_to="variable", values_to="mean_rsqrd")

# true model: non-mutagenic rec
m_r2_sem_2 <- merge(r2_nonmut_rec, models, by="model") %>%
              group_by(model, scale, DAG) %>%
              mutate(mean_rsqrd_pi=mean(pi_neu_sim),
                     sd_rsqrd_pi=sd(pi_neu_sim),
                     mean_rsqrd_B=mean(B),
                     sd_rsqrd_B=sd(B)) %>% ungroup %>%
              distinct(model, scale, DAG, .keep_all=TRUE) %>%
              pivot_longer(., cols=c("mean_rsqrd_B", 
                                     "mean_rsqrd_pi"),
                           names_to="variable", values_to="mean_rsqrd")

# adding "indicator" variable for source of pi
m_r2_sem_1$pi_source <- ifelse(m_r2_sem_1$DAG == 2, "neu", "tot")
m_r2_sem_2$pi_source <- ifelse(m_r2_sem_2$DAG == 1, "neu", "tot")

# plotting mutagenic rec data
for(e in unique(m_r2_sem_1$num_exons)) {
  for(ars in unique(m_r2_sem_1$avg_rec_spans)) {
    for(srr in unique(m_r2_sem_1$shapes_rate_r)) {
      for(mru in unique(m_r2_sem_1$mult_r_u)) {
        p1 <- ggplot(data=filter(m_r2_sem_1, 
                                 num_exons==e, 
                                 avg_rec_spans==ars, 
                                 shapes_rate_r==srr, 
                                 mult_r_u==mru),
             aes(x=scale,
                 y=mean_rsqrd, 
                 shape=as.factor(pi_source), 
                 color=variable)) +
             geom_point(size=3) + geom_line() + theme_bw() + 
             facet_grid(shapes_rate_u~avg_mut_spans, labeller=label_both) +
             scale_shape_manual(values=c(0, 1, 2)) +
             scale_color_manual(values=c("brown1", "green4", "purple3")) +
             scale_y_continuous(breaks=pretty_breaks()) +
             scale_x_log10() + 
             labs(title="Variance explained in endogenous variables",
                  x="Scale (window size)", y=expression(paste("R"^"2")),
                  shape="pi", color="Variable") +
             theme(axis.title=element_text(size=16), 
                   axis.text=element_text(size=12), 
                   axis.text.x=element_text(size=12),
                   legend.text=element_text(size=16),
                   legend.title=element_text(size=16),
                   legend.position="bottom")
        
        fout <- paste("SEM/r-sqrd_mutagenic_DAG_exons", e, 
                      "_ars", ars, 
                      "_srr", srr, 
                      "_mru", mru,
                      ".png", sep="")
        save_plot(fout, p1, base_height=10, base_width=18)
      }
    }
  }
}

# plotting non-mutagenic rec data
for(e in unique(m_r2_sem_2$num_exons)) {
  for(ars in unique(m_r2_sem_2$avg_rec_spans)) {
    for(srr in unique(m_r2_sem_2$shapes_rate_r)) {
      for(mru in unique(m_r2_sem_2$mult_r_u)) {
        p1 <- ggplot(data=filter(m_r2_sem_2, 
                                 num_exons==e, 
                                 avg_rec_spans==ars, 
                                 shapes_rate_r==srr, 
                                 mult_r_u==mru),
             aes(x=scale, 
                 y=mean_rsqrd, 
                 shape=as.factor(pi_source), 
                 color=variable)) +
             geom_point(size=3) + geom_line() + theme_bw() + 
             facet_grid(shapes_rate_u~avg_mut_spans, labeller=label_both) +
             scale_shape_manual(values=c(0, 1, 2)) +
             scale_color_manual(values=c("brown1", "purple3")) +
             scale_y_continuous(breaks=pretty_breaks()) +
             scale_x_log10() + 
             labs(title="Variance explained in endogenous variables",
                  x="Scale (window size)", y=expression(paste("R"^"2")),
                  shape="pi", color="Variable") +
             theme(axis.title=element_text(size=16), 
                   axis.text=element_text(size=12), 
                   axis.text.x=element_text(size=12),
                   legend.text=element_text(size=16),
                   legend.title=element_text(size=16),
                   legend.position="bottom")
        
        fout <- paste("SEM/r-sqrd_non-mutagenic_DAG_exons", e, 
                       "_ars", ars, 
                       "_srr", srr, 
                       "_mru", mru,
                       ".png", sep="")
        save_plot(fout, p1, base_height=10, base_width=18)
      }
    }
  }
}
```

Path Coefficients
Assembling the tables

```{r path_coeff-SEM-1, include=F, message=F, warning=F}

# path coefficients under DAG 1
pc_dag1 <- list(length=num_models)
pb <- txtProgressBar(min=0, max=num_models, style=3)
for(m in 1:num_models) {
  setTxtProgressBar(pb, m)
  pc_reps <- list(length=num_reps)
  for(rep in 1:num_reps) {
    df_10kb <- fread(paste("model_", m, 
                           "/rep_", rep, 
                           "/lavaan_DAG1_10kb.csv", sep=""))
    df_10kb$DAG <- 1
    df_10kb$model <- m
    df_10kb$replicate <- rep
    df_10kb$scale <- 1e+4
    
    df_100kb <- fread(paste("model_", m, "/rep_", rep, 
                            "/lavaan_DAG1_100kb.csv", sep=""))
    df_100kb$DAG <- 1
    df_100kb$model <- m
    df_100kb$replicate <- rep
    df_100kb$scale <- 1e+5
    
    df_1Mb <- fread(paste("model_", m, 
                          "/rep_", rep, 
                          "/lavaan_DAG1_1Mb.csv", sep=""))
    df_1Mb$DAG <- 1
    df_1Mb$model <- m
    df_1Mb$replicate <- rep
    df_1Mb$scale <- 1e+6
    
    pc_reps[[rep]] <- rbind.data.frame(df_10kb,
                                       df_100kb,
                                       df_1Mb)
  }
  
  tbl <- do.call(rbind.data.frame, pc_reps)
  pc_dag1[[m]] <- tbl
}
close(pb)

pc_sem_dag1 <- do.call(rbind.data.frame, pc_dag1)
pc_sem_dag1 <- filter(pc_sem_dag1, label != "")

# path coefficients under DAG 2
pc_dag2 <- list(length=num_models)
pb <- txtProgressBar(min=0, max=num_models, style=3)
for(m in 1:num_models) {
  setTxtProgressBar(pb, m)
  pc_reps <- list(length=num_reps)
  for(rep in 1:num_reps) {
    df_10kb <- fread(paste("model_", m,
                           "/rep_", rep, 
                           "/lavaan_DAG2_10kb.csv", sep=""))
    df_10kb$DAG <- 2
    df_10kb$model <- m
    df_10kb$replicate <- rep
    df_10kb$scale <- 1e+4
    
    df_100kb <- fread(paste("model_", m,
                            "/rep_", rep, 
                            "/lavaan_DAG2_100kb.csv", sep=""))
    df_100kb$DAG <- 2
    df_100kb$model <- m
    df_100kb$replicate <- rep
    df_100kb$scale <- 1e+5
    
    df_1Mb <- fread(paste("model_", m,
                          "/rep_", rep, 
                          "/lavaan_DAG2_1Mb.csv", sep=""))
    df_1Mb$DAG <- 2
    df_1Mb$model <- m
    df_1Mb$replicate <- rep
    df_1Mb$scale <- 1e+6
    
    pc_reps[[rep]] <- rbind.data.frame(df_10kb, 
                                       df_100kb, 
                                       df_1Mb)
  }
  
  tbl <- do.call(rbind.data.frame, pc_reps)
  pc_dag2[[m]] <- tbl
}
close(pb)

pc_sem_dag2 <- do.call(rbind.data.frame, pc_dag2)
pc_sem_dag2 <- filter(pc_sem_dag2, label != "")

# path coefficients under DAG 3
pc_dag3 <- list(length=num_models)
pb <- txtProgressBar(min=0, max=num_models, style=3)
for(m in 1:num_models) {
  setTxtProgressBar(pb, m)
  pc_reps <- list(length=num_reps)
  for(rep in 1:num_reps) {
    df_10kb <- fread(paste("model_", m, 
                           "/rep_", rep, 
                           "/lavaan_DAG3_10kb.csv", sep=""))
    df_10kb$DAG <- 3
    df_10kb$model <- m
    df_10kb$replicate <- rep
    df_10kb$scale <- 1e+4
    
    df_100kb <- fread(paste("model_", m,
                            "/rep_", rep, 
                            "/lavaan_DAG3_100kb.csv", sep=""))
    df_100kb$DAG <- 3
    df_100kb$model <- m
    df_100kb$replicate <- rep
    df_100kb$scale <- 1e+5
    
    df_1Mb <- fread(paste("model_", m, 
                          "/rep_", rep, 
                          "/lavaan_DAG3_1Mb.csv", sep=""))
    df_1Mb$DAG <- 3
    df_1Mb$model <- m
    df_1Mb$replicate <- rep
    df_1Mb$scale <- 1e+6
    
    pc_reps[[rep]] <- rbind.data.frame(df_10kb,
                                       df_100kb,
                                       df_1Mb)
  }
  
  tbl <- do.call(rbind.data.frame, pc_reps)
  pc_dag3[[m]] <- tbl
}
close(pb)

pc_sem_dag3 <- do.call(rbind.data.frame, pc_dag3)
pc_sem_dag3 <- filter(pc_sem_dag3, label != "")

# path coefficients under DAG 4
pc_dag4 <- list(length=num_models)
pb <- txtProgressBar(min=0, max=num_models, style=3)
for(m in 1:num_models) {
  setTxtProgressBar(pb, m)
  pc_reps <- list(length=num_reps)
  for(rep in 1:num_reps) {
    df_10kb <- fread(paste("model_", m, 
                           "/rep_", rep, 
                           "/lavaan_DAG4_10kb.csv", sep=""))
    df_10kb$DAG <- 4
    df_10kb$model <- m
    df_10kb$replicate <- rep
    df_10kb$scale <- 1e+4
    
    df_100kb <- fread(paste("model_", m, 
                            "/rep_", rep, 
                            "/lavaan_DAG4_100kb.csv", sep=""))
    df_100kb$DAG <- 4
    df_100kb$model <- m
    df_100kb$replicate <- rep
    df_100kb$scale <- 1e+5
    
    df_1Mb <- fread(paste("model_", m,
                          "/rep_", rep, 
                          "/lavaan_DAG4_1Mb.csv", sep=""))
    df_1Mb$DAG <- 4
    df_1Mb$model <- m
    df_1Mb$replicate <- rep
    df_1Mb$scale <- 1e+6
    
    pc_reps[[rep]] <- rbind.data.frame(df_10kb,
                                       df_100kb, 
                                       df_1Mb)
  }
  
  tbl <- do.call(rbind.data.frame, pc_reps)
  pc_dag4[[m]] <- tbl
}
close(pb)

pc_sem_dag4 <- do.call(rbind.data.frame, pc_dag4)
pc_sem_dag4 <- filter(pc_sem_dag4, label != "")

# putting it together
pc_nonmut_rec <- rbind.data.frame(pc_sem_dag1, pc_sem_dag3) # B, pi endogenous
pc_mut_rec <- rbind.data.frame(pc_sem_dag2, pc_sem_dag4) # u, B, pi endogenous

fwrite(pc_nonmut_rec, "pc_SEM_nonmut_rec.csv")
fwrite(pc_mut_rec, "pc_SEM_mut_rec.csv")
```

Filtering the tables

```{r path_coeff-SEM-2, include=F, message=F, warning=F}

# checking for model convergence in lavaan (NA's in the pvalue table)
m_pval_sem <- fread("m_pval_sem.csv.gz")

# little d-tour to see which models do not converge
m_pval_sem$converged <- ifelse(!is.na(m_pval_sem$pval), T, F) 
p <- ggplot(data=m_pval_sem,
             aes(x=model, 
                y=converged, 
                color=as.factor(shapes_rate_u), 
                shape=as.factor(avg_mut_spans))) +
      geom_point(size=1.5) + theme_bw() + 
      facet_wrap(~scale, ncol=1, labeller=label_both) +
      scale_shape_manual(values=c(0, 1, 2, 3)) +
      scale_color_manual(values=c("brown1", "purple3", "green4", "cyan3")) +
      scale_x_continuous(breaks=c(0, 432, 864, 1296)) + 
      labs(title="Convergence of SEM fitting by ML",
           x="Model ID", y="lavaan converged?",
           color="Shape Rate Mu", shape="Avg. Mut. Span") +
      theme(axis.title=element_text(size=16), 
            axis.text=element_text(size=12), 
            axis.text.x=element_text(size=12),
            legend.text=element_text(size=16),
            legend.title=element_text(size=16),
            legend.position="bottom")
save_plot("SEM/lavaan_convergence_1.png", p, base_height=8, base_width=16)

m_pval_sem$pi_source <- ifelse(m_pval_sem$DAG == 2, "neu", "tot")
m_pval_sem$pi_source <- ifelse(m_pval_sem$DAG == 1, "neu", "tot")
q <- ggplot(data=filter(m_pval_sem, converged==F),
             aes(x=shapes_rate_u, 
                 y=avg_mut_spans, 
                 color=as.factor(scale), 
                 shape=as.factor(pi_source))) +
      geom_point(size=2.5) + theme_bw() + 
      facet_grid(shapes_rate_r~avg_rec_spans, labeller=label_both) +
      scale_shape_manual(values=c(8, 0)) +
      scale_color_manual(values=c("brown1", "purple3", "green4")) +
      scale_x_continuous(breaks=unique(m_pval_sem$shapes_rate_u)) +
      scale_y_log10(breaks=unique(m_pval_sem$avg_mut_spans)) + 
      labs(title="Failure of convergence by lavaan",
           x="Shape=Rate Gamma Mut. dist.", y="Avg Mut. Spans",
           color="Scale", shape="Pi") +
      theme(axis.title=element_text(size=16), 
            axis.text=element_text(size=12), 
            axis.text.x=element_text(size=12),
            legend.text=element_text(size=16),
            legend.title=element_text(size=16),
            legend.position="bottom")
save_plot("SEM/lavaan_convergence_2.png", q, base_height=8, base_width=16)

pc_nonmut_rec <- fread("pc_SEM_nonmut_rec.csv")
pc_mut_rec <- fread("pc_SEM_mut_rec.csv")

pc_nonmut_rec <- left_join(pc_nonmut_rec,
                           m_pval_sem,
                           by=c("model", "replicate", "scale", "DAG"))

pc_mut_rec <- left_join(pc_mut_rec,
                        m_pval_sem,
                        by=c("model", "replicate", "scale", "DAG"))

########################################################
# focusing on pi as the outcome
# filtering out NA's due to lack of lavaan model convergence
pc_nonmut_rec_pi <- filter(na.omit(pc_nonmut_rec),
                           lhs=="pi_neu_sim" | 
                           lhs=="sim_pi" |
                           lhs=="mu_B_pi")

pc_mut_rec_pi <- filter(na.omit(pc_mut_rec),
                        lhs=="pi_neu_sim" |
                        lhs=="sim_pi" | 
                        lhs=="mu_B_pi")

# combinig pvalues across replicates to plot significance later on
pc_nonmut_rec_pi <- pc_nonmut_rec_pi %>% group_by(model, rhs, scale, DAG) %>%
                    mutate(fisher_c=combine.test(pvalue, 1:10, "fisher", na.rm=T)) 

pc_mut_rec_pi <- pc_mut_rec_pi %>% group_by(model, rhs, scale, DAG) %>%
                 mutate(fisher_c=combine.test(pvalue, 1:10, "fisher", na.rm=T)) 

# true model: mutagenic rec
m_pc_sem_p_mutrec <- pc_mut_rec_pi %>%
                     group_by(model, rhs, scale, DAG) %>%
                     mutate(mean_pc=mean(est),
                            sd_pc=sd(est)) %>% ungroup %>%
                      distinct(model, rhs, scale, DAG, .keep_all=TRUE) 

# true model: non-mutagenic rec
m_pc_sem_p_nonmutrec <- pc_nonmut_rec_pi %>%
                        group_by(model, rhs, scale, DAG) %>%
                        mutate(mean_pc=mean(est),
                               sd_pc=sd(est)) %>% ungroup %>%
                        distinct(model, rhs, scale, DAG, .keep_all=TRUE) 

# adding "indicator" variable for source of pi
m_pc_sem_p_mutrec$pi_source <- ifelse(m_pc_sem_p_mutrec$DAG == 2, 
                                      "neu", "tot")
m_pc_sem_p_nonmutrec$pi_source <- ifelse(m_pc_sem_p_nonmutrec$DAG == 1, 
                                         "neu", "tot")

fwrite(m_pc_sem_p_mutrec, "pc_sem_pi_mutrec.csv.gz")
fwrite(m_pc_sem_p_nonmutrec, "pc_sem_pi_nonmutrec.csv.gz")

########################################################
# focusing on B as the outcome
# filtering out NA's due to lack of lavaan model convergence
pc_nonmut_rec_B <- filter(na.omit(pc_nonmut_rec),
                          lhs=="B")
pc_mut_rec_B <- filter(na.omit(pc_mut_rec),
                       lhs=="B")

# combinig pvalues across replicates to plot significance later on
pc_nonmut_rec_B <- pc_nonmut_rec_B %>% group_by(model, rhs, scale, DAG) %>%
                   mutate(fisher_c=combine.test(pvalue, 1:10, "fisher", na.rm=T)) 

pc_mut_rec_B <- pc_mut_rec_B %>% group_by(model, rhs, scale, DAG) %>%
                mutate(fisher_c=combine.test(pvalue, 1:10, "fisher", na.rm=T)) 

# true model: mutagenic rec
m_pc_sem_B_mutrec <- pc_mut_rec_B %>%
                     group_by(model, rhs, scale, DAG) %>%
                     mutate(mean_pc=mean(est),
                            sd_pc=sd(est)) %>% ungroup %>%
                     distinct(model, rhs, scale, DAG, .keep_all=TRUE) 

# true model: non-mutagenic rec
m_pc_sem_B_nonmutrec <- pc_nonmut_rec_B %>%
                        group_by(model, rhs, scale, DAG) %>%
                        mutate(mean_pc=mean(est),
                               sd_pc=sd(est)) %>% ungroup %>%
                        distinct(model, rhs, scale, DAG, .keep_all=TRUE) 

fwrite(m_pc_sem_B_mutrec, "pc_sem_B_mutrec.csv.gz")
fwrite(m_pc_sem_B_nonmutrec, "pc_sem_B_nonmutrec.csv.gz")

########################################################
# focusing on the indirect effects of rec on pi 
# (mutagenic rec only, assessing the Przeworski's regime)
# filtering out NA's due to lack of lavaan model convergence
pc_rec_pi <- filter(na.omit(pc_mut_rec), 
                    lhs=="rec_B_pi" | lhs=="rec_mu_pi")

# combinig pvalues across replicates to plot significance later on
pc_rec_pi <- pc_rec_pi %>% group_by(model, rhs, scale, DAG) %>%
             mutate(fisher_c=combine.test(pvalue, 1:10, "fisher", na.rm=T)) 

# true model: mutagenic rec
m_pc_sem_rec_pi <- pc_rec_pi %>%
                   group_by(model, rhs, scale, DAG) %>%
                   mutate(mean_pc=mean(est),
                          sd_pc=sd(est)) %>% ungroup %>%
                   distinct(model, rhs, scale, DAG, .keep_all=TRUE) 

m_pc_sem_rec_pi$pi_source <- ifelse(m_pc_sem_rec_pi$DAG == 2, "neu", "tot")

fwrite(m_pc_sem_rec_pi, "pc_sem_recpi.csv.gz")
```

Plotting path coefficients to pi

```{r pc-plots-SEM-pi, include=F, message=F, warning=F}

m_pc_sem_p_mutrec <- fread("pc_sem_pi_mutrec.csv.gz")
m_pc_sem_p_nonmutrec <- fread("pc_sem_pi_nonmutrec.csv.gz")
cv_models <- fread("cv_models.csv")

# mutagenic rec, DAG 2, B
p1 <- ggplot(data=filter(m_pc_sem_p_mutrec, 
                         fisher_c < 0.01,
                         rhs=="B",
                         lhs=="pi_neu_sim"), # DAG 2
            aes(x=model, 
                y=mean_pc, 
                color=as.factor(shapes_rate_u), 
                shape=as.factor(avg_mut_spans))) +
      geom_point(size=1.5) + theme_bw() + 
      geom_hline(aes(yintercept=0), 
                 linetype="dashed", color="grey1") +
      facet_wrap(~scale, ncol=1, labeller=label_both) +
      scale_shape_manual(values=c(0, 1, 2, 3)) +
      scale_color_manual(values=c("brown1", "purple3", "green4", "cyan3")) +
      scale_y_continuous(breaks=pretty_breaks()) +
      scale_x_continuous(breaks=c(0, 432, 864, 1296)) + 
      labs(title=paste("Structural Equation Modeling, DAG 2 (mutagenic rec.)",
                       "average PC across 10 replicates"),
           x="Model ID", y="Path Coefficient B->pi_neu",
           color="Shape Rate Mu", shape="Avg. Mut. Span") +
      theme(axis.title=element_text(size=16), 
            axis.text=element_text(size=12), 
            axis.text.x=element_text(size=12),
            legend.text=element_text(size=16),
            legend.title=element_text(size=16),
            legend.position="bottom")
save_plot("SEM/pc_B_pi_neu_sim_DAG2.png", p1, base_height=10, base_width=16)

# mutagenic rec, DAG 4, B
p2 <- ggplot(data=filter(m_pc_sem_p_mutrec, 
                         fisher_c < 0.01,
                         rhs=="B",
                         lhs=="sim_pi"), # DAG 4
            aes(x=model, 
                y=mean_pc, 
                color=as.factor(shapes_rate_u), 
                shape=as.factor(avg_mut_spans))) +
      geom_point(size=1.5) + theme_bw() + 
      geom_hline(aes(yintercept=0), 
                 linetype="dashed", color="grey1") +
      facet_wrap(~scale, ncol=1, labeller=label_both) +
      scale_shape_manual(values=c(0, 1, 2, 3)) +
      scale_color_manual(values=c("brown1", "purple3", "green4", "cyan3")) +
      scale_y_continuous(breaks=pretty_breaks()) +
      scale_x_continuous(breaks=c(0, 432, 864, 1296)) + 
      labs(title=paste("Structural Equation Modeling, DAG 4 (mutagenic rec.)",
                       "average PC across 10 replicates"),
           x="Model ID", y="Path Coefficient B->pi_tot",
           color="Shape Rate Mu", shape="Avg. Mut. Span") +
      theme(axis.title=element_text(size=16), 
            axis.text=element_text(size=12), 
            axis.text.x=element_text(size=12),
            legend.text=element_text(size=16),
            legend.title=element_text(size=16),
            legend.position="bottom")
save_plot("SEM/pc_B_pi_tot_sim_DAG4.png", p2, base_height=10, base_width=16)

p3 <- ggplot(data=filter(m_pc_sem_p_nonmutrec, 
                         fisher_c < 0.01,
                         rhs=="B",
                         lhs=="sim_pi"), # DAG 3
            aes(x=model, 
                y=mean_pc, 
                color=as.factor(shapes_rate_u), 
                shape=as.factor(avg_mut_spans))) +
      geom_point(size=1.5) + theme_bw() + 
      geom_hline(aes(yintercept=0), 
                 linetype="dashed", color="grey1") +
      facet_wrap(~scale, ncol=1, labeller=label_both) +
      scale_shape_manual(values=c(0, 1, 2, 3)) +
      scale_color_manual(values=c("brown1", "purple3", "green4", "cyan3")) +
      scale_y_continuous(breaks=pretty_breaks()) +
      scale_x_continuous(breaks=c(0, 432, 864, 1296)) + 
      labs(title=paste("Structural Equation Modeling, DAG 3 (non-mutagenic rec.)",
                       "average PC across 10 replicates"),
           x="Model ID", y="Path Coefficient B->pi_tot",
           color="Shape Rate Mu", shape="Avg. Mut. Span") +
      theme(axis.title=element_text(size=16), 
            axis.text=element_text(size=12), 
            axis.text.x=element_text(size=12),
            legend.text=element_text(size=16),
            legend.title=element_text(size=16),
            legend.position="bottom")
save_plot("SEM/pc_B_pi_tot_sim_DAG3.png", p1, base_height=10, base_width=16)

# mutagenic rec, DAG 1, B
p4 <- ggplot(data=filter(m_pc_sem_p_mutrec, 
                         fisher_c < 0.01,
                         rhs=="B",
                         lhs=="sim_pi"), # DAG 1
            aes(x=model, 
                y=mean_pc, 
                color=as.factor(shapes_rate_u), 
                shape=as.factor(avg_mut_spans))) +
      geom_point(size=1.5) + theme_bw() + 
      geom_hline(aes(yintercept=0), 
                 linetype="dashed", color="grey1") +
      facet_wrap(~scale, ncol=1, labeller=label_both) +
      scale_shape_manual(values=c(0, 1, 2, 3)) +
      scale_color_manual(values=c("brown1", "purple3", "green4", "cyan3")) +
      scale_y_continuous(breaks=pretty_breaks()) +
      scale_x_continuous(breaks=c(0, 432, 864, 1296)) + 
      labs(title=paste("Structural Equation Modeling, DAG 1 (mutagenic rec.)",
                       "average PC across 10 replicates"),
           x="Model ID", y="Path Coefficient B->pi_tot",
           color="Shape Rate Mu", shape="Avg. Mut. Span") +
      theme(axis.title=element_text(size=16), 
            axis.text=element_text(size=12), 
            axis.text.x=element_text(size=12),
            legend.text=element_text(size=16),
            legend.title=element_text(size=16),
            legend.position="bottom")
save_plot("SEM/pc_B_pi_tot_sim_DAG1.png", p4, base_height=10, base_width=16)

# mutagenic rec, DAG 2, mu
p5 <- ggplot(data=filter(m_pc_sem_p_mutrec, 
                         fisher_c < 0.01,
                         rhs=="avg_mut",
                         lhs=="pi_neu_sim"), # DAG 2
            aes(x=model, 
                y=mean_pc, 
                color=as.factor(shapes_rate_u), 
                shape=as.factor(avg_mut_spans))) +
      geom_point(size=1.5) + theme_bw() + 
      geom_hline(aes(yintercept=0), 
                 linetype="dashed", color="grey1") +
      facet_wrap(~scale, ncol=1, labeller=label_both) +
      scale_shape_manual(values=c(0, 1, 2, 3)) +
      scale_color_manual(values=c("brown1", "purple3", "green4", "cyan3")) +
      scale_y_continuous(breaks=pretty_breaks()) +
      scale_x_continuous(breaks=c(0, 432, 864, 1296)) + 
      labs(title=paste("Structural Equation Modeling, DAG 2 (mutagenic rec.)",
                       "average PC across 10 replicates"),
           x="Model ID", y="Path Coefficient mu->pi_neu",
           color="Shape Rate Mu", shape="Avg. Mut. Span") +
      theme(axis.title=element_text(size=16), 
            axis.text=element_text(size=12), 
            axis.text.x=element_text(size=12),
            legend.text=element_text(size=16),
            legend.title=element_text(size=16),
            legend.position="bottom")
save_plot("SEM/pc_mu_pi_neu_sim_DAG2.png", p5, base_height=10, base_width=16)

# mutagenic rec, DAG 4, mu
p6 <- ggplot(data=filter(m_pc_sem_p_mutrec, 
                         fisher_c < 0.01,
                         rhs=="avg_mut",
                         lhs=="sim_pi"), # DAG 4
            aes(x=model, 
                y=mean_pc, 
                color=as.factor(shapes_rate_u), 
                shape=as.factor(avg_mut_spans))) +
      geom_point(size=1.5) + theme_bw() + 
      geom_hline(aes(yintercept=0), 
                 linetype="dashed", color="grey1") +
      facet_wrap(~scale, ncol=1, labeller=label_both) +
      scale_shape_manual(values=c(0, 1, 2, 3)) +
      scale_color_manual(values=c("brown1", "purple3", "green4", "cyan3")) +
      scale_y_continuous(breaks=pretty_breaks()) +
      scale_x_continuous(breaks=c(0, 432, 864, 1296)) + 
      labs(title=paste("Structural Equation Modeling, DAG 4 (mutagenic rec.)",
                       "average PC across 10 replicates"),
           x="Model ID", y="Path Coefficient mu->pi_tot",
           color="Shape Rate Mu", shape="Avg. Mut. Span") +
      theme(axis.title=element_text(size=16), 
            axis.text=element_text(size=12), 
            axis.text.x=element_text(size=12),
            legend.text=element_text(size=16),
            legend.title=element_text(size=16),
            legend.position="bottom")
save_plot("SEM/pc_mu_pi_tot_sim_DAG4.png", p6, base_height=10, base_width=16)

# non-mutagenic rec, DAG 1, mu
p7 <- ggplot(data=filter(m_pc_sem_p_nonmutrec,
                         fisher_c < 0.01,
                         rhs=="avg_mut",
                         lhs=="pi_neu_sim"), # DAG 1
            aes(x=model, 
                y=mean_pc, 
                color=as.factor(shapes_rate_u), 
                shape=as.factor(avg_mut_spans))) +
      geom_point(size=1.5) + theme_bw() + 
      geom_hline(aes(yintercept=0), 
                 linetype="dashed", color="grey1") +
      facet_wrap(~scale, ncol=1, labeller=label_both) +
      scale_shape_manual(values=c(0, 1, 2, 3)) +
      scale_color_manual(values=c("brown1", "purple3", "green4", "cyan3")) +
      scale_y_continuous(breaks=pretty_breaks()) +
      scale_x_continuous(breaks=c(0, 432, 864, 1296)) + 
      labs(title=paste("Structural Equation Modeling, DAG 1 (non-mutagenic rec.)",
                       "average PC across 10 replicates"),
           x="Model ID", y="Path Coefficient mu->pi_neu",
           color="Shape Rate Mu", shape="Avg. Mut. Span") +
      theme(axis.title=element_text(size=16), 
            axis.text=element_text(size=12), 
            axis.text.x=element_text(size=12),
            legend.text=element_text(size=16),
            legend.title=element_text(size=16),
            legend.position="bottom")
save_plot("SEM/pc_mu_pi_neu_sim_DAG1.png", p7, base_height=10, base_width=16)

# non-mutagenic rec, DAG 3, mu
p8 <- ggplot(data=filter(m_pc_sem_p_nonmutrec,
                         fisher_c < 0.01,
                         rhs=="avg_mut",
                         lhs=="sim_pi"), # DAG 3
            aes(x=model, 
                y=mean_pc, 
                color=as.factor(shapes_rate_u), 
                shape=as.factor(avg_mut_spans))) +
      geom_point(size=1.5) + theme_bw() + 
      geom_hline(aes(yintercept=0), 
                 linetype="dashed", color="grey1") +
      facet_wrap(~scale, ncol=1, labeller=label_both) +
      scale_shape_manual(values=c(0, 1, 2, 3)) +
      scale_color_manual(values=c("brown1", "purple3", "green4", "cyan3")) +
      scale_y_continuous(breaks=pretty_breaks()) +
      scale_x_continuous(breaks=c(0, 432, 864, 1296)) + 
      labs(title=paste("Structural Equation Modeling, DAG 3 (non-mutagenic rec.)",
                       "average PC across 10 replicates"),
           x="Model ID", y="Path Coefficient mu->pi_tot",
           color="Shape Rate Mu", shape="Avg. Mut. Span") +
      theme(axis.title=element_text(size=16), 
            axis.text=element_text(size=12), 
            axis.text.x=element_text(size=12),
            legend.text=element_text(size=16),
            legend.title=element_text(size=16),
            legend.position="bottom")
save_plot("SEM/pc_mu_pi_tot_sim_DAG3.png", p8, base_height=10, base_width=16)

# non-mutagenic rec, DAG 1, mu_B_pi
p9 <- ggplot(data=filter(m_pc_sem_p_nonmutrec,
                         fisher_c < 0.01, 
                         lhs=="mu_B_pi",
                         DAG==1),
            aes(x=model, 
                y=mean_pc, 
                color=as.factor(shapes_rate_u), 
                shape=as.factor(avg_mut_spans))) +
      geom_point(size=1.5) + theme_bw() + 
      geom_hline(aes(yintercept=0), 
                 linetype="dashed", color="grey1") +
      facet_wrap(~scale, ncol=1, labeller=label_both) +
      scale_shape_manual(values=c(0, 1, 2, 3)) +
      scale_color_manual(values=c("brown1", "purple3", "green4", "cyan3")) +
      scale_y_continuous(breaks=pretty_breaks()) +
      scale_x_continuous(breaks=c(0, 432, 864, 1296)) + 
      labs(title=paste("Structural Equation Modeling, DAG 1 (non-mutagenic rec.)",
                       "average PC across 10 replicates"),
           x="Model ID", y="Path Coefficient mu->B->pi_neu",
           color="Shape Rate Mu", shape="Avg. Mut. Span") +
      theme(axis.title=element_text(size=16), 
            axis.text=element_text(size=12), 
            axis.text.x=element_text(size=12),
            legend.text=element_text(size=16),
            legend.title=element_text(size=16),
            legend.position="bottom")
save_plot("SEM/pc_mu_B_pi_neu_sim_DAG1.png", p9, base_height=10, base_width=16)

# non-mutagenic rec, DAG 3, mu_B_pi
p10 <- ggplot(data=filter(m_pc_sem_p_nonmutrec, 
                          fisher_c < 0.01,
                          lhs=="mu_B_pi",
                          DAG==3),
            aes(x=model, 
                y=mean_pc, 
                color=as.factor(shapes_rate_u), 
                shape=as.factor(avg_mut_spans))) +
      geom_point(size=1.5) + theme_bw() + 
      geom_hline(aes(yintercept=0), 
                 linetype="dashed", color="grey1") +
      facet_wrap(~scale, ncol=1, labeller=label_both) +
      scale_shape_manual(values=c(0, 1, 2, 3)) +
      scale_color_manual(values=c("brown1", "purple3", "green4", "cyan3")) +
      scale_y_continuous(breaks=pretty_breaks()) +
      scale_x_continuous(breaks=c(0, 432, 864, 1296)) + 
      labs(title=paste("Structural Equation Modeling, DAG 3 (non-mutagenic rec.)",
                       "average PC across 10 replicates"),
           x="Model ID", y="Path Coefficient mu->B->pi_tot",
           color="Shape Rate Mu", shape="Avg. Mut. Span") +
      theme(axis.title=element_text(size=16), 
            axis.text=element_text(size=12), 
            axis.text.x=element_text(size=12),
            legend.text=element_text(size=16),
            legend.title=element_text(size=16),
            legend.position="bottom")
save_plot("SEM/pc_mu_B_pi_tot_sim_DAG3.png", p9, base_height=10, base_width=16)

# mutagenic rec, DAG 2, mu_B_pi
p11 <- ggplot(data=filter(m_pc_sem_p_mutrec,
                          fisher_c < 0.01, 
                          lhs=="mu_B_pi",
                          DAG==2),
            aes(x=model, 
                y=mean_pc, 
                color=as.factor(shapes_rate_u), 
                shape=as.factor(avg_mut_spans))) +
      geom_point(size=1.5) + theme_bw() + 
      geom_hline(aes(yintercept=0), 
                 linetype="dashed", color="grey1") +
      facet_wrap(~scale, ncol=1, labeller=label_both) +
      scale_shape_manual(values=c(0, 1, 2, 3)) +
      scale_color_manual(values=c("brown1", "purple3", "green4", "cyan3")) +
      scale_y_continuous(breaks=pretty_breaks()) +
      scale_x_continuous(breaks=c(0, 432, 864, 1296)) + 
      labs(title=paste("Structural Equation Modeling, DAG 2 (mutagenic rec.)",
                       "average PC across 10 replicates"),
           x="Model ID", y="Path Coefficient mu->B->pi_neu",
           color="Shape Rate Mu", shape="Avg. Mut. Span") +
      theme(axis.title=element_text(size=16), 
            axis.text=element_text(size=12), 
            axis.text.x=element_text(size=12),
            legend.text=element_text(size=16),
            legend.title=element_text(size=16),
            legend.position="bottom")
save_plot("SEM/pc_mu_B_pi_neu_sim_DAG2.png", p11, base_height=10, base_width=16)

# mutagenic rec, DAG 4, mu_B_pi
p12 <- ggplot(data=filter(m_pc_sem_p_mutrec, 
                          fisher_c < 0.01,
                          lhs=="mu_B_pi",
                          DAG==4),
            aes(x=model, 
                y=mean_pc, 
                color=as.factor(shapes_rate_u), 
                shape=as.factor(avg_mut_spans))) +
      geom_point(size=1.5) + theme_bw() + 
      geom_hline(aes(yintercept=0), 
                 linetype="dashed", color="grey1") +
      facet_wrap(~scale, ncol=1, labeller=label_both) +
      scale_shape_manual(values=c(0, 1, 2, 3)) +
      scale_color_manual(values=c("brown1", "purple3", "green4", "cyan3")) +
      scale_y_continuous(breaks=pretty_breaks()) +
      scale_x_continuous(breaks=c(0, 432, 864, 1296)) + 
      labs(title=paste("Structural Equation Modeling, DAG 4 (mutagenic rec.)",
                       "average PC across 10 replicates"),
           x="Model ID", y="Path Coefficient mu->B->pi_tot",
           color="Shape Rate Mu", shape="Avg. Mut. Span") +
      theme(axis.title=element_text(size=16), 
            axis.text=element_text(size=12), 
            axis.text.x=element_text(size=12),
            legend.text=element_text(size=16),
            legend.title=element_text(size=16),
            legend.position="bottom")
save_plot("SEM/pc_mu_B_pi_tot_sim_DAG4.png", p12, base_height=10, base_width=16)

# mutagenic rec DAGs
for(e in unique(m_pc_sem_p_mutrec$num_exons)) {
  for(ars in unique(m_pc_sem_p_mutrec$avg_rec_spans)) {
    for(srr in unique(m_pc_sem_p_mutrec$shapes_rate_r)) {
      for(mru in unique(m_pc_sem_p_mutrec$mult_r_u)) {
        p1 <- ggplot(data=filter(m_pc_sem_p_mutrec, 
                                 fisher_c < 0.01,
                                 num_exons==e, 
                                 avg_rec_spans==ars, 
                                 shapes_rate_r==srr, 
                                 mult_r_u==mru),
             aes(x=scale, 
                 y=mean_pc, 
                 shape=as.factor(pi_source), 
                 color=rhs,
                 label=ifelse(fisher_c < 0.05, "*", ""))) +
             geom_text(vjust=0.8, size=10) +
             geom_point(size=3) + geom_line() + theme_bw() + 
             geom_hline(aes(yintercept=0), linetype="dashed", color="grey1") +
             facet_grid(shapes_rate_u~avg_mut_spans, labeller=label_both) +
             scale_shape_manual(values=c(0, 1, 2)) +
             scale_color_manual(values=c("brown1", "green4", "purple3"),
                                labels=c(expression(mu), "B", "u>B>pi")) +
             scale_x_log10() + scale_y_continuous(breaks=pretty_breaks()) +
             labs(title="Path coefficients (effects on pi, mutagenic DAG)",
                  x="Scale (window size)", y="Path Coefficient",
                  shape="pi", color="Variable") +
             theme(axis.title=element_text(size=16), 
                   axis.text=element_text(size=12), 
                   axis.text.x=element_text(size=12),
                   legend.text=element_text(size=16),
                   legend.title=element_text(size=16),
                   legend.position="bottom")
        
        fout <- paste("SEM/pcoeff_pi_mutagenic_DAG_exons", e, 
                      "_ars", ars, 
                      "_srr", srr, 
                      "_mru", mru,
                      ".png", sep="")
        save_plot(fout, p1, base_height=10, base_width=18)
      }
    }
  }
}

# non-mutagenic rec DAGs
for(e in unique(m_pc_sem_p_nonmutrec$num_exons)) {
  for(ars in unique(m_pc_sem_p_nonmutrec$avg_rec_spans)) {
    for(srr in unique(m_pc_sem_p_nonmutrec$shapes_rate_r)) {
      for(mru in unique(m_pc_sem_p_nonmutrec$mult_r_u)) {
        p1 <- ggplot(data=filter(m_pc_sem_p_nonmutrec, 
                                 num_exons==e, 
                                 avg_rec_spans==ars, 
                                 shapes_rate_r==srr, 
                                 mult_r_u==mru),
             aes(x=scale, 
                 y=mean_pc, 
                 shape=as.factor(pi_source), 
                 color=rhs)) +
             geom_point(size=3) + geom_line() + theme_bw() + 
             geom_hline(aes(yintercept=0), linetype="dashed", color="grey1") +
             facet_grid(shapes_rate_u~avg_mut_spans, labeller=label_both) +
             scale_shape_manual(values=c(0, 1, 2)) +
             scale_color_manual(values=c("brown1", "green4", "purple3"),
                                labels=c(expression(mu), "B", "u>B>pi")) +
             scale_y_continuous(breaks=pretty_breaks()) +
             scale_x_log10() + 
             labs(title=paste("Path coefficients",
                              "(effects on pi, non-mutagenic DAG)",
                              "average PC across 10 replicates"),
                  x="Scale (window size)", y="Path Coefficient",
                  shape="pi", color="Variable") +
             theme(axis.title=element_text(size=16), 
                   axis.text=element_text(size=12), 
                   axis.text.x=element_text(size=12),
                   legend.text=element_text(size=16),
                   legend.title=element_text(size=16),
                   legend.position="bottom")
        
        fout <- paste("SEM/pcoeff_pi_non-mutagenic_DAG_exons", e, 
                       "_ars", ars, 
                       "_srr", srr, 
                       "_mru", mru,
                       ".png", sep="")
        save_plot(fout, p1, base_height=10, base_width=18)
      }
    }
  }
}

# adding ratio of CVs
pc_sem_mutrec_cv_models <- left_join(m_pc_sem_p_mutrec, cv_models)

for(e in unique(pc_sem_mutrec_cv_models$num_exons)) {
  for(ars in unique(pc_sem_mutrec_cv_models$avg_rec_spans)) {
    for(srr in unique(pc_sem_mutrec_cv_models$shapes_rate_r)) {
      for(mru in unique(pc_sem_mutrec_cv_models$mult_r_u)) {
        p <- ggplot(data=filter(pc_sem_mutrec_cv_models,
                                num_exons==e, 
                                avg_rec_spans==ars, 
                                shapes_rate_r==srr, 
                                mult_r_u==mru),
                     aes(x=cv_u_cv_B, 
                         y=mean_pc, 
                         color=rhs,
                         shape=pi_source,
                         label=ifelse(fisher_c < 0.05, "*", ""))) +
          geom_text(vjust=0.8, size=10) +
          geom_line() + geom_point(size=4) + theme_bw() + 
          facet_grid(avg_mut_spans~scale, labeller=label_both) +
          geom_hline(aes(yintercept=0), linetype="dashed", color="grey1") +
          scale_shape_manual(values=c(0, 1, 2), name="pi")+
          scale_color_discrete(name=NULL, labels=c("u", "B", "u>B>pi"), 
                               type=c("brown1", "green4", "purple3")) +
          scale_x_continuous(breaks=c(1, 10, 100), trans="log10") +
          scale_y_continuous(breaks=pretty_breaks()) +
          labs(title=paste("Path coefficients",
                           "(effects on pi, non-mutagenic DAG)",
                           "average PC across 10 replicates"),
               x="CV(u) / CV(B)", y=" Mean path coefficient") +
          theme(axis.title=element_text(size=16), 
                axis.text=element_text(size=12), 
                axis.text.x=element_text(size=12),
                legend.text=element_text(size=16),
                legend.title=element_text(size=16),
                legend.position="bottom")
        
        fout <- paste("SEM/sem_pc_cvs_exons", e, 
                      "_ars", ars, 
                      "_srr", srr, 
                      "_mru", mru,
                      ".png", sep="")
        save_plot(fout, p, base_height=10, base_width=18)
      }
    }
  }
}
                  
# getting ratios of path coefficients
df <- pivot_wider(dplyr::select(pc_sem_mutrec_cv_models, # mutagenic rec models
                                -c(model,
                                   lhs,
                                   op, 
                                   rhs,
                                   est,
                                   z,
                                   se,
                                   pvalue,
                                   ci.lower,
                                   ci.upper,
                                   fisher_c,
                                   sd_pc,
                                   replicate)),
                  names_from=label, 
                  values_from=mean_pc)                  

names(df)[(ncol(df)-2):ncol(df)] <- c("B", "mut", "mu_B_pi")
df$ratio <- df$mut / df$B
  
# plotting ratios of path coefficients (standardized log-transformed variables)
p <- ggplot(data=filter(df, DAG==2), # neutral pi
                     aes(x=cv_u_cv_B, 
                         y=ratio, 
                         color=as.factor(100*num_exons*exon_lengths/L),
                         shape=as.factor(scale))) +
          geom_line() + geom_point(size=4) + theme_bw() + 
          geom_hline(yintercept=1, linetype="dashed", color="grey1") +
          geom_vline(xintercept=1, linetype="dashed", color="grey1") +
          facet_grid(shapes_rate_r~avg_rec_spans, labeller=label_both) +
          scale_shape_manual(values=c(0, 1, 2, 3), name="Bin Size")+
          scale_color_discrete(name="Exon %",
                               type=c("brown1", "purple3", "green4")) +
          scale_x_continuous(breaks=c(0.25, 1, 10, 50), trans="log10") +
          scale_y_continuous(breaks=c(0.1, 1, 10), trans="log10") +
          labs(title="Ratios of SEM path coefficients, DAG 2 (mut. rec., Pi Neu.)",
               x="CV(u) / CV(B)", y="PC(u) / PC(B)") +
          theme(axis.title=element_text(size=16), 
                axis.text=element_text(size=12), 
                axis.text.x=element_text(size=12),
                legend.text=element_text(size=16),
                legend.title=element_text(size=16),
                legend.position="bottom")
save_plot("SEM/ratios_sem_pc_cv.png", p, base_height=12, base_width=18)
```

Plotting path coefficients to B 

```{r pc-plots-SEM-B, include=F, message=F, warning=F}

m_pc_sem_B_mutrec <- fread("pc_sem_B_mutrec.csv.gz")
m_pc_sem_B_nonmutrec <- fread("pc_sem_B_nonmutrec.csv.gz")

# mutagenic rec, DAGs 2 (pi_neu) and 4 (pi_tot)
p1 <- ggplot(data=filter(m_pc_sem_B_mutrec, 
                         fisher_c < 0.01),
            aes(x=model, 
                y=mean_pc, 
                color=as.factor(shapes_rate_u), 
                shape=rhs)) +
      geom_point(size=1.5) + theme_bw() + 
      geom_hline(aes(yintercept=0), 
                 linetype="dashed", color="grey1") +
      facet_grid(scale~DAG, labeller=label_both) +
      scale_shape_manual(values=c(0, 1, 2)) +
      scale_color_manual(values=c("brown1", "purple3", "green4", "cyan3")) +
      scale_y_continuous(breaks=pretty_breaks()) +
      scale_x_continuous(breaks=c(0, 432, 864, 1296)) + 
      labs(title=paste("Structural Equation Modeling, DAG 2 (mutagenic rec.)",
                       "average PC across 10 replicates"),
           x="Model ID", y="Path Coefficient to B",
           color="Shape Rate Mu", shape="Var") +
      theme(axis.title=element_text(size=16), 
            axis.text=element_text(size=12), 
            axis.text.x=element_text(size=12),
            legend.text=element_text(size=16),
            legend.title=element_text(size=16),
            legend.position="bottom")
save_plot("SEM/pc_to_B_nonmutrec_DAGs.png", p1, base_height=10, base_width=16)

# non-mutagenic rec, DAGs 1 (pi_neu) and 3 (pi_tot)
q1 <- ggplot(data=filter(m_pc_sem_B_nonmutrec,
                         fisher_c < 0.01),
            aes(x=model, 
                y=mean_pc, 
                color=as.factor(shapes_rate_u), 
                shape=rhs)) +
      geom_point(size=1.5) + theme_bw() + 
      geom_hline(aes(yintercept=0), 
                 linetype="dashed", color="grey1") +
      facet_grid(scale~DAG, labeller=label_both) +
      scale_shape_manual(values=c(0, 1, 2)) +
      scale_color_manual(values=c("brown1", "purple3", "green4", "cyan3")) +
      scale_y_continuous(breaks=pretty_breaks()) +
      scale_x_continuous(breaks=c(0, 432, 864, 1296)) + 
      labs(title=paste("Structural Equation Modeling, DAG 4 (non-mutagenic rec.)",
                       "average PC across 10 replicates"),
           x="Model ID", y="Path Coefficient to B",
           color="Shape Rate Mu", shape="Var") +
      theme(axis.title=element_text(size=16), 
            axis.text=element_text(size=12), 
            axis.text.x=element_text(size=12),
            legend.text=element_text(size=16),
            legend.title=element_text(size=16),
            legend.position="bottom")
save_plot("SEM/pc_to_B_mutrec_DAGs.png", q1, base_height=10, base_width=16)

# effect on B is independent from source pi, keep total only
m_pc_sem_B <- rbind.data.frame(m_pc_sem_B_mutrec, m_pc_sem_B_nonmutrec)
m_pc_sem_B <- filter(m_pc_sem_B, DAG==3 | DAG==4) 
m_pc_sem_B$mutagenic_rec <- ifelse(m_pc_sem_B$DAG == 4, TRUE, FALSE)

# plotting mutagenic rec data
for(e in unique(m_pc_sem_B$num_exons)) {
  for(ars in unique(m_pc_sem_B$avg_rec_spans)) {
    for(srr in unique(m_pc_sem_B$shapes_rate_r)) {
      for(mru in unique(m_pc_sem_B$mult_r_u)) {
        p1 <- ggplot(data=filter(m_pc_sem_B, 
                                 num_exons==e, 
                                 avg_rec_spans==ars, 
                                 shapes_rate_r==srr, 
                                 mult_r_u==mru),
             aes(x=scale,
                 y=mean_pc,
                 color=rhs, 
                 shape=mutagenic_rec,
                 label=ifelse(fisher_c < 0.05, "*", ""))) +
             geom_text(vjust=0.8, size=10) +
             geom_point(size=3) + geom_line() + theme_bw() + 
             geom_hline(aes(yintercept=0), linetype="dashed", color="grey1") +
             facet_grid(shapes_rate_u~avg_mut_spans, labeller=label_both) +
             scale_shape_manual(values=c(0, 1, 2)) +
             scale_color_manual(values=c("brown1", "green4", "purple3"),
                                labels=c(expression(mu), "r")) +
             scale_y_continuous(breaks=pretty_breaks()) +
             scale_x_log10() + 
             labs(title="Direct effects on B, average PC across 10 replicates",
                  x="Scale (window size)", y="Path Coefficient",
                  color="Variable", shape="mutagenic_rec") +
             theme(axis.title=element_text(size=16), 
                   axis.text=element_text(size=12), 
                   axis.text.x=element_text(size=12),
                   legend.text=element_text(size=16),
                   legend.title=element_text(size=16),
                   legend.position="bottom")
        
        fout <- paste("SEM/pcoeff_B_exons", e, 
                      "_ars", ars, 
                      "_srr", srr, 
                      "_mru", mru,
                      ".png", sep="")
        save_plot(fout, p1, base_height=10, base_width=18)
      }
    }
  }
}
```

Plotting path coefficients from recombination to pi

```{r pc-plots-SEM-recpi, include=F, message=F, warning=F}

m_pc_sem_recpi <- fread("pc_sem_recpi.csv.gz")

# here we are only interesting in mutagenic DAGs, to contrast 2 paths:
# rec>B>pi and rec>mu>pi
p1 <- ggplot(data=filter(m_pc_sem_recpi, 
                         fisher_c < 0.01),
            aes(x=model, 
                y=mean_pc, 
                color=as.factor(mult_r_u), 
                shape=lhs)) +
      geom_point(size=1.5) + theme_bw() + 
      geom_hline(aes(yintercept=0), 
                 linetype="dashed", color="grey1") +
      facet_grid(scale~DAG, labeller=label_both) +
      scale_shape_manual(values=c(0, 1)) +
      scale_color_manual(values=c("brown1", "purple3", "green4", "cyan3")) +
      scale_y_continuous(breaks=pretty_breaks()) +
      scale_x_continuous(breaks=c(0, 432, 864, 1296)) + 
      labs(title=paste("Structural Equation Modeling, mutagenic rec.,",
                       "average PC across 10 replicates"),
           x="Model ID", y="Path Coefficient to pi",
           color="Mult. Rec. Mu", shape="Var") +
      theme(axis.title=element_text(size=16), 
            axis.text=element_text(size=12), 
            axis.text.x=element_text(size=12),
            legend.text=element_text(size=16),
            legend.title=element_text(size=16),
            legend.position="bottom")
save_plot("SEM/pc_rec_pi_DAG2.png", p1, base_height=10, base_width=16)

for(e in unique(m_pc_sem_recpi$num_exons)) {
  for(ars in unique(m_pc_sem_recpi$avg_rec_spans)) {
    for(srr in unique(m_pc_sem_recpi$shapes_rate_r)) {
      for(mru in unique(m_pc_sem_recpi$mult_r_u)) {
        p1 <- ggplot(data=filter(m_pc_sem_recpi, 
                                 num_exons==e, 
                                 avg_rec_spans==ars, 
                                 shapes_rate_r==srr, 
                                 mult_r_u==mru),
             aes(x=scale, 
                 y=mean_pc, 
                 shape=as.factor(pi_source), 
                 color=lhs,
                 label=ifelse(fisher_c < 0.05, "*", ""))) +
             geom_text(vjust=0.8, size=10) +
             geom_point(size=3) + geom_line() + theme_bw() + 
             geom_hline(aes(yintercept=0), linetype="dashed", color="grey1") +
             facet_grid(shapes_rate_u~avg_mut_spans, labeller=label_both) +
             scale_shape_manual(values=c(0, 1, 2)) +
             scale_color_manual(values=c("brown1", "green4")) +
             scale_y_continuous(breaks=pretty_breaks()) +
             scale_x_log10() + 
             labs(title=paste("Effects of recombination on pi",
                              "(finding the Przeworski regime)"),
                  x="Scale (window size)", y="Path Coefficient",
                  color="Variable", shape="pi") +
             theme(axis.title=element_text(size=16), 
                   axis.text=element_text(size=12), 
                   axis.text.x=element_text(size=12),
                   legend.text=element_text(size=16),
                   legend.title=element_text(size=16),
                   legend.position="bottom")
        
        fout <- paste("SEM/pcoeff_recpi_exons", e, 
                      "_ars", ars, 
                      "_srr", srr, 
                      "_mru", mru,
                      ".png", sep="")
        save_plot(fout, p1, base_height=10, base_width=18)
      }
    }
  }
}
```

Building a linear regression that models SEM path coefficients (mu, B to pi)
as functions of parameters shaping chromosome landscapes.

```{r meta-lm-SEM-pc-1, include=F, message=F, warning=F}

# path coefficients to pi, mutagenic DAGs
m_pc_sem_p_mutrec <- fread("pc_sem_pi_mutrec.csv.gz")

# first we standardize the variables
df <- pivot_wider(dplyr::select(m_pc_sem_p_mutrec, -c(model,
                                                      lhs,
                                                      op, 
                                                      rhs,
                                                      est,
                                                      z,
                                                      se,
                                                      pvalue,
                                                      ci.lower,
                                                      ci.upper,
                                                      fisher_c,
                                                      sd_pc,
                                                      replicate)),
                  names_from=label, 
                  values_from=mean_pc)
names(df)[(ncol(df)-2):ncol(df)] <- c("B", "mut", "mu_B_pi")

## 10 kb
std_pc_10kb <- as.data.frame(df %>%
                             filter(., scale==1e+4) %>% 
                             dplyr::select(c(B, 
                                             mut,
                                             mu_B_pi,
                                             num_exons,
                                             mult_r_u,
                                             shapes_rate_u,
                                             shapes_rate_r,
                                             avg_rec_spans,
                                             avg_mut_spans)) %>%
                            apply(., 2, function(x) (x-mean(x)) / sd(x)))

m_B_10kb <- lm(B ~ (num_exons + shapes_rate_u + shapes_rate_r + mult_r_u +
                   avg_mut_spans + avg_rec_spans)^2, data=std_pc_10kb)
m_u_10kb <- lm(mut ~ (num_exons + shapes_rate_u + shapes_rate_r + mult_r_u +
                     avg_mut_spans + avg_rec_spans)^2, data=std_pc_10kb)
m_uBp_10kb <- lm(mu_B_pi ~ (num_exons + shapes_rate_u + shapes_rate_r + mult_r_u +
                           avg_mut_spans + avg_rec_spans)^2, data=std_pc_10kb)

## 100 kb
std_pc_100kb <- as.data.frame(df %>%
                              filter(., scale==1e+5) %>% 
                              dplyr::select(c(B, 
                                              mut,
                                              mu_B_pi,
                                              num_exons,
                                              mult_r_u,
                                              shapes_rate_u,
                                              shapes_rate_r,
                                              avg_rec_spans,
                                              avg_mut_spans)) %>%
                              apply(., 2, function(x) (x-mean(x)) / sd(x)))

m_B_100kb <- lm(B ~ (num_exons + shapes_rate_u + shapes_rate_r + mult_r_u +
                   avg_mut_spans + avg_rec_spans)^2, data=std_pc_100kb)
m_u_100kb <- lm(mut ~ (num_exons + shapes_rate_u + shapes_rate_r + mult_r_u +
                     avg_mut_spans + avg_rec_spans)^2, data=std_pc_100kb)
m_uBp_100kb <- lm(mu_B_pi ~ (num_exons + shapes_rate_u + shapes_rate_r + mult_r_u +
                           avg_mut_spans + avg_rec_spans)^2, data=std_pc_100kb)

## 1 Mb
std_pc_1Mb <- as.data.frame(df %>%
                            filter(., scale==1e+6) %>% 
                            dplyr::select(c(B, 
                                            mut,
                                            mu_B_pi,
                                            num_exons,
                                            mult_r_u,
                                            shapes_rate_u,
                                            shapes_rate_r,
                                            avg_rec_spans,
                                            avg_mut_spans)) %>%
                            apply(., 2, function(x) (x-mean(x)) / sd(x)))

m_B_1Mb <- lm(B ~ (num_exons + shapes_rate_u + shapes_rate_r + mult_r_u +
                   avg_mut_spans + avg_rec_spans)^2, data=std_pc_1Mb)
m_u_1Mb <- lm(mut ~ (num_exons + shapes_rate_u + shapes_rate_r + mult_r_u + 
                     avg_mut_spans + avg_rec_spans)^2, data=std_pc_1Mb)
m_uBp_1Mb <- lm(mu_B_pi ~ (num_exons + shapes_rate_u + shapes_rate_r + mult_r_u +
                           avg_mut_spans + avg_rec_spans)^2, data=std_pc_1Mb)

# mut
meta_lm_u <- rbind.data.frame(m_u_10kb$coefficients,
                              m_u_100kb$coefficients,
                              m_u_1Mb$coefficients)
names(meta_lm_u) <- names(m_u_10kb$coefficients)

meta_lm_u <- dplyr::select(meta_lm_u, c("shapes_rate_u",
                                        "avg_mut_spans",
                                        "num_exons",
                                        "mult_r_u",
                                        "shapes_rate_r",
                                        "avg_rec_spans"))

meta_lm_u$bin_size <- c(1e+4, 1e+5, 1e+6)
meta_lm_u$response <- "Mu"

mmlm_mu <- pivot_longer(meta_lm_u, cols=names(meta_lm_u)[1:(ncol(meta_lm_u)-2)],
                        names_to="var", values_to="coeff")

# B
meta_lm_B <- rbind.data.frame(m_B_10kb$coefficients,
                              m_B_100kb$coefficients,
                              m_B_1Mb$coefficients)
names(meta_lm_B) <- names(m_B_10kb$coefficients)

meta_lm_B <- dplyr::select(meta_lm_B, c("shapes_rate_u",
                                        "avg_mut_spans",
                                        "num_exons",
                                        "mult_r_u",
                                        "shapes_rate_r",
                                        "avg_rec_spans"))

meta_lm_B$bin_size <- c(1e+4, 1e+5, 1e+6)
meta_lm_B$response <- "B"

mmlm_B <- pivot_longer(meta_lm_B, cols=names(meta_lm_B)[1:(ncol(meta_lm_B)-2)],
                       names_to="var", values_to="coeff")

mmlm_tbl <- rbind.data.frame(mmlm_mu, mmlm_B)

# TODO add real data
meta_lm <- ggplot(data=mmlm_tbl, 
                  aes(x=var, 
                      y=coeff, 
                      shape=as.factor(bin_size),
                      color=response)) +
  geom_point(size=4) + theme_bw() +
  geom_hline(yintercept=0, linetype="dashed") +
  scale_y_continuous(breaks=pretty_breaks()) +
  scale_color_manual(values=c("brown1", "green4"),
                              labels=c("B", expression(mu))) +
  scale_shape_manual(values=c(0, 1, 2, 4), name="Bin Size") +
  labs(title="Regression of Path Coefficient of 'Response' on pi",
       x="Variable", y="Coefficient") +
  theme(axis.title=element_text(size=16), 
        axis.text=element_text(size=12), 
        axis.text.x=element_text(angle=90, size=12, vjust=0.5, hjust=1.0),
        legend.text=element_text(size=16),
        legend.title=element_text(size=16),
        legend.position="bottom")
save_plot("meta_lm_pi_SEM.png", meta_lm, base_height=8, base_width=10)

# mu -> B -> pi
## coeffs
meta_lm_uBp_c <- rbind.data.frame(m_uBp_10kb$coefficients,
                                  m_uBp_100kb$coefficients,
                                  m_uBp_1Mb$coefficients)

names(meta_lm_uBp_c) <- names(m_uBp_10kb$coefficients)
meta_lm_uBp_c$bin_size <- c(1e+4, 1e+5, 1e+6)
meta_lm_uBp_c$path <- "u->B->pi"

mmlm_uBp_c <- pivot_longer(meta_lm_uBp_c, 
                           cols=names(meta_lm_uBp_c)[1:(ncol(meta_lm_uBp_c)-2)],
                           names_to="var", values_to="coeff")

## pvals
meta_lm_uBp_p <- rbind.data.frame(summary(m_uBp_10kb)$coefficients[,4], 
                                  summary(m_uBp_100kb)$coefficients[,4], 
                                  summary(m_uBp_1Mb)$coefficients[,4])

names(meta_lm_uBp_p) <- names(m_uBp_10kb$coefficients)
meta_lm_uBp_p$bin_size <- c(1e+4, 1e+5, 1e+6)
meta_lm_uBp_p$path <- "u->B->pi"

mmlm_uBp_p <- pivot_longer(meta_lm_uBp_p, 
                           cols=names(meta_lm_uBp_p)[1:(ncol(meta_lm_uBp_p)-2)],
                           names_to="var", values_to="pval")

mmlm_uBp <- left_join(mmlm_uBp_c, mmlm_uBp_p)

p <- ggplot(data=filter(mmlm_uBp, pval < 0.01), 
            aes(x=var, 
                y=coeff, 
                shape=as.factor(bin_size))) +
  geom_point(size=4) + theme_bw() +
  geom_hline(yintercept=0, linetype="dashed") +
  scale_y_continuous(breaks=pretty_breaks()) +
  scale_shape_manual(values=c(0, 1, 2, 4), name="Bin Size") +
  labs(title="Regression of Path Coefficient mu -> B -> pi",
       x="Variable", y="Coefficient") +
  theme(axis.title=element_text(size=16), 
        axis.text=element_text(size=12), 
        axis.text.x=element_text(angle=90, size=12, vjust=0.5, hjust=1.0),
        legend.text=element_text(size=16),
        legend.title=element_text(size=16),
        legend.position="bottom")
```

Building a linear regression that models SEM path coefficients (rec. to pi)
as a function of parameters shaping chromosome landscapes.

```{r meta-lm-SEM-pc-2, include=F, message=F, warning=F}

# path coefficients from rec to pi
m_pc_sem_recpi <- fread("pc_sem_recpi.csv.gz")

# first we standardize the variables
df <- pivot_wider(dplyr::select(m_pc_sem_recpi, -c(model,
                                                   lhs,
                                                   op, 
                                                   rhs,
                                                   est,
                                                   z,
                                                   se,
                                                   pvalue,
                                                   ci.lower,
                                                   ci.upper,
                                                   fisher_c,
                                                   sd_pc,
                                                   replicate)),
                  names_from=label, 
                  values_from=mean_pc)

## 10 kb
std_pc_10kb <- as.data.frame(df %>%
                            filter(., scale==1e+4) %>% 
                            dplyr::select(c(rec_B_pi, 
                                            rec_mu_pi,
                                            num_exons,
                                            mult_r_u,
                                            shapes_rate_u,
                                            shapes_rate_r,
                                            avg_rec_spans,
                                            avg_mut_spans)) %>%
                            apply(., 2, function(x) (x-mean(x)) / sd(x)))

m_rBp_10kb <- lm(rec_B_pi ~ (num_exons + shapes_rate_u + 
                             shapes_rate_r + mult_r_u +
                             avg_mut_spans + avg_rec_spans)^2,
                 data=std_pc_10kb)
m_rup_10kb <- lm(rec_mu_pi ~ (num_exons + shapes_rate_u +
                              shapes_rate_r + mult_r_u +
                              avg_mut_spans + avg_rec_spans)^2, 
                 data=std_pc_10kb)

summary(m_rBp_10kb)
summary(m_rup_10kb)

## 100 kb
std_pc_100kb <- as.data.frame(df %>%
                            filter(., scale==1e+5) %>% 
                            dplyr::select(c(rec_B_pi, 
                                            rec_mu_pi,
                                            num_exons,
                                            mult_r_u,
                                            shapes_rate_u,
                                            shapes_rate_r,
                                            avg_rec_spans,
                                            avg_mut_spans)) %>%
                            apply(., 2, function(x) (x-mean(x)) / sd(x)))

m_rBp_100kb <- lm(rec_B_pi ~ (num_exons + shapes_rate_u + 
                              shapes_rate_r + mult_r_u +
                              avg_mut_spans + avg_rec_spans)^2, 
                  data=std_pc_100kb)
m_rup_100kb <- lm(rec_mu_pi ~ (num_exons + shapes_rate_u + 
                               shapes_rate_r + mult_r_u +
                               avg_mut_spans + avg_rec_spans)^2, 
                  data=std_pc_100kb)

summary(m_rBp_100kb)
summary(m_rup_100kb)

## 1 Mb
std_pc_1Mb <- as.data.frame(df %>%
                            filter(., scale==1e+6) %>% 
                            dplyr::select(c(rec_B_pi, 
                                            rec_mu_pi,
                                            num_exons,
                                            mult_r_u,
                                            shapes_rate_u,
                                            shapes_rate_r,
                                            avg_rec_spans,
                                            avg_mut_spans)) %>%
                            apply(., 2, function(x) (x-mean(x)) / sd(x)))

m_rBp_1Mb <- lm(rec_B_pi ~ (num_exons + shapes_rate_u + 
                            shapes_rate_r + mult_r_u +
                            avg_mut_spans + avg_rec_spans)^2,
                data=std_pc_1Mb)
m_rup_1Mb <- lm(rec_mu_pi ~ (num_exons + shapes_rate_u + 
                             shapes_rate_r + mult_r_u +
                             avg_mut_spans + avg_rec_spans)^2, 
                data=std_pc_1Mb)

summary(m_rBp_1Mb)
summary(m_rup_1Mb)

# rec -> B -> pi
m_rBp <- rbind.data.frame(m_rBp_10kb$coefficients,
                          m_rBp_100kb$coefficients,
                          m_rBp_1Mb$coefficients)
names(m_rBp) <- names(m_rBp_10kb$coefficients)

m_rBp <- dplyr::select(m_rBp, c("shapes_rate_u",
                                "avg_mut_spans",
                                "mult_r_u",
                                "num_exons",
                                "shapes_rate_r",
                                "avg_rec_spans"))

m_rBp$bin_size <- c(1e+4, 1e+5, 1e+6)
m_rBp$path <- "rec->B->pi"

mmlm_rBp <- pivot_longer(m_rBp, cols=names(m_rBp)[1:(ncol(m_rBp)-2)],
                        names_to="var", values_to="coeff")

# rec > mu > pi
m_rup <- rbind.data.frame(m_rup_10kb$coefficients,
                          m_rup_100kb$coefficients,
                          m_rup_1Mb$coefficients)
names(m_rup) <- names(m_rup_10kb$coefficients)

m_rup <- dplyr::select(m_rup, c("shapes_rate_u",
                                "avg_mut_spans",
                                "num_exons",
                                "mult_r_u",
                                "shapes_rate_r",
                                "avg_rec_spans"))

m_rup$bin_size <- c(1e+4, 1e+5, 1e+6)
m_rup$path <- "rec->mu->pi"

mmlm_rup <- pivot_longer(m_rup, cols=names(m_rup)[1:(ncol(m_rup)-2)],
                         names_to="var", values_to="coeff")

mmlm_tbl <- rbind.data.frame(mmlm_rBp, mmlm_rup)

# TODO add real data
p <- ggplot(data=mmlm_tbl, 
            aes(x=var, 
                y=coeff, 
                shape=as.factor(bin_size),
                color=path)) +
  geom_point(size=4) + theme_bw() +
  geom_hline(yintercept=0, linetype="dashed") +
  scale_y_continuous(breaks=pretty_breaks()) +
  scale_color_manual(name=NULL, values=c("brown1", "green4")) +
  scale_shape_manual(values=c(0, 1, 2, 4), name="Bin Size") +
  labs(title="Regression of total coefficient along 'Path' to pi",
       x="Variable", y="Coefficient") +
  theme(axis.title=element_text(size=16), 
        axis.text=element_text(size=12), 
        axis.text.x=element_text(angle=90, size=12, vjust=0.5, hjust=1.0),
        legend.text=element_text(size=16),
        legend.title=element_text(size=16),
        legend.position="bottom")
save_plot("meta_lm_rec_pi_SEM.png", p, base_height=8, base_width=10)
```

# Human maps

Here we use the Roulette mutation maps, pyrho recombination maps as well as
a the positions of functionally constrained elements in the human genome
(build 38 and SFS-based DFEs) to predict B-values and pi along autosomes

Without interference, we can treat the maps as "true" and 
avoid modeling error in their estimation.

Preparing the data

```{r human-maps, include=F, message=F, warning=F}

human_maps_10kb <- data.frame()
human_maps_100kb <- data.frame()
human_maps_1Mb <- data.frame()

pb <- txtProgressBar(min=0, max=22, style=3)
for(c in 1:22) {
  setTxtProgressBar(pb, c)
  tbl_10kb <- fread(paste("real_data_maps/equilibrium_human_predictions/",
                          "functional/no_interf_correction/",
                          "B_raw_functional_YRI_chr", c,
                          "_10kb.csv.gz", sep=""))
  tbl_10kb$chrom <- c
  human_maps_10kb <- rbind.data.frame(human_maps_10kb, tbl_10kb)
  
  tbl_100kb <- fread(paste("real_data_maps/equilibrium_human_predictions/",
                           "functional/no_interf_correction/",
                           "B_raw_functional_YRI_chr", c,
                           "_100kb.csv.gz", sep=""))
  tbl_100kb$chrom <- c
  human_maps_100kb <- rbind.data.frame(human_maps_100kb, tbl_100kb)
  
  tbl_1Mb <- fread(paste("real_data_maps/equilibrium_human_predictions/",
                         "functional/no_interf_correction/",
                         "B_raw_functional_YRI_chr", c,
                         "_1Mb.csv.gz", sep=""))
  tbl_1Mb$chrom <- c
  human_maps_1Mb <- rbind.data.frame(human_maps_1Mb, tbl_1Mb)
}
close(pb)

# facilitates calling
names(human_maps_10kb)[1] <- "chr" 
names(human_maps_100kb)[1] <- "chr" 
names(human_maps_1Mb)[1] <- "chr" 

# introducing mutation noise in expected pi
S = 100 # diploid sample size

## 10 kb
pi_sim <- numeric(length=nrow(human_maps_10kb))
pi_sim_neu <- numeric(length=nrow(human_maps_10kb))
pi_sim_del <- numeric(length=nrow(human_maps_10kb))
for(i in 1:S) { 
    pi_sim <- pi_sim + rbinom(size=1e+4, 
                              n=nrow(human_maps_10kb),
                              prob=human_maps_10kb$exp_pi) / 1e+4
    
    pi_sim_neu <- pi_sim_neu + rbinom(size=1e+4, 
                                      n=nrow(human_maps_10kb),
                                      prob=human_maps_10kb$exp_neu_pi) / 1e+4
    
    pi_sim_del <- pi_sim_del + rbinom(size=1e+4, 
                                      n=nrow(human_maps_10kb),
                                      prob=human_maps_10kb$exp_del_pi) / 1e+4
}
     
human_maps_10kb$pi_sim_tot <- pi_sim / S
human_maps_10kb$pi_sim_neu <- pi_sim_neu / S
human_maps_10kb$pi_sim_del <- pi_sim_del / S

human_maps_10kb$coverage <- human_maps_10kb$num_sites / 1e+4
human_maps_10kb$prop_del_sites <- human_maps_10kb$del_sites / human_maps_10kb$num_sites
fwrite(human_maps_10kb, "real_data_maps/human_maps_10kb.csv.gz")

## 100kb
pi_sim <- numeric(length=nrow(human_maps_100kb))
pi_sim_neu <- numeric(length=nrow(human_maps_100kb))
pi_sim_del <- numeric(length=nrow(human_maps_100kb))
for(i in 1:S) { 
    pi_sim <- pi_sim + rbinom(size=1e+5, 
                              n=nrow(human_maps_100kb),
                              prob=human_maps_100kb$exp_pi) / 1e+5
    
    pi_sim_neu <- pi_sim_neu + rbinom(size=1e+5, 
                                      n=nrow(human_maps_100kb),
                                      prob=human_maps_100kb$exp_neu_pi) / 1e+5
    
    pi_sim_del <- pi_sim_del + rbinom(size=1e+5, 
                                      n=nrow(human_maps_100kb),
                                      prob=human_maps_100kb$exp_del_pi) / 1e+5
}
     
human_maps_100kb$pi_sim_tot <- pi_sim / S
human_maps_100kb$pi_sim_neu <- pi_sim_neu / S
human_maps_100kb$pi_sim_del <- pi_sim_del / S

human_maps_100kb$coverage <- human_maps_100kb$num_sites / 1e+5
human_maps_100kb$prop_del_sites <- human_maps_100kb$del_sites / human_maps_100kb$num_sites
fwrite(human_maps_100kb, "real_data_maps/human_maps_100kb.csv.gz")

## 1 Mb
pi_sim <- numeric(length=nrow(human_maps_1Mb))
pi_sim_neu <- numeric(length=nrow(human_maps_1Mb))
pi_sim_del <- numeric(length=nrow(human_maps_1Mb))
for(i in 1:S) { 
    pi_sim <- pi_sim + rbinom(size=1e+6, 
                              n=nrow(human_maps_1Mb),
                              prob=human_maps_1Mb$exp_pi) / 1e+6
    
    pi_sim_neu <- pi_sim_neu + rbinom(size=1e+6, 
                                      n=nrow(human_maps_1Mb),
                                      prob=human_maps_1Mb$exp_neu_pi) / 1e+6
    
    pi_sim_del <- pi_sim_del + rbinom(size=1e+6, 
                                      n=nrow(human_maps_1Mb),
                                      prob=human_maps_1Mb$exp_del_pi) / 1e+6
}
     
human_maps_1Mb$pi_sim_tot <- pi_sim / S
human_maps_1Mb$pi_sim_neu <- pi_sim_neu / S
human_maps_1Mb$pi_sim_del <- pi_sim_del / S

human_maps_1Mb$coverage <- human_maps_1Mb$num_sites / 1e+6
human_maps_1Mb$prop_del_sites <- human_maps_1Mb$del_sites / human_maps_1Mb$num_sites

fwrite(human_maps_1Mb, "real_data_maps/human_maps_1Mb.csv.gz")

# summarizing in terms of AC and CV
human_summaries_10kb <- human_maps_10kb %>% group_by(chr) %>%
                            mutate(cv_mut=sd(avg_mut)/mean(avg_mut)) %>%
                            mutate(cv_rec=sd(avg_rec)/mean(avg_rec)) %>%
                            mutate(ac_mut=as.numeric(unlist(acf(avg_mut,
                                                                lag=1, 
                                                                pl=F))[2])) %>%
                            mutate(ac_rec=as.numeric(unlist(acf(avg_rec,
                                                                lag=1, 
                                                                pl=F))[2])) %>%
                            distinct(chr, .keep_all=T)
human_summaries_10kb$scale <- 1e+4

human_summaries_100kb <- human_maps_100kb %>% group_by(chr) %>%
                            mutate(cv_mut=sd(avg_mut)/mean(avg_mut)) %>%
                            mutate(cv_rec=sd(avg_rec)/mean(avg_rec)) %>%
                            mutate(ac_mut=as.numeric(unlist(acf(avg_mut,
                                                                lag=1, 
                                                                pl=F))[2])) %>%
                            mutate(ac_rec=as.numeric(unlist(acf(avg_rec,
                                                                lag=1, 
                                                                pl=F))[2])) %>%
                            distinct(chr, .keep_all=T)
human_summaries_100kb$scale <- 1e+5

human_summaries_1Mb <- human_maps_1Mb %>% group_by(chr) %>%
                       mutate(cv_mut=sd(avg_mut)/mean(avg_mut)) %>%
                       mutate(cv_rec=sd(avg_rec)/mean(avg_rec)) %>%
                       mutate(ac_mut=as.numeric(unlist(acf(avg_mut,
                                                           lag=1, 
                                                           pl=F))[2])) %>%
                       mutate(ac_rec=as.numeric(unlist(acf(avg_rec,
                                                                lag=1, 
                                                                pl=F))[2])) %>%
                       distinct(chr, .keep_all=T)
human_summaries_1Mb$scale <- 1e+6

human_summaries <- rbind.data.frame(human_summaries_10kb,
                                    human_summaries_100kb,
                                    human_summaries_1Mb)
human_summaries$chr <- paste("chr", human_summaries$chr, sep="") # easier viz
```

Finding the parameter combination which results in simulated maps
most similar to the real ones 

Preparing the simulated maps

```{r sd-ac-maps, include=F, message=F, warning=F}

# collecting simulated maps
muts_10kb <- as.data.frame(matrix(nrow=num_models*num_reps, ncol=1e+4))
recs_10kb <- as.data.frame(matrix(nrow=num_models*num_reps, ncol=1e+4))

muts_100kb <- as.data.frame(matrix(nrow=num_models*num_reps, ncol=1e+3))
recs_100kb <- as.data.frame(matrix(nrow=num_models*num_reps, ncol=1e+3))

muts_1Mb <- as.data.frame(matrix(nrow=num_models*num_reps, ncol=1e+2))
recs_1Mb <- as.data.frame(matrix(nrow=num_models*num_reps, ncol=1e+2))

pb <- txtProgressBar(min=0, max=num_models, style=3)
for(m in 1:num_models) {
  setTxtProgressBar(pb, m)
  for(rep in 1:num_reps) {
    maps_1kb <- fread(paste("model_", m,
                            "/rep_", rep, 
                            "/maps_1kb.csv.gz", sep=""))
    
    maps_1kb$bin_1kb <- 1:nrow(maps_1kb)
    maps_1kb$bin_10kb <- (maps_1kb$bin_1kb - 1) %/% 10
    maps_1kb$bin_100kb <- (maps_1kb$bin_1kb - 1) %/% 100
    maps_1kb$bin_1Mb <- (maps_1kb$bin_1kb - 1) %/% 1000
    
    maps_10kb <- maps_1kb %>% group_by(bin_10kb) %>% 
                 summarise_at(c("avg_mut", "avg_rec"), mean)
    maps_100kb <- maps_1kb %>% group_by(bin_100kb) %>% 
                  summarise_at(c("avg_mut", "avg_rec"), mean)
    maps_1Mb <- maps_1kb %>% group_by(bin_1Mb) %>% 
                summarise_at(c("avg_mut", "avg_rec"), mean)
    
    idx <- rep + (m - 1) * num_reps
    
    muts_10kb[idx,] <- maps_10kb$avg_mut
    recs_10kb[idx,] <- maps_10kb$avg_rec

    muts_100kb[idx,] <- maps_100kb$avg_mut
    recs_100kb[idx,] <- maps_100kb$avg_rec

    muts_1Mb[idx,] <- maps_1Mb$avg_mut
    recs_1Mb[idx,] <- maps_1Mb$avg_rec
  }
}
close(pb)

# mutation maps
names(muts_10kb) <- paste("bin", 1:ncol(muts_10kb), sep="_")
muts_10kb$model <- sort(rep(1:num_models, num_reps))
muts_10kb$replicate <- rep(1:10, num_models)

names(muts_100kb) <- paste("bin", 1:ncol(muts_100kb), sep="_")
muts_100kb$model <- sort(rep(1:num_models, num_reps))
muts_100kb$replicate <- rep(1:10, num_models)

names(muts_1Mb) <- paste("bin", 1:ncol(muts_1Mb), sep="_")
muts_1Mb$model <- sort(rep(1:num_models, num_reps))
muts_1Mb$replicate <- rep(1:10, num_models)
  
# recombination maps
names(recs_10kb) <- paste("bin", 1:ncol(recs_10kb), sep="_")
recs_10kb$model <- sort(rep(1:num_models, num_reps))
recs_10kb$replicate <- rep(1:10, num_models)

names(recs_100kb) <- paste("bin", 1:ncol(recs_100kb), sep="_")
recs_100kb$model <- sort(rep(1:num_models, num_reps))
recs_100kb$replicate <- rep(1:10, num_models)

names(recs_1Mb) <- paste("bin", 1:ncol(recs_1Mb), sep="_")
recs_1Mb$model <- sort(rep(1:num_models, num_reps))
recs_1Mb$replicate <- rep(1:10, num_models)

fwrite(muts_10kb, "mut_maps_10kb_models.csv.gz")
fwrite(muts_100kb, "mut_maps_100kb_models.csv.gz")
fwrite(muts_1Mb, "mut_maps_1Mb_models.csv.gz")
  
fwrite(recs_10kb, "rec_maps_10kb_models.csv.gz")
fwrite(recs_100kb, "rec_maps_100kb_models.csv.gz")
fwrite(recs_1Mb, "rec_maps_1Mb_models.csv.gz")
```

Plotting 2D maps (AC x CV), simulated and real

```{r sd-ac-plots, include=F, message=F, warning=F}

muts_10kb <- fread("mut_maps_10kb_models.csv.gz")
muts_100kb <- fread("mut_maps_100kb_models.csv.gz")
muts_1Mb <- fread("mut_maps_1Mb_models.csv.gz")
  
recs_10kb <- fread("rec_maps_10kb_models.csv.gz")
recs_100kb <- fread("rec_maps_100kb_models.csv.gz")
recs_1Mb <- fread("rec_maps_1Mb_models.csv.gz")

# mutation maps
## 10 kb
muts_10kb$replicate <- rep(1:10, num_models)
muts_10kb_models <- left_join(muts_10kb, models, by="model") %>%
                    pivot_longer(cols=starts_with("bin"), 
                                  values_to="mu",
                                  names_to="bin")

muts_10kb_models <- muts_10kb_models %>% group_by(model, replicate) %>%
                                         mutate(cv_mut=sd(mu)/mean(mu)) %>%
                                         distinct(model, 
                                                  replicate, 
                                                  .keep_all=T) %>%
                                         dplyr::select(., -bin)

muts_10kb_models$ac_mut <- apply(muts_10kb[,1:100], 1, # 1st order autocorrelation
                       function(x) as.numeric(unlist(acf(x, lag=1, pl=F))[2]))
muts_10kb_models$scale <- 1e+4

## 100 kb
muts_100kb$replicate <- rep(1:10, num_models)
muts_100kb_models <- left_join(muts_100kb, models, by="model") %>%
                     pivot_longer(cols=starts_with("bin"), 
                                  values_to="mu",
                                  names_to="bin")

muts_100kb_models <- muts_100kb_models %>% group_by(model, replicate) %>%
                                           mutate(cv_mut=sd(mu)/mean(mu)) %>%
                                           distinct(model, replicate, .keep_all=T) %>%
                                           dplyr::select(., -bin)

muts_100kb_models$ac_mut <- apply(muts_100kb[,1:100], 1, # 1st order autocorrelation
                        function(x) as.numeric(unlist(acf(x, lag=1, pl=F))[2]))
muts_100kb_models$scale <- 1e+5

## 1 Mb
muts_1Mb$replicate <- rep(1:10, num_models)
muts_1Mb_models <- left_join(muts_1Mb, models, by="model") %>%
                   pivot_longer(cols=starts_with("bin"), 
                                values_to="mu",
                                names_to="bin")

muts_1Mb_models <- muts_1Mb_models %>% group_by(model, replicate) %>%
                                       mutate(cv_mut=sd(mu)/mean(mu)) %>%
                                       distinct(model, replicate, .keep_all=T) %>%
                                       dplyr::select(., -bin)

muts_1Mb_models$ac_mut <- apply(muts_1Mb[,1:100], 1, # 1st order autocorrelation
                      function(x) as.numeric(unlist(acf(x, lag=1, pl=F))[2]))
muts_1Mb_models$scale <- 1e+6

muts <- rbind.data.frame(muts_10kb_models, 
                         muts_100kb_models, 
                         muts_1Mb_models)
  
p <- ggplot(data=muts, 
            aes(y=cv_mut, 
                x=ac_mut, 
                color=as.factor(shapes_rate_u),
                shape=as.factor(avg_mut_spans))) +
 geom_point(size=3) + theme_bw() +
 geom_point(data=human_summaries, 
            aes(x=ac_mut, y=cv_mut),
            shape=4, color="black") +
 #geom_text(data=human_summaries, aes(label=chr), vjust=0.8, size=10) +
 facet_grid(scale~mult_r_u, labeller=label_both) +
 scale_shape_manual(values=c(0, 1, 2), name="Avg. Mut. Spans") + 
 # https://stats.stackexchange.com/questions/185425/how-to-determine-the-critical-values-of-acf
 geom_vline(xintercept=-1.96 / sqrt(100-1), color="grey", linetype="dashed") + 
 geom_vline(xintercept=1.96 / sqrt(100-1), color="grey", linetype="dashed") + 
 scale_color_manual(values=c("brown1", "green4", "purple3", "cyan3"),
                    name="Shape=Rate Gamma") + 
 scale_y_continuous(breaks=pretty_breaks()) +
 scale_x_continuous(breaks=pretty_breaks()) +
 labs(title=NULL, x="AC (mutation rates)", y="CV (mutation rates)") +
 theme(axis.title=element_text(size=16), 
       axis.text=element_text(size=12), 
       axis.text.x=element_text(size=12),
       legend.text=element_text(size=16),
       legend.title=element_text(size=16),
       legend.position="bottom")
save_plot("2D_plot_mut.png", p, base_height=12, base_width=18)

# recombination maps
## 10 kb
recs_10kb$replicate <- rep(1:10, num_models)
recs_10kb_models <- left_join(recs_10kb, models, by="model") %>%
                    pivot_longer(cols=starts_with("bin"), 
                                 values_to="rec",
                                 names_to="bin")

recs_10kb_models <- recs_10kb_models %>% group_by(model, replicate) %>%
                                         mutate(cv_rec=sd(rec)/mean(rec)) %>%
                                         distinct(model, replicate, .keep_all=T) %>%
                                         dplyr::select(., -bin)

recs_10kb_models$ac_rec <- apply(recs_10kb[,1:100], 1, # 1st order autocorrelation
                       function(x) as.numeric(unlist(acf(x, lag=1, pl=F))[2]))
recs_10kb_models$scale <- 1e+4

## 100 kb
recs_100kb$replicate <- rep(1:10, num_models)
recs_100kb_models <- left_join(recs_100kb, models, by="model") %>%
                     pivot_longer(cols=starts_with("bin"), 
                                  values_to="rec",
                                  names_to="bin")

recs_100kb_models <- recs_100kb_models %>% group_by(model, replicate) %>%
                                           mutate(cv_rec=sd(rec)/mean(rec)) %>%
                                           distinct(model, replicate, .keep_all=T) %>%
                                           dplyr::select(., -bin)

recs_100kb_models$ac_rec <- apply(recs_100kb[,1:100], 1, # 1st order autocorrelation
                        function(x) as.numeric(unlist(acf(x, lag=1, pl=F))[2]))
recs_100kb_models$scale <- 1e+5

## 1 Mb
recs_1Mb$replicate <- rep(1:10, num_models)
recs_1Mb_models <- left_join(recs_1Mb, models, by="model") %>%
                   pivot_longer(cols=starts_with("bin"), 
                                values_to="rec",
                                names_to="bin")

recs_1Mb_models <- recs_1Mb_models %>% group_by(model, replicate) %>%
                                       mutate(cv_rec=sd(rec)/mean(rec)) %>%
                                       distinct(model, replicate, .keep_all=T) %>%
                                       dplyr::select(., -bin)

recs_1Mb_models$ac_rec <- apply(recs_1Mb[,1:100], 1, # 1st order autocorrelation
                      function(x) as.numeric(unlist(acf(x, lag=1, pl=F))[2]))
recs_1Mb_models$scale <- 1e+6

recs <- rbind.data.frame(recs_10kb_models, recs_100kb_models, recs_1Mb_models)

q <- ggplot(data=recs, 
            aes(y=cv_rec, 
                x=ac_rec, 
                color=as.factor(shapes_rate_r),
                shape=as.factor(avg_rec_spans))) +
 geom_point(size=3) + theme_bw() +
 geom_point(data=human_summaries, 
            aes(x=ac_rec, y=cv_rec),
            shape=4, color="black") +
 #geom_text(data=human_summaries, aes(label=chr), vjust=0.8, size=10) +
 facet_grid(scale~mult_r_u, labeller=label_both) +
 scale_shape_manual(values=c(0, 1, 2), name="Avg. Mut. Spans") + 
 # https://stats.stackexchange.com/questions/185425/how-to-determine-the-critical-values-of-acf
 geom_vline(xintercept=-1.96 / sqrt(100-1), color="grey", linetype="dashed") + 
 geom_vline(xintercept=1.96 / sqrt(100-1), color="grey", linetype="dashed") + 
 scale_color_manual(values=c("brown1", "green4", "purple3", "cyan3"),
                    name="Shape=Rate Gamma") + 
 scale_y_continuous(breaks=pretty_breaks()) +
 scale_x_continuous(breaks=pretty_breaks()) +
 labs(title=NULL, x="AC (recombination rates)", y="CV (recombination rates)") +
 theme(axis.title=element_text(size=16), 
       axis.text=element_text(size=12), 
       axis.text.x=element_text(size=12),
       legend.text=element_text(size=16),
       legend.title=element_text(size=16),
       legend.position="bottom")
save_plot("2D_plot_rec.png", q, base_height=12, base_width=18)
```

Basic visualization

```{r sd-ac-plots, include=F, message=F, warning=F}

for(chrom in c(1:18, 20:22)) {
  dt_1Mb <- filter(human_maps_1Mb, chr==chrom, coverage > 0.75) %>% 
            dplyr::select(., c(prop_del_sites,
                               avg_mut,
                               avg_rec,
                               B,
                               exp_pi,
                               avg_pi,
                               num_sites)) 
            
  p <- ggpairs(dt_1Mb)
  save_plot(paste("Correlations/pairs_1Mb_chr", chrom, ".png", sep=""), p,
            base_height=10, base_width=12)
}

for(chrom in c(1:18, 20:22)) {
  dt_100kb <- filter(human_maps_100kb, chr==chrom, coverage > 0.75) %>% 
            dplyr::select(., c(prop_del_sites,
                               avg_mut,
                               avg_rec,
                               B,
                               exp_pi,
                               avg_pi,
                               num_sites)) 
            
  p <- ggpairs(dt_100kb)
  save_plot(paste("Correlations/pairs_100kb_chr", chrom, ".png", sep=""), p,
            base_height=10, base_width=12)
}

for(chrom in c(1:18, 20:22)) {
  dt_10kb <- filter(human_maps_10kb, chr==chrom, coverage > 0.75) %>% 
             dplyr::select(., c(prop_del_sites,
                                avg_mut,
                                avg_rec,
                                B,
                                exp_pi,
                                avg_pi,
                                num_sites)) 
            
  p <- ggpairs(dt_10kb)
  save_plot(paste("Correlations/pairs_10kb_chr", chrom, ".png", sep=""), p,
            base_height=10, base_width=12)
}
```

## d-separation

### Simulated pi

NOTES:
We filter for missing (obs) data and coverage to match real data tests

Specifying the models with sim_pi

```{r dags-human-maps, include=F, eval=T}

D1 <- DAG(
  pi_sim_tot~B,
  pi_sim_tot~avg_mut,
  B~avg_mut,
  B~avg_rec
)

D2 <- DAG(
  pi_sim_tot~B,
  pi_sim_tot~avg_mut,
  B~avg_mut,
  B~avg_rec,
  B~prop_del_sites
)

D3 <- DAG(
  pi_sim_tot~B,
  pi_sim_tot~avg_mut,
  B~avg_mut,
  B~avg_rec,
  B~prop_del_sites,
  avg_mut~avg_rec
)

D4 <- DAG(
  pi_sim_tot~B,
  pi_sim_tot~avg_mut,
  B~avg_mut,
  B~avg_rec,
  B~prop_del_sites,
  avg_mut~avg_rec,
  # 2nd order selection on the mutation landscape
  avg_mut~prop_del_sites
)

D5 <- DAG(
  pi_sim_tot~B,
  pi_sim_tot~avg_mut,
  B~avg_mut,
  B~avg_rec,
  B~prop_del_sites,
  avg_mut~avg_rec,
  # 2nd order selection on the mutation landscape
  avg_mut~prop_del_sites,
  avg_rec~prop_del_sites
)

models_list <- list(length=5)
models_list[[1]] <- D1
models_list[[2]] <- D2
models_list[[3]] <- D3
models_list[[4]] <- D4
models_list[[5]] <- D5
```

Testing the models

```{r dsep-humans, include=F, eval=T}

num_chrom <- 22
nboot <- 1e+3 # sampling the dataframe for reducing autocorrelation
  
for(m in 1:length(models_list)) {
  
  basis_set <- basiSet(models_list[[m]])
  weights <- rep(1, length=length(basis_set))

  #dsep_c_10kb <- as.data.frame(matrix(nrow=nboot, ncol=num_chrom))
  dsep_c_100kb <- as.data.frame(matrix(nrow=nboot, ncol=num_chrom))
  dsep_c_1Mb <- as.data.frame(matrix(nrow=nboot, ncol=num_chrom))
  
  #names(dsep_c_10kb) <-  paste("chr_", 1:num_chrom, sep="")
  names(dsep_c_100kb) <- paste("chr_", 1:num_chrom, sep="")
  names(dsep_c_1Mb) <-  paste("chr_", 1:num_chrom, sep="")
   
  for(c in 1:num_chrom) {
    
    cat(paste("DAG", m, "chromosome", c))
    
    #df_10kb <- filter(human_maps_10kb, 
    #                  chr==c,
    #                  coverage > 0.75, 
    #                  !is.na(avg_pi)) 
    df_100kb <- filter(human_maps_100kb,
                       chr==c, 
                       coverage > 0.75, 
                       !is.na(avg_pi)) 
    df_1Mb <- filter(human_maps_1Mb, 
                     chr==c,
                     coverage > 0.5, 
                     !is.na(avg_pi)) 
    
    pb <- txtProgressBar(min=0, max=nboot, style=3) 
    for(b in 1:nboot) {
      
      setTxtProgressBar(pb, b)
  
      #samp_10kb <- sample_n(df_10kb, nrow(df_10kb), replace=T)
      samp_100kb <- sample_n(df_100kb, nrow(df_100kb), replace=T)
      samp_1Mb <- sample_n(df_1Mb, nrow(df_1Mb), replace=T)
      
      #pvals_10kb <- rep(NA, length(basis_set))
      pvals_100kb <- rep(NA, length(basis_set))
      pvals_1Mb <- rep(NA, length(basis_set))
  
      for(i in 1:length(basis_set)) {
        
        var1 <- basis_set[[i]][1]
        var2 <- basis_set[[i]][2]
        
        cond <- NULL
        if(length(basis_set[[i]]) > 2) {
          cond <- basis_set[[i]][3:length(basis_set[[i]])]
        }
        
        if(length(basis_set[[i]]) > 2) {
            #pvals_10kb[i] <- ppcor::pcor.test(samp_10kb[, ..var1], 
            #                                  samp_10kb[, ..var2], 
            #                                  samp_10kb[, ..cond],
            #                                  method="kendall")$p.value
            
            pvals_100kb[i] <- ppcor::pcor.test(samp_100kb[, ..var1], 
                                               samp_100kb[, ..var2], 
                                               samp_100kb[, ..cond],
                                               method="kendall")$p.value
            
            pvals_1Mb[i] <- ppcor::pcor.test(samp_1Mb[, ..var1], 
                                             samp_1Mb[, ..var2], 
                                             samp_1Mb[, ..cond],
                                             method="kendall")$p.value
        } else {
          #pvals_10kb[i] <- cor.test(samp_10kb[, ..var1][[1]],
          #                          samp_10kb[, ..var2][[1]],
          #                          method="kendall",
          #                          exact=F)$p.value
          
          pvals_100kb[i] <- cor.test(samp_100kb[, ..var1][[1]],
                                     samp_100kb[, ..var2][[1]],
                                     method="kendall",
                                     exact=F)$p.value
          
          pvals_1Mb[i] <- cor.test(samp_1Mb[, ..var1][[1]],
                                   samp_1Mb[, ..var2][[1]],
                                   method="kendall",
                                   exact=F)$p.value
        }
      }
      close(pb)
      
      #dsep_c_10kb[b, c] <- combine.test(pvals_10kb, weights, "fisher", na.rm=T)
      dsep_c_100kb[b, c] <- combine.test(pvals_100kb, weights, "fisher", na.rm=T)
      dsep_c_1Mb[b, c] <- combine.test(pvals_1Mb, weights, "fisher", na.rm=T)
    }
  }
  
  #fwrite(dsep_c_10kb, paste("real_data_maps/dsep_10kb_pisim_m",
  #                          i, ".csv.gz", sep=""))
  fwrite(dsep_c_100kb, paste("real_data_maps/dsep_100kb_pisim_m",
                             m, ".csv.gz", sep=""))
  fwrite(dsep_c_1Mb, paste("real_data_maps/dsep_1Mb_pisim_m",
                            m, ".csv.gz", sep=""))
}
```

Processing files

```{r d-sep-plot, include=F, message=F, warning=F}

# combines tables from previous chunks
## 100 kb
pval_dsep1_100kb <- fread("real_data_maps/dsep_100kb_pisim_m1.csv.gz") 
pval_dsep2_100kb <- fread("real_data_maps/dsep_100kb_pisim_m2.csv.gz") 
pval_dsep3_100kb <- fread("real_data_maps/dsep_100kb_pisim_m3.csv.gz") 
pval_dsep4_100kb <- fread("real_data_maps/dsep_100kb_pisim_m4.csv.gz") 
pval_dsep5_100kb <- fread("real_data_maps/dsep_100kb_pisim_m5.csv.gz") 

pval_dsep1_100kb$DAG <- 1
pval_dsep2_100kb$DAG <- 2
pval_dsep3_100kb$DAG <- 3
pval_dsep4_100kb$DAG <- 4
pval_dsep5_100kb$DAG <- 5

pval_dsep1_100kb$boot <- 1:nrow(pval_dsep1_100kb)
pval_dsep2_100kb$boot <- 1:nrow(pval_dsep2_100kb)
pval_dsep3_100kb$boot <- 1:nrow(pval_dsep3_100kb)
pval_dsep4_100kb$boot <- 1:nrow(pval_dsep4_100kb)
pval_dsep5_100kb$boot <- 1:nrow(pval_dsep5_100kb)

dsep_human_maps_100kb <- rbind.data.frame(pval_dsep1_100kb,
                                        pval_dsep2_100kb,
                                        pval_dsep3_100kb,
                                        pval_dsep4_100kb,
                                        pval_dsep5_100kb)
names(dsep_human_maps_100kb) <- c(as.character(1:22), "DAG", "boot")
fwrite(dsep_human_maps_100kb,
       "real_data_maps/dsep_human_maps_100kb_pisim.csv.gz")

## 1 Mb
pval_dsep1_1Mb <- fread("real_data_maps/dsep_1Mb_pisim_m1.csv.gz") 
pval_dsep2_1Mb <- fread("real_data_maps/dsep_1Mb_pisim_m2.csv.gz") 
pval_dsep3_1Mb <- fread("real_data_maps/dsep_1Mb_pisim_m3.csv.gz") 
pval_dsep4_1Mb <- fread("real_data_maps/dsep_1Mb_pisim_m4.csv.gz") 
pval_dsep5_1Mb <- fread("real_data_maps/dsep_1Mb_pisim_m5.csv.gz") 

pval_dsep1_1Mb$DAG <- 1
pval_dsep2_1Mb$DAG <- 2
pval_dsep3_1Mb$DAG <- 3
pval_dsep4_1Mb$DAG <- 4
pval_dsep5_1Mb$DAG <- 5

pval_dsep1_1Mb$boot <- 1:nrow(pval_dsep1_1Mb)
pval_dsep2_1Mb$boot <- 1:nrow(pval_dsep2_1Mb)
pval_dsep3_1Mb$boot <- 1:nrow(pval_dsep3_1Mb)
pval_dsep4_1Mb$boot <- 1:nrow(pval_dsep4_1Mb)
pval_dsep5_1Mb$boot <- 1:nrow(pval_dsep5_1Mb)

dsep_human_maps_1Mb <- rbind.data.frame(pval_dsep1_1Mb,
                                        pval_dsep2_1Mb,
                                        pval_dsep3_1Mb,
                                        pval_dsep4_1Mb,
                                        pval_dsep5_1Mb)
names(dsep_human_maps_1Mb) <- c(as.character(1:22), "DAG", "boot")
fwrite(dsep_human_maps_1Mb,
       "real_data_maps/dsep_human_maps_1Mb_pisim.csv.gz")
```

### Observed pi

Specifying the models with sim_pi

```{r dags-human-maps, include=F, eval=T}

D1 <- DAG(
  pi_avg~B,
  pi_avg~avg_mut,
  B~avg_mut,
  B~avg_rec
)

D2 <- DAG(
  pi_avg~B,
  pi_avg~avg_mut,
  B~avg_mut,
  B~avg_rec,
  B~prop_del_sites
)

D3 <- DAG(
  pi_avg~B,
  pi_avg~avg_mut,
  B~avg_mut,
  B~avg_rec,
  B~prop_del_sites,
  avg_mut~avg_rec
)

D4 <- DAG(
  pi_avg~B,
  pi_avg~avg_mut,
  B~avg_mut,
  B~avg_rec,
  B~prop_del_sites,
  avg_mut~avg_rec,
  # 2nd order selection on the mutation landscape
  avg_mut~prop_del_sites
)

D5 <- DAG(
  pi_avg~B,
  pi_avg~avg_mut,
  B~avg_mut,
  B~avg_rec,
  B~prop_del_sites,
  avg_mut~avg_rec,
  # 2nd order selection on the mutation landscape
  avg_mut~prop_del_sites,
  avg_rec~prop_del_sites
)

models_list <- list(length=5)
models_list[[1]] <- D1
models_list[[2]] <- D2
models_list[[3]] <- D3
models_list[[4]] <- D4
models_list[[5]] <- D5
```

Testing the models

```{r dsep-humans, include=F, eval=T}

num_chrom <- 22
nboot <- 1e+3 # sampling the dataframe for reducing autocorrelation
  
for(m in 1:length(models_list)) {
  
  basis_set <- basiSet(models_list[[m]])
  weights <- rep(1, length=length(basis_set))

  #dsep_c_10kb <- as.data.frame(matrix(nrow=nboot, ncol=num_chrom))
  dsep_c_100kb <- as.data.frame(matrix(nrow=nboot, ncol=num_chrom))
  dsep_c_1Mb <- as.data.frame(matrix(nrow=nboot, ncol=num_chrom))
  
  #names(dsep_c_10kb) <-  paste("chr_", 1:num_chrom, sep="")
  names(dsep_c_100kb) <- paste("chr_", 1:num_chrom, sep="")
  names(dsep_c_1Mb) <-  paste("chr_", 1:num_chrom, sep="")
   
  for(c in 1:num_chrom) {
    
    cat(paste("DAG", m, "chromosome", c))
    
    #df_10kb <- filter(human_maps_10kb, 
    #                  chr==c,
    #                  coverage > 0.75, 
    #                  !is.na(avg_pi)) 
    df_100kb <- filter(human_maps_100kb,
                       chr==c, 
                       coverage > 0.75, 
                       !is.na(avg_pi)) 
    df_1Mb <- filter(human_maps_1Mb, 
                     chr==c,
                     coverage > 0.5, 
                     !is.na(avg_pi)) 
    
    pb <- txtProgressBar(min=0, max=nboot, style=3) 
    for(b in 1:nboot) {
      
      setTxtProgressBar(pb, b)
  
      #samp_10kb <- sample_n(df_10kb, nrow(df_10kb), replace=T)
      samp_100kb <- sample_n(df_100kb, nrow(df_100kb), replace=T)
      samp_1Mb <- sample_n(df_1Mb, nrow(df_1Mb), replace=T)
      
      #pvals_10kb <- rep(NA, length(basis_set))
      pvals_100kb <- rep(NA, length(basis_set))
      pvals_1Mb <- rep(NA, length(basis_set))
  
      for(i in 1:length(basis_set)) {
        
        var1 <- basis_set[[i]][1]
        var2 <- basis_set[[i]][2]
        
        cond <- NULL
        if(length(basis_set[[i]]) > 2) {
          cond <- basis_set[[i]][3:length(basis_set[[i]])]
        }
        
        if(length(basis_set[[i]]) > 2) {
            #pvals_10kb[i] <- ppcor::pcor.test(samp_10kb[, ..var1], 
            #                                  samp_10kb[, ..var2], 
            #                                  samp_10kb[, ..cond],
            #                                  method="kendall")$p.value
            
            pvals_100kb[i] <- ppcor::pcor.test(samp_100kb[, ..var1], 
                                               samp_100kb[, ..var2], 
                                               samp_100kb[, ..cond],
                                               method="kendall")$p.value
            
            pvals_1Mb[i] <- ppcor::pcor.test(samp_1Mb[, ..var1], 
                                             samp_1Mb[, ..var2], 
                                             samp_1Mb[, ..cond],
                                             method="kendall")$p.value
        } else {
          #pvals_10kb[i] <- cor.test(samp_10kb[, ..var1][[1]],
          #                          samp_10kb[, ..var2][[1]],
          #                          method="kendall",
          #                          exact=F)$p.value
          
          pvals_100kb[i] <- cor.test(samp_100kb[, ..var1][[1]],
                                     samp_100kb[, ..var2][[1]],
                                     method="kendall",
                                     exact=F)$p.value
          
          pvals_1Mb[i] <- cor.test(samp_1Mb[, ..var1][[1]],
                                   samp_1Mb[, ..var2][[1]],
                                   method="kendall",
                                   exact=F)$p.value
        }
      }
      close(pb)
      
      #dsep_c_10kb[b, c] <- combine.test(pvals_10kb, weights, "fisher", na.rm=T)
      dsep_c_100kb[b, c] <- combine.test(pvals_100kb, weights, "fisher", na.rm=T)
      dsep_c_1Mb[b, c] <- combine.test(pvals_1Mb, weights, "fisher", na.rm=T)
    }
  }
  
  #fwrite(dsep_c_10kb, paste("real_data_maps/dsep_10kb_piobs_m",
  #                          i, ".csv.gz", sep=""))
  fwrite(dsep_c_100kb, paste("real_data_maps/dsep_100kb_piobs_m",
                             m, ".csv.gz", sep=""))
  fwrite(dsep_c_1Mb, paste("real_data_maps/dsep_1Mb_piobs_m",
                            m, ".csv.gz", sep=""))
}
```

Processing files

```{r d-sep-plot, include=F, message=F, warning=F}

# combines tables from previous chunks
## 100 kb
pval_dsep1_100kb <- fread("real_data_maps/dsep_100kb_piobs_m1.csv.gz") 
pval_dsep2_100kb <- fread("real_data_maps/dsep_100kb_piobs_m2.csv.gz") 
pval_dsep3_100kb <- fread("real_data_maps/dsep_100kb_piobs_m3.csv.gz") 
pval_dsep4_100kb <- fread("real_data_maps/dsep_100kb_piobs_m4.csv.gz") 
pval_dsep5_100kb <- fread("real_data_maps/dsep_100kb_piobs_m5.csv.gz") 

pval_dsep1_100kb$DAG <- 1
pval_dsep2_100kb$DAG <- 2
pval_dsep3_100kb$DAG <- 3
pval_dsep4_100kb$DAG <- 4
pval_dsep5_100kb$DAG <- 5

pval_dsep1_100kb$boot <- 1:nrow(pval_dsep1_100kb)
pval_dsep2_100kb$boot <- 1:nrow(pval_dsep2_100kb)
pval_dsep3_100kb$boot <- 1:nrow(pval_dsep3_100kb)
pval_dsep4_100kb$boot <- 1:nrow(pval_dsep4_100kb)
pval_dsep5_100kb$boot <- 1:nrow(pval_dsep5_100kb)

dsep_human_maps_100kb <- rbind.data.frame(pval_dsep1_100kb,
                                        pval_dsep2_100kb,
                                        pval_dsep3_100kb,
                                        pval_dsep4_100kb,
                                        pval_dsep5_100kb)
names(dsep_human_maps_100kb) <- c(as.character(1:22), "DAG", "boot")
fwrite(dsep_human_maps_100kb,
       "real_data_maps/dsep_human_maps_100kb_piobs.csv.gz")

## 1 Mb
pval_dsep1_1Mb <- fread("real_data_maps/dsep_1Mb_piobs_m1.csv.gz") 
pval_dsep2_1Mb <- fread("real_data_maps/dsep_1Mb_piobs_m2.csv.gz") 
pval_dsep3_1Mb <- fread("real_data_maps/dsep_1Mb_piobs_m3.csv.gz") 
pval_dsep4_1Mb <- fread("real_data_maps/dsep_1Mb_piobs_m4.csv.gz") 
pval_dsep5_1Mb <- fread("real_data_maps/dsep_1Mb_piobs_m5.csv.gz") 

pval_dsep1_1Mb$DAG <- 1
pval_dsep2_1Mb$DAG <- 2
pval_dsep3_1Mb$DAG <- 3
pval_dsep4_1Mb$DAG <- 4
pval_dsep5_1Mb$DAG <- 5

pval_dsep1_1Mb$boot <- 1:nrow(pval_dsep1_1Mb)
pval_dsep2_1Mb$boot <- 1:nrow(pval_dsep2_1Mb)
pval_dsep3_1Mb$boot <- 1:nrow(pval_dsep3_1Mb)
pval_dsep4_1Mb$boot <- 1:nrow(pval_dsep4_1Mb)
pval_dsep5_1Mb$boot <- 1:nrow(pval_dsep5_1Mb)

dsep_human_maps_1Mb <- rbind.data.frame(pval_dsep1_1Mb,
                                        pval_dsep2_1Mb,
                                        pval_dsep3_1Mb,
                                        pval_dsep4_1Mb,
                                        pval_dsep5_1Mb)
names(dsep_human_maps_1Mb) <- c(as.character(1:22), "DAG", "boot")
fwrite(dsep_human_maps_1Mb,
       "real_data_maps/dsep_human_maps_1Mb_piobs.csv.gz")
```

Combining simulated and observed pi, then plotting

```{r d-sep-plot, include=F, message=F, warning=F}

# combines tables from previous chunks
pval_dsep_pisim <- fread("real_data_maps/dsep_human_maps_1Mb_pisim.csv.gz") 
pval_dsep_piobs <- fread("real_data_maps/dsep_human_maps_1Mb_piobs.csv.gz") 

pval_dsep_pisim$pi <- "sim"
pval_dsep_piobs$pi <- "obs"

dsep_human_maps_1Mb <- rbind.data.frame(pval_dsep_pisim,
                                        pval_dsep_piobs)

m_dsep_human_1Mb <- pivot_longer(dsep_human_maps_1Mb,
                                 cols=1:num_chrom,
                                 names_to="chr", values_to="pval")
m_dsep_human_1Mb$chr <- as.numeric(m_dsep_human_1Mb$chr)

# testing the trend of avg pvalue vs chromosome id (~inverse of chr size)
x <- m_dsep_human_1Mb %>% group_by(DAG, pi, chr) %>% 
                          mutate(avg_p=mean(pval)) %>%
                          distinct(DAG, pi, chr, .keep_all=T) %>%
                          dplyr::select(DAG, pi, chr, avg_p)
y <- x %>% group_by(DAG, pi) %>% 
           mutate(cor_pval_chrID_est=cor.test(avg_p, chr, method="kendall")$estimate,
                  cor_pval_chrID_pval=cor.test(avg_p, chr, method="kendall")$p.value) %>% 
           distinct(DAG, pi, .keep_all=T) %>%
           dplyr::select(., c(DAG, pi, cor_pval_chrID_est, cor_pval_chrID_pval))

# helping viz
xsim <- filter(m_dsep_human_1Mb, pi=="sim") %>% mutate(chr_p=chr - 0.1)  
xobs <- filter(m_dsep_human_1Mb, pi=="obs") %>% mutate(chr_p=chr + 0.1)  

m_dsep_human_1Mb <- rbind.data.frame(xsim, xobs)
m_dsep_human_1Mb$chr_p <- as.factor(m_dsep_human_1Mb$chr_p)
levels(m_dsep_human_1Mb$chr_p) <- as.character(sort(rep(1:22, 2)))

num_boot <- max(m_dsep_human_1Mb$boot)
p1 <- ggplot(data=filter(m_dsep_human_1Mb), 
             aes(x=chr_p, 
                 y=-log10(pval), 
                 color=pi)) +
      geom_hline(aes(yintercept=-log10(0.05/(num_boot * num_chrom))), 
                 linetype="dashed", color="red") +
      #geom_violin(trim=FALSE, fill="white") + # gets misaligned by levels() <-
      geom_boxplot(width=0.25, fill="white", alpha=0.25)+
      facet_wrap(~DAG, ncol=1, labeller=label_both) + theme_bw() + 
      scale_y_continuous(breaks=pretty_breaks()) +
      scale_x_discrete(breaks=as.character(sort(rep(1:22, 2)))) + 
      scale_color_manual(values=c("purple3", "seagreen4")) +
      guides(color=guide_legend(override.aes=list(alpha=1, shape=16, size=1))) +
      labs(title=paste("d-separation tests => H0: data based on human maps", 
                       "(1Mb) fits DAG model (1000 bootstrap samples)"),
           x="Chromosome", y="-log10(p-value)") +
      theme(axis.title=element_text(size=16), 
            axis.text=element_text(size=12), 
            axis.text.x=element_text(size=12),
            legend.text=element_text(size=16),
            legend.title=element_text(size=16),
            legend.position="bottom")
save_plot("Manhattan/manhattan_d-sep_human_1Mb.png", p1, 
          base_height=10, base_width=15)

m_dsep_human_1Mb$minuslog10pval <- -log10(m_dsep_human_1Mb$pval)

pisim_tbl=filter(m_dsep_human_1Mb, pi=="sim")
compare_means(pval ~ DAG,  data=pisim_tbl)
my_comparisons <- list(c("1", "2"),
                       c("2", "3"),
                       c("3", "4"),
                       c("4", "5"))

p2 <- ggpubr::ggboxplot(data=pisim_tbl,
                x="DAG", 
                y="minuslog10pval",
                legend="bottom",
                ylab="-log10(pval)",
                title=paste("Difference in d-sep p-values (1000 bootstrap samples)", 
                       "between pairs of increasingly complex DAGs (sim. pi)")) +
      facet_wrap(~chr, nrow=4, labeller=label_both) +
      scale_y_continuous(breaks=pretty_breaks()) +
      stat_compare_means(comparisons=my_comparisons) + theme_bw() + 
      theme(plot.title=element_text(size=26),
            axis.title=element_text(size=24), 
            axis.text=element_text(size=16),
            strip.text.x=element_text(size=14))
save_plot("pairwise_d-sep_human_DAGs_pisim_1Mb.png", p2, 
          base_height=20, base_width=30)

piobs_tbl=filter(m_dsep_human_1Mb, pi=="obs")
compare_means(pval ~ DAG,  data=piobs_tbl)
my_comparisons <- list(c("1", "2"),
                       c("2", "3"),
                       c("3", "4"),
                       c("4", "5"))

p3 <- ggpubr::ggboxplot(data=piobs_tbl,
                x="DAG", 
                y="minuslog10pval",
                legend="bottom",
                ylab="-log10(pval)",
                title=paste("Difference in d-sep p-values (1000 bootstrap samples)", 
                       "between pairs of increasingly complex DAGs (obs. pi)")) +
      facet_wrap(~chr, nrow=4, labeller=label_both) +
      scale_y_continuous(breaks=pretty_breaks()) +
      stat_compare_means(comparisons=my_comparisons) + theme_bw() + 
      theme(plot.title=element_text(size=26),
            axis.title=element_text(size=24), 
            axis.text=element_text(size=16),
            strip.text.x=element_text(size=14))
save_plot("pairwise_d-sep_human_DAGs_piobs_1Mb.png", p3, 
          base_height=20, base_width=30)
```

## Path Models

Here we assume that the maps (roulette, pyrho, moments++) are perfect measures
of the true mutation, recombination and B-value maps.

This is of course unrealistic. 
We later relax these assumptions with latent variables.
 
Testing model fitting at the 1 Mb scale, keeping 1 in every w windows to
reduce auto-correlation.

TODO: use modIndices()

```{r SEM-human-maps-fw, include=F, message=F, warning=F}

human_maps_1Mb$coverage <- human_maps_1Mb$num_sites / 1e+6 
human_maps_1Mb$prop_del_sites <- human_maps_1Mb$del_sites / 
  human_maps_1Mb$num_sites

for(w in 1:5) {
  cat(paste("fitting SEMs with filter w = ", w, sep=""))
  
  human_maps_1Mb$bin <- human_maps_1Mb$chromStart %/% w * 1e+6 
  dat_1Mb <- human_maps_1Mb %>% 
             group_by(chr) %>% # chr 19 too few data points
             filter(., del_sites > 0, coverage > 0.75, chr!=19) %>% 
             na.exclude() %>% 
             distinct(bin, .keep_all=T)
  
  fwrite(dat_1Mb, paste("SEM/dat_1Mb_w", w, ".csv.gz", sep="")) # storing
  
  model_1 <- "
  avg_pi ~ start(0.25)*c1*B
  avg_pi ~ start(0.25)*c2*avg_mut
  B ~ start(-0.25)*c3*avg_mut
  avg_mut ~~ 0*B
  "
  
  m1gls <- sem(model_1,
               data=dat_1Mb,
               fixed.x=F,
               std.ov=T,
               estimator="GLS",
               group="chr", 
               group.equal="none")
  
  fwrite(parameterestimates(m1gls, ci=F), paste("SEM/m1gls_est_w", 
                                                w, 
                                                ".csv", sep=""))
  fwrite(t(as.data.frame(fitMeasures(m1gls))), paste("SEM/m1gls_fit_w",
                                                     w,
                                                     ".csv", sep=""))
  fwrite(as.data.frame(inspect(m1gls, "rsquare")), paste("SEM/m1gls_r2_w",
                                                   w,
                                                   ".csv", sep="")) 
  
  m1boot <- sem(model_1,
                data=dat_1Mb,
                fixed.x=F,
                std.ov=T,
                estimator="ML",
                test="boot",
                se="boot",
                group="chr", 
                group.equal="none")
  
  fwrite(parameterestimates(m1boot,
                            boot.ci.type="norm",
                            ci=T,
                            standardized=T), paste("SEM/m1boot_est_w", 
                                                   w,
                                                   ".csv", sep=""))
  fwrite(t(as.data.frame(fitMeasures(m1boot))), paste("SEM/m1boot_fit_w",
                                                w,
                                                ".csv", sep=""))
  fwrite(as.data.frame(inspect(m1boot, "rsquare")), paste("SEM/m1boot_r2_w",
                                                          w,
                                                          ".csv", sep=""))
  
  model_2 <- "
  avg_pi ~ start(0.25)*c1*B
  avg_pi ~ start(0.25)*c2*avg_mut
  B ~ start(-0.25)*c3*avg_mut
  avg_mut ~~ B # covariance due to un-modelled rec rate
  "
  
  m2gls <- sem(model_2,
               data=dat_1Mb,
               fixed.x=F,
               std.ov=T,
               estimator="GLS",
               group="chr", 
               group.equal="regressions")
  
  fwrite(parameterestimates(m2gls, ci=F), paste("SEM/m2gls_est_w", 
                                                w, 
                                                ".csv", sep=""))
  fwrite(t(as.data.frame(fitMeasures(m2gls))), paste("SEM/m2gls_fit_w",
                                                     w,
                                                     ".csv", sep=""))
  fwrite(as.data.frame(inspect(m2gls, "rsquare")), paste("SEM/m2gls_r2_w",
                                                   w,
                                                   ".csv", sep="")) 
  
  m2boot <- sem(model_2,
               data=dat_1Mb,
               fixed.x=F,
               std.ov=T,
               estimator="ML",
               test="boot",
               se="boot",
               group="chr", 
               group.equal="regressions")
  
  fwrite(parameterestimates(m2boot,
                            boot.ci.type="norm",
                            ci=T,
                            standardized=T), paste("SEM/m2boot_est_w", 
                                                   w,
                                                   ".csv", sep=""))
  fwrite(t(as.data.frame(fitMeasures(m2boot))), paste("SEM/m2boot_fit_w",
                                                w,
                                                ".csv", sep=""))
  fwrite(as.data.frame(inspect(m2boot, "rsquare")), paste("SEM/m2boot_r2_w",
                                                          w,
                                                          ".csv", sep="")) 
  
  # incorporating rec map
  model_3 <- "
  avg_pi ~ start(0.75)*c1*B
  avg_pi ~ start(0.75)*c2*avg_mut
  B ~ start(-0.25)*c3*avg_mut
  B ~ start(0.1)*c4*avg_rec
  avg_mut ~ start(0.1)*c5*avg_rec
  "
  
  m3gls <- sem(model_3,
               data=dat_1Mb,
               fixed.x=F,
               std.ov=T,
               estimator="GLS",
               group="chr", 
               group.equal="none")
  
  fwrite(parameterestimates(m3gls, ci=F), paste("SEM/m3gls_est_w", 
                                                w, 
                                                ".csv", sep=""))
  fwrite(t(as.data.frame(fitMeasures(m3gls))), paste("SEM/m3gls_fit_w",
                                                     w,
                                                     ".csv", sep=""))
  fwrite(as.data.frame(inspect(m3gls, "rsquare")), paste("SEM/m3gls_r2_w",
                                                   w,
                                                   ".csv", sep=""))  
  
  m3boot <- sem(model_3,
               data=dat_1Mb,
               fixed.x=F,
               std.ov=T,
               estimator="ML",
               se="boot",
               test="boot", 
               group="chr", 
               group.equal="regressions")
  
  fwrite(parameterestimates(m3boot,
                            boot.ci.type="norm",
                            ci=T,
                            standardized=T), paste("SEM/m3boot_est_w", 
                                                   w,
                                                   ".csv", sep=""))
  fwrite(t(as.data.frame(fitMeasures(m3boot))), paste("SEM/m3boot_fit_w",
                                                w,
                                                ".csv", sep=""))
  fwrite(as.data.frame(inspect(m3boot, "rsquare")), paste("SEM/m3boot_r2_w",
                                                          w,
                                                          ".csv", sep="")) 
  
  # incorporating del sites (normalized by number of accessible sites)
  model_4 <- "
  avg_pi ~ start(0.75)*c1*B
  avg_pi ~ start(0.75)*c2*avg_mut
  B ~ start(-0.25)*c3*avg_mut
  B ~ start(-0.1)*c4*prop_del_sites
  avg_mut ~~ 0*prop_del_sites
  avg_mut ~~ 0*B
  "
  
  m4gls <- sem(model_4,
               data=dat_1Mb,
               fixed.x=F,
               std.ov=T,
               estimator="GLS",
               group="chr", 
               group.equal="none")
  
  fwrite(parameterestimates(m4gls, ci=F), paste("SEM/m4gls_est_w", 
                                                w, 
                                                ".csv", sep=""))
  fwrite(t(as.data.frame(fitMeasures(m4gls))), paste("SEM/m4gls_fit_w",
                                                     w,
                                                     ".csv", sep=""))
  fwrite(as.data.frame(inspect(m4gls, "rsquare")), paste("SEM/m4gls_r2_w",
                                                   w,
                                                   ".csv", sep="")) 
  
  m4boot <- sem(model_4,
               data=dat_1Mb,
               fixed.x=F,
               std.ov=T,
               estimator="ML",
               se="boot",
               test="boot", 
               group="chr", 
               group.equal="regressions")
  
  fwrite(parameterestimates(m4boot,
                            boot.ci.type="norm",
                            ci=T,
                            standardized=T), paste("SEM/m4boot_est_w", 
                                                   w,
                                                   ".csv", sep=""))
  fwrite(t(as.data.frame(fitMeasures(m4boot))), paste("SEM/m4boot_fit_w",
                                                w,
                                                ".csv", sep=""))
  fwrite(as.data.frame(inspect(m4boot, "rsquare")), paste("SEM/m4boot_r2_w",
                                                          w,
                                                          ".csv", sep="")) 
  
  model_5 <- "
  avg_pi ~ start(0.75)*c1*B
  avg_pi ~ start(0.75)*c2*avg_mut
  B ~ start(-0.25)*c3*avg_mut
  B ~ start(-0.1)*c4*prop_del_sites
  avg_mut ~ start(0.25)*c6*avg_rec
  avg_mut ~~ prop_del_sites
  "
  
  m5gls <- sem(model_5,
               data=dat_1Mb,
               fixed.x=F,
               std.ov=T,
               estimator="GLS",
               group="chr", 
               group.equal="none")
  
  fwrite(parameterestimates(m5gls, ci=F), paste("SEM/m5gls_est_w", 
                                                w, 
                                                ".csv", sep=""))
  fwrite(t(as.data.frame(fitMeasures(m5gls))), paste("SEM/m5gls_fit_w",
                                                     w,
                                                     ".csv", sep=""))
  fwrite(as.data.frame(inspect(m5gls, "rsquare")), paste("SEM/m5gls_r2_w",
                                                   w,
                                                   ".csv", sep=""))  
  
  m5boot <- sem(model_5,
               data=dat_1Mb,
               fixed.x=F,
               std.ov=T,
               estimator="ML",
               se="boot",
               test="boot", 
               group="chr", 
               group.equal="regressions")
  
  fwrite(parameterestimates(m5boot,
                            boot.ci.type="norm",
                            ci=T,
                            standardized=T), paste("SEM/m5boot_est_w", 
                                                   w,
                                                   ".csv", sep=""))
  fwrite(t(as.data.frame(fitMeasures(m5boot))), paste("SEM/m5boot_fit_w",
                                                w,
                                                ".csv", sep=""))
  fwrite(as.data.frame(inspect(m5boot, "rsquare")), paste("SEM/m5boot_r2_w",
                                                          w,
                                                          ".csv", sep="")) 
  
  model_6 <- "
  avg_pi ~ start(0.25)*c1*B
  avg_pi ~ start(0.25)*c2*avg_mut
  B ~ start(-0.25)*c3*avg_mut
  B ~ start(-0.1)*c4*prop_del_sites
  B ~ start(0.25)*c5*avg_rec
  avg_mut ~ start(0.25)*c6*avg_rec
  avg_mut ~~ prop_del_sites
  avg_mut ~~ B
  "
  
  m6gls <- sem(model_6,
               data=dat_1Mb,
               fixed.x=F,
               std.ov=T,
               estimator="GLS",
               group="chr", 
               group.equal="none")
  
  fwrite(parameterestimates(m6gls, ci=F), paste("SEM/m6gls_est_w", 
                                                w, 
                                                ".csv", sep=""))
  fwrite(t(as.data.frame(fitMeasures(m6gls))), paste("SEM/m6gls_fit_w",
                                                     w,
                                                     ".csv", sep=""))
  fwrite(as.data.frame(inspect(m6gls, "rsquare")), paste("SEM/m6gls_r2_w",
                                                   w,
                                                   ".csv", sep="")) 
  
  m6boot <- sem(model_6,
               data=dat_1Mb,
               fixed.x=F,
               std.ov=T,
               estimator="ML",
               test="boot",
               se="boot",
               group="chr", 
               group.equal="none")
  
  fwrite(parameterestimates(m6boot,
                            boot.ci.type="norm",
                            ci=T,
                            standardized=T), paste("SEM/m6boot_est_w", 
                                                   w,
                                                   ".csv", sep=""))
  fwrite(t(as.data.frame(fitMeasures(m6boot))), paste("SEM/m6boot_fit_w",
                                                w,
                                                ".csv", sep=""))
  fwrite(as.data.frame(inspect(m6boot, "rsquare")), paste("SEM/m6boot_r2_w",
                                                          w,
                                                          ".csv", sep="")) 
}
```


## Measurement models

```{r measure-mut, include=F, message=F, warning=F}

num_chrom <- 22

# roulette
roulette <- list(length=num_chrom)
for(c in 1:num_chrom) {
  roulette[[c]] <- fread(paste("~/Data/bgs_lmr/human_data/data/",
                               "mutation_tables/roulette_tbls/",
                               "roulette_tbl_chr", c, ".csv.gz", sep=""))
}
roulette_1kb <- do.call(rbind.data.frame, roulette)

names(roulette_1kb)[4:ncol(roulette_1kb)] <- paste(names(roulette_1kb)[4:ncol(roulette_1kb)],
                                                   "roulette", sep="_")

# Carlson et al 2018
carlson <- list(length=num_chrom)
for(c in 1:num_chrom) {
  carlson[[c]] <- fread(paste("~/Data/bgs_lmr/human_data/data/",
                              "mutation_tables/carlson_tbls/",
                              "carlson_tbl_chr", c, ".csv.gz", sep=""))
}
carlson_1kb <- do.call(rbind.data.frame, carlson)

names(carlson_1kb)[4:ncol(carlson_1kb)] <- paste(names(carlson_1kb)[4:ncol(carlson_1kb)],
                                               "carlson", sep="_")

# gnomad
gnomad <- list(length=num_chrom)
for(c in 1:num_chrom) {
  gnomad[[c]] <- fread(paste("~/Data/bgs_lmr/human_data/data/",
                             "mutation_tables/gnomad_tbls/",
                             "gnomad_tbl_chr", c, ".csv.gz", sep=""))
}
gnomad_1kb <- do.call(rbind.data.frame, gnomad)

names(gnomad_1kb)[4:ncol(gnomad_1kb)] <- paste(names(gnomad_1kb)[4:ncol(gnomad_1kb)],
                                               "gnomad", sep="_")

maps_1kb <- cbind.data.frame(roulette_1kb,
                             carlson_1kb[,4:ncol(carlson_1kb)],
                             gnomad_1kb[,4:ncol(gnomad_1kb)])

maps_1kb <- maps_1kb %>% group_by(chrom) %>% 
                         mutate(bin_10kb=(1:n() - 1) %/% 10,
                                bin_100kb=(1:n() - 1) %/% 100,
                                bin_1Mb=(1:n() - 1) %/% 1000)


maps_10kb <- maps_1kb %>%
             group_by(chrom, bin_10kb) %>% 
             mutate(mean_mut_roulette=mean(avg_mut_roulette),
                    mean_mut_masked_roulette=mean(avg_mut_masked_roulette),
                    mean_mut_carlson=mean(avg_mut_carlson),
                    mean_mut_masked_carlson=mean(avg_mut_masked_carlson),
                    mean_mut_gnomad=mean(avg_mut_gnomad),
                    mean_mut_masked_gnomad=mean(avg_mut_masked_gnomad),
                    sum_sites_roulette=sum(num_sites_roulette),
                    sum_sites_masked_roulette=sum(num_sites_masked_roulette),
                    sum_sites_carlson=sum(num_sites_carlson),
                    sum_sites_masked_carlson=sum(num_sites_masked_carlson),
                    sum_sites_gnomad=sum(num_sites_gnomad),
                    sum_sites_masked_gnomad=sum(num_sites_masked_gnomad),
                    chromStart=min(chromStart),
                    chromEnd=max(chromEnd)) %>%
             distinct(bin_10kb, .keep_all=T) %>%
             ungroup() %>%
             dplyr::select(., c(chrom, chromStart, chromEnd, 
                                mean_mut_roulette,
                                mean_mut_masked_roulette,
                                mean_mut_masked_roulette,
                                mean_mut_carlson,
                                mean_mut_masked_carlson,
                                mean_mut_gnomad,
                                mean_mut_masked_gnomad,
                                sum_sites_roulette,
                                sum_sites_masked_roulette,
                                sum_sites_carlson,
                                sum_sites_masked_carlson,
                                sum_sites_gnomad,
                                sum_sites_masked_gnomad))
    
maps_100kb <- maps_1kb %>%
             group_by(chrom, bin_100kb) %>% 
             mutate(mean_mut_roulette=mean(avg_mut_roulette),
                    mean_mut_masked_roulette=mean(avg_mut_masked_roulette),
                    mean_mut_carlson=mean(avg_mut_carlson),
                    mean_mut_masked_carlson=mean(avg_mut_masked_carlson),
                    mean_mut_gnomad=mean(avg_mut_gnomad),
                    mean_mut_masked_gnomad=mean(avg_mut_masked_gnomad),
                    sum_sites_roulette=sum(num_sites_roulette),
                    sum_sites_masked_roulette=sum(num_sites_masked_roulette),
                    sum_sites_carlson=sum(num_sites_carlson),
                    sum_sites_masked_carlson=sum(num_sites_masked_carlson),
                    sum_sites_gnomad=sum(num_sites_gnomad),
                    sum_sites_masked_gnomad=sum(num_sites_masked_gnomad),
                    chromStart=min(chromStart),
                    chromEnd=max(chromEnd)) %>%
             distinct(bin_100kb, .keep_all=T) %>%
             ungroup() %>%
             dplyr::select(., c(chrom, chromStart, chromEnd, 
                                mean_mut_roulette,
                                mean_mut_masked_roulette,
                                mean_mut_masked_roulette,
                                mean_mut_carlson,
                                mean_mut_masked_carlson,
                                mean_mut_gnomad,
                                mean_mut_masked_gnomad,
                                sum_sites_roulette,
                                sum_sites_masked_roulette,
                                sum_sites_carlson,
                                sum_sites_masked_carlson,
                                sum_sites_gnomad,
                                sum_sites_masked_gnomad))

maps_1Mb <- maps_1kb %>%
             group_by(chrom, bin_1Mb) %>% 
             mutate(mean_mut_roulette=mean(avg_mut_roulette),
                    mean_mut_masked_roulette=mean(avg_mut_masked_roulette),
                    mean_mut_carlson=mean(avg_mut_carlson),
                    mean_mut_masked_carlson=mean(avg_mut_masked_carlson),
                    mean_mut_gnomad=mean(avg_mut_gnomad),
                    mean_mut_masked_gnomad=mean(avg_mut_masked_gnomad),
                    sum_sites_roulette=sum(num_sites_roulette),
                    sum_sites_masked_roulette=sum(num_sites_masked_roulette),
                    sum_sites_carlson=sum(num_sites_carlson),
                    sum_sites_masked_carlson=sum(num_sites_masked_carlson),
                    sum_sites_gnomad=sum(num_sites_gnomad),
                    sum_sites_masked_gnomad=sum(num_sites_masked_gnomad),
                    chromStart=min(chromStart),
                    chromEnd=max(chromEnd)) %>%
             distinct(bin_1Mb, .keep_all=T) %>%
             ungroup() %>%
             dplyr::select(., c(chrom, chromStart, chromEnd, 
                                mean_mut_roulette,
                                mean_mut_masked_roulette,
                                mean_mut_masked_roulette,
                                mean_mut_carlson,
                                mean_mut_masked_carlson,
                                mean_mut_gnomad,
                                mean_mut_masked_gnomad,
                                sum_sites_roulette,
                                sum_sites_masked_roulette,
                                sum_sites_carlson,
                                sum_sites_masked_carlson,
                                sum_sites_gnomad,
                                sum_sites_masked_gnomad))

maps_10kb$mean_mut_roulette[which(maps_10kb$mean_mut_roulette==0)] <- NA
maps_10kb$mean_mut_masked_roulette[which(maps_10kb$mean_mut_masked_roulette==0)] <- NA
maps_10kb$mean_mut_carlson[which(maps_10kb$mean_mut_carlson==0)] <- NA
maps_10kb$mean_mut_masked_carlson[which(maps_10kb$mean_mut_masked_carlson==0)] <- NA
maps_10kb$mean_mut_gnomad[which(maps_10kb$mean_mut_gnomad==0)] <- NA
maps_10kb$mean_mut_masked_gnomad[which(maps_10kb$mean_mut_masked_gnomad==0)] <- NA

maps_100kb$mean_mut_roulette[which(maps_100kb$mean_mut_roulette==0)] <- NA
maps_100kb$mean_mut_masked_roulette[which(maps_100kb$mean_mut_masked_roulette==0)] <- NA
maps_100kb$mean_mut_carlson[which(maps_100kb$mean_mut_carlson==0)] <- NA
maps_100kb$mean_mut_masked_carlson[which(maps_100kb$mean_mut_masked_carlson==0)] <- NA
maps_100kb$mean_mut_gnomad[which(maps_100kb$mean_mut_gnomad==0)] <- NA
maps_100kb$mean_mut_masked_gnomad[which(maps_100kb$mean_mut_masked_gnomad==0)] <- NA

maps_1Mb$mean_mut_roulette[which(maps_1Mb$mean_mut_roulette==0)] <- NA
maps_1Mb$mean_mut_masked_roulette[which(maps_1Mb$mean_mut_masked_roulette==0)] <- NA
maps_1Mb$mean_mut_carlson[which(maps_1Mb$mean_mut_carlson==0)] <- NA
maps_1Mb$mean_mut_masked_carlson[which(maps_1Mb$mean_mut_masked_carlson==0)] <- NA
maps_1Mb$mean_mut_gnomad[which(maps_1Mb$mean_mut_gnomad==0)] <- NA
maps_1Mb$mean_mut_masked_gnomad[which(maps_1Mb$mean_mut_masked_gnomad==0)] <- NA

# testing measurement model 1 Mb
names(human_maps_1Mb)[1] <- "chrom"
df_1Mb <- left_join(maps_1Mb, human_maps_1Mb, by=c("chrom",
                                                   "chromStart",
                                                   "chromEnd"))
tbl_1Mb <- filter(df_1Mb, 
                  coverage > 0.75,
                  chrom != 19,
                  mean_mut_masked_carlson > 1e-8) # removes one weird outlier

p <- ggpairs(dplyr::select(tbl_1Mb, c(mean_mut_masked_carlson,
                                      mean_mut_masked_roulette,
                                      mean_mut_masked_gnomad,
                                      avg_pi)))

mm_model <- "
mu=~mean_mut_masked_roulette + mean_mut_masked_carlson + mean_mut_masked_gnomad + avg_pi
"

fit <- sem(model=mm_model,
           data=tbl_1Mb,
           std.ov=T,
           group="chrom")

summary(fit)
predict(fit)
inspect(fit, "r2")
parameterEstimates(fit)
modificationIndices(fit)
```

# Real Data

```{r real-data, include=F, message=F, warning=F}

num_chr <- 22

tbl_1kb <- list(length=num_chr)
tbl_10kb <- list(length=num_chr)
tbl_100kb <- list(length=num_chr)
tbl_1Mb <- list(length=num_chr)

# YRI pop
for(i in 1:num_chr) {
  tbl_1kb[[i]] <- fread(paste("B_tbl_functional_YRI_chr", i, "_1kb.csv.gz", sep=""))
  tbl_10kb[[i]] <- fread(paste("B_tbl_functional_YRI_chr", i, "_10kb.csv.gz", sep=""))
  tbl_100kb[[i]] <- fread(paste("B_tbl_functional_YRI_chr", i, "_100kb.csv.gz", sep=""))
  tbl_1Mb[[i]] <- fread(paste("B_tbl_functional_YRI_chr", i, "_1Mb.csv.gz", sep=""))
  
  tbl_1kb[[i]]$'#chrom' <- i
  tbl_10kb[[i]]$'#chrom' <- i
  tbl_100kb[[i]]$'#chrom' <- i
  tbl_1Mb[[i]]$'#chrom' <- i
}

big_tbl_1kb <- do.call(rbind.data.frame, tbl_1kb)
big_tbl_10kb <- do.call(rbind.data.frame, tbl_10kb)
big_tbl_100kb <- do.call(rbind.data.frame, tbl_100kb)
big_tbl_1Mb <- do.call(rbind.data.frame, tbl_1Mb)

names(big_tbl_1kb)[1] <- "chr"
names(big_tbl_10kb)[1] <- "chr"
names(big_tbl_100kb)[1] <- "chr"
names(big_tbl_1Mb)[1] <- "chr"


# mut x rec
p1 <- ggplot(data=big_tbl_1kb, aes(y=avg_rec, x=avg_mut)) +
  geom_point(size=3, shape=1) + theme_bw() + geom_smooth() +
  scale_y_continuous(breaks=pretty_breaks()) +
  scale_x_continuous(breaks=pretty_breaks()) +
  labs(title="1kb", x=expression(mu), y="r") +
  theme(axis.title=element_text(size=16), 
        axis.text=element_text(size=12), 
        axis.text.x=element_text(size=12),
        legend.text=element_text(size=16),
        legend.title=element_text(size=16),
        legend.position="bottom")

p2 <- ggplot(data=big_tbl_10kb, aes(y=avg_rec, x=avg_mut)) +
  geom_point(size=3, shape=1) + theme_bw() + geom_smooth() +
  scale_y_continuous(breaks=pretty_breaks()) +
  scale_x_continuous(breaks=pretty_breaks()) +
  labs(title="10kb", x=expression(mu), y="r") +
  theme(axis.title=element_text(size=16), 
        axis.text=element_text(size=12), 
        axis.text.x=element_text(size=12),
        legend.text=element_text(size=16),
        legend.title=element_text(size=16),
        legend.position="bottom")

p3 <- ggplot(data=big_tbl_100kb, aes(y=avg_rec, x=avg_mut)) +
  geom_point(size=3, shape=1) + theme_bw() + geom_smooth() +
  scale_y_continuous(breaks=pretty_breaks()) +
  scale_x_continuous(breaks=pretty_breaks()) +
  labs(title="100kb", x=expression(mu), y="r") +
  theme(axis.title=element_text(size=16), 
        axis.text=element_text(size=12), 
        axis.text.x=element_text(size=12),
        legend.text=element_text(size=16),
        legend.title=element_text(size=16),
        legend.position="bottom")

p4 <- ggplot(data=big_tbl_1Mb, aes(y=avg_rec, x=avg_mut)) +
  geom_point(size=3, shape=1) + theme_bw() + geom_smooth() +
  scale_y_continuous(breaks=pretty_breaks()) +
  scale_x_continuous(breaks=pretty_breaks()) +
  labs(title="1Mb", x=expression(mu), y="r") +
  theme(axis.title=element_text(size=16), 
        axis.text=element_text(size=12), 
        axis.text.x=element_text(size=12),
        legend.text=element_text(size=16),
        legend.title=element_text(size=16),
        legend.position="bottom")

cp1 <- plot_grid(p1, p2, p3, p4, nrow=2)
save_plot("cor_r_u.svg", cp1)
save_plot("cor_r_u.pdf", cp1)

tbl_mut_rec <- as.data.frame(matrix(ncol=4, nrow=3))
names(tbl_mut_rec) <- c("pearson", "spearman", "kendall", "scale")
tbl_mut_rec$scale <- c(1e+4, 1e+5, 1e+6)

tbl_mut_rec[1, 1] <- cor.test(big_tbl_1kb$avg_rec, big_tbl_1kb$avg_mut, method="pearson")$estimate
tbl_mut_rec[2, 1] <- cor.test(big_tbl_10kb$avg_rec, big_tbl_10kb$avg_mut, method="pearson")$estimate
tbl_mut_rec[3, 1] <- cor.test(big_tbl_100kb$avg_rec, big_tbl_100kb$avg_mut, method="pearson")$estimate
tbl_mut_rec[4, 1] <- cor.test(big_tbl_1Mb$avg_rec, big_tbl_1Mb$avg_mut, method="pearson")$estimate

tbl_mut_rec[1, 2] <- cor.test(big_tbl_1kb$avg_rec, big_tbl_1kb$avg_mut, method="spearman")$estimate
tbl_mut_rec[2, 2] <- cor.test(big_tbl_10kb$avg_rec, big_tbl_10kb$avg_mut, method="spearman")$estimate
tbl_mut_rec[3, 2] <- cor.test(big_tbl_100kb$avg_rec, big_tbl_100kb$avg_mut, method="spearman")$estimate
tbl_mut_rec[4, 2] <- cor.test(big_tbl_1Mb$avg_rec, big_tbl_1Mb$avg_mut, method="spearman")$estimate

tbl_mut_rec[1, 3] <- cor.test(big_tbl_1kb$avg_rec, big_tbl_1kb$avg_mut, method="kendall")$estimate
tbl_mut_rec[2, 3] <- cor.test(big_tbl_10kb$avg_rec, big_tbl_10kb$avg_mut, method="kendall")$estimate
tbl_mut_rec[3, 3] <- cor.test(big_tbl_100kb$avg_rec, big_tbl_100kb$avg_mut, method="kendall")$estimate
tbl_mut_rec[4, 3] <- cor.test(big_tbl_1Mb$avg_rec, big_tbl_1Mb$avg_mut, method="kendall")$estimate

m_cor <- pivot_longer(tbl_mut_rec, cols=c("pearson", "spearman"), names_to="method", values_to="cor")

q1 <- ggplot(data=m_cor, aes(y=cor, x=scale, color=method)) +
  geom_point(size=3) + geom_line() + theme_bw() +
  scale_y_continuous(breaks=pretty_breaks(), limits=c(0, 1)) +
  scale_x_log10(breaks=c(1e+4, 1e+5, 1e+6)) +
  labs(title="Correlation mut x rec", x="scale", y="coeff.") +
  theme(axis.title=element_text(size=16), 
        axis.text=element_text(size=12), 
        axis.text.x=element_text(size=12),
        legend.text=element_text(size=16),
        legend.title=element_text(size=16),
        legend.position="bottom")

tbl_B_rec <- as.data.frame(matrix(ncol=4, nrow=3))
names(tbl_B_rec) <- c("pearson", "spearman", "kendall", "scale")
tbl_B_rec$scale <- c(1e+4, 1e+5, 1e+6)

tbl_B_rec[1, 1] <- cor.test(big_tbl_1kb$avg_rec, big_tbl_1kb$B, method="pearson")$estimate
tbl_B_rec[2, 1] <- cor.test(big_tbl_10kb$avg_rec, big_tbl_10kb$B, method="pearson")$estimate
tbl_B_rec[3, 1] <- cor.test(big_tbl_100kb$avg_rec, big_tbl_100kb$B, method="pearson")$estimate
tbl_B_rec[4, 1] <- cor.test(big_tbl_1Mb$avg_rec, big_tbl_1Mb$B, method="pearson")$estimate

tbl_B_rec[1, 2] <- cor.test(big_tbl_1kb$avg_rec, big_tbl_1kb$B, method="spearman")$estimate
tbl_B_rec[2, 2] <- cor.test(big_tbl_10kb$avg_rec, big_tbl_10kb$B, method="spearman")$estimate
tbl_B_rec[3, 2] <- cor.test(big_tbl_100kb$avg_rec, big_tbl_100kb$B, method="spearman")$estimate
tbl_B_rec[4, 2] <- cor.test(big_tbl_1Mb$avg_rec, big_tbl_1Mb$B, method="spearman")$estimate

tbl_B_rec[1, 3] <- cor.test(big_tbl_1kb$avg_rec, big_tbl_1kb$B, method="kendall")$estimate
tbl_B_rec[2, 3] <- cor.test(big_tbl_10kb$avg_rec, big_tbl_10kb$B, method="kendall")$estimate
tbl_B_rec[3, 3] <- cor.test(big_tbl_100kb$avg_rec, big_tbl_100kb$B, method="kendall")$estimate
tbl_B_rec[4, 3] <- cor.test(big_tbl_1Mb$avg_rec, big_tbl_1Mb$B, method="kendall")$estimate

m_cor <- pivot_longer(tbl_B_rec, cols=c("pearson", "spearman"), names_to="method", values_to="cor")

q1 <- ggplot(data=m_cor, aes(y=cor, x=scale, color=method)) +
  geom_point(size=3) + geom_line() + theme_bw() +
  scale_y_continuous(breaks=pretty_breaks(), limits=c(0, 1)) +
  scale_x_log10(breaks=c(1e+4, 1e+5, 1e+6)) +
  labs(title="Correlation B x rec", x="scale", y="coeff.") +
  theme(axis.title=element_text(size=16), 
        axis.text=element_text(size=12), 
        axis.text.x=element_text(size=12),
        legend.text=element_text(size=16),
        legend.title=element_text(size=16),
        legend.position="bottom")


tbl_B_mut <- as.data.frame(matrix(ncol=4, nrow==3))
names(tbl_B_mut) <- c("pearson", "spearman", "kendall", "scale")
tbl_B_mut$scale <- c(1e+4, 1e+5, 1e+6)

tbl_B_mut[1, 1] <- cor.test(big_tbl_1kb$avg_mut, big_tbl_1kb$B, method="pearson")$estimate
tbl_B_mut[2, 1] <- cor.test(big_tbl_10kb$avg_mut, big_tbl_10kb$B, method="pearson")$estimate
tbl_B_mut[3, 1] <- cor.test(big_tbl_100kb$avg_mut, big_tbl_100kb$B, method="pearson")$estimate
tbl_B_mut[4, 1] <- cor.test(big_tbl_1Mb$avg_mut, big_tbl_1Mb$B, method="pearson")$estimate

tbl_B_mut[1, 2] <- cor.test(big_tbl_1kb$avg_mut, big_tbl_1kb$B, method="spearman")$estimate
tbl_B_mut[2, 2] <- cor.test(big_tbl_10kb$avg_mut, big_tbl_10kb$B, method="spearman")$estimate
tbl_B_mut[3, 2] <- cor.test(big_tbl_100kb$avg_mut, big_tbl_100kb$B, method="spearman")$estimate
tbl_B_mut[4, 2] <- cor.test(big_tbl_1Mb$avg_mut, big_tbl_1Mb$B, method="spearman")$estimate

tbl_B_mut[1, 3] <- cor.test(big_tbl_1kb$avg_mut, big_tbl_1kb$B, method="kendall")$estimate
tbl_B_mut[2, 3] <- cor.test(big_tbl_10kb$avg_mut, big_tbl_10kb$B, method="kendall")$estimate
tbl_B_mut[3, 3] <- cor.test(big_tbl_100kb$avg_mut, big_tbl_100kb$B, method="kendall")$estimate
tbl_B_mut[4, 3] <- cor.test(big_tbl_1Mb$avg_mut, big_tbl_1Mb$B, method="kendall")$estimate


p1 <- ggplot(data=big_tbl_1kb, aes(y=del_sites, x=avg_mut, alpha=0.25)) +
  geom_point(size=3, shape=1) + theme_bw() + geom_smooth() +
  scale_y_continuous(breaks=pretty_breaks()) +
  scale_x_continuous(breaks=pretty_breaks()) +
  labs(title=NULL, x=expression(mu), y="# del. sites") +
  theme(axis.title=element_text(size=16), 
        axis.text=element_text(size=12), 
        axis.text.x=element_text(size=12),
        legend.text=element_text(size=16),
        legend.title=element_text(size=16),
        legend.position="bottom")

p2 <- ggplot(data=big_tbl_10kb, aes(y=del_sites, x=avg_mut, alpha=0.25)) +
  geom_point(size=3, shape=1) + theme_bw() + geom_smooth() +
  scale_y_continuous(breaks=pretty_breaks()) +
  scale_x_continuous(breaks=pretty_breaks()) +
  labs(title=NULL, x=expression(mu), y="# del. sites") +
  theme(axis.title=element_text(size=16), 
        axis.text=element_text(size=12), 
        axis.text.x=element_text(size=12),
        legend.text=element_text(size=16),
        legend.title=element_text(size=16),
        legend.position="bottom")

p3 <- ggplot(data=big_tbl_100kb, aes(y=del_sites, x=avg_mut, alpha=0.25)) +
  geom_point(size=3, shape=1) + theme_bw() + geom_smooth() +
  scale_y_continuous(breaks=pretty_breaks()) +
  scale_x_continuous(breaks=pretty_breaks()) +
  labs(title=NULL, x=expression(mu), y="# del. sites") +
  theme(axis.title=element_text(size=16), 
        axis.text=element_text(size=12), 
        axis.text.x=element_text(size=12),
        legend.text=element_text(size=16),
        legend.title=element_text(size=16),
        legend.position="bottom")

p4 <- ggplot(data=big_tbl_1Mb, aes(y=del_sites, x=avg_mut, alpha=0.25)) +
  geom_point(size=3, shape=1) + theme_bw() + geom_smooth() +
  scale_y_continuous(breaks=pretty_breaks()) +
  scale_x_continuous(breaks=pretty_breaks()) +
  labs(title=NULL, x=expression(mu), y="# del. sites") +
  theme(axis.title=element_text(size=16), 
        axis.text=element_text(size=12), 
        axis.text.x=element_text(size=12),
        legend.text=element_text(size=16),
        legend.title=element_text(size=16),
        legend.position="bottom")
ggsave("u_delsites.svg", p1, dpi=1000)

# u x del sites
tbl_del_sites_mut <- as.data.frame(matrix(ncol=4, nrow=3))
names(tbl_del_sites_mut) <- c("pearson", "spearman", "kendall", "scale")
tbl_del_sites_mut$scale <- c(1e+4, 1e+5, 1e+6)

tbl_del_sites_mut[1, 1] <- cor.test(big_tbl_1kb$avg_mut, big_tbl_1kb$del_sites, method="pearson")$estimate
tbl_del_sites_mut[2, 1] <- cor.test(big_tbl_10kb$avg_mut, big_tbl_10kb$del_sites, method="pearson")$estimate
tbl_del_sites_mut[3, 1] <- cor.test(big_tbl_100kb$avg_mut, big_tbl_100kb$del_sites, method="pearson")$estimate
tbl_del_sites_mut[4, 1] <- cor.test(big_tbl_1Mb$avg_mut, big_tbl_1Mb$del_sites, method="pearson")$estimate

tbl_del_sites_mut[1, 2] <- cor.test(big_tbl_1kb$avg_mut, big_tbl_1kb$del_sites, method="spearman")$estimate
tbl_del_sites_mut[2, 2] <- cor.test(big_tbl_10kb$avg_mut, big_tbl_10kb$del_sites, method="spearman")$estimate
tbl_del_sites_mut[3, 2] <- cor.test(big_tbl_100kb$avg_mut, big_tbl_100kb$del_sites, method="spearman")$estimate
tbl_del_sites_mut[4, 2] <- cor.test(big_tbl_1Mb$avg_mut, big_tbl_1Mb$del_sites, method="spearman")$estimate

tbl_del_sites_mut[1, 3] <- cor.test(big_tbl_1kb$avg_mut, big_tbl_1kb$del_sites, method="kendall")$estimate
tbl_del_sites_mut[2, 3] <- cor.test(big_tbl_10kb$avg_mut, big_tbl_10kb$del_sites, method="kendall")$estimate
tbl_del_sites_mut[3, 3] <- cor.test(big_tbl_100kb$avg_mut, big_tbl_100kb$del_sites, method="kendall")$estimate
tbl_del_sites_mut[4, 3] <- cor.test(big_tbl_1Mb$avg_mut, big_tbl_1Mb$del_sites, method="kendall")$estimate

m_cor <- pivot_longer(tbl_del_sites_mut, cols=c("pearson", "spearman"), names_to="method", values_to="cor")

q1 <- ggplot(data=m_cor, aes(y=cor, x=scale, color=method)) +
  geom_point(size=3) + geom_line() + theme_bw() +
  scale_y_continuous(breaks=pretty_breaks()) +
  scale_x_log10(breaks=c(1e+4, 1e+5, 1e+6)) +
  labs(title="Correlation u x # del sites", x="scale", y="coeff.") +
  theme(axis.title=element_text(size=16), 
        axis.text=element_text(size=12), 
        axis.text.x=element_text(size=12),
        legend.text=element_text(size=16),
        legend.title=element_text(size=16),
        legend.position="bottom")

# B x del sites
tbl_del_sites_B <- as.data.frame(matrix(ncol=4, nrow=3))
names(tbl_del_sites_B) <- c("pearson", "spearman", "kendall", "scale")
tbl_del_sites_B$scale <- c(1e+4, 1e+5, 1e+6)

tbl_del_sites_B[1, 1] <- cor.test(big_tbl_1kb$B, big_tbl_1kb$del_sites, method="pearson")$estimate
tbl_del_sites_B[2, 1] <- cor.test(big_tbl_10kb$B, big_tbl_10kb$del_sites, method="pearson")$estimate
tbl_del_sites_B[3, 1] <- cor.test(big_tbl_100kb$B, big_tbl_100kb$del_sites, method="pearson")$estimate
tbl_del_sites_B[4, 1] <- cor.test(big_tbl_1Mb$B, big_tbl_1Mb$del_sites, method="pearson")$estimate

tbl_del_sites_B[1, 2] <- cor.test(big_tbl_1kb$B, big_tbl_1kb$del_sites, method="spearman")$estimate
tbl_del_sites_B[2, 2] <- cor.test(big_tbl_10kb$B, big_tbl_10kb$del_sites, method="spearman")$estimate
tbl_del_sites_B[3, 2] <- cor.test(big_tbl_100kb$B, big_tbl_100kb$del_sites, method="spearman")$estimate
tbl_del_sites_B[4, 2] <- cor.test(big_tbl_1Mb$B, big_tbl_1Mb$del_sites, method="spearman")$estimate

tbl_del_sites_B[1, 3] <- cor.test(big_tbl_1kb$B, big_tbl_1kb$del_sites, method="kendall")$estimate
tbl_del_sites_B[2, 3] <- cor.test(big_tbl_10kb$B, big_tbl_10kb$del_sites, method="kendall")$estimate
tbl_del_sites_B[3, 3] <- cor.test(big_tbl_100kb$B, big_tbl_100kb$del_sites, method="kendall")$estimate
tbl_del_sites_B[4, 3] <- cor.test(big_tbl_1Mb$B, big_tbl_1Mb$del_sites, method="kendall")$estimate

m_cor <- pivot_longer(tbl_del_sites_B, cols=c("pearson", "spearman"), names_to="method", values_to="cor")

q1 <- ggplot(data=m_cor, aes(y=cor, x=scale, color=method)) +
  geom_point(size=3) + geom_line() + theme_bw() +
  scale_y_continuous(breaks=pretty_breaks()) +
  scale_x_log10(breaks=c(1e+4, 1e+5, 1e+6)) +
  labs(title="Correlation B x # del sites", x="scale", y="coeff.") +
  theme(axis.title=element_text(size=16), 
        axis.text=element_text(size=12), 
        axis.text.x=element_text(size=12),
        legend.text=element_text(size=16),
        legend.title=element_text(size=16),
        legend.position="bottom")
q1

# B x del pi0
tbl_del_sites_B <- as.data.frame(matrix(ncol=4, nrow=3))
names(tbl_del_sites_B) <- c("pearson", "spearman", "kendall", "scale")
tbl_del_sites_B$scale <- c(1e+4, 1e+5, 1e+6)

tbl_del_sites_B[1, 1] <- cor.test(big_tbl_1kb$B, big_tbl_1kb$del_sites, method="pearson")$estimate
tbl_del_sites_B[2, 1] <- cor.test(big_tbl_10kb$B, big_tbl_10kb$del_sites, method="pearson")$estimate
tbl_del_sites_B[3, 1] <- cor.test(big_tbl_100kb$B, big_tbl_100kb$del_sites, method="pearson")$estimate
tbl_del_sites_B[4, 1] <- cor.test(big_tbl_1Mb$B, big_tbl_1Mb$del_sites, method="pearson")$estimate

tbl_del_sites_B[1, 2] <- cor.test(big_tbl_1kb$B, big_tbl_1kb$del_sites, method="spearman")$estimate
tbl_del_sites_B[2, 2] <- cor.test(big_tbl_10kb$B, big_tbl_10kb$del_sites, method="spearman")$estimate
tbl_del_sites_B[3, 2] <- cor.test(big_tbl_100kb$B, big_tbl_100kb$del_sites, method="spearman")$estimate
tbl_del_sites_B[4, 2] <- cor.test(big_tbl_1Mb$B, big_tbl_1Mb$del_sites, method="spearman")$estimate

tbl_del_sites_B[1, 3] <- cor.test(big_tbl_1kb$B, big_tbl_1kb$del_sites, method="kendall")$estimate
tbl_del_sites_B[2, 3] <- cor.test(big_tbl_10kb$B, big_tbl_10kb$del_sites, method="kendall")$estimate
tbl_del_sites_B[3, 3] <- cor.test(big_tbl_100kb$B, big_tbl_100kb$del_sites, method="kendall")$estimate
tbl_del_sites_B[4, 3] <- cor.test(big_tbl_1Mb$B, big_tbl_1Mb$del_sites, method="kendall")$estimate

m_cor <- pivot_longer(tbl_del_sites_B, cols=c("pearson", "spearman"), names_to="method", values_to="cor")

q1 <- ggplot(data=m_cor, aes(y=cor, x=scale, color=method)) +
  geom_point(size=3) + geom_line() + theme_bw() +
  scale_y_continuous(breaks=pretty_breaks()) +
  scale_x_log10(breaks=c(1e+4, 1e+5, 1e+6)) +
  labs(title="Correlation B x # del sites", x="scale", y="coeff.") +
  theme(axis.title=element_text(size=16), 
        axis.text=element_text(size=12), 
        axis.text.x=element_text(size=12),
        legend.text=element_text(size=16),
        legend.title=element_text(size=16),
        legend.position="bottom")
q1
```