---
title: "Mutation Rate Variation Pt. II (paper 2)"
author: "Gustavo V. Barroso"
date: "`r Sys.Date()`"
output:
  pdf_document:
  toc: true
number_sections: true
toc_depth: 1
---
  
```{r setup, include=F, message=F, warning=F}

knitr::opts_chunk$set(echo=TRUE)

library(tidyverse)
library(bigsnpr) # for seq_log
library(data.table)

knitr::opts_knit$set(root.dir="~/Data/bgs_lmr/sim_data/lookup_tables/")
```

# Creating folders and files

This builds a look-up table on a dense parameter grid that will allow for fast 
assembly of whole chromosomes

```{r, params, include=F, eval=T, results=F, message=F, warning=F}

Ns <- paste(10000, sep=";") # from present to past
Ts <- paste(0, sep=";") # from present (0) to past, in generations
uR <- 1e-8 # B-values are independent of uR from the mut map, let's assume flat
# rounding is required so that directory names don't diverge due to precision
uL <- round(seq_log(5e-11, 5e-8, 3*11+1), digits=15)
r <- round(c(0, seq_log(1e-8, 1e-1, 71)), digits=12)
s <- round(-c(0, seq_log(1e-6, 1e-3, 3*11+1)), digits=12)

params <- setDT(crossing(Ns, Ts, r, s, uL, uR))
n_models <- nrow(params)

fwrite(as.data.frame(r), "r.csv", col.names=F)
fwrite(as.data.frame(s), "s.csv", col.names=F)
fwrite(as.data.frame(uL), "uL.csv", col.names=F)
fwrite(params, "params.csv")
```

```{bash dirs, engine.opts='-i', include=F, eval=T, results=F, warning=F}

# NOTE: this chunk requires that the directories do not already exist
source ~/.bashrc
  
cat r.csv | while IFS=, read -r r
do
  mkdir r_$r
  cat s.csv | while IFS=, read -r s
  do
    mkdir r_$r/s_$s
  done
done

find -mindepth 2 -type d -exec cp opt.bpp {} \;

# makes directory to store purely neutral models to get pi0 predictions
mkdir demo
cp opt.bpp demo
```

Creates files that will be used as input for moments++

```{r, demes, include=F, eval=T, results=F, message=F, warning=F}

params <- fread("params.csv")
n_models <- nrow(params)

# writes demes files specifying models (one demography, r X s)
options(digits=16) # precision
for(i in 1:n_models) {
  
  if(i %% 10000 == 0) {
    print(Sys.time())
    cat(paste(i, "\n"))
  }
  
  Ns <- as.numeric(strsplit(as.character(params[i,]$Ns), ";")[[1]])
  Ts <- as.integer(strsplit(as.character(params[i,]$Ts), ";")[[1]])
  uL <- as.numeric(params[i,]$uL)
  uR <- as.numeric(params[i,]$uR)
  r <- as.numeric(params[i,]$r)
  s <- as.numeric(params[i,]$s)
  
  name <- paste("model_", i, sep="")
  sink(paste("r_", r, "/s_", s,"/", name, ".yaml", sep=""))
  
  cat("time_units: generations\n")
  cat("demes:\n")
  cat("  - name: X\n")
  cat("    epochs:\n")
  for(j in length(Ts):1) {
    cat("      - end_time: ")
    cat(Ts[j])
    cat("\n")
    cat("        start_size: ")
    cat(Ns[j])
    cat("\n")
  }
  cat("metadata:\n")
  cat("  - name: mutation\n")
  cat("    left_factor: " )
  cat(uL / uR)
  cat("\n")
  cat("    epochs:\n")
  cat("      - end_time: 0\n")
  cat("        rates: [")
  cat(uR)
  cat("]\n")
  cat("  - name: recombination\n")
  cat("    epochs:\n")
  cat("      - end_time: 0\n")
  cat("        rates: [")
  cat(r)
  cat("]\n")
  cat("  - name: selection\n")
  cat("    start_time: .inf\n")
  cat("    epochs:\n")
  cat("      - end_time: 0\n")
  cat("        rates: [")
  cat(format(s, scientific=F))
  cat("]\n")
  
  sink()
}

# creates the (purely neutral) demographic model
# single u* and r because we're only interested in pi0 here (for B-values)
Ns <- as.numeric(strsplit(as.character(params[i,]$Ns), ";")[[1]])
Ts <- as.numeric(strsplit(as.character(params[i,]$Ts), ";")[[1]])
uR <- as.numeric(unique(params$uR))

sink(paste("demo/model_neutral.yaml", sep=""))
cat("time_units: generations\n")
cat("demes:\n")
cat("  - name: X\n")
cat("    epochs:\n")
for(j in length(Ts):1) {
  cat("      - end_time: ")
  cat(Ts[j])
  cat("\n")
  cat("        start_size: ")
  cat(Ns[j])
  cat("\n")
}
cat("metadata:\n")
cat("  - name: mutation\n")
cat("    left_factor: 1\n") 
cat("    epochs:\n")
cat("      - end_time: 0\n")
cat("        rates: [")
cat(uR)
cat("]\n")
cat("  - name: recombination\n")
cat("    epochs:\n")
cat("      - end_time: 0\n")
cat("        rates: [0]\n")
cat("  - name: selection\n")
cat("    start_time: .inf\n")
cat("    epochs:\n")
cat("      - end_time: 0\n")
cat("        rates: [0]\n")
sink()
```

# Running moments++

Computes summary statistics for each parameter combination in lookup_tbl

```{bash moments++, engine.opts='-i', include=F, eval=F, results=F, warning=F}
#!/bin/bash

i=1
nr=$(wc -l r.csv | cut -d ' ' -f 1)

for r in r_*
do
  date
  echo "$i of $nr"
  cd $r
  for s in s_*
  do
    cd $s
    for f in model_*.yaml
    do
	    N=$(sed '6q;d' $f | cut -d : -f 2 | cut -d ' ' -f 2)
      x=$(sed '21q;d' $f | cut -d : -f 2 | cut -d [ -f 2 | cut -d ] -f 1)
      s=$(echo "scale=10; $x" | bc)
	    a=$(echo "scale=10; $N * $s" | bc)
      if [[ $(echo "$a < -1" | bc) == 1 ]]
	    then
	      b=${a%.*}
	      Ns=$(($b * -2))
	      if [[ $Ns -lt 30 ]]
	      then
  	      momentspp params=opt.bpp F=$f O=70 > /dev/null
  	    elif [[ $Ns -lt 70 ]]
  	    then 
  	      momentspp params=opt.bpp F=$f O=150 > /dev/null
  	    elif [[ $Ns -lt 150 ]]
  	    then    
  	      momentspp params=opt.bpp F=$f O=300 > /dev/null
  	    else
  	      momentspp params=opt.bpp F=$f O=400 > /dev/null
  	    fi
      else
	      momentspp params=opt.bpp F=$f O=10 > /dev/null
	    fi
    done
    cd ..
  done
  cd ..
  i=$(($i + 1))
done

# get pi0 over time
cd demo
for f in *yaml
do 
  momentspp params=opt.bpp F=$f O=2
done
```

Collects lookup table with Hr values on the parameter grid

```{r, grid, include=F, eval=F, results=F, message=F, warning=F}

params <- fread("params.csv")
n_models <- nrow(params)

lookup_r <- unique(params$r)
lookup_s <- unique(params$s)

# Het stats over time (include steady-state and present-time)
hrs <- as.data.frame(matrix(nrow=n_models, ncol=2))
hls <- as.data.frame(matrix(nrow=n_models, ncol=2))

names(hrs) <- c("Hr", "Order")
names(hls) <- c("Hl", "Order")

for(i in 1:length(lookup_r)) {
  cat(paste(i, "\n"))
  print(Sys.time())
  for(j in 1:length(lookup_s)) {
    # one fname per uL in params.csv
    fnames <- list.files(paste("r_", lookup_r[i], "/s_", lookup_s[j], sep=""),
                           pattern="*_expectations.txt", full.names=TRUE)
    
    idx <- 0  
    for(l in 1:length(fnames)) {
      moms <- fread(fnames[l])
      str_sides <- strsplit(fnames[l], "_O_")[[1]]
      idx <- as.numeric(strsplit(str_sides[1], "model_")[[1]][2])
      order <- as.numeric(strsplit(str_sides[2], "_e")[[1]][1])
      left <- filter(moms, V1=="Hl_0_0")
      right <- filter(moms, V1=="Hr_0_0")
    
      hls[idx,] <- c(t(dplyr::select(left, V3)), order)
      hrs[idx,] <- c(t(dplyr::select(right, V3)), order)
    }
  }
}

lookup_tbl <- cbind.data.frame(params, hls)
lookup_tbl$Generation <- 0 # we work with steady-state in this project
lookup_tbl$Hr <- hrs$Hr

# get pi0 from demography
fnames <- list.files("demo", pattern="*_expectations.txt", full.names=TRUE)
moms <- fread(fnames[1])
str_sides <- strsplit(fnames[1], "_O_")[[1]]
order <- as.numeric(strsplit(str_sides[2], "_e")[[1]][1])
lookup_tbl$pi0 <- filter(moms, V1=="Hr_0_0")$V3

lookup_tbl$B <- lookup_tbl$Hr / lookup_tbl$pi0
lookup_tbl$piN_pi0 <- lookup_tbl$Hl / lookup_tbl$pi0
lookup_tbl$piN_piS <- lookup_tbl$Hl / lookup_tbl$Hr

fwrite(lookup_tbl, "lookup_tbl.csv.gz", compress="gzip") 
```

# Building chr models

Builds a table of models from which to build genomic landscapes (bgs_maps.R)
and compute expected nucleotide diversity (bgs_pi.R)

```{r, landscapes, include=F, eval=F, results=F, message=F, warning=F}

# Gamma dist params
shapes_rate_u <- c(3, 10, 30, 100)
shapes_rate_r <- c(1, 3, 10)
mean_r <- 1e-8
mean_u <- 1e-8

# Geom dist params
avg_rec_spans <- c(1e+3, 1e+4, 1e+5)
avg_mut_spans <- c(5e+1, 2e+2, 1e+3, 1e+4, 1e+5)
exon_lengths <- 1000
num_exons <- c(600, 1500, 3000)
L <- 3e+7

models <- setDT(crossing(L, num_exons, exon_lengths,
                         mean_r, mean_u, shapes_rate_u, shapes_rate_r,
                         avg_rec_spans, avg_mut_spans))
models$model <- 1:nrow(models)
n_models <- nrow(models)

fwrite(models, "models.csv")
```

```{bash folders, engine.opts='-i', include=F, eval=F, results=F, warning=F}

source ~/.bashrc

for((i=1;i<=n_models;i++))
do 
  mkdir model_$i
done

for((i=1;i<=n_models;i++))
do
  cd model_$i
  
  for((j=1;j<=5;j++))
  do
    mkdir rep_$j
  done
  
  cd ..
done
```

# TODO 
Add Python chunk with fwdpy11 simulations
