---
title: "Benchmarking Moments++"
author: "Gustavo V. Barroso"
date: "`r Sys.Date()`"
output:
  pdf_document:
  toc: true
number_sections: true
toc_depth: 3
---
  
```{r setup, include=F, message=F, warning=F}

knitr::opts_chunk$set(echo=TRUE)

library(tidyverse)
library(cowplot)
library(stringr)
library(RColorBrewer)
library(scales)
library(plotly)
library(MASS)
library(lmtest)
library(nlme)
library(car)
library(bigsnpr)
library(data.table)
library(GenomicRanges)

library(reticulate)
use_python("/usr/bin/python3")

scale.4d <- function(x) sprintf("%.4f", x)

knitr::opts_knit$set(root.dir="~/Devel/momentspp/benchmarking/")
```

This is an empty python chunk.
Apparently this must exist in order for reticulate to be able to read
R data frames in future python chunks.

```{python}
```

This creates the directories

```{bash run-moments++, engine.opts='-i', include=F, eval=F, results=F, message=F, warning=F}

source ~/.bashrc

mkdir neutrality
mkdir selection

cd neutrality
mkdir migration
mkdir no_migration

cd ../selection
mkdir migration
mkdir no_migration
mkdir single-pop

cd single-pop
mkdir bench
mkdir simple_bgs
mkdir landscapes

cd simple_bgs
mkdir mult_neutral
mkdir single_neutral
mkdir single_neutral_single_selected

cd ../../../
find -type d -exec cp opt.bpp {} \;
cp sim_sel.bpp selection/single-pop/bench
```

# Neutrality

Here we benchmark moments++ against moments.LD when s=0

## No-migration

### Building Demes files

```{r, demes-neutral-2-pop, include=F, eval=F, results=F, message=F, warning=F}

# parameters
N1 <- 10000
N2 <- c(N1, N1 / 10)
u <- 1e-7
s <- 0
r  <- c(1e-4, 1e-5, 1e-6, 1e-7, 0)
split_time <- c(500, 2000)

params <- crossing(N1, N2, u, r, s, split_time)
params$scenario <- 1:nrow(params)
n_models <- nrow(params)

write.csv(params, "neutrality/no_migration/params.csv", quote=F, row.names=F)

Sys.setenv(num_models=n_models)

py$params <- params # stores df to be used in python sections
py$n_models <- as.integer(n_models + 1) # because python is indexed from zero

# writes files in strict Demes format (no metadata)
for(i in 1:n_models) {
  
  name <- paste("model_", i, sep="")
  sink(paste("neutrality/no_migration/", name, "_demes.yaml", sep=""))
  
  cat("time_units: generations\n")
  cat("demes:\n")
  cat("  - name: X\n")
  cat("    epochs:\n")
  cat("      - end_time: ")
  cat(params$split_time[i])
  cat("\n")
  cat("        start_size: ")
  cat(params$N1[i])
  cat("\n")
  cat("  - name: A\n")
  cat("    ancestors: [X]\n")
  cat("    epochs:\n")
  cat("      - end_time: 0\n")
  cat("        start_size: ")
  cat(params$N1[i])
  cat("\n")
  cat("  - name: B\n")
  cat("    ancestors: [X]\n")
  cat("    epochs:\n")
  cat("      - end_time: 0\n")
  cat("        start_size: ")
  cat(params$N2[i])
  cat("\n")
  
  sink()
}

# writes files with enriched metadata, moments++ can parse it
for(i in 1:n_models) {
  
  name <- paste("model_", i, sep="")
  sink(paste("neutrality/no_migration/", name, ".yaml", sep=""))
  
  cat("time_units: generations\n")
  cat("demes:\n")
  cat("  - name: X\n")
  cat("    epochs:\n")
  cat("      - end_time: ")
  cat(params$split_time[i])
  cat("\n")
  cat("        start_size: ")
  cat(params$N1[i])
  cat("\n")
  cat("  - name: A\n")
  cat("    ancestors: [X]\n")
  cat("    epochs:\n")
  cat("      - end_time: 0\n")
  cat("        start_size: ")
  cat(params$N1[i])
  cat("\n")
  cat("  - name: B\n")
  cat("    ancestors: [X]\n")
  cat("    epochs:\n")
  cat("      - end_time: 0\n")
  cat("        start_size: ")
  cat(params$N2[i])
  cat("\n")
  cat("metadata:\n")
  cat("  - name: mutation\n")
  cat("    epochs:\n")
  cat("      - end_time: 0\n")
  cat("        rates: [")
  cat(params$u[i])
  cat("]\n")
  cat("  - name: recombination\n")
  cat("    epochs:\n")
  cat("      - end_time: 0\n")
  cat("        rates: [")
  cat(params$r[i])
  cat("]\n")
  cat("  - name: selection\n")
  cat("    start_time: .inf\n")
  cat("    epochs:\n")
  cat("      - end_time: 0\n")
  cat("        rates: [0]\n")
  
  sink()
}

```

### Running moments++

```{bash create-dirs, engine.opts='-i', include=F, eval=F, results=F, message=F, warning=F}

source ~/.bashrc

cd neutrality/no_migration

for((i=1;i<=$num_models;i++))
do
  momentspp params=opt.bpp F=model_$i.yaml O=1
done

cd ../..
```

### Running moments.LD

```{python run-momentsLD, include=F, eval=F, results=F}

import moments.LD
import demes
import sys, os
import numpy as np

def simulate_ld(f, sampled_demes, u, r):
    g = demes.load(f)
    y = moments.LD.LDstats.from_demes(g, sampled_demes, r=r, u=u)
    return y
  
def write_column(y, out_file):
    mld, mh = moments.LD.Util.moment_names(y.num_pops)
    # write D2
    for m, v in zip(mld, y[0]):
        if m.split("_")[0] == "DD":
            out_file.write(str(v) + "\n")
    # write Dz
    for m, v in zip(mld, y[0]):
        if m.split("_")[0] == "Dz":
            out_file.write(str(v) + "\n")
    # write H
    for m, v in zip(mh, y[-1]):
        out_file.write(str(v) + "\n")
    # write pi2
    for m, v in zip(mld, y[0]):
        if m.split("_")[0] == "pi2":
            out_file.write(str(v) + "\n")

for i in range(1, n_models):
  with open(f"neutrality/no_migration/moments_LD_model_{i}.csv", "w+") as fout:
      f = f"neutrality/no_migration/model_{i}_demes.yaml"
      pops = ["A", "B"]
      u = params["u"][i - 1] # python is indexed from zero
      r = params["r"][i - 1] # python is indexed from zero
      y = simulate_ld(f, pops, u, r)
      stats = moments.LD.Util.moment_names(len(pops))
      l = ",".join(stats[0] + stats[1]) + "\n"
      fout.write(l)
      l = ",".join([str(_) for _ in y[0]] + [str(_) for _ in y[1]]) + "\n"
      fout.write(l)
```

### Plots

```{r neutral-bench, include=T, eval=T, results=F, message=F, warning=F}

core_stats <- c("DD_1_1", 
                "DD_1_2", 
                "DD_2_2",
                "Dr_1_1_(1-2p1)^1",
                "Dr_1_1_(1-2p2)^1",
                "Dr_1_2_(1-2p2)^1",
                "Dr_2_1_(1-2p1)^1", 
                "Dr_2_1_(1-2p2)^1",
                "Dr_2_2_(1-2p2)^1",
                "pi2_1_1_1_1", 
                "pi2_1_1_1_2",
                "pi2_1_1_2_2", 
                "pi2_1_2_1_2", 
                "pi2_1_2_2_2", 
                "pi2_2_2_2_2",
                "Hl_1_1", 
                "Hl_1_2",
                "Hl_2_2")

# constant Ne
mpp <- numeric()
for(i in 1:n_models) {
  vals <- read.table(paste("neutrality/no_migration/model_", i, "_expectations.txt", sep=""))
  vals <- subset(vals, V1 %in% core_stats)
  vals <- slice(vals, 1:9, 13:18, 10:12) # to match moments.LD output and plot
  mpp <- c(mpp, vals$V3)
}

tbl <- as.data.frame(mpp)
tbl$stats <- core_stats
scenario <- NULL
for(i in 1:n_models) { 
  scenario <- c(scenario, rep(i, length(core_stats)))
}
tbl$scenario <- scenario

mLD <- numeric()
for(i in 1:n_models) {
  vals <- as.numeric(read.csv(paste("neutrality/no_migration/moments_LD_model_", i, ".csv", sep=""), header=T))
  mLD <- c(mLD, vals)
}

tbl$mLD <- mLD
tbl$ratio <- tbl$mpp / tbl$mLD

molten_tbl <- pivot_longer(tbl, cols=starts_with("m"))

p1 <- ggplot(data=molten_tbl, aes(x=stats, y=ratio, shape=as.factor(scenario)))
p1 <- p1 + geom_point(size=3) + theme_bw()
p1 <- p1 + geom_hline(yintercept=0.99, linetype=2)
p1 <- p1 + geom_hline(yintercept=1.01, linetype=2)
p1 <- p1 + labs(title="++ / LD", x="Moment", y="Ratio", shape="Model")
p1 <- p1 + scale_shape_manual(values=(0:(n_models - 1)))
p1 <- p1 + theme(axis.title=element_text(size=12),
                 axis.text=element_text(size=10),
                 axis.text.x=element_text(angle=90, size=8, vjust=0.5, hjust=1.0),
                 legend.position="bottom")

save_plot("neutrality/no_migration/ratios_no-mig.png", p1, base_height=12, base_width=12)
```

## Migration (TODO)


# Selection 1-population
## Benchmark 

Here we benchmark against moments.TwoLocus and TwoLocusSim

### Building Demes files

```{r, demes-selected-1-pop, include=F, eval=F, results=F, message=F, warning=F}

N <- c(1000, 10000)
u <- 1e-8
r  <- c(1e-4, 1e-5, 1e-6, 1e-7, 1e-8)
s <- c(0, -1e-04, -5e-4, -1e-3)

params <- crossing(N, u, r, s)
params$scenario <- 1:nrow(params)
n_models <- nrow(params)

write.csv(params, "selection/single-pop/bench/params.csv", quote=F, row.names=F)

Sys.setenv(num_models=n_models)

py$params <- params # stores df to be used in python sections
py$n_models <- as.integer(n_models + 1) # because python is indexed from zero

for(i in 1:n_models) {
  
  name <- paste("model_", i, sep="")
  sink(paste("selection/single-pop/bench/", name, ".yaml", sep=""))
  
  cat("time_units: generations\n")
  cat("demes:\n")
  cat("  - name: X\n")
  cat("    epochs:\n")
  cat("      - end_time: 0\n")
  cat("        start_size: ")
  cat(params$N[i])
  cat("\n")
  cat("metadata:\n")
  cat("  - name: mutation\n")
  cat("    epochs:\n")
  cat("      - end_time: 0\n")
  cat("        rates: [")
  cat(params$u[i])
  cat("]\n")
  cat("  - name: recombination\n")
  cat("    epochs:\n")
  cat("      - end_time: 0\n")
  cat("        rates: [")
  cat(params$r[i])
  cat("]\n")
  cat("  - name: selection\n")
  cat("    start_time: .inf\n")
  cat("    epochs:\n")
  cat("      - end_time: 0\n")
  cat("        rates: [")
  cat(params$s[i])
  cat("]\n")
  
  sink()
}

```

### Running TwoLocusSim

```{bash run-twoLocusSim, engine.opts='-i', include=F, eval=F, results=F, message=F, warning=F}

source ~/.bashrc

cd selection/single-pop/bench

sed 1d params.csv | while IFS=, read -r N u r s m
do
  twoLocusSim params=sim_sel.bpp N=$N u=$u r=$r s=$s label=$m tag=rep_1 &
  twoLocusSim params=sim_sel.bpp N=$N u=$u r=$r s=$s label=$m tag=rep_2 &
  twoLocusSim params=sim_sel.bpp N=$N u=$u r=$r s=$s label=$m tag=rep_3 &
  twoLocusSim params=sim_sel.bpp N=$N u=$u r=$r s=$s label=$m tag=rep_4 &
  twoLocusSim params=sim_sel.bpp N=$N u=$u r=$r s=$s label=$m tag=rep_5 &
  twoLocusSim params=sim_sel.bpp N=$N u=$u r=$r s=$s label=$m tag=rep_6 &
  twoLocusSim params=sim_sel.bpp N=$N u=$u r=$r s=$s label=$m tag=rep_7 &
  twoLocusSim params=sim_sel.bpp N=$N u=$u r=$r s=$s label=$m tag=rep_8 &
  twoLocusSim params=sim_sel.bpp N=$N u=$u r=$r s=$s label=$m tag=rep_9 &
  twoLocusSim params=sim_sel.bpp N=$N u=$u r=$r s=$s label=$m tag=rep_10;
  
  wait
  
  twoLocusSim params=sim_sel.bpp N=$N u=$u r=$r s=$s label=$m tag=rep_11 &
  twoLocusSim params=sim_sel.bpp N=$N u=$u r=$r s=$s label=$m tag=rep_12 &
  twoLocusSim params=sim_sel.bpp N=$N u=$u r=$r s=$s label=$m tag=rep_13 &
  twoLocusSim params=sim_sel.bpp N=$N u=$u r=$r s=$s label=$m tag=rep_14 &
  twoLocusSim params=sim_sel.bpp N=$N u=$u r=$r s=$s label=$m tag=rep_15 &
  twoLocusSim params=sim_sel.bpp N=$N u=$u r=$r s=$s label=$m tag=rep_16 &
  twoLocusSim params=sim_sel.bpp N=$N u=$u r=$r s=$s label=$m tag=rep_17 &
  twoLocusSim params=sim_sel.bpp N=$N u=$u r=$r s=$s label=$m tag=rep_18 &
  twoLocusSim params=sim_sel.bpp N=$N u=$u r=$r s=$s label=$m tag=rep_19 &
  twoLocusSim params=sim_sel.bpp N=$N u=$u r=$r s=$s label=$m tag=rep_20
  
  wait 
  
  twoLocusSim params=sim_sel.bpp N=$N u=$u r=$r s=$s label=$m tag=rep_21 &
  twoLocusSim params=sim_sel.bpp N=$N u=$u r=$r s=$s label=$m tag=rep_22 &
  twoLocusSim params=sim_sel.bpp N=$N u=$u r=$r s=$s label=$m tag=rep_23 &
  twoLocusSim params=sim_sel.bpp N=$N u=$u r=$r s=$s label=$m tag=rep_24 &
  twoLocusSim params=sim_sel.bpp N=$N u=$u r=$r s=$s label=$m tag=rep_25 &
  twoLocusSim params=sim_sel.bpp N=$N u=$u r=$r s=$s label=$m tag=rep_26 &
  twoLocusSim params=sim_sel.bpp N=$N u=$u r=$r s=$s label=$m tag=rep_27 &
  twoLocusSim params=sim_sel.bpp N=$N u=$u r=$r s=$s label=$m tag=rep_28 &
  twoLocusSim params=sim_sel.bpp N=$N u=$u r=$r s=$s label=$m tag=rep_29 &
  twoLocusSim params=sim_sel.bpp N=$N u=$u r=$r s=$s label=$m tag=rep_30
done

cd ../../..

```

### Running moments++

```{bash run-moments++-1-pop, engine.opts='-i', include=F, eval=T, results=F, message=F, warning=F}

source ~/.bashrc

cd selection/single-pop/bench

for((i=1;i<=$num_models;i++))
do
  momentspp params=opt.bpp F=model_$i.yaml L=model_$i O=25
done

cd ../../..
```

### Running moments.TwoLocus

For benchmarking the single population case, we use equilibrium demography:

```{python run-momentsTwoLocus, include=F, eval=F, results=F}

import moments.TwoLocus
import sys, os
import numpy as np
    
for i in range(1, n_models):
  with open(f"selection/single-pop/bench/moments_twoLocus_n40_model_{i}.csv", "w+") as fout:
      n_samples = 40
      
      # i - 1 because python is indexed from zero
      Ne = params["N"][i - 1] 
      u = params["u"][i - 1] 
      r = params["r"][i - 1] 
      s = params["s"][i - 1]
      
      gammas = [2 * Ne * s, 2 * Ne * s, 0] # sel on [AB, Ab, aB] haplotypes, resp
      rho = 4 * Ne * r
      theta = 4 * Ne * u
      
      F = moments.TwoLocus.Demographics.equilibrium(n_samples, 
                                                    rho=rho, 
                                                    theta=theta, 
                                                    sel_params=gammas)
        
      fs = moments.Spectrum(moments.LinearSystem_1D.steady_state_1D(n_samples, gamma=2 * Ne * s) * theta)    
      fs_left = moments.Spectrum(F[0, :, 0])
      fs_right = moments.Spectrum(F[0, 0, :])
      
      fout.write(f"D^2 = {F.D2():}\n")
      fout.write(f"Dz = {F.Dz():}\n")
      fout.write(f"Hl = {fs_left.pi()}\n")
      fout.write(f"Hr = {fs_right.pi()}\n")
      fout.write(f"pi2 = {F.pi2():}\n")
      
      assert np.allclose(fs_left, fs)

for i in range(1, n_models):
  with open(f"selection/single-pop/bench/moments_twoLocus_n60_model_{i}.csv", "w+") as fout:
      n_samples = 60
      
      # i - 1 because python is indexed from zero
      Ne = params["N"][i - 1] 
      u = params["u"][i - 1] 
      r = params["r"][i - 1] 
      s = params["s"][i - 1]
      
      gammas = [2 * Ne * s, 2 * Ne * s, 0] # sel on [AB, Ab, aB] haplotypes, resp
      rho = 4 * Ne * r
      theta = 4 * Ne * u
      
      F = moments.TwoLocus.Demographics.equilibrium(n_samples, 
                                                    rho=rho, 
                                                    theta=theta, 
                                                    sel_params=gammas)
        
      fs = moments.Spectrum(moments.LinearSystem_1D.steady_state_1D(n_samples, gamma=2 * Ne * s) * theta)    
      fs_left = moments.Spectrum(F[0, :, 0])
      fs_right = moments.Spectrum(F[0, 0, :])
      
      fout.write(f"D^2 = {F.D2():}\n")
      fout.write(f"Dz = {F.Dz():}\n")
      fout.write(f"Hl = {fs_left.pi()}\n")
      fout.write(f"Hr = {fs_right.pi()}\n")
      fout.write(f"pi2 = {F.pi2():}\n")
      
      assert np.allclose(fs_left, fs)
```

### Plots

```{r, plots-selected-1-pop, include=T, eval=F, results=F, message=F, warning=F}

commom_stats <- c("DD_0_0",
                  "Dr_0_0_(1-2p0)^1",
                  "Hl_0_0", 
                  "Hr_0_0",
                  "pi2_0_0_0_0")

# from moments++
expectation <- numeric()
for(i in 1:n_models) {
  vals <- read.table(paste("selection/single-pop/bench/model_", i, "_expectations.txt", sep=""))
  vals <- subset(vals, V1 %in% commom_stats)$V3
  expectation <- c(expectation, vals)
}

tbl <- as.data.frame(expectation)
tbl$stats <- commom_stats
scenario <- NULL
for(i in 1:n_models) { 
  scenario <- c(scenario, rep(i, length(commom_stats)))
}
tbl$scenario <- scenario

mpp <- full_join(tbl, params, by="scenario")
mpp$alpha <- 2 * mpp$s * mpp$N

# moments.TwoLocus was fitted with 2 sample sizes, n=40 and n=60
# n = 40
expectation <- numeric()
for(i in 1:n_models) {
  vals <- read.table(paste("selection/single-pop/bench/moments_twoLocus_n40_model_", i, ".csv", sep=""))$V3
  expectation <- c(expectation, vals)
}

tbl_py_n40 <- as.data.frame(expectation)
tbl_py_n40$stats <- commom_stats
tbl_py_n40$scenario <- scenario
tbl_py_n40$sample_size <- 40
tbl_py_n40 <- full_join(tbl_py_n40, params, by="scenario")

# from 2p(1-p) to p(1-p)
tbl_py_n40[tbl_py_n40$stats=="Hl_0_0", 1] <- tbl_py_n40[tbl_py_n40$stats=="Hl_0_0", 1] / 2
tbl_py_n40[tbl_py_n40$stats=="Hr_0_0", 1] <- tbl_py_n40[tbl_py_n40$stats=="Hr_0_0", 1] / 2

# n = 60
expectation <- numeric()
for(i in 1:n_models) {
  vals <- read.table(paste("selection/single-pop/bench/moments_twoLocus_n60_model_", i, ".csv", sep=""))$V3
  expectation <- c(expectation, vals)
}

tbl_py_n60 <- as.data.frame(expectation)
tbl_py_n60$stats <- commom_stats
tbl_py_n60$scenario <- scenario
tbl_py_n60$sample_size <- 60
tbl_py_n60 <- full_join(tbl_py_n60, params, by="scenario")

# from 2p(1-p) to p(1-p)
tbl_py_n60[tbl_py_n60$stats=="Hl_0_0", 1] <- tbl_py_n60[tbl_py_n60$stats=="Hl_0_0", 1] / 2
tbl_py_n60[tbl_py_n60$stats=="Hr_0_0", 1] <- tbl_py_n60[tbl_py_n60$stats=="Hr_0_0", 1] / 2

tbl_py_ns <- bind_rows(tbl_py_n40, tbl_py_n60)
tbl_pp <- bind_rows(mpp, mpp)

tbl_pp <- tbl_pp %>% mutate(ratio = expectation / tbl_py_ns$expectation)
tbl_pp$sample_size <- tbl_py_ns$sample_size
  
p1 <- ggplot(data=filter(tbl_pp, stats==commom_stats[1]), aes(x=r, y=ratio, shape=as.factor(N), color=as.factor(sample_size)))
p1 <- p1 + geom_point(size=3) + geom_line() + theme_bw() + facet_wrap(~s, nrow=1)
p1 <- p1 + scale_shape_manual(values=0:length((unique(tbl_pp))))
p1 <- p1 + scale_x_log10(breaks=r) + scale_y_continuous(labels=scale.4d)
p1 <- p1 + labs(title=NULL, x=NULL, y=expression(D^2))
p1 <- p1 + geom_hline(yintercept=1, linetype="dashed", color="red")
p1 <- p1 + theme(axis.title=element_text(size=12),
                 axis.text=element_text(size=10),
                 axis.title.x=element_blank(),
                 axis.text.x=element_blank(),
                 axis.ticks.x=element_blank(),
                 legend.position="none")
p1

p2 <- ggplot(data=filter(tbl_pp, stats==commom_stats[2]), aes(x=r, y=ratio, shape=as.factor(N), color=as.factor(sample_size)))
p2 <- p2 + geom_point(size=3) + geom_line() + theme_bw() + facet_wrap(~s, nrow=1)
p2 <- p2 + scale_shape_manual(values=0:length((unique(tbl_pp))))
p2 <- p2 + scale_x_log10(breaks = r) + scale_y_continuous(labels = scale.4d)
p2 <- p2 + labs(title=NULL, x=NULL, y=expression(Dz))
p2 <- p2 + geom_hline(yintercept=1, linetype="dashed", color="red")
p2 <- p2 + theme(axis.title=element_text(size=12),
                 axis.text=element_text(size=10),
                 axis.title.x=element_blank(),
                 axis.text.x=element_blank(),
                 axis.ticks.x=element_blank(),
                 strip.text.x = element_blank(),
                 legend.position="none")
p2

p3 <- ggplot(data=filter(tbl_pp, stats==commom_stats[3]), aes(x=r, y=ratio, shape=as.factor(N), color=as.factor(sample_size)))
p3 <- p3 + geom_point(size=3) + geom_line() + theme_bw() + facet_wrap(~s, nrow=1)
p3 <- p3 + scale_shape_manual(values=0:length((unique(tbl_pp))))
p3 <- p3 + scale_x_log10(breaks = r) + scale_y_continuous(labels = scale.4d)
p3 <- p3 + labs(title=NULL, x=NULL, y=expression(H[l]))
p3 <- p3 + geom_hline(yintercept=1, linetype="dashed", color="red")
p3 <- p3 + theme(axis.title=element_text(size=12),
                 axis.text=element_text(size=10),
                 axis.title.x=element_blank(),
                 axis.text.x=element_blank(),
                 axis.ticks.x=element_blank(),
                 strip.text.x = element_blank(),
                 legend.position="none")
p3

p4 <- ggplot(data=filter(tbl_pp, stats==commom_stats[4]), aes(x=r, y=ratio, shape=as.factor(N), color=as.factor(sample_size)))
p4 <- p4 + geom_point(size=3) + geom_line() + theme_bw() + facet_wrap(~s, nrow=1)
p4 <- p4 + scale_shape_manual(values=0:length((unique(tbl_pp))))
p4 <- p4 + scale_x_log10(breaks = r) + scale_y_continuous(labels = scale.4d)
p4 <- p4 + labs(title=NULL, x=NULL, y=expression(H[r]))
p4 <- p4 + geom_hline(yintercept=1, linetype="dashed", color="red")
p4 <- p4 + theme(axis.title=element_text(size=12),
                 axis.text=element_text(size=10),
                 axis.title.x=element_blank(),
                 axis.text.x=element_blank(),
                 axis.ticks.x=element_blank(),
                 strip.text.x = element_blank(),
                 legend.position="none")
p4

p5 <- ggplot(data=filter(tbl_pp, stats==commom_stats[5]), aes(x=r, y=ratio, shape=as.factor(N), color=as.factor(sample_size))) 
p5 <- p5 + geom_point(size=3) + geom_line() + theme_bw() + facet_wrap(~s, nrow=1)
p5 <- p5 + scale_shape_manual(values=0:length((unique(tbl_pp))))
p5 <- p5 + scale_x_log10(breaks = r) + scale_y_continuous(breaks=pretty_breaks(), labels = scale.4d)
p5 <- p5 + labs(title=NULL, x="r", y=expression(pi[2]))
p5 <- p5 + geom_hline(yintercept=1, linetype="dashed", color="red")
p5 <- p5 + theme(axis.title=element_text(size=12),
                 axis.text=element_text(size=10),
                 strip.text.x = element_blank(),
                 axis.text.x=element_text(angle=90, size=12, vjust=0.5, hjust=1.0),
                 legend.position="bottom")
p5

moms_comp_ratios <- plot_grid(p1, p2, p3, p4, p5, ncol=1, rel_heights=c(1, 1, 1, 1, 1.85))
save_plot("selection/single-pop/bench/moms_++_vs_py_ratios.png", moms_comp_ratios, base_height=10, base_width=20)
  
# appending two-locus simulations (performed w/ twoLocusSim binary from momentspp)
n_reps <- 30
twoLocusSim <- numeric()
for(i in 1:n_models) {
  rp <- numeric()
  for(j in 1:n_reps) {
    vals <- read.table(paste("selection/single-pop/bench/tls_", i, "_rep_", j, ".txt", sep=""))$V3
    rp <- c(rp, vals)
  }
  twoLocusSim <- c(twoLocusSim, rp)
}

tls <- as.data.frame(twoLocusSim)
tls$stats <- commom_stats # not needed, output stats are in same order

reps <- NULL
for(i in 1:n_reps) { 
  reps <- c(reps, rep(i, length(commom_stats)))
}
tls$reps <- reps

scenario <- NULL
for(i in 1:n_models) { 
  scenario <- c(scenario, rep(i, length(commom_stats) * n_reps))
}
tls$scenario <- scenario

df <- cbind.data.frame(mpp, tbl_py_ns$expectation)
names(df)[1] <- "moments++"
names(df)[9] <- "momentsTwoLocus"

df_all <- full_join(df, tls, by=c("stats", "scenario")) 

sd <- df_all %>% group_by(scenario, stats) %>% summarize(sd = sd(twoLocusSim))
rep_means <- df_all %>% group_by(scenario, stats) %>% summarize(avg = mean(twoLocusSim))

df_all <- full_join(df_all, rep_means, by=c("stats", "scenario")) 
df_all <- full_join(df_all, sd, by=c("stats", "scenario")) 
df_m <- pivot_longer(df_all, cols=c("moments++", "momentsTwoLocus", "twoLocusSim"), names_to="Method")

dd <- filter(df_m, stats==commom_stats[1])
p6 <- ggplot() + geom_errorbar(data=filter(dd, Method=="twoLocusSim"), aes(x=r, ymax=avg + 2*sd, ymin=avg - 2*sd, color=as.factor(N))) + facet_wrap(~s, nrow=1)
p6 <- p6 + geom_point(data=filter(dd, Method!="twoLocusSim"), aes(x=r, y=value, color=as.factor(N), shape=Method, size=1)) + theme_bw()
p6 <- p6 + scale_shape_manual(values=c(0, 1))
p6 <- p6 + scale_x_log10(breaks = r) + scale_y_log10()
p6 <- p6 + labs(x="r", y=expression(D^2))
p6 <- p6 + scale_color_manual(values = brewer.pal(n=3, name="Dark2"))
p6 <- p6 + theme(axis.title=element_text(size=16),
                 axis.text=element_text(size=12),
                 axis.text.x=element_text(angle=90, size=12, vjust=0.5, hjust=1.0),
                 legend.position="bottom")
p6
save_plot("selection/single-pop/bench/moms_++_vs_py_dd.png", p6, base_height=10, base_width=20)

dz <- filter(df_m, stats==commom_stats[2])
p7 <- ggplot() + geom_errorbar(data=filter(dz, Method=="twoLocusSim"), aes(x=r, ymax=avg + 2*sd, ymin=avg - 2*sd, color=as.factor(N))) + facet_wrap(~s, nrow=1)
p7 <- p7 + geom_point(data=filter(dz, Method!="twoLocusSim"), aes(x=r, y=value, color=as.factor(N), shape=Method, size=1)) + theme_bw()
p7 <- p7 + scale_shape_manual(values=c(0, 1))
p7 <- p7 + scale_x_log10(breaks = r) + scale_y_log10()
p7 <- p7 + labs(x="r", y=expression(Dz))
p7 <- p7 + scale_color_manual(values = brewer.pal(n=3, name="Dark2"))
p7 <- p7 + theme(axis.title=element_text(size=16),
                 axis.text=element_text(size=12),
                 axis.text.x=element_text(angle=90, size=12, vjust=0.5, hjust=1.0),
                 legend.position="bottom")
p7
save_plot("selection/single-pop/bench/moms_++_vs_py_dz.png", p7, base_height=10, base_width=20)

hl <- filter(df_m, stats==commom_stats[3])
p8 <- ggplot() + geom_errorbar(data=filter(hl, Method=="twoLocusSim"), aes(x=r, ymax=avg + 2*sd, ymin=avg - 2*sd, color=as.factor(N))) + facet_wrap(~s, nrow=1)
p8 <- p8 + geom_point(data=filter(hl, Method!="twoLocusSim"), aes(x=r, y=value, color=as.factor(N), shape=Method, size=1)) + theme_bw()
p8 <- p8 + scale_shape_manual(values=c(0, 1))
p8 <- p8 + scale_x_log10(breaks=r) + scale_y_log10()
p8 <- p8 + labs(x="r", y=expression(H[l]))
p8 <- p8 + scale_color_manual(values = brewer.pal(n=3, name="Dark2"))
p8 <- p8 + theme(axis.title=element_text(size=16),
                 axis.text=element_text(size=12),
                 axis.text.x=element_text(angle=90, size=12, vjust=0.5, hjust=1.0),
                 legend.position="bottom")
p8
save_plot("selection/single-pop/bench/moms_++_vs_py_hl.png", p8, base_height=10, base_width=20)

hr <- filter(df_m, stats==commom_stats[4])
p9 <- ggplot() + geom_errorbar(data=filter(hr, Method=="twoLocusSim"), aes(x=r, ymax=avg + 2*sd, ymin=avg - 2*sd, color=as.factor(N))) + facet_wrap(~s, nrow=1)
p9 <- p9 + geom_point(data=filter(hr, Method!="twoLocusSim"), aes(x=r, y=value, color=as.factor(N), shape=Method, size=1)) + theme_bw()
p9 <- p9 + scale_shape_manual(values=c(0, 1))
p9 <- p9 + scale_x_log10(breaks=r) + scale_y_log10()
p9 <- p9 + labs(x="r", y=expression(H[r]))
p9 <- p9 + scale_color_manual(values = brewer.pal(n=3, name="Dark2"))
p9 <- p9 + theme(axis.title=element_text(size=16),
                 axis.text=element_text(size=12),
                 axis.text.x=element_text(angle=90, size=12, vjust=0.5, hjust=1.0),
                 legend.position="bottom")
p9
save_plot("selection/single-pop/bench/moms_++_vs_py_hr.png", p9, base_height=10, base_width=20)

pi2 <- filter(df_m, stats==commom_stats[5])
p10 <- ggplot() + geom_errorbar(data=filter(pi2, Method=="twoLocusSim"), aes(x=r, ymax=avg + 2*sd, ymin=avg - 2*sd, color=as.factor(N))) + facet_wrap(~s, nrow=1)
p10 <- p10 + geom_point(data=filter(pi2, Method!="twoLocusSim"), aes(x=r, y=value, color=as.factor(N), shape=Method, size=1)) + theme_bw()
p10 <- p10 + scale_shape_manual(values=c(0, 1))
p10 <- p10 + scale_x_log10(breaks=r) + scale_y_log10()
p10 <- p10 + labs(x="r", y=expression(pi[2]))
p10 <- p10 + scale_color_manual(values = brewer.pal(n=3, name="Dark2"))
p10 <- p10 + theme(axis.title=element_text(size=16),
                 axis.text=element_text(size=12),
                 axis.text.x=element_text(angle=90, size=12, vjust=0.5, hjust=1.0),
                 legend.position="bottom")
p10
save_plot("selection/single-pop/bench/moms_++_vs_py_pi2.png", p10, base_height=10, base_width=20)

```

## Simple BGS model

### Building Demes files

Here we model BGS experienced by a single neutral locus
at the edge of a chromosome with w selected sites of length l

```{r, demes-BGS-single, include=F, eval=T, results=F, message=F, warning=F}

N <- c(10000)
u <- c(1e-8)
r <- unique(sort(1:10 %o% 10^(-4:-8)))
s <- -seq_log(from=1e-5, to=1e-3, length.out=200)

# number of independent genomic windows, each corresponding to a two-locus system
w <- c(1)
# length of each constrained locus
l <- c(1)

params <- crossing(N, u, r, s, w, l)
params$pi0 <- 2 * params$N * params$u
params$alpha <- 2 * params$N * params$s
params$scenario <- 1:nrow(params)
n_models <- nrow(params)

write.csv(params, "selection/single-pop/simple_bgs/single_neutral/params.csv", quote=F, row.names=F)

Sys.setenv(num_models=n_models)

for(i in 1:n_models) {
  
  w <- params$w[i] # number of selected non-recombining genes

  for(j in 1:w) { # for all "left" loci
  
      name <- paste("model_", i, "_l_", j, sep="")
      sink(paste("selection/single-pop/simple_bgs/single_neutral/", name, ".yaml", sep=""))
      
      cat("time_units: generations\n")
      cat("demes:\n")
      cat("  - name: X\n")
      cat("    epochs:\n")
      cat("      - end_time: 0\n")
      cat("        start_size: ")
      cat(as.integer(params$N[i]))
      cat("\n")
      cat("metadata:\n")
      cat("  - name: mutation\n")
      cat("    left_factor: ")
      cat(params$l[i])
      cat("\n")
      cat("    epochs:\n")
      cat("      - end_time: 0\n")
      cat("        rates: [")
      cat(params$u[i])
      cat("]\n")
      cat("  - name: recombination\n")
      cat("    epochs:\n")
      cat("      - end_time: 0\n")
      cat("        rates: [")
      cat(j * params$r[i])
      cat("]\n")
      cat("  - name: selection\n")
      cat("    start_time: .inf\n")
      cat("    epochs:\n")
      cat("      - end_time: 0\n")
      cat("        rates: [")
      cat(params$s[i])
      cat("]\n")
      
      sink()
  }
}

```

### Running moments++

```{bash run-moments++BGS-single, engine.opts='-i', include=F, eval=T, results=F, message=F, warning=F}

source ~/.bashrc

cd selection/single-pop/simple_bgs/single_neutral/

for m in *.yaml
do
  momentspp params=opt.bpp F=$m O=25
done

cd ../../../..
```

### Plots

```{r, plots-BGS-single, include=T, eval=T, results=F, message=F, warning=F}

params <- read.table("selection/single-pop/simple_bgs/single_neutral/params.csv", sep=",", header=T)
n_models <- nrow(params)
hr_BGS <- numeric(length=n_models)
hk95 <- numeric(length=n_models)
factors <- numeric(length=n_models)

# for landscape plots
factors_tbl <- as.data.frame(matrix(nrow=n_models, ncol=w))

for(i in 1:n_models) {
  
  w <- params$w[i] # number of selected non-recombining genes
  pi0 <- params$pi0[i]  
  one_minus_f <- numeric(length=w)
  hudson_kaplan_95 <- numeric(length=w)
  
  for(j in 1:w) { # left locus (selected)
  
    name <- paste("model_", i, "_l_", j, "_expectations.txt", sep="")
    expectations <- read.table(paste("selection/single-pop/simple_bgs/single_neutral/", name, sep=""))
      
    hr <- expectations[expectations$V1 == "Hr_0_0",]$V3
    one_minus_f[j] <- hr / pi0
    hudson_kaplan_95[j] <- 1 - (params$u[i] * params$l[i] * -1 * params$s[i]) / (2 * (-params$s[i] + j * params$r[i])^2)
  }
  
  factors[i] <- prod(one_minus_f)
  hr_BGS[i] <- pi0 * factors[i]
  hk95[i] <- pi0 * prod(hudson_kaplan_95)
    
  factors_tbl[i,] <- one_minus_f # stores to later plot relative reduction in diversity
}

params$factors <- factors
params$momentspp <- hr_BGS
params$hk95 <- hk95

# visualizing the accuracy of the approximation ul = L * u for the left locus
tbl <- filter(params, N==10000, w==1, r>0) # looking at single N to simplify

tbl$ul <- tbl$u * tbl$l
x <- filter(tbl, l==1)["factors"]
y <- numeric()
for(i in 1:(nrow(tbl) / length(unique(tbl$l)))) {
  for(j in unique(params$l)) {
    y <- c(y, x[i,] ^ j)
  }
}
tbl$scaled_factors <- y
tbl$scaled_hr <- tbl$scaled_factors * tbl$pi0
tbl$ratio <- tbl$momentspp / tbl$scaled_hr
molten_tbl <- pivot_longer(tbl, cols=c("momentspp", "scaled_hr"), names_to="method")

c <- 1
plot_list <- list(length=length(unique(molten_tbl$l)))
for(i in unique(molten_tbl$l)) {
  p <- ggplot(data=filter(molten_tbl, l==i, u==1e-8), aes(x=r, y=1e+4*value, shape=method))
  p <- p + geom_line() + geom_point(size=3) + facet_wrap(~s, nrow=1) + theme_bw()
  p <- p + geom_hline(data=filter(molten_tbl, ul==i), aes(yintercept=1e+4*pi0), linetype="dashed")
  p <- p + scale_shape_manual(values=c(0, 1))
  p <- p + scale_x_log10(breaks=unique(molten_tbl$r))
  p <- p + scale_y_log10(breaks=pretty_breaks(), labels=function(x) sprintf("%.4f", x))
  
  if(c==1) {
    p <- p + labs(title="moments++ Hr x 10,000, rows = L, cols = s", x="r", y=expression(H[r]))
    p <- p + theme(axis.title=element_text(size=16), 
                   axis.text=element_text(size=12), 
                   strip.text.x=element_text(size=16), 
                   axis.title.x=element_blank(),
                   axis.text.x=element_blank(),
                   axis.ticks.x=element_blank(),
                   legend.position="none")
  } else if(c==length(unique(molten_tbl$l))) {
    p <- p + labs(title=NULL, x="r", y=expression(H[r]))
    p <- p + theme(axis.title=element_text(size=16),
                   axis.text=element_text(size=12), 
                   strip.text.x=element_blank(), 
                   axis.text.x=element_text(angle=90, size=12, vjust=0.5, hjust=1.0),
                   legend.text=element_text(size=16), 
                   legend.position="bottom")
  } else {
    p <- p + labs(title=NULL, x="r", y=expression(H[r]))
    p <- p + theme(axis.title=element_text(size=16), 
                   axis.text=element_text(size=12), 
                   strip.text.x=element_blank(), 
                   axis.title.x=element_blank(),
                   axis.text.x=element_blank(),
                   axis.ticks.x=element_blank(),
                   legend.position="none")
  }
  
  plot_list[[c]] <- p
  c <- c + 1
}

bgs_plot_1 <- plot_grid(plotlist=plot_list, ncol=1, rel_heights=c(1, 1.1))
save_plot("selection/single-pop/simple_bgs/bgs_single_neutral_scaling_1.png", bgs_plot_1, base_height=12, base_width=16)

c <- 1
plot_list <- list(length=length(unique(tbl$l)))
tbl2 <- filter(tbl, l>1, u==1e-8)
for(i in unique(tbl2$l)) {
  p <- ggplot(data=filter(tbl2, l==i), aes(x=r, y=ratio, shape=as.factor(alpha)))
  p <- p + geom_line() + geom_point(size=3) + theme_bw()
  p <- p + scale_shape_manual(values=1:length(unique(tbl2$alpha)))
  p <- p + scale_x_log10(breaks=unique(tbl2$r))
  p <- p + scale_y_log10(breaks=pretty_breaks(), labels=function(x) sprintf("%.4f", x))
  
  if(c==1) {
    p <- p + labs(title="Hr / scaled_hr", x="r", y="Ratio")
    p <- p + theme(axis.title=element_text(size=16), 
                   axis.text=element_text(size=12), 
                   strip.text.x=element_text(size=16), 
                   axis.title.x=element_blank(),
                   axis.text.x=element_blank(),
                   axis.ticks.x=element_blank(),
                   legend.position="none")
  } else if(c==length(unique(tbl2$l))) {
    p <- p + labs(title=NULL, x="r", y="Ratio")
    p <- p + theme(axis.title=element_text(size=16),
                   axis.text=element_text(size=12), 
                   strip.text.x=element_blank(), 
                   axis.text.x=element_text(angle=90, size=12, vjust=0.5, hjust=1.0),
                   legend.text=element_text(size=16), 
                   legend.position="bottom")
  } else {
    p <- p + labs(title=NULL, x="r", y="Ratio")
    p <- p + theme(axis.title=element_text(size=16), 
                   axis.text=element_text(size=12), 
                   strip.text.x=element_blank(), 
                   axis.title.x=element_blank(),
                   axis.text.x=element_blank(),
                   axis.ticks.x=element_blank(),
                   legend.position="none")
  }
  
  plot_list[[c]] <- p
  c <- c + 1
}

bgs_plot_2 <- plot_grid(plotlist=plot_list, ncol=1, rel_heights=c(1, 1, 1.1))
save_plot("selection/single-pop/simple_bgs/bgs_single_neutral_scaling_2.png", bgs_plot_2, base_height=12, base_width=16)

# plots of moments++
df <- filter(params, w==1, u==1e-8)
for(i in unique(df$N)) {
  
  df2 <- filter(df, N==i)
  c <- 1
  plot_list <- list(length=length(unique(df2$l)))
  
  for(j in unique(df2$l)) {
    p <- ggplot(data=filter(df2, l==j), aes(x=s, y=momentspp))
    p <- p + geom_line() + geom_point(size=3) + facet_wrap(~r, nrow=1) + theme_bw()
    p <- p + geom_hline(data=filter(df2, l==j), aes(yintercept=pi0), linetype="dashed")
    p <- p + scale_x_continuous(breaks=pretty_breaks())
    p <- p + scale_y_log10(breaks=pretty_breaks(), labels=function(x) format(x, scientific=T))
  
    if(c==1) {
      p <- p + labs(title="Hr moments++, rows = uL, cols = r", x="s", y=expression(H[r]))
      p <- p + theme(axis.title=element_text(size=16), 
                     axis.text=element_text(size=12), 
                     strip.text.x=element_text(size=16), 
                     axis.title.x=element_blank(),
                     axis.text.x=element_blank(),
                     axis.ticks.x=element_blank(),
                     legend.position="none")
    } else if(c==length(unique(df2$l))) {
      p <- p + labs(title=NULL, x="s", y=expression(H[r]))
      p <- p + theme(axis.title=element_text(size=16),
                     axis.text=element_text(size=12), 
                     strip.text.x=element_blank(), 
                     axis.text.x=element_text(angle=90, size=12, vjust=0.5, hjust=1.0),
                     legend.text=element_text(size=16), 
                     legend.position="bottom")
    } else {
      p <- p + labs(title=NULL, x="s", y=expression(H[r]))
      p <- p + theme(axis.title=element_text(size=16), 
                     axis.text=element_text(size=12), 
                     strip.text.x=element_blank(), 
                     axis.title.x=element_blank(),
                     axis.text.x=element_blank(),
                     axis.ticks.x=element_blank(),
                     legend.position="none")
    }
  
    plot_list[[c]] <- p
    c <- c + 1
  }
  
  bgs_plot_3 <- plot_grid(plotlist=plot_list, ncol=1, rel_heights=c(1.15, 1, 1, 1, 1.75))
  save_plot(paste("selection/single-pop/simple_bgs/bgs_single_neutral_mpp_A_N_", i, ".png", sep=""), bgs_plot_3, base_height=12, base_width=16)
}

c <- 1
plot_list <- list(length=length(unique(df$l)))
for(i in unique(df$l)) {
  p <- ggplot(data=filter(df, l==i), aes(x=s, y=factors, color=as.factor(N)))
  p <- p + geom_line() + geom_point(size=3) + facet_wrap(~r, nrow=1) + theme_bw()
  p <- p + geom_hline(yintercept=1, linetype="dashed")
  p <- p + scale_x_continuous(breaks=pretty_breaks())
  p <- p + scale_y_log10(breaks=pretty_breaks(), labels=function(x) sprintf("%.3f", x))
  
  if(c==1) {
    p <- p + labs(title="Two-locus Reduction Factors BGS, rows = uL, cols = r", x="s", y=expression(H[r]))
    p <- p + theme(axis.title=element_text(size=16), 
                   axis.text=element_text(size=12), 
                   strip.text.x=element_text(size=16), 
                   axis.title.x=element_blank(),
                   axis.text.x=element_blank(),
                   axis.ticks.x=element_blank(),
                   legend.position="none")
  } else if(c==length(unique(params$l))) {
    p <- p + labs(title=NULL, x="s", y=expression(H[r]))
    p <- p + theme(axis.title=element_text(size=16),
                   axis.text=element_text(size=12), 
                   strip.text.x=element_blank(), 
                   axis.text.x=element_text(angle=90, size=12, vjust=0.5, hjust=1.0),
                   legend.text=element_text(size=16), 
                   legend.position="bottom")
  } else {
    p <- p + labs(title=NULL, x="s", y=expression(H[r]))
    p <- p + theme(axis.title=element_text(size=16), 
                   axis.text=element_text(size=12), 
                   strip.text.x=element_blank(), 
                   axis.title.x=element_blank(),
                   axis.text.x=element_blank(),
                   axis.ticks.x=element_blank(),
                   legend.position="none")
  }
  
  plot_list[[c]] <- p
  c <- c + 1
}

bgs_plot_2 <- plot_grid(plotlist=plot_list, ncol=1, rel_heights=c(1.15, 1, 1, 1.75))
save_plot("selection/single-pop/simple_bgs/bgs_single_neutral_mpp_B.png", bgs_plot_2, base_height=12, base_width=16)

names(factors_tbl) <- 1:ncol(factors_tbl)
windows <- bind_cols(params, factors_tbl)
molten_windows <- pivot_longer(windows, cols=(ncol(params) + 1):ncol(windows), names_to="window")
molten_windows$window <- as.numeric(molten_windows$window)

c <- 3
for(i in unique(params$u)) {
  p <- ggplot(data=filter(molten_windows, u==i, w==100), aes(x=window, y=value, color=as.factor(s)))
  p <- p + geom_point(size=0.5) + facet_grid(N~r) + theme_bw()
  p <- p + scale_x_discrete(breaks=pretty_breaks())
  p <- p + scale_y_continuous(breaks=pretty_breaks())
  p <- p + labs(title=paste("Two-locus Hr BGS windows, mu", i, ", cols = r, rows = N", sep=""), x="Window", y=expression(1-f[j]))
  p <- p + theme(axis.title=element_text(size=16),
                 axis.text=element_text(size=12),
                 axis.text.x=element_text(angle=90, size=12, vjust=0.5, hjust=1.0),
                 legend.position="bottom")
  
  save_plot(paste("selection/single-pop/simple_bgs/bgs_single_neutral_mpp_", c, ".png", sep=""), p, base_height=8, base_width=15)
  
  c <- c + 1
}

# plots of hudson & kaplan 1995
for(i in unique(df$N)) {
  
  df2 <- filter(df, N==i)
  c <- 1
  plot_list <- list(length=length(unique(df2$l)))
  
  for(j in unique(df2$l)) {
    p <- ggplot(data=filter(df2, l==j), aes(x=s, y=hk95))
    p <- p + geom_line() + geom_point(size=3) + facet_wrap(~r, nrow=1) + theme_bw()
    p <- p + geom_hline(data=filter(df2, l==j), aes(yintercept=pi0), linetype="dashed")
    p <- p + scale_x_continuous(breaks=pretty_breaks())
    p <- p + scale_y_log10(breaks=pretty_breaks(), labels=function(x) format(x, scientific=T))
  
    if(c==1) {
      p <- p + labs(title="Hr HK95, rows = L, cols = r", x="s", y="HK95")
      p <- p + theme(axis.title=element_text(size=16), 
                     axis.text=element_text(size=12), 
                     strip.text.x=element_text(size=16), 
                     axis.title.x=element_blank(),
                     axis.text.x=element_blank(),
                     axis.ticks.x=element_blank(),
                     legend.position="none")
    } else if(c==length(unique(params$l))) {
      p <- p + labs(title=NULL, x="s", y="HK95")
      p <- p + theme(axis.title=element_text(size=16),
                     axis.text=element_text(size=12), 
                     strip.text.x=element_blank(), 
                     axis.text.x=element_text(angle=90, size=12, vjust=0.5, hjust=1.0),
                     legend.text=element_text(size=16), 
                     legend.position="bottom")
    } else {
      p <- p + labs(title=NULL, x="s", y="HK95")
      p <- p + theme(axis.title=element_text(size=16), 
                     axis.text=element_text(size=12), 
                     strip.text.x=element_blank(), 
                     axis.title.x=element_blank(),
                     axis.text.x=element_blank(),
                     axis.ticks.x=element_blank(),
                     legend.position="none")
    }
  
    plot_list[[c]] <- p
    c <- c + 1
  }
  
  bgs_plot_4 <- plot_grid(plotlist=plot_list, ncol=1, rel_heights=c(1.15, 1, 1, 1, 1.75))
  save_plot(paste("selection/single-pop/simple_bgs/bgs_single_neutral_hk95_N_", i, ".png", sep=""), bgs_plot_4, base_height=12, base_width=24)
}

# compare moments++ and HK95
for(i in unique(df$N)) {
  
  df2 <- pivot_longer(filter(df, i==N), cols=c("momentspp", "hk95"), names_to="method")
  c <- 1
  plot_list <- list(length=length(unique(df2$l)))
  
  for(j in unique(df$l)) {
    p <- ggplot(data=filter(df2, l==j), aes(x=s, y=value, shape=method))
    p <- p + geom_line() + geom_point(size=3) + facet_wrap(~r, nrow=1) + theme_bw()
    p <- p + geom_hline(data=filter(df2, u==j), aes(yintercept=pi0, color=as.factor(N)), linetype="dashed")
    p <- p + scale_shape_manual(values=c(0, 1, 2))
    p <- p + scale_x_continuous(breaks=pretty_breaks())
    p <- p + scale_y_log10(breaks=pretty_breaks(), labels=function(x) format(x, scientific=T))
  
    if(c==1) {
      p <- p + labs(title="Hr methods, rows = ul, cols = r", x="s", y="Hr")
      p <- p + theme(axis.title=element_text(size=16), 
                     axis.text=element_text(size=12), 
                     strip.text.x=element_text(size=16), 
                     axis.title.x=element_blank(),
                     axis.text.x=element_blank(),
                     axis.ticks.x=element_blank(),
                     legend.position="none")
    } else if(c==length(unique(params$l))) {
      p <- p + labs(title=NULL, x="s", y=expression(H[r]))
      p <- p + theme(axis.title=element_text(size=16),
                     axis.text=element_text(size=12), 
                     strip.text.x=element_blank(), 
                     axis.text.x=element_text(angle=90, size=12, vjust=0.5, hjust=1.0),
                     legend.text=element_text(size=16), 
                     legend.position="bottom")
    } else {
      p <- p + labs(title=NULL, x="s", y=expression(H[r]))
      p <- p + theme(axis.title=element_text(size=16), 
                     axis.text=element_text(size=12), 
                     strip.text.x=element_blank(), 
                     axis.title.x=element_blank(),
                     axis.text.x=element_blank(),
                     axis.ticks.x=element_blank(),
                     legend.position="none")
    }
  
    plot_list[[c]] <- p
    c <- c + 1
  }
  
  bgs_plot_5 <- plot_grid(plotlist=plot_list, ncol=1, rel_heights=c(1.15, 1, 1, 1.75))
  save_plot(paste("selection/single-pop/simple_bgs/bgs_mpp_hk95_N_", i, ".png", sep=""), bgs_plot_5, base_height=12, base_width=24)
}

# checking the relationship between the s and the relevant width of BGS (~r)
df3 <- distinct(dplyr::select(filter(molten_windows, w==1, r>0), -window))

p <- ggplot(data=filter(df3, l==1), aes(x=r, y=value, color=as.factor(s), shape=as.factor(u)))
p <- p + geom_line() + geom_point(size=3) + facet_wrap(~N, nrow=2) + theme_bw()
p <- p + geom_vline(data=filter(df3, l==1), aes(xintercept=-s, color=as.factor(s)), linetype="dashed")
p <- p + scale_shape_manual(values=c(0, 1, 2))
p <- p + scale_color_brewer(palette = "Accent")
p <- p + scale_x_log10(breaks=unique(df3$r))
p <- p + scale_y_log10(breaks=pretty_breaks(), labels=function(x) format(x, scientific=T))
p <- p + labs(title="Diversity Reduction Factors, L=1, W=1", x="r", y=expression(1-f[j]))
p <- p + theme(axis.title=element_text(size=12),
               axis.text=element_text(size=10),
               axis.text.x=element_text(angle=90, size=8, vjust=0.5, hjust=1.0),
               legend.position="bottom")
save_plot("selection/single-pop/simple_bgs/s_vs_r_A.png", p, base_height=9, base_width=12)

df4 <- filter(molten_windows, w==100, r==1e-5)
df4$rec <- df4$r * df4$window
df4$cum_red <- cumprod(df4$value)

p <- ggplot(data=filter(df4, l==1), aes(x=rec, y=value, color=as.factor(s), linetype=as.factor(u)))
p <- p + geom_line() + facet_wrap(~N, nrow=2) 
p <- p + theme(panel.grid.major=element_blank(), panel.grid.minor=element_blank(),
               panel.background=element_blank(), axis.line=element_line(colour="black"))
p <- p + geom_vline(data=filter(df4, l==1), aes(xintercept=-s, color=as.factor(s)), linetype="dashed")
p <- p + scale_shape_manual(values=c(0, 1))
p <- p + scale_color_brewer(palette = "Dark2")
p <- p + scale_x_continuous(breaks=unique(df4$rec))
p <- p + scale_y_log10(breaks=pretty_breaks(), labels=function(x) format(x, scientific=T))
p <- p + labs(title="Diversity Reduction Factors, L=1, W=100", x="cum. rec. rate.", y=expression(1-f[j]))
p <- p + theme(axis.title=element_text(size=12),
               axis.text=element_text(size=10),
               axis.text.x=element_blank(),
               axis.ticks.x=element_blank(),
               legend.position="bottom")
save_plot("selection/single-pop/simple_bgs/s_vs_r_B.png", p, base_height=9, base_width=12)
```

### Building Demes files

We now model multiple neutral loci, interdigitated with 
constrained, non-recombining loci of length l

```{r, demes-BGS-mult-1, include=F, eval=F, results=F, message=F, warning=F}

N <- 10000
u <- c(1e-7, 1e-8, 1e-9)
r <- c(1e-5, 1e-6, 1e-7, 1e-8)
s <- c(-1e-5, -1e-4, -1e-3)

# number of independent genomic windows, each corresponding to a two-locus system
w <- 100
# length of each genomic block
l <- 10

params <- crossing(N, u, r, s, w, l)
params$pi0 <- 2 * params$N * params$u
params$alpha <- 2 * params$N * params$s
params$scenario <- 1:nrow(params)
n_models <- nrow(params)

write.csv(params, "selection/single-pop/simple_bgs/mult_neutral/params.csv", quote=F, row.names=F)

Sys.setenv(n_models=n_models)
```

```{bash mk-mult-dirs, engine.opts='-i', include=F, eval=F, results=F, message=F, warning=F}

source ~/.bashrc

cd selection/single-pop/simple_bgs/mult_neutral/

for((i=1;i<=$n_models;i++));
do
  mkdir model_$i
  cp opt.bpp model_$i
done

cd ../../../..
```

```{r, demes-BGS-mult-2, include=F, eval=F, results=F, message=F, warning=F}

for(i in 1:n_models) {
  
  for(j in 1:w) { # left locus
  
    for(k in 1:w) { # right locus
      
      name <- paste("model_", i, "_l_", j, "_r_", k, sep="")
      sink(paste("selection/single-pop/simple_bgs/mult_neutral/model_", i, "/", name, ".yaml", sep=""))
      
      total_r <- (abs(j-k) + 1) * params$r[i]
      
      cat("time_units: generations\n")
      cat("demes:\n")
      cat("  - name: X\n")
      cat("    epochs:\n")
      cat("      - end_time: 0\n")
      cat("        start_size: ")
      cat(as.integer(params$N[i]))
      cat("\n")
      cat("metadata:\n")
      cat("  - name: mutation\n")
      cat("    left_factor: ")
      cat(params$l[i])
      cat("\n")
      cat("    epochs:\n")
      cat("      - end_time: 0\n")
      cat("        rates: [")
      cat(params$u[i])
      cat("]\n")
      cat("  - name: recombination\n")
      cat("    epochs:\n")
      cat("      - end_time: 0\n")
      cat("        rates: [")
      cat(total_r)
      cat("]\n")
      cat("  - name: selection\n")
      cat("    start_time: .inf\n")
      cat("    epochs:\n")
      cat("      - end_time: 0\n")
      cat("        rates: [")
      cat(params$s[i])
      cat("]\n")
      
      sink()
    }
  }
}

```

### Running moments++

```{bash run-moments++BGS-mult, engine.opts='-i', include=F, eval=F, results=F, message=F, warning=F}

source ~/.bashrc

cd selection/single-pop/simple_bgs/mult_neutral/

for((i=1;i<=$n_models;i++));
do
  cd model_$i
  for m in *.yaml
  do
    momentspp params=opt.bpp F=$m O=25
  done
  cd ..
done

cd ../../../..
```

### Plots

```{r, plots-BGS-mult, include=T, eval=F, results=F, message=F, warning=F}

params <- read.table("selection/single-pop/simple_bgs/mult_neutral/params.csv", sep=",", header=T)
n_models <- nrow(params)

# for landscape plots
hr_tbl <- as.data.frame(matrix(nrow=n_models, ncol=w))

for(i in 1:n_models) {
  
  cat(paste("Retrieving data from model ", i, "...\n"))
  
  for(j in 1:w) { # right locus
    
    one_minus_f <- numeric(length=w)
      
    for(k in 1:w) { # left locus
      
      name <- paste("model_", i, "_l_", k, "_r_", j, "_expectations.txt", sep="")
      expectations <- read.table(paste("selection/single-pop/simple_bgs/mult_neutral/model_", i, "/", name, sep=""))
      
      hr <- expectations[expectations$V1 == "Hr_0_0",]$V3
      one_minus_f[k] <- hr / params$pi0[i]
    }
    
    hr_BGS[j] <- params$pi0[i] * prod(one_minus_f)
  }
  
  hr_tbl[i,] <- hr_BGS
}

names(hr_tbl) <- 1:ncol(hr_tbl)
windows <- bind_cols(params, hr_tbl)
molten_windows <- pivot_longer(windows, cols=(ncol(params) + 1):ncol(windows), names_to="window")
molten_windows$window <- as.numeric(molten_windows$window)

p1 <- ggplot(data=filter(molten_windows, u==1e-9), aes(x=window, y=value, color=as.factor(alpha)))
p1 <- p1 + geom_point(size=1) + facet_wrap(~r, nrow=1) + theme_bw()
p1 <- p1 + scale_x_discrete(breaks=pretty_breaks())
p1 <- p1 + scale_y_continuous(breaks=pretty_breaks())
p1 <- p1 + geom_hline(yintercept=unique(filter(molten_windows, u==1e-9)$pi0), linetype="dashed")
p1 <- p1 + labs(title="Two-locus Hr BGS windows, mu=1e-9, cols = r", x="Window", y=expression(H[r]))
p1 <- p1 + theme(axis.title=element_text(size=16),
                 axis.text=element_text(size=12),
                 axis.text.x=element_text(angle=90, size=12, vjust=0.5, hjust=1.0),
                 legend.position="bottom")

save_plot("selection/single-pop/simple_bgs/bgs_mult_neutral_1.png", p1, base_height=12, base_width=20)

p2 <- ggplot(data=filter(molten_windows, u==1e-8), aes(x=window, y=value, color=as.factor(alpha)))
p2 <- p2 + geom_point(size=1) + facet_wrap(~r, nrow=1) + theme_bw()
p2 <- p2 + scale_x_discrete(breaks=pretty_breaks())
p2 <- p2 + scale_y_continuous(breaks=pretty_breaks())
p2 <- p2 + geom_hline(yintercept=unique(filter(molten_windows, u==1e-8)$pi0), linetype="dashed")
p2 <- p2 + labs(title="Two-locus Hr BGS windows, mu=1e-8, cols = r", x="Window", y=expression(H[r]))
p2 <- p2 + theme(axis.title=element_text(size=16),
                 axis.text=element_text(size=12),
                 axis.text.x=element_text(angle=90, size=12, vjust=0.5, hjust=1.0),
                 legend.position="bottom")

save_plot("selection/single-pop/simple_bgs/bgs_mult_neutral_2.png", p2, base_height=12, base_width=20)

p3 <- ggplot(data=filter(molten_windows, u==1e-7), aes(x=window, y=value, color=as.factor(alpha)))
p3 <- p3 + geom_point(size=1) + facet_wrap(~r, nrow=1) + theme_bw()
p3 <- p3 + scale_x_discrete(breaks=pretty_breaks())
p3 <- p3 + scale_y_continuous(breaks=pretty_breaks())
p3 <- p3 + geom_hline(yintercept=unique(filter(molten_windows, u==1e-7)$pi0), linetype="dashed")
p3 <- p3 + labs(title="Two-locus Hr BGS windows, mu=1e-7, cols = r", x="Window", y=expression(H[r]))
p3 <- p3 + theme(axis.title=element_text(size=16),
                 axis.text=element_text(size=12),
                 axis.text.x=element_text(angle=90, size=12, vjust=0.5, hjust=1.0),
                 legend.position="bottom")

save_plot("selection/single-pop/simple_bgs/bgs_mult_neutral_3.png", p3, base_height=12, base_width=20)
```

## BGS with Genomic Landscapes

### Building Demes files

This builds a look-up table on a dense parameter grid that will allow for fast 
assembly of whole chromosomes

```{r, BGS-mutvar-params, include=F, eval=F, results=F, message=F, warning=F}

N <- 10000
lookup_r <- seq_log(from=1e-15, to=1e-2, length.out=200) # no need to round, will using spline interpolation
#lookup_ul <- round(seq_log(from=1e-9, to=1e-7, length.out=30), digits=11)
lookup_ur <- round(seq_log(from=1e-9, to=1e-7, length.out=30), digits=11)
lookup_s <- round(-seq_log(from=1e-5, to=1e-3, length.out=100), digits=11)

# for discretizing Gamma distributions and make look-up faster
dt_u <- data.table(lookup_ul)
dt_r <- data.table(lookup_r)
dt_s <- data.table(lookup_s)

setkey(dt_u, lookup_ul)
setkey(dt_r, lookup_r)
setkey(dt_s, lookup_s)

params <- setDT(crossing(N, lookup_r, lookup_ur, lookup_s))
params$scenario <- 1:nrow(params)
n_models <- nrow(params)

fwrite(as.data.frame(lookup_r), "selection/single-pop/landscapes/r.csv", quote=F, row.names=F, col.names=F)
fwrite(as.data.frame(lookup_ul), "selection/single-pop/landscapes/ul.csv", quote=F, row.names=F, col.names=F)
fwrite(as.data.frame(lookup_ur), "selection/single-pop/landscapes/ur.csv", quote=F, row.names=F, col.names=F)
fwrite(as.data.frame(lookup_s), "selection/single-pop/landscapes/s.csv", quote=F, row.names=F, col.names=F)
fwrite(params, "selection/single-pop/landscapes/params.csv", quote=F, row.names=F)

Sys.setenv(num_models=n_models)

```

```{bash make_dirs_mutvar, engine.opts='-i', include=F, eval=F, results=F, message=F, warning=F}

source ~/.bashrc

cd selection/single-pop/landscapes/

cat r.csv | while IFS=, read -r r
do
  mkdir r_$r
done

for d in r_*
do
  cat ul.csv | while IFS=, read -r ul
  do
    mkdir $d/ul_$ul
  done
done

for d in r_*
do
  for e in $d/ul_*
  do
    cat ur.csv | while IFS=, read -r ur
    do
      mkdir $e/ur_$ur
    done
  done
done

cd ../../..
```

```{r, BGS-mutvar-demes, include=F, eval=F, results=F, message=F, warning=F}

for(i in 1:n_models) {
  
  if(i %% 10000 == 0) {
    cat(paste(i, "\n"))
  }
  
  N <- as.integer(params["N"][i,])
  r <- as.numeric(params["r"][i,])
  ul <- as.numeric(params["ul"][i,])
  ur <- as.numeric(params["ur"][i,])
  s <- as.numeric(params["s"][i,])
  f <- ul / ur
  
  name <- paste("model_", i, sep="")
  sink(paste("selection/single-pop/landscapes/r_", r, "/ul_", ul, "/ur_", ur, "/", name, ".yaml", sep=""))
    
  cat("time_units: generations\n")
  cat("demes:\n")
  cat("  - name: X\n")
  cat("    epochs:\n")
  cat("      - end_time: 0\n")
  cat("        start_size: ")
  cat(N)
  cat("\n")
  cat("metadata:\n")
  cat("  - name: mutation\n")
  cat("    left_factor: ")
  cat(f)
  cat("\n")
  cat("    epochs:\n")
  cat("      - end_time: 0\n")
  cat("        rates: [")
  cat(ur)
  cat("]\n")
  cat("  - name: recombination\n")
  cat("    epochs:\n")
  cat("      - end_time: 0\n")
  cat("        rates: [")
  cat(r)
  cat("]\n")
  cat("  - name: selection\n")
  cat("    start_time: .inf\n")
  cat("    epochs:\n")
  cat("      - end_time: 0\n")
  cat("        rates: [")
  cat(s)
  cat("]\n")
      
  sink()
}

```

### Running moments++

Gets the lookup table

```{bash run-moments++BGS-mutvar, engine.opts='-i', include=F, eval=F, results=F, message=F, warning=F}

source ~/.bashrc

cd selection/single-pop/landscapes/

for r in r_*
do 
  cd $r
  for ul in ul_*
  do 
    cd $ul
    for ur in ur_*
    do 
      cd $ur
      for f in model_*
      do
	      momentspp params=opt.bpp F=$f O=25 > /dev/null
      done
      cd ..
    done
    cd ..
  done
  cd ..
done

cd ../../..
```

### Analyses

Collect lookup table with Hr values on the parameter grid

```{r, BGS-mutvar-grid, include=F, eval=F, results=F, message=F, warning=F}

params <- fread("~/../../media/gvbarroso/extradrive1/landscapes/params.csv", sep=",", header=T)
params$pi0 <- 2 * params$N * params$lookup_ur
n_models <- nrow(params)
c <- 1
hr <- numeric(length=n_models)

for(i in r) {
  cat(paste(c, "\n"))
  print(Sys.time())
  for(j in ul) {
    for(k in ur) {
      for(l in 1:length(s)) {
        
        moms <- read.table(paste("~/../../media/gvbarroso/extradrive1/landscapes/r_", i, "/ul_", j, "/ur_", k, "/model_", as.integer(c), "_expectations.txt", sep=""))
        hr[c] <- moms[moms$V1 == "Hr_0_0",]$V3

        c <- c + 1
      }
    }
  }
}

params$Hr <- hr
fwrite(params, "~/../../media/gvbarroso/extradrive1/landscapes/lookup_tbl.csv", row.names=F, quote=F, sep=",")
```

```{r, BGS-mutvar-landscapes, include=F, eval=F, results=F, message=F, warning=F}

#setNumericRounding(0) # for binary search with data.table

lookup_tbl <- setDT(fread("~/../../media/gvbarroso/extradrive1/landscapes/lookup_tbl.csv"))
setkey(lookup_tbl, lookup_ul, lookup_ur, lookup_s, lookup_r)

# constants
N <- unique(lookup_tbl$N)
num_constrained <- 1000 # number of (non-recombining) selected segment in chr

# Gamma dist params
scale_sel <- c(2.5e-5, 5e-5, 1e-4, 2.5e-4, 5e-4)
shape_sel <- c(10, 20)
scales_mu <- c(1e-8)
shapes_mu <- c(1, 3, 10, 30, 1e+9)
scales_rec <- c(1e-8)
shapes_rec <- c(1, 3, 10, 30, 1e+9)

# Geom dist params
avg_rec_spans <- c(1e+3, 1e+4, 1e+5, 1e+6)
avg_mut_spans <- c(1e+3, 1e+4, 1e+5, 1e+6)
avg_neutral_lengths <- 1e+4
exon_lengths <- c(100, 1000)  
models <- setDT(crossing(N, num_constrained, shape_sel, scale_sel, shapes_mu, scales_mu, shapes_rec, scales_rec, 
                avg_rec_spans, avg_mut_spans, avg_neutral_lengths, exon_lengths))
models$scenario <- 1:nrow(models)
n_models <- nrow(models)

fwrite(models, "~/../../media/gvbarroso/extradrive1/landscapes/models.csv", row.names=F, quote=F, sep=",")

# make this for loop a function of jump_length and bin_size
jump_length <- 1e+3
bin_size <- 1e+5

for(i in 1:n_models) {

  cat(paste("Building chromosome(s) for model ", i, "...\n", sep=""))
  
  ncsl <- rgeom(n=models$num_constrained[i] + 1, prob=1/models$avg_neutral_lengths[i])
  csl <- rep(models$exon_lengths[i], models$num_constrained[i]) 
  L <- sum(csl) + sum(ncsl)
  
  ss <- -rgamma(n=length(csl), shape=models$shape_sel[i], scale=models$scale_sel[i]/models$shape_sel[i])
  alpha <- 2 * models$N[i] * ss
  
  while(any(alpha < -20)) { 
    cat("Selection is too strong in at least one gene! Redrawing...\n")
    ss <- -rgamma(n=length(csl), shape=models$shape_sel[i], scale=models$scale_sel[i]/models$shape_sel[i])
    alpha <- 2 * models$N[i] * ss
  }
  
  disc_ss <- dt_s[dt_s[J(ss), roll="nearest", which=T]]$lookup_s
  
  intervals <- c(rbind(ncsl, csl))[-2*length(ncsl)]
  intervals <- setDT(bind_cols(chr="chr1", dplyr::lag(cumsum(intervals), n=1, default=0), dplyr::lag(cumsum(intervals), n=0, default=0)))
  names(intervals) <- c("chr", "start", "end")
  intervals$s <- c(rbind(rep(0, length(csl)), disc_ss), 0)
  setkey(intervals, start, end)
  
  dt_neutral <- filter(intervals[,2:4], s==0)
  setkey(dt_neutral, start, end)
  dt_exons <- filter(intervals, s < 0)
  dt_exons$u <- 1.083e-08 # it is inside lookup_ul 
  setkey(dt_exons, start, end)

  rec_spans <- rgeom(n=ceiling(L/models$avg_rec_spans[i]), prob=1/models$avg_rec_spans[i])
  while(sum(rec_spans) < L) {
    rec_spans <- c(rec_spans, rgeom(n=1, prob=1/models$avg_rec_spans[i]))
  }
  if(sum(rec_spans) > L) {
    rec_spans <- rec_spans[cumsum(rec_spans) < L]
    rec_spans <- c(rec_spans, L - sum(rec_spans))
  }
  rec_spans <- rec_spans[-which(rec_spans==0)]
  
  rmap <- setDT(bind_cols(chr="chr1", dplyr::lag(cumsum(rec_spans), n=1, default=0), dplyr::lag(cumsum(rec_spans), n=0, default=0)))
  names(rmap) <- c("chr", "start", "end")
  rs <- rgamma(n=nrow(rmap), shape=models$shapes_rec[i], scale=models$scales_rec[i]/models$shapes_rec[i])
  disc_rs <- dt_r[dt_r[J(rs), roll="nearest", which=T]]$lookup_r
  rmap$r <- disc_rs
  setkey(rmap, start, end)
  
  mut_spans <- rgeom(n=ceiling(L/models$avg_mut_spans[i]), prob=1/models$avg_mut_spans[i])
  while(sum(mut_spans) < L) {
    mut_spans <- c(mut_spans, rgeom(n=1, prob=1/models$avg_mut_spans[i]))
  }
  if(sum(mut_spans) > L) {
    mut_spans <- mut_spans[cumsum(mut_spans) < L]
    mut_spans <- c(mut_spans, L - sum(mut_spans))
  }
  mut_spans <- mut_spans[-which(mut_spans==0)]
  
  mmap <- setDT(bind_cols(chr="chr1", dplyr::lag(cumsum(mut_spans), n=1, default=0), dplyr::lag(cumsum(mut_spans), n=0, default=0)))
  names(mmap) <- c("chr", "start", "end")
  mus <- rgamma(n=nrow(mmap), shape=models$shapes_mu[i], scale=models$scales_mu[i]/models$shapes_mu[i])
  disc_mus <- dt_u[dt_u[J(mus), roll="nearest", which=T]]$lookup_ul
  mmap$u <- disc_mus
  setkey(mmap, start, end)
  
  # finding mu for each sampled pos and the cumulative recombination map over the useful intervals
  samp_pos <- unlist(apply(dt_neutral, 1, function(x) seq(from=x[1], to=x[2], by=jump_length))) + 1
  linear_pos <- sort(c(samp_pos, dt_exons$start + models$exon_lengths[i]/2))
  # (the following searches take some time:)
  focal_mu <- unlist(lapply(linear_pos, function(pos) mmap[J(pos), roll=T]$u))
  focal_r <- unlist(lapply(linear_pos, function(pos) rmap[J(pos), roll=T]$r))
  cum_rec <- numeric(length=length(linear_pos))
  cum_rec[1] <- focal_r[1]
  for(j in 2:length(linear_pos)) { cum_rec[j] <- cum_rec[j-1] + (linear_pos[j] - linear_pos[j-1]) * focal_r[j] }
  
  pos_dt <- setDT(bind_cols(linear_pos, focal_mu, cum_rec))
  names(pos_dt) <- c("position", "focal_mu", "cumrec")
  setkey(pos_dt, position)
  
  # binning for linear models
  bins <- rep(bin_size, floor(L / bin_size))
  bins <- bind_cols(chr="chr1", dplyr::lag(cumsum(bins), n=1, default=0), dplyr::lag(cumsum(bins), n=0, default=0))
  names(bins) <- c("chr", "start", "end")
  bins$start <- bins$start + 1 # so that GRanges understands the intervals -> NOT in BEDGRAPH format
  gr_bins <- makeGRangesFromDataFrame(bins)
  
  # binning recombination maps
  rmap$start <- rmap$start + 1 # so that GRanges understands the intervals -> NOT in BEDGRAPH format
  rec_gr <- makeGRangesFromDataFrame(rmap, keep.extra.columns=T)
  hits <- findOverlaps(query=gr_bins, subject=rec_gr) 
  hit_list <- as.data.frame(hits)
  names(hit_list) <- c("bins", "rec_bins")
  rmap$rec_bins <- 1:nrow(rmap)
  tmpr <- merge(rmap, hit_list)
  rec_bins <- tmpr %>% group_by(bins) %>% mutate(avg_rec=weighted.mean(r, end-start))
  rec_bins <- rec_bins %>% distinct(bins, .keep_all=T)
  
  # binning mutation maps
  mmap$start <- mmap$start + 1 # so that GRanges understands the intervals -> NOT in BEDGRAPH format
  mut_gr <- makeGRangesFromDataFrame(mmap, keep.extra.columns=T)
  hits <- findOverlaps(query=gr_bins, subject=mut_gr) 
  hit_list <- as.data.frame(hits)
  names(hit_list) <- c("bins", "mut_bins")
  mmap$mut_bins <- 1:nrow(mmap)
  tmpm <- merge(mmap, hit_list)
  mut_bins <- tmpm %>% group_by(bins) %>% mutate(avg_mut=weighted.mean(u, end-start))
  mut_bins <- mut_bins %>% distinct(bins, .keep_all=T)
  
  # binning DFE
  intervals$start <- intervals$start + 1 # so that GRanges understands the intervals -> NOT in BEDGRAPH format
  dfe_gr <- makeGRangesFromDataFrame(intervals, keep.extra.columns=T)
  hits <- findOverlaps(query=gr_bins, subject=dfe_gr) 
  hit_list <- as.data.frame(hits)
  names(hit_list) <- c("bins", "dfe")
  intervals$dfe <- 1:nrow(intervals)
  tmpd <- merge(intervals, hit_list)
  # avg_s encompasses density of constrained sites
  dfe_bins <- tmpd %>% group_by(bins) %>% mutate(avg_s=weighted.mean(s, end-start)) 
  dfe_bins <- dfe_bins %>% distinct(bins, .keep_all=T)
  
  maps_bins <- bind_cols(bins, rec_bins["avg_rec"], mut_bins["avg_mut"], dfe_bins["avg_s"])
  
  # pre-compute recombination distances from sampled neutral sites and exons:
  cr_samp <- pos_dt[position %in% samp_pos,]
  cr_exons <- pos_dt[position %in% (dt_exons$start + models$exon_lengths[i]/2),]
  x <- as.data.frame(abs(outer(cr_samp$cumrec, cr_exons$cumrec, "-")))
  nrow(x)
  length(cr_samp$cumrec) * length(cr_exons$cumrec)
  # TODO pickup here
  y <- which(x<1e-3)
  
  
  
  
  
  BGS <- numeric(length=length(samp_pos))
  Rprof() 
  pb <- txtProgressBar(min=1, max=length(samp_pos), style=3)
  #system.time({
  for(j in 1:length(samp_pos)) { 
    
    #setTxtProgressBar(pb, j)
    system.time({
    pos <- samp_pos[j]
    cumrec_to_pos <- pos_dt[pos_dt$position==pos,]$cumrec
    focal_ur <- pos_dt[position==pos]$focal_mu
    pi0 <- 2 * N * focal_ur
    B <- 1 
    c <- 1
    idx <- dt_exons[J(pos), roll="nearest", which=T] # index of closest exon
    
    # TODO perform search over relevant exons based on table x computed above
    for(k in idx:nrow(dt_exons)) { # loop to the right
      
      focal_s <- dt_exons[k,]$s
      total_r <- abs(pos_dt[pos_dt$position==dt_exons[k,]$start + models$exon_lengths[i] / 2,]$cumrec - cumrec_to_pos)

      if(total_r < 1e-3) {
        # TODO implement interpolation
        closest_r <- lookup_r[which.min(abs(lookup_r - total_r))] 
        hr <- lookup_tbl[.(focal_ur, focal_s, closest_r)]$Hr
        B <- B * (hr / pi0) ^ models$exon_lengths[i] 
        
        c <- c + 1
      } else {
          break
      }
    }
    
    if(idx > 1) {
      for(k in idx:1) { # loop to the left

        focal_s <- dt_exons[k,]$s
        total_r <- abs(pos_dt[pos_dt$position==dt_exons[k,]$start + models$exon_lengths[i] / 2,]$cumrec - cumrec_to_pos)

        if(total_r < 1e-3) {
          # TODO implement interpolation
          closest_r <- lookup_r[which.min(abs(lookup_r - total_r))]
          hr <- lookup_tbl[.(focal_ur, focal_s, closest_r)]$Hr
          B <- B * (hr / pi0) ^ models$exon_lengths[i]

          c <- c + 1
        } else {
            break
        }
      }
    }
    
    cat(paste("pos =", pos, "\t", c, "exons\n"))
    BGS[j] <- pi0 * B
    })
  }
  Rprof(NULL)
  #View(summaryRprof()$by.total)
  View(summaryRprof()$by.self)
  
  pos_hr_pairs <- bind_cols(neu_pos, hr_list)
  # TODO bin
  
  windows <- bind_cols(mus, rs, ss, .name_repair="unique")
  names(windows) <- c("mus", "recs", "dfe")
  windows$pi0 <- 2 * models$N[i] * windows$mus
  windows$alpha <- alpha
  
  write.csv(windows, paste("selection/single-pop/mut_rate_var/model_", i, "/_windows.csv", sep=""), quote=F, row.names=F)
  
  landscapes[i,] <- hr_BGS

  windows$bin <- 1:nrow(windows)
  molten_bins <- pivot_longer(windows, cols=c("mus", "recs", "alpha", "pi0", "hr_BGS"), names_to="variable")
  
  
  comb <- plot_grid(p1, p2, p3, p4, ncol=1)
  save_plot(paste("selection/single-pop/mut_rate_var/model_", i, "_landscapes.png", sep=""), comb, base_height=10, base_width=12)
      
  # standardizing
  windows$ss_pi_BGS <- (windows$hr_BGS - mean(windows$hr_BGS)) / sd(windows$hr_BGS)
  windows$ss_pi0 <- (windows$pi0 - mean(windows$pi0)) / sd(windows$pi0)
  windows$ss_mus <- (windows$mus - mean(windows$mus)) / sd(windows$mus)
  windows$ss_recs <- (windows$recs - mean(windows$recs)) / sd(windows$recs)
  windows$ss_dfe <- (windows$dfe - mean(windows$dfe)) / sd(windows$dfe)
  
  m.pi.1 <- lm(ss_pi_BGS ~ ss_mus + ss_recs + ss_dfe + ss_mus:ss_dfe + ss_recs:ss_dfe, data=windows)
  
  summary(m.pi.1)

  anova.pi <- Anova(m.pi.1)
  apiss <- anova.pi$"Sum Sq"
  anova.pi$VarExp <- apiss / sum(apiss)

  r2[i, 1] <- (anova.pi$VarExp[1] + anova.pi$VarExp[2] + anova.pi$VarExp[3] + anova.pi$VarExp[4] + anova.pi$VarExp[5]) * 100
  r2[i, 2] <- anova.pi$VarExp[1] * 100
  r2[i, 3] <- anova.pi$VarExp[2] * 100
  r2[i, 4] <- anova.pi$VarExp[3] * 100
  r2[i, 5] <- anova.pi$VarExp[4] * 100
  r2[i, 6] <- anova.pi$VarExp[5] * 100
}

names(landscapes) <- 1:ncol(landscapes)
tbl <- bind_cols(params, r2, landscapes)
tbl$scale_sel <- -tbl$scale_sel
write.csv(tbl, "selection/single-pop/mut_rate_var/tbl.csv", quote=F, row.names=F)

molten_windows <- pivot_longer(tbl, cols=(ncol(params) + ncol(r2) + 1):ncol(tbl), names_to="window")
molten_windows$window <- as.numeric(molten_windows$window)


r2 <- as.data.frame(matrix(nrow=n_models, ncol=6))
colnames(r2) <- c("Total_R2", "Mutation_R2", "Recombination_R2", "Selection_R2", "Mu:Sel_R2", "Rec:Sel_R2")

# visualizing raw landscapes
c=1
for(i in unique(molten_windows$N)) {
  for(j in unique(molten_windows$scales_mu)) {
    for(k in unique(molten_windows$scales_rec)) {
      
      p <- ggplot(data=filter(molten_windows, N==i, scales_mu==j, scales_rec==k), aes(x=window, y=value, color=as.factor(scale_sel)))
      p <- p + geom_line() + facet_grid(shapes_rec~shapes_mu) + theme_bw()
      p <- p + scale_x_discrete(breaks=pretty_breaks())
      p <- p + scale_y_continuous(breaks=pretty_breaks())
      p <- p + labs(title=paste("Hr landscapes, N=", i, ", mu=", j, ", r=", k, " cols~1/var_mu, rows~1/var_rec", sep=""), x="Window", y=expression(H[r]))
      p <- p + theme(axis.title=element_text(size=16),
                     axis.text=element_text(size=12),
                     axis.text.x=element_text(angle=90, size=12, vjust=0.5, hjust=1.0),
                     legend.position="bottom")
      save_plot(paste("selection/single-pop/Hr_landscapes_", c, ".png", sep=""), p, base_height=10, base_width=15)
      
      c <- c + 1
    }
  }
}

# zooming in plots with tiny mut rate variation
p1 <- ggplot(data=filter(molten_windows, N==1000, shapes_mu==1e+6, scales_mu==1e-9), aes(x=window, y=value, color=as.factor(scale_sel)))
p1 <- p1 + geom_line() + facet_grid(shapes_rec~scales_rec) + theme_bw()
p1 <- p1 + scale_x_discrete(breaks=pretty_breaks())
p1 <- p1 + scale_y_continuous(breaks=pretty_breaks())
p1 <- p1 + labs(title="Hr landscapes, N=1000, shape_mu=1e+6, scale_mu=1e-9, rows~1/var_rec, cols~mean_rec", x="Window", y=expression(H[r]))
p1 <- p1 + theme(axis.title=element_text(size=16),
                 axis.text=element_text(size=12),
                 axis.text.x=element_text(angle=90, size=12, vjust=0.5, hjust=1.0),
                 legend.position="bottom")
save_plot(paste("selection/single-pop/mut_rate_var/Hr_landscapes_", c + 1, ".png", sep=""), p1, base_height=10, base_width=15)

p2 <- ggplot(data=filter(molten_windows, N==1000, shapes_mu==1e+6, scales_mu==1e-8), aes(x=window, y=value, color=as.factor(scale_sel)))
p2 <- p2 + geom_line() + facet_grid(shapes_rec~scales_rec) + theme_bw()
p2 <- p2 + scale_x_discrete(breaks=pretty_breaks())
p2 <- p2 + scale_y_continuous(breaks=pretty_breaks())
p2 <- p2 + labs(title="Hr landscapes, N=1000, shape_mu=1e+6, scale_mu=1e-8, rows~1/var_rec, cols~mean_rec", x="Window", y=expression(H[r]))
p2 <- p2 + theme(axis.title=element_text(size=16),
                 axis.text=element_text(size=12),
                 axis.text.x=element_text(angle=90, size=12, vjust=0.5, hjust=1.0),
                 legend.position="bottom")
save_plot(paste("selection/single-pop/mut_rate_var/Hr_landscapes_", c + 2, ".png", sep=""), p2, base_height=10, base_width=15)

p3 <- ggplot(data=filter(molten_windows, N==10000, shapes_mu==1e+6, scales_mu==1e-9), aes(x=window, y=value, color=as.factor(scale_sel)))
p3 <- p3 + geom_line() + facet_grid(shapes_rec~scales_rec) + theme_bw()
p3 <- p3 + scale_x_discrete(breaks=pretty_breaks())
p3 <- p3 + scale_y_continuous(breaks=pretty_breaks())
p3 <- p3 + labs(title="Hr landscapes, N=10000, shape_mu=1e+6, scale_mu=1e-9, rows~1/var_rec, cols~mean_rec", x="Window", y=expression(H[r]))
p3 <- p3 + theme(axis.title=element_text(size=16),
                 axis.text=element_text(size=12),
                 axis.text.x=element_text(angle=90, size=12, vjust=0.5, hjust=1.0),
                 legend.position="bottom")
save_plot(paste("selection/single-pop/mut_rate_var/Hr_landscapes_", c + 3, ".png", sep=""), p3, base_height=10, base_width=15)

p4 <- ggplot(data=filter(molten_windows, N==10000, shapes_mu==1e+6, scales_mu==1e-8), aes(x=window, y=value, color=as.factor(scale_sel)))
p4 <- p4 + geom_line() + facet_grid(shapes_rec~scales_rec) + theme_bw()
p4 <- p4 + scale_x_discrete(breaks=pretty_breaks())
p4 <- p4 + scale_y_continuous(breaks=pretty_breaks())
p4 <- p4 + labs(title="Hr landscapes, N=10000, shape_mu=1e+6, scale_mu=1e-8, rows~1/var_rec, cols~mean_rec", x="Window", y=expression(H[r]))
p4 <- p4 + theme(axis.title=element_text(size=16),
                 axis.text=element_text(size=12),
                 axis.text.x=element_text(angle=90, size=12, vjust=0.5, hjust=1.0),
                 legend.position="bottom")
save_plot(paste("selection/single-pop/mut_rate_var/Hr_landscapes_", c + 4, ".png", sep=""), p4, base_height=10, base_width=15)

# visualizing R^2
molten_tbl <- pivot_longer(tbl, cols=c("Total_R2", "Mutation_R2", "Recombination_R2", "Selection_R2", "Mu:Sel_R2", "Rec:Sel_R2"), names_to="variable")

# TODO change filter parameters (mu) after new set of evolutionary parameters is used

p5 <- ggplot(data=filter(molten_tbl, N==1000, scales_mu==1e-9), aes(x=shapes_mu, y=value, shape=variable, color=as.factor(scale_sel)))
p5 <- p5 + geom_line() + facet_grid(shapes_rec~scales_rec)+ theme_bw()
p5 <- p5 + geom_point(aes(shape=variable), size=4)
p5 <- p5 + scale_shape_manual(values=c(0, 1, 2, 3, 4, 8), name = "",
                              labels=c(expression(mu:alpha), expression(mu), expression(r:alpha), expression(r), expression(alpha),"Total"))
p5 <- p5 + scale_x_continuous(breaks=unique(molten_tbl$shapes_mu), trans="log10") 
p5 <- p5 + scale_y_continuous(breaks=pretty_breaks())
p5 <- p5 + labs(title = "Var Explained, N=1000, mu=1e-9, rows~1/var_rec, cols~mean_rec", x="~1/var_mu", y="Variance Explained (%)")
p5 <- p5 + theme(axis.title=element_text(size=16), 
                 axis.text=element_text(size=12), 
                 axis.text.x=element_text(angle=90, size=12, vjust=0.5, hjust=1.0),
                 legend.text=element_text(size=16), legend.position="bottom")
save_plot("selection/single-pop/R2_plot_1.png", p5, base_height=10, base_width=15)

p6 <- ggplot(data=filter(molten_tbl, N==1000, scales_mu==1e-8), aes(x=shapes_mu, y=value, shape=variable, color=as.factor(scale_sel)))
p6 <- p6 + geom_line() + facet_grid(shapes_rec~scales_rec)+ theme_bw()
p6 <- p6 + geom_point(aes(shape=variable), size=4)
p6 <- p6 + scale_shape_manual(values=c(0, 1, 2, 3, 4, 8), name = "",
                              labels=c(expression(mu:alpha), expression(mu), expression(r:alpha), expression(r), expression(alpha),"Total"))
p6 <- p6 + scale_x_continuous(breaks=unique(molten_tbl$shapes_mu), trans="log10") 
p6 <- p6 + scale_y_continuous(breaks=pretty_breaks())
p6 <- p6 + labs(title = "Var Explained, N=1000, mu=1e-8, rows~1/var_rec, cols~mean_rec", x="~1/var_mu", y="Variance Explained (%)")
p6 <- p6 + theme(axis.title=element_text(size=16), 
                 axis.text=element_text(size=12), 
                 axis.text.x=element_text(angle=90, size=12, vjust=0.5, hjust=1.0),
                 legend.text=element_text(size=16), legend.position="bottom")
save_plot("selection/single-pop/R2_plot_2.png", p6, base_height=10, base_width=15)

p7 <- ggplot(data=filter(molten_tbl, N==10000, scales_mu==1e-9), aes(x=shapes_mu, y=value, shape=variable, color=as.factor(scale_sel)))
p7 <- p7 + geom_line() + facet_grid(shapes_rec~scales_rec)+ theme_bw()
p7 <- p7 + geom_point(aes(shape=variable), size=4)
p7 <- p7 + scale_shape_manual(values=c(0, 1, 2, 3, 4, 8), name = "",
                              labels=c(expression(mu:alpha), expression(mu), expression(r:alpha), expression(r), expression(alpha),"Total"))
p7 <- p7 + scale_x_continuous(breaks=unique(molten_tbl$shapes_mu), trans="log10") 
p7 <- p7 + scale_y_continuous(breaks=pretty_breaks())
p7 <- p7 + labs(title = "Var Explained, N=10000, mu=1e-9, rows~1/var_rec, cols~mean_rec", x="~1/var_mu", y="Variance Explained (%)")
p7 <- p7 + theme(axis.title=element_text(size=16), 
                 axis.text=element_text(size=12), 
                 axis.text.x=element_text(angle=90, size=12, vjust=0.5, hjust=1.0),
                 legend.text=element_text(size=16), legend.position="bottom")
save_plot("selection/single-pop/R2_plot_3.png", p7, base_height=10, base_width=15)

p8 <- ggplot(data=filter(molten_tbl, N==10000, scales_mu==1e-8), aes(x=shapes_mu, y=value, shape=variable, color=as.factor(scale_sel)))
p8 <- p8 + geom_line() + facet_grid(shapes_rec~scales_rec)+ theme_bw()
p8 <- p8 + geom_point(aes(shape=variable), size=4)
p8 <- p8 + scale_shape_manual(values=c(0, 1, 2, 3, 4, 8), name = "",
                              labels=c(expression(mu:alpha), expression(mu), expression(r:alpha), expression(r), expression(alpha),"Total"))
p8 <- p8 + scale_x_continuous(breaks=unique(molten_tbl$shapes_mu), trans="log10") 
p8 <- p8 + scale_y_continuous(breaks=pretty_breaks())
p8 <- p8 + labs(title = "Var Explained, N=10000, mu=1e-8, rows~1/var_rec, cols~mean_rec", x="~1/var_mu", y="Variance Explained (%)")
p8 <- p8 + theme(axis.title=element_text(size=16), 
                 axis.text=element_text(size=12), 
                 axis.text.x=element_text(angle=90, size=12, vjust=0.5, hjust=1.0),
                 legend.text=element_text(size=16), legend.position="bottom")
save_plot("selection/single-pop/R2_plot_4.png", p8, base_height=10, base_width=15)
```

# Selection 2-populations

## No-migration

### Building Demes files

This first set of model is meant to visualize statistics for two populations,
potentially with different selective coefficients between them

```{r, demes-selected-2-pop, include=F, eval=F, results=F, message=F, warning=F}

N1 <- 10000
N2 <- c(N1, N1 / 10)
u <- 1e-8
r  <- c(1e-5, 1e-6, 1e-7, 0)
s1 <- c(0, -1e-04, -1e-3)
s2 <- c(0, -1e-04, -1e-3)
sanc <- c(0, -1e-3)
split_time <- c(400, 2000) # in generations

params <- crossing(N1, N2, u, r, s1, s2, sanc, split_time)
params$scenario <- 1:nrow(params)
n_models <- nrow(params)

write.csv(params, "selection/no_migration/params.csv", quote=F, row.names=F)

Sys.setenv(num_models=n_models)

py$params <- params # stores df to be used in python sections
py$n_models <- as.integer(n_models + 1) # because python is indexed from zero

for(i in 1:n_models) {
  
  name <- paste("model_", i, sep="")
  sink(paste("selection/no_migration/", name, ".yaml", sep=""))
  
  cat("time_units: generations\n")
  cat("demes:\n")
  cat("  - name: X\n")
  cat("    epochs:\n")
  cat("      - end_time: ")
  cat(params$split_time[i])
  cat("\n")
  cat("        start_size: ")
  cat(params$N1[i])
  cat("\n")
  cat("  - name: A\n")
  cat("    ancestors: [X]\n")
  cat("    epochs:\n")
  cat("      - end_time: 0\n")
  cat("        start_size: ")
  cat(params$N1[i])
  cat("\n")
  cat("  - name: B\n")
  cat("    ancestors: [X]\n")
  cat("    epochs:\n")
  cat("      - end_time: 0\n")
  cat("        start_size: ")
  cat(params$N2[i])
  cat("\n")
  cat("metadata:\n")
  cat("  - name: mutation\n")
  cat("    epochs:\n")
  cat("      - end_time: 0\n")
  cat("        rates: [")
  cat(params$u[i])
  cat("]\n")
  cat("  - name: recombination\n")
  cat("    epochs:\n")
  cat("      - end_time: 0\n")
  cat("        rates: [")
  cat(params$r[i])
  cat("]\n")
  cat("  - name: selection\n")
  cat("    start_time: .inf\n")
  cat("    epochs:\n")
  cat("      - end_time: ")
  cat(params$split_time[i])
  cat("\n")
  cat("        rates: [")
  cat(params$sanc[i])
  cat("]\n")
  cat("      - end_time: 0\n")
  cat("        rates: [")
  cat(params$s1[i])
  cat(", ")
  cat(params$s2[i])
  cat("]\n")
  
  sink()
}

```

### Running moments++

```{bash run-moments++-2-pops, engine.opts='-i', include=F, eval=F, results=F, message=F, warning=F}

source ~/.bashrc

cd selection/no_migration

for((i=1;i<=$num_models;i++))
do
  momentspp params=opt.bpp FILE=model_$i.yaml LABEL=model_$i ORDER=25
done

cd ../..
```

### Plots 

TODO assimilate split time into plots

```{r, plots-selected-2-pop, include=T, eval=F, results=F, message=F, warning=F}

core_stats <- c("DD_1_1", 
                "DD_1_2", 
                "DD_2_2",
                "Dr_1_1_(1-2p1)^1",
                "Dr_1_1_(1-2p2)^1",
                "Dr_1_2_(1-2p2)^1",
                "Dr_2_1_(1-2p1)^1", 
                "Dr_2_1_(1-2p2)^1",
                "Dr_2_2_(1-2p2)^1",
                "Hl_1_1", 
                "Hl_1_2",
                "Hl_2_1",
                "Hl_2_2",
                "Hr_1_1", 
                "Hr_1_2",
                "Hr_2_2",
                "pi2_1_1_1_1", 
                "pi2_1_1_1_2",
                "pi2_1_1_2_2", 
                "pi2_1_2_1_2", 
                "pi2_1_2_2_2", 
                "pi2_2_1_1_2", 
                "pi2_2_2_1_1",
                "pi2_2_2_2_2")

# to be on the very safe side...
core_stats <- sort(core_stats)

expectation <- numeric()
for(i in 1:n_models) {
  vals <- read.table(paste("selection/no_migration/model_", i, "_expectations.txt", sep=""))
  vals <- subset(vals, V1 %in% core_stats)
  expectation <- c(expectation, vals$V3)
}

tbl <- as.data.frame(expectation)
tbl$stats <- core_stats
scenario <- NULL
for(i in 1:n_models) { 
  scenario <- c(scenario, rep(i, length(core_stats)))
}
tbl$scenario <- scenario

df <- full_join(tbl, params, by="scenario")

# transforms variables
df$alpha_1 <- 2 * df$s1 * df$N1
df$alpha_2 <- 2 * df$s2 * df$N2

# symmetric population sizes
df_sym_N <- df[df$N1==df$N2,]
df_sym_N[df_sym_N$r==0,]$r <- 1e-12 # for log transformation

# pi2 within
pa1 <- ggplot(data=subset(filter(df_sym_N, stats %in% c("pi2_1_1_1_1", "pi2_2_2_2_2")), sanc == 0), aes(x=r, y=expectation, color=stats)) 
pa1 <- pa1 + geom_point(size=3, shape=0) + geom_line() + theme_bw() + facet_grid(+alpha_1~alpha_2)
pa1 <- pa1 + scale_x_log10(breaks=c(1e-12, 1e-6, 1e-5, 1e-4, 1e-3))
pa1 <- pa1 + scale_y_log10(breaks=c(1e-6, 2e-6, 3e-6, 4e-6, 5e-6))
pa1 <- pa1 + scale_shape_manual(values=c(0))
pa1 <- pa1 + scale_color_brewer(palette = "Dark2")
pa1 <- pa1 + labs(title="Symmetric Population Sizes", x="r", y="Expectation")
pa1 <- pa1 + theme(axis.title=element_text(size=16),
                 axis.text=element_text(size=12),
                 axis.text.x=element_text(angle=90, size=12, vjust=0.5, hjust=1.0),
                 legend.position="bottom")
pa1
save_plot("selection/no_migration/pi2_within_no-mig.png", pa1, base_height=12, base_width=12)

# pi2 cross-pop
pa2 <- ggplot(data=subset(filter(df_sym_N, stats %in% c("pi2_1_1_1_2", "pi2_1_1_2_2", "pi2_1_2_1_2", "pi2_2_1_1_2", "pi2_1_2_2_2", "pi2_2_2_1_1")), sanc == 0),
              aes(x=r, y=expectation, color=stats)) 
pa2 <- pa2 + geom_point(size=3, shape=0) + geom_line() + theme_bw() + facet_grid(+alpha_1~alpha_2)
pa2 <- pa2 + scale_x_log10(breaks=c(1e-12, 1e-6, 1e-5, 1e-4, 1e-3))
pa2 <- pa2 + scale_y_log10(breaks=c(1e-6, 2e-6, 3e-6, 4e-6, 5e-6, 1e-5))
pa2 <- pa2 + scale_color_brewer(palette = "Dark2")
pa2 <- pa2 + labs(title="Symmetric Population Sizes", x="r", y="Expectation")
pa2 <- pa2 + theme(axis.title=element_text(size=16),
                 axis.text=element_text(size=12),
                 axis.text.x=element_text(angle=90, size=12, vjust=0.5, hjust=1.0),
                 legend.position="bottom")
pa2
save_plot("selection/no_migration/pi2_cross_no-mig.png", pa2, base_height=12, base_width=12)

# Dr
pb <- ggplot(data=subset(filter(df_sym_N, str_detect(stats, "^Dr")), sanc == 0), aes(x=r, y=expectation, color=stats)) 
pb <- pb + geom_point(size=3, shape=0) + geom_line() + theme_bw() + facet_grid(+alpha_1~alpha_2)
pb <- pb + scale_x_log10(breaks=c(1e-12, 1e-6, 1e-5, 1e-4, 1e-3))
pb <- pb + scale_color_brewer(palette = "Dark2")
pb <- pb + labs(title="Symmetric Population Sizes", x="r", y="Expectation")
pb <- pb + theme(axis.title=element_text(size=16),
                 axis.text=element_text(size=12),
                 axis.text.x=element_text(angle=90, size=12, vjust=0.5, hjust=1.0),
                 legend.position="bottom")
pb
save_plot("selection/no_migration/Dr_no-mig.png", pd, base_height=12, base_width=12)

# Hets
pc <- ggplot(data=subset(filter(df_sym_N, str_detect(stats, "^H")), sanc == 0), aes(x=r, y=expectation, color=stats)) 
pc <- pc + geom_point(size=3, shape=0) + geom_line() + theme_bw() + facet_grid(+alpha_1~alpha_2)
pc <- pc + scale_x_log10(breaks=c(1e-12, 1e-6, 1e-5, 1e-4, 1e-3))
pc <- pc + scale_y_log10(breaks=pretty_breaks())
pc <- pc + scale_color_brewer(palette = "Dark2")
pc <- pc + labs(title="Symmetric Population Sizes", x="r", y="Expectation")
pc <- pc + theme(axis.title=element_text(size=16),
                 axis.text=element_text(size=12),
                 axis.text.x=element_text(angle=90, size=12, vjust=0.5, hjust=1.0),
                 legend.position="bottom")
pc
save_plot("selection/no_migration/Hets_no-mig.png", pd, base_height=12, base_width=12)

# Hr
pd <- ggplot(data=filter(df_sym_N, stats %in% c("Hr_1_1", "Hr_2_2")), aes(x=r, y=expectation, color=stats, shape=as.factor(sanc))) 
pd <- pd + geom_point(size=3) + geom_line() + theme_bw() + facet_grid(+alpha_1~alpha_2)
pd <- pd + scale_x_log10(breaks=c(1e-12, 1e-6, 1e-5, 1e-4, 1e-3))
pd <- pd + scale_y_log10(breaks=pretty_breaks())
pd <- pd + scale_color_brewer(palette = "Dark2")
pd <- pd + scale_shape_manual(values=c(0, 1))
pd <- pd + labs(title="Symmetric Population Sizes", x="r", y="Expectation")
pd <- pd + theme(axis.title=element_text(size=16),
                 axis.text=element_text(size=12),
                 axis.text.x=element_text(angle=90, size=12, vjust=0.5, hjust=1.0),
                 legend.position="bottom")
pd
save_plot("selection/no_migration/Hr_no-mig.png", pd, base_height=12, base_width=12)

# Hl
pe <- ggplot(data=subset(filter(df_sym_N, str_detect(stats, "^Hl")), sanc == 0), aes(x=alpha_1, y=expectation, color=stats, shape=as.factor(alpha_2))) 
pe <- pe + geom_point(size=3) + geom_line() + theme_bw() #+ facet_wrap(~alpha_2, nrow=1)
pe <- pe + scale_y_log10(breaks=pretty_breaks())
pe <- pe + scale_color_brewer(palette = "Dark2")
pe <- pe + scale_shape_manual(values=0:length(levels(as.factor(df_sym_N$alpha_2))))
pe <- pe + labs(title="Symmetric Population Sizes", x="alpha_1", y="Expectation")
pe <- pe + theme(axis.title=element_text(size=16),
                 axis.text=element_text(size=12),
                 axis.text.x=element_text(angle=90, size=12, vjust=0.5, hjust=1.0),
                 legend.position="bottom")
pe
save_plot("selection/no_migration/Hl_no-mig.png", pe, base_height=12, base_width=12)

pf <- plot_ly(data=subset(filter(df_sym_N, stats %in% c("Hl_1_2", "Hl_2_1")), sanc == 0), x=~alpha_1, y=~alpha_2, z=~expectation, type="scatter3d", 
              mode="markers", color=~stats)
pf

# D1D2
d1d2 <- filter(df_sym_N, stats %in% "DD_1_2") 
d1d2$ratio <- d1d2$expectation / subset(df_sym_N, stats %in% "pi2_1_2_1_2")$expectation

# save ratio D1D2/ pi2_1212 for later, plot D1D2 expectation now
p1 <- ggplot(data=d1d2, aes(x=r, y=expectation, shape=as.factor(sanc))) + facet_grid(+alpha_1~alpha_2)
p1 <- p1 + geom_point(size=3) + geom_line() + theme_bw()
p1 <- p1 + scale_x_log10(breaks=c(1e-12, 1e-6, 1e-5, 1e-4, 1e-3))
p1 <- p1 + scale_y_log10(breaks=c(5e-8, 1e-7, 2e-7, 5e-7, 1e-6))
p1 <- p1 + scale_shape_manual(values=c(0, 1))
p1 <- p1 + labs(title="Symmetric Population Sizes", x="r", y="E[D1D2]")
p1 <- p1 + theme(axis.title=element_text(size=16),
                 axis.text=element_text(size=12),
                 axis.text.x=element_text(angle=90, size=12, vjust=0.5, hjust=1.0),
                 legend.position="bottom")
p1
save_plot("selection/no_migration/Covar_D_no-mig_1.png", p1, base_height=12, base_width=12)

d1d2 <- d1d2[d1d2$alpha_1==d1d2$alpha_2,] # symmetric alpha

p2 <- ggplot(data=d1d2, aes(x=alpha_1, y=expectation, color=as.factor(r), shape=as.factor(sanc)))
p2 <- p2 + geom_point(size=3) + geom_line() + theme_bw()
p2 <- p2 + scale_color_brewer(palette="Dark2")
p2 <- p2 + scale_shape_manual(values=c(0, 1))
p2 <- p2 + scale_y_log10(breaks=c(1e-7, 2e-7, 4e-7, 6e-7, 1e-6))
p2 <- p2 + labs(title="Symmetric Population Sizes and Selection coefficients", x="2Ns", y="E[D1D2]")
p2 <- p2 + theme(axis.title=element_text(size=16),
                 axis.text=element_text(size=12),
                 axis.text.x=element_text(angle=90, size=12, vjust=0.5, hjust=1.0),
                 legend.position="bottom")
p2
save_plot("selection/no_migration/Covar_D_no-mig_2.png", p2, base_height=12, base_width=12)

# Hr
hr11 <- subset(df_sym_N, stats %in% "Hr_1_1") 

p3 <- ggplot(data=hr11, aes(x=r, y=expectation * 1e+3, color=as.factor(alpha_1), shape=as.factor(sanc)))
p3 <- p3 + geom_point(size=3) + geom_line() + theme_bw()
p3 <- p3 + scale_color_brewer(palette="Dark2")
p3 <- p3 + scale_shape_manual(values=c(0, 1))
p3 <- p3 + scale_x_log10(breaks=c(1e-12, 1e-6, 1e-5, 1e-4))
p3 <- p3 + scale_y_log10(breaks=pretty_breaks())
p3 <- p3 + labs(title="Symmetric Population Sizes", x="r", y="E[Hr_11] (x 10^-3)")
p3 <- p3 + theme(axis.title=element_text(size=16),
                 axis.text=element_text(size=12),
                 axis.text.x=element_text(angle=90, size=12, vjust=0.5, hjust=1.0),
                 legend.position="bottom")
p3
save_plot("selection/no_migration/Hr_11.png", p3, base_height=12, base_width=12)

# Fst at the Left Locus
Hl_12 <- subset(df_sym_N, stats %in% "Hl_1_2")
Hl_21 <- subset(df_sym_N, stats %in% "Hl_2_1")
Hl_11 <- subset(df_sym_N, stats %in% "Hl_1_1")
Hl_22 <- subset(df_sym_N, stats %in% "Hl_2_2")

pi_between <- (Hl_12$expectation + Hl_21$expectation) / 2 
pi_within <- (Hl_11$expectation + Hl_22$expectation) / 2 

fst <- Hl_12 # just to copy the data frame
fst$expectation <- 1 - pi_within / pi_between # we'll be looking at this one

p4 <- plot_ly(data=fst, x=~alpha_1, y=~alpha_2, z=~expectation, type="scatter3d", 
              mode="markers", color=~expectation)
p4

# Fst at the Right Locus
Hr_12 <- subset(df_sym_N, stats %in% "Hr_1_2")
#Hr_21 <- subset(df_sym_N, stats %in% "Hr_2_1")
Hr_11 <- subset(df_sym_N, stats %in% "Hr_1_1")
Hr_22 <- subset(df_sym_N, stats %in% "Hr_2_2")

pi_between <- Hr_12$expectation #(Hr_12$expectation + Hr_21$expectation) / 2 
pi_within <- (Hr_11$expectation + Hr_22$expectation) / 2 

fst <- Hr_12
fst$expectation <- 1 - pi_within / pi_between

p5 <- ggplot(data=fst, aes(x=r, y=expectation, shape=as.factor(sanc))) + facet_grid(+alpha_1~alpha_2)
p5 <- p5 + geom_point(size=3) + geom_line() + theme_bw()
p5 <- p5 + scale_shape_manual(values=c(0, 1))
p5 <- p5 + scale_x_log10(breaks=c(1e-12, 1e-6, 1e-5, 1e-4, 1e-3))
p5 <- p5 + scale_y_log10(breaks=pretty_breaks())
p5 <- p5 + labs(title="Symmetric Population Sizes", x="r", y="Fst neutral locus")
p5 <- p5 + theme(axis.title=element_text(size=16),
                 axis.text=element_text(size=12),
                 axis.text.x=element_text(angle=90, size=12, vjust=0.5, hjust=1.0),
                 legend.position="bottom")

save_plot("selection/no_migration/Fst_no-mig_right_2.png", p5, base_height=12, base_width=12)

fst <- fst[fst$alpha_1==fst$alpha_2,]

p6 <- ggplot(data=fst, aes(x=alpha_1, y=expectation, color=as.factor(r), shape=as.factor(sanc)))
p6 <- p6 + geom_point(size=3) + geom_line() + theme_bw()
p6 <- p6 + scale_color_brewer(palette="Dark2")
p6 <- p6 + scale_shape_manual(values=c(0, 1))
p6 <- p6 + scale_y_log10(breaks=pretty_breaks())
p6 <- p6 + labs(title="Symmetric Population Sizes and Selection coefficients", x="2Ns", y="Fst neutral locus")
p6 <- p6 + theme(axis.title=element_text(size=16),
                 axis.text=element_text(size=12),
                 axis.text.x=element_text(angle=90, size=12, vjust=0.5, hjust=1.0),
                 legend.position="bottom")
p6
save_plot("selection/no_migration/Fst_no-mig_right_1.png", p6, base_height=12, base_width=12)

# all population sizes
df_cpy <- df
df_cpy[df_cpy$r==0,]$r <- 1e-12 # for log transformation

# D1D2
d1d2 <- subset(df_cpy, stats %in% "DD_1_2") 
d1d2$ratio <- d1d2$expectation / subset(df_cpy, stats %in% "pi2_1_2_1_2")$expectation
  
p7 <- ggplot(data=d1d2, aes(x=r, y=expectation, color=as.factor(N2), shape=as.factor(sanc))) + facet_grid(+s1~s2)
p7 <- p7 + geom_point(size=3) + geom_line() + theme_bw()
p7 <- p7 + scale_color_brewer(palette="Dark2")
p7 <- p7 + scale_shape_manual(values=c(0, 1))
p7 <- p7 + scale_x_log10(breaks=c(1e-12, 1e-6, 1e-5, 1e-4, 1e-3))
p7 <- p7 + scale_y_log10(breaks=c(1e-7, 2e-7, 4e-7, 6e-7, 1e-6))
p7 <- p7 + labs(title="All Population Sizes", x="r", y="E[D1D2]")
p7 <- p7 + theme(axis.title=element_text(size=16),
                 axis.text=element_text(size=12),
                 axis.text.x=element_text(angle=90, size=12, vjust=0.5, hjust=1.0),
                 legend.position="bottom")
p7
save_plot("selection/no_migration/Covar_D_no-mig_3.png", p7, base_height=12, base_width=12)

# Fst at the Left Locus
Hl_12 <- subset(df_cpy, stats %in% "Hl_1_2")
Hl_21 <- subset(df_cpy, stats %in% "Hl_2_1")
Hl_11 <- subset(df_cpy, stats %in% "Hl_1_1")
Hl_22 <- subset(df_cpy, stats %in% "Hl_2_2")

pi_between <- (Hl_12$expectation + Hl_21$expectation) / 2 
pi_within <- (Hl_11$expectation + Hl_22$expectation) / 2 

fst <- Hl_12 # just to copy the data frame
fst$expectation <- 1 - pi_within / pi_between # we'll be looking at this one

p8 <- plot_ly(data=fst, x=~s1, y=~s2, z=~expectation, type="scatter3d", mode="markers", color=~N2)
p8
```

## Migration (TODO)
