#!/usr/bin/env Rscript
args=commandArgs(trailingOnly=T)

library(R.utils)
library(tidyverse)
library(data.table)
library(pracma) # for cubicspline()
library(cowplot)
library(scales)

# This script does the heavy-lifting of the pipeline to simulate genomic
# landscapes. Here we compute B-values for sampled sites,
# interpolate them and multiply by local mu to generate predictions of
# pairwise diversity along the chromosome that was simulated by bgs_maps.R

num_iter <- 4 # for updating B-values due to interference selection
jump_length <- as.numeric(args[1]) # sampling distance for interpolating B-vals

lookup_tbl <- fread("../../lookup_tbl.csv.gz") # generated by setup_models.Rmd

N <- unique(lookup_tbl$N) # should be of length 1
u <- unique(lookup_tbl$u) # should be of length 1
bin_size <- 1e+3 # "basal" scale, should be the same as in bgs_maps.R

dt_r <- data.table(unique(lookup_tbl[,3]))
dt_s <- data.table(unique(lookup_tbl[,4]))

setkey(dt_r, lookup_r)
setkey(dt_s, lookup_s)

# loads tables generated by bgs_maps.R
mmap <- fread("mmap.csv")
rmap <- fread("rmap.csv")
smap <- fread("smap.csv")

dt_neutral <- filter(smap[,2:4], s==0)
dt_exons <- filter(smap[,2:4], s<0)

setkey(mmap, start, end)
setkey(rmap, start, end)
setkey(smap, start, end)
setkey(dt_neutral, start, end)
setkey(dt_exons, start, end)
setkey(lookup_tbl, lookup_r, lookup_s)

exon_lengths <- unique((dt_exons$end - dt_exons$start)) # should be of length 1
L <- mmap$end[nrow(mmap)] # seq length, same for all maps

# finding mu for each sampled position
neut_pos <- unlist(apply(dt_neutral, 1, 
              function(x) seq(from=x[1], to=x[2], by=jump_length))) + 1
linear_pos <- sort(c(neut_pos, dt_exons$start + exon_lengths / 2))

focal_mu <- unlist(lapply(linear_pos, function(pos) mmap[J(pos), roll=T]$u))

pos_dt <- setDT(bind_cols(linear_pos, focal_mu))
pos_dt$idx <- 1:nrow(pos_dt)
names(pos_dt) <- c("position", "focal_mu", "idx")
setkey(pos_dt, position)

# thinning sampled neutral sites clustered away from exons
# heuristic: downsample to a maximum of 5 sampled neutral sites between exons
pos_dt$neutral <- pos_dt$position %in% neut_pos
runs <- rle(pos_dt$neutral)
runs <- cbind.data.frame(runs$lengths, cumsum(runs$lengths), runs$values)  
names(runs) <- c("run_length", "end_idx", "val")
short_runs <- filter(runs, run_length <= 5)
long_runs <- filter(runs, run_length > 5)

short_runs$start_idx <- short_runs$end_idx - short_runs$run_length + 1
short_runs <- dplyr::select(short_runs, c(start_idx, end_idx))
short_pos <- unique(unlist(apply(short_runs, 1,
               function(x) seq(from=x[1], to=x[2], by=1))))
  
long_runs$start_idx <- long_runs$end_idx - long_runs$run_length + 1
long_runs <- dplyr::select(long_runs, c(start_idx, end_idx))

get_5_pts <- function(y) {
  # y[1] = start index, y[2] = end index of run
  mid <- (y[1] + y[2]) %/% 2
  q1 <- (mid + y[1]) %/% 2
  q2 <- (y[2] + mid) %/% 2
  return(c(y[1], q1, mid, q2, y[2]))
}

thinned_neut <- c(apply(long_runs, 1, get_5_pts))
thinned_pos <- sort(c(thinned_neut, short_pos))

pos_dt <- pos_dt[thinned_pos,] # thinning
linear_pos <- pos_dt$position # updates after thinning

B_values <- rep(1, length(linear_pos)) # init 
tbl <- as.data.frame(matrix(ncol=num_iter, nrow=length(B_values))) # for viz

for(i in 1:num_iter) {
  # approximates cumulative rec by looking at r at sampled sites only
  focal_r <- unlist(lapply(linear_pos, function(pos) rmap[J(pos), roll=T]$r)) 
  focal_r <- focal_r * B_values
  cum_rec <- numeric(length=length(linear_pos))
  cum_rec[1] <- focal_r[1]

  for(j in 2:length(linear_pos)) {
    cum_rec[j] <- cum_rec[j-1] + (linear_pos[j] - linear_pos[j-1]) * focal_r[j]
  }
  pos_dt$cumrec <- cum_rec

  # pre-computes "effective" rec. dist. between sampled sites and exons
  cr_exons <- pos_dt[position %in% (dt_exons$start + exon_lengths / 2),]
  prd <- as.data.frame(2 * N * abs(outer(pos_dt$cumrec, cr_exons$cumrec, "-")))
  names(prd) <- 1:ncol(prd)
  
  erd <- prd # "effective" rec. distance is inversely proportional to alpha
  for(j in 1:ncol(erd)) { erd[,j] <- erd[,j] / abs(2 * N * dt_exons$s[j]) }
  relevant_exons <- apply(erd, 2, function(x) x < 10 * N * 1e-3)
  exons_per_samp_site <- apply(relevant_exons, 1, function(x) which(x))
  
  getB <- function(focal_exon, focal_samp) { # arguments are site indices
    total_r <- prd[focal_samp, focal_exon] / (2 * N) 
    if(total_r > 1e-2) { return(1) } # not in lookup_tbl, too high anyway
    else {
      exon_pos <- dt_exons$start[focal_exon] + exon_lengths / 2 # midpoint pos
      idx <- pos_dt[.(exon_pos)]$idx # index within pos_dt
      dt_exons[focal_exon,]$s <- dt_exons[focal_exon,]$s * B_values[idx]
      focal_s <- dt_exons[focal_exon,]$s
      
      # lookup_tbl grid is fine-grained enough, we don't need interpolation
      closest_r <- dt_r[dt_r[.(total_r), roll="nearest", which=T]]
      closest_s <- dt_s[dt_s[.(focal_s), roll="nearest", which=T]]
      
      # correction for interference on u is applied by scaling exon lengths
      hr <- lookup_tbl[.(closest_r, closest_s)]$Hr # lookup_tbl has single u val
      return((hr / (2 * N * u)) ^ (exon_lengths * B_values[idx]))
    }
  }
  
  tmp <- B_values # temporary copy
  cat(paste("Computing B's...(iteration ", i, " of ", num_iter, ")\n", sep=""))
  pb <- txtProgressBar(min=1, max=length(exons_per_samp_site), style=3)
  for(k in 1:length(exons_per_samp_site)) {
    setTxtProgressBar(pb, k)
    if(length(exons_per_samp_site[[k]]) > 0) {
      B <- unlist(lapply(exons_per_samp_site[[k]], getB, focal_samp=k))
      tmp[k] <- cumprod(B)[length(B)]
    }
    else {
      tmp[k] <- 1 
    }
  }
  close(pb)

  cat(tail(tmp))
  B_values <- tmp
  tbl[,i] <- B_values
}

# interpolating B-values for every site 1:L can also take some time
B_map <- cubicspline(linear_pos, B_values, 1:L)
ones <- rep(1, L)

# uses closed intervals for mmap & rmap (+1 start coordinates)
pi0s <- invisible(apply(mmap, 1, function(x)
                  ones[(as.numeric(x[2])+1):as.numeric(x[3])] *
                  as.numeric(x[4]) * 2 * N))
pis <- invisible(apply(mmap, 1, function(x)
                 B_map[(as.numeric(x[2])+1):as.numeric(x[3])] *
                 as.numeric(x[4]) * 2 * N))

pi0s <- unlist(pi0s)
pis <- unlist(pis)
hrmap <- setDT(bind_cols(1:L, pis, pi0s, B_map))
names(hrmap) <- c("Pos", "Hr", "pi0", "B")
fwrite(hrmap, "hrmap.csv.gz", compress="gzip") 

# binning 
hrmap$bin <- ((hrmap$Pos - 1) %/% bin_size)
hrmap_1kb <- hrmap %>% group_by(bin) %>%
             summarize(avg_pi=mean(Hr), avg_pi0=mean(pi0), avg_B=mean(B))
hrmap_1kb$bin_10kb <- hrmap_1kb$bin %/% 10
hrmap_1kb$bin_100kb <- hrmap_1kb$bin %/% 100
hrmap_10kb <- hrmap_1kb %>% group_by(bin_10kb) %>% 
              summarise_at(c("avg_pi", "avg_pi0", "avg_B"), mean)
hrmap_100kb <- hrmap_1kb %>% group_by(bin_100kb) %>% 
               summarise_at(c("avg_pi", "avg_pi0", "avg_B"), mean)

fwrite(hrmap_1kb, "hrmap_1kb.csv")
fwrite(hrmap_10kb, "hrmap_10kb.csv")
fwrite(hrmap_100kb, "hrmap_100kb.csv")

# visualizing the iterative correction for interference selection
names(tbl) <- c(paste("iter_", rep(1:num_iter), sep=""))
tbl$pos <- as.integer(linear_pos)
m_tbl <- pivot_longer(tbl, cols=starts_with("iter"), names_to="Iteration")

pa <- ggplot(data=m_tbl[1e+3:3e+3,], aes(x=pos, y=value, color=Iteration)) + 
  geom_point(aes(alpha=0.5)) + theme_bw() + geom_line() + 
  scale_x_continuous(breaks=pretty_breaks()) +
  scale_y_log10(breaks=pretty_breaks()) + guides(alpha="none") + 
  labs(title="Iterative correction for interference", x="Pos", y="B-value") +
  theme(axis.title=element_text(size=16), 
        axis.text=element_text(size=12), 
        axis.text.x=element_text(size=12),
        legend.text=element_text(size=16),
        legend.title=element_text(size=16),
        legend.position="bottom")

save_plot("Interf.png", pa, base_height=8, base_width=16)

# visualizing the validity of threshold for spotting "relevant" exons
rex <- apply(prd, 2, function(x) x < 4 * N * 1e-2)
ex400 <- apply(rex, 1, function(x) which(x))

Bl400 <- unlist(lapply(ex400[[1]], getB, focal_samp=1))
Bq400 <- unlist(lapply(ex400[[length(neut_pos)/4]], getB,
                       focal_samp=length(neut_pos) %/% 4))
Bm400 <- unlist(lapply(ex400[[length(neut_pos)/2]], getB,
                       focal_samp=length(neut_pos) %/% 2))

Blr <- unlist(lapply(exons_per_samp_site[[1]], getB, focal_samp=1))
Bqr <- unlist(lapply(exons_per_samp_site[[length(neut_pos)/4]], getB,
                     focal_samp=length(neut_pos) %/% 4))
Bmr <- unlist(lapply(exons_per_samp_site[[length(neut_pos)/2]], getB,
                     focal_samp=length(neut_pos) %/% 2))

xl <- cbind.data.frame(as.numeric(prd[1, as.numeric(names(Bl400))]), Bl400)
xq <- cbind.data.frame(as.numeric(prd[1, as.numeric(names(Bq400))]), Bq400)
xm <- cbind.data.frame(as.numeric(prd[1, as.numeric(names(Bm400))]), Bm400)

names(xl) <- c("rec", "B")
names(xq) <- c("rec", "B")
names(xm) <- c("rec", "B")

xl$CummBval <- cumprod(xl$B)
xq$CummBval <- cumprod(xq$B)
xm$CummBval <- cumprod(xm$B)

xl$relevant <- xl$B %in% Blr
xq$relevant <- xq$B %in% Bqr
xm$relevant <- xm$B %in% Bmr

pl <- ggplot(data=xl, aes(x=rec, y=B, color=relevant)) +
  geom_point() + theme_bw() + scale_y_log10() +
  labs(title="B-values Left", x=NULL, y=NULL) +
  theme(axis.title=element_text(size=16),
        axis.text=element_text(size=12),
        axis.text.x=element_text(angle=90, size=12, vjust=0.5, hjust=1.0),
        legend.position="none")

pq <- ggplot(data=xq, aes(x=rec, y=B, color=relevant)) +
  geom_point() + theme_bw() + scale_y_log10() +
  labs(title="B-values 1/4 chr", x=NULL, y=NULL) +
  theme(axis.title=element_text(size=16),
        axis.text=element_text(size=12),
        axis.text.x=element_text(angle=90, size=12, vjust=0.5, hjust=1.0),
        legend.position="none")

pm <- ggplot(data=xm, aes(x=rec, y=B, color=relevant)) +
  geom_point() + theme_bw() + scale_y_log10() +
  labs(title="B-values 1/2 chr", x=NULL, y=NULL) +
  theme(axis.title=element_text(size=16),
        axis.text=element_text(size=12),
        axis.text.x=element_text(angle=90, size=12, vjust=0.5, hjust=1.0),
        legend.position="none")

ql <- ggplot(data=xl, aes(x=rec, y=CummBval, color=relevant)) + geom_point() +
  theme_bw() + scale_y_log10(limits=c(min(xl$CummBval), 1)) +
  labs(title="Cumm. B-value, Left", x="Cummulative Rho", y="B") +
  theme(axis.title=element_text(size=16),
        axis.text=element_text(size=12),
        axis.text.x=element_text(angle=90, size=12, vjust=0.5, hjust=1.0),
        legend.position="bottom")

qq <- ggplot(data=xq, aes(x=rec, y=CummBval, color=relevant)) + 
  geom_point() + theme_bw() + scale_y_log10() +
  labs(title="Cumm. B-value, 1/4 chr", x="Cummulative Rho", y="B") +
  theme(axis.title=element_text(size=16),
        axis.text=element_text(size=12),
        axis.text.x=element_text(angle=90, size=12, vjust=0.5, hjust=1.0),
        legend.position="bottom")

qm <- ggplot(data=xm, aes(x=rec, y=CummBval, color=relevant)) +
  geom_point() + theme_bw() + scale_y_log10() +
  labs(title="Cumm. B-value, 1/2 chr", x="Cummulative Rho", y="B") +
  theme(axis.title=element_text(size=16),
        axis.text=element_text(size=12),
        axis.text.x=element_text(angle=90, size=12, vjust=0.5, hjust=1.0),
        legend.position="bottom")

cp <- plot_grid(pl, pq, pm, ql, qq, qm, align='v', nrow=2)
save_plot("Bvals.png", cp, base_height=10, base_width=15)
