---
title: "Mutation Rate Variation Pt. II"
author: "Gustavo V. Barroso"
date: "`r Sys.Date()`"
output:
  pdf_document:
  toc: true
number_sections: true
toc_depth: 1
---
  
```{r setup, include=F, message=F, warning=F}

knitr::opts_chunk$set(echo=TRUE)

library(tidyverse)
library(bigsnpr) # for seq_log
library(data.table)

knitr::opts_knit$set(root.dir="~/Devel/momentspp/selection/single-pop/")
```

# Creating folders and files

This builds a look-up table on a dense parameter grid that will allow for fast 
assembly of whole chromosomes

```{r, params, include=F, eval=T, results=F, message=F, warning=F}

N <- 10000
u <- 1e-8 # constant among exons as well as "fake" neutral u used to get B-vals
lookup_r <- round(seq_log(from=1e-12, to=1e-2, length.out=500), digits=15)
lookup_s <- round(-seq_log(from=1e-5, to=1e-3, length.out=200), digits=11)

# for discretizing Gamma distributions
dt_r <- data.table(lookup_r)
dt_s <- data.table(lookup_s)

setkey(dt_r, lookup_r)
setkey(dt_s, lookup_s)

params <- setDT(crossing(N, u, lookup_r, lookup_s))
params$scenario <- 1:nrow(params)
n_models <- nrow(params)

fwrite(dt_r, "r.csv", quote=F, row.names=F, col.names=F)
fwrite(dt_s, "s.csv", quote=F, row.names=F, col.names=F)
fwrite(params, "params.csv", quote=F, row.names=F)

```

```{bash dirs, engine.opts='-i', include=F, eval=T, results=F, warning=F}

source ~/.bashrc
  
cat r.csv | while IFS=, read -r r
do
  mkdir r_$r
done

```

Creates demes files that will be used as input for moments++

```{r, demes, include=F, eval=T, results=F, message=F, warning=F}

for(i in 1:n_models) {
  
  if(i %% 10000 == 0) {
    cat(paste(i, "\n"))
  }
  
  N <- as.integer(params[i,]$N)
  u <- as.numeric(params[i,]$u)
  r <- as.numeric(params[i,]$lookup_r)
  s <- as.numeric(params[i,]$lookup_s)
  
  name <- paste("model_", i, sep="")
  sink(paste("r_", r, "/", name, ".yaml", sep=""))
  
  cat("time_units: generations\n")
  cat("demes:\n")
  cat("  - name: X\n")
  cat("    epochs:\n")
  cat("      - end_time: 0\n")
  cat("        start_size: ")
  cat(N)
  cat("\n")
  cat("metadata:\n")
  cat("  - name: mutation\n")
  cat("    left_factor: 1\n")
  cat("    epochs:\n")
  cat("      - end_time: 0\n")
  cat("        rates: [")
  cat(u)
  cat("]\n")
  cat("  - name: recombination\n")
  cat("    epochs:\n")
  cat("      - end_time: 0\n")
  cat("        rates: [")
  cat(r)
  cat("]\n")
  cat("  - name: selection\n")
  cat("    start_time: .inf\n")
  cat("    epochs:\n")
  cat("      - end_time: 0\n")
  cat("        rates: [")
  cat(s)
  cat("]\n")
  
  sink()
}

```

# Running moments++

Computes summary statistics for each parameter combination in lookup_tbl

```{bash moments++, engine.opts='-i', include=F, eval=T, results=F, warning=F}

source ~/.bashrc

for r in r_*
do 
  cd $r
  do
	  momentspp params=opt.bpp F=$f O=25 > /dev/null
  done
  cd ..
done

```

Collects lookup table with Hr values on the parameter grid

```{r, grid, include=F, eval=T, results=F, message=F, warning=F}

params <- fread("params.csv", sep=",", header=T)
n_models <- nrow(params)
c <- 1
hr <- numeric(length=n_models)

for(i in 1:length(lookup_r)) {
  for(j in 1:length(lookup_s)) {
    moms <- read.table(paste("r_", lookup_r[i], 
                             "/model_", as.integer(c), 
                             "_expectations.txt", sep=""))
    hr[c] <- moms[moms$V1 == "Hr_0_0",]$V3
    c <- c + 1
  }
}

params$Hr <- hr
fwrite(params, "lookup_tbl.csv.gz", compress="gzip") # used by other scripts

```

Builds a table of models from which to build genomic landscapes (bgs_maps.R)
and compute expected nucleotide diversity (bgs_pi.R)

```{r, landscapes, include=F, eval=T, results=F, message=F, warning=F}

lookup_tbl <- fread("lookup_tbl.csv.gz")
setkey(lookup_tbl, lookup_r, lookup_s)

lookup_r <- unique(lookup_tbl$lookup_r)
lookup_s <- unique(lookup_tbl$lookup_s)

# for discretizing Gamma distributions and make look-up faster
dt_r <- data.table(lookup_r)
dt_s <- data.table(lookup_s)

setkey(dt_r, lookup_r)
setkey(dt_s, lookup_s)

# constants
N <- unique(lookup_tbl$N)
u <- unique(lookup_tbl$u)
num_constrained <- 1000 # number of (non-recombining) selected segment in chr

# Gamma dist params
scale_sel <- c(5e-5, 2.5e-4)
shape_sel <- c(10)
scales_mu <- c(1e-8)
shapes_mu <- c(3, 10, 30, 100)
scales_rec <- c(1e-6, 1e-7, 1e-8)
shapes_rec <- c(1)

# Geom dist params
avg_rec_spans <- c(1e+4)
avg_mut_spans <- c(1e+4, 1e+5, 5e+5)
avg_neutral_lengths <- c(1e+4)
exon_lengths <- c(1000)  
models <- setDT(crossing(N, num_constrained, 
                         shape_sel, scale_sel, 
                         shapes_mu, scales_mu, 
                         shapes_rec, scales_rec, 
                         avg_rec_spans, avg_mut_spans,
                         avg_neutral_lengths, exon_lengths))
models$model <- 1:nrow(models)
n_models <- nrow(models)

fwrite(models, "models.csv.gz", compress="gzip")
```
